// @c8y/ngx-components/esm2020/core/search/highlight.component.mjs
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { escape, escapeRegExp } from 'lodash-es';
import * as i0 from "@angular/core";
export class HighlightComponent {
    constructor() {
        this.pattern = '';
        this.text = '';
        this.elementClass = 'text-info';
        this.shouldTrimPattern = false;
        this.highlightedText = '';
    }
    ngOnChanges() {
        const textNoHtml = escape(this.text || '');
        const patternEventuallyTrimmed = this.shouldTrimPattern ? this.pattern?.trim() : this.pattern;
        const patternNoHtml = escape(patternEventuallyTrimmed || '');
        const patternForRegex = escapeRegExp(patternNoHtml).replace(/\\\*/g, '.*');
        this.highlightedText = textNoHtml.replace(new RegExp(patternForRegex, 'i'), `<strong class="${this.elementClass}">$&</strong>`);
    }
}
HighlightComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HighlightComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
HighlightComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: HighlightComponent, selector: "c8y-highlight", inputs: { pattern: "pattern", text: "text", elementClass: "elementClass", shouldTrimPattern: "shouldTrimPattern" }, viewQueries: [{ propertyName: "container", first: true, predicate: ["text"], descendants: true, read: ElementRef, static: true }], usesOnChanges: true, ngImport: i0, template: '<span [innerHTML]="highlightedText"></span>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HighlightComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-highlight',
                    template: '<span [innerHTML]="highlightedText"></span>'
                }]
        }], propDecorators: { container: [{
                type: ViewChild,
                args: ['text', { read: ElementRef, static: true }]
            }], pattern: [{
                type: Input
            }], text: [{
                type: Input
            }], elementClass: [{
                type: Input
            }], shouldTrimPattern: [{
                type: Input
            }] } });
