// @c8y/ngx-components/esm2020/core/search/inventory-search.service.mjs
import { Injectable } from '@angular/core';
import { InventoryService, QueriesUtil } from '@c8y/client';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
export class InventorySearchService {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
        this.DEFAULT_PAGE_SIZE = 50;
        this.queriesUtil = new QueriesUtil();
    }
    /**
     * Returns the full-text search results.
     *
     * @param term The search term.
     * @param pagination The currently used pagination.
     */
    search(term, pagination = { currentPage: 1, pageSize: this.DEFAULT_PAGE_SIZE }) {
        return this.inventoryService.list({
            text: term,
            withTotalPages: true,
            pageSize: pagination.pageSize,
            withChildren: false,
            currentPage: pagination.currentPage || 1
        });
    }
    /**
     * Returns group search results by name.
     *
     * @param term The search term - name of group.
     * @param pagination The currently used pagination.
     */
    searchGroups(term, pagination = { currentPage: 1, pageSize: this.DEFAULT_PAGE_SIZE }) {
        const searchFilterQuery = this.getGroupsSearchFilterQuery();
        const query = this.queriesUtil.addAndFilter(searchFilterQuery, { name: `*${term}*` });
        return this.inventoryService.listQuery(query, {
            withTotalPages: true,
            pageSize: pagination.pageSize,
            withChildren: false,
            currentPage: pagination.currentPage || 1
        });
    }
    /**
     * Will return only valid group assets and filter out
     * none useful inventories (e.g. c8y_JsonSchema).
     * @param data All managed objects that should be filtered.
     */
    filterOnlyGroups(data) {
        return data.filter(mo => mo.c8y_IsDeviceGroup || mo.c8y_IsDynamicGroup);
    }
    getGroupsSearchFilterQuery() {
        return {
            __filter: {
                __and: [
                    { __not: { __has: 'c8y_IsDynamicGroup.invisible' } },
                    {
                        __or: [{ __has: 'c8y_IsDeviceGroup' }, { __has: 'c8y_IsDynamicGroup' }]
                    }
                ]
            }
        };
    }
}
InventorySearchService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: InventorySearchService, deps: [{ token: i1.InventoryService }], target: i0.ɵɵFactoryTarget.Injectable });
InventorySearchService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: InventorySearchService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: InventorySearchService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.InventoryService }]; } });
