// @c8y/ngx-components/esm2020/core/search/search.component.mjs
import { Component, Input, Output, ViewChild, EventEmitter } from '@angular/core';
import { SearchService } from './search.service';
import * as i0 from "@angular/core";
import * as i1 from "./search.service";
export class SearchComponent {
    constructor(searchService) {
        this.searchService = searchService;
        this.icon = 'search';
        this.priority = 0;
        this.search = new EventEmitter();
        this.term = '';
    }
    onSearch() {
        this.search.emit(this);
    }
    ngAfterViewInit() {
        this.template = undefined;
        this.viewInitTimeout = setTimeout(() => this.searchService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.searchService.remove(this);
    }
}
SearchComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchComponent, deps: [{ token: i1.SearchService }], target: i0.ɵɵFactoryTarget.Component });
SearchComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: SearchComponent, selector: "c8y-search", inputs: { name: "name", icon: "icon", priority: "priority", term: "term" }, outputs: { search: "search" }, viewQueries: [{ propertyName: "template", first: true, predicate: ["template"], descendants: true }], ngImport: i0, template: "<ng-template #template>\n  <ng-content></ng-content>\n</ng-template>" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-search', template: "<ng-template #template>\n  <ng-content></ng-content>\n</ng-template>" }]
        }], ctorParameters: function () { return [{ type: i1.SearchService }]; }, propDecorators: { name: [{
                type: Input
            }], icon: [{
                type: Input
            }], priority: [{
                type: Input
            }], template: [{
                type: ViewChild,
                args: ['template', { static: false }]
            }], search: [{
                type: Output
            }], term: [{
                type: Input
            }] } });
