// @c8y/ngx-components/esm2020/core/search/search.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { SearchOutletComponent } from './search-outlet.component';
import { SearchComponent } from './search.component';
import { FormsModule } from '@angular/forms';
import { HighlightComponent } from './highlight.component';
import { SearchResultEmptyComponent } from './search-result-empty.component';
import { SearchInputComponent } from './search-input.component';
import { SelectModule } from '../select/select.module';
import { ListGroupModule } from '../list-group/list-group.module';
import { DeviceStatusModule } from '../device-status/device-status.module';
import { A11yModule } from '@angular/cdk/a11y';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * The angular module definition for the search.
 * @exports SearchComponent
 * @exports SearchOutletComponent
 * @exports HighlightComponent A component to highlight search results.
 */
export class SearchModule {
}
SearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, declarations: [SearchOutletComponent,
        SearchComponent,
        HighlightComponent,
        SearchResultEmptyComponent,
        SearchInputComponent], imports: [A11yModule,
        FormsModule,
        BsDropdownModule,
        CommonModule,
        SelectModule,
        ListGroupModule,
        DeviceStatusModule, i1.RouterModule], exports: [SearchOutletComponent,
        SearchComponent,
        HighlightComponent,
        SearchResultEmptyComponent,
        SearchInputComponent] });
SearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, imports: [A11yModule,
        FormsModule,
        BsDropdownModule,
        CommonModule,
        SelectModule,
        ListGroupModule,
        DeviceStatusModule,
        RouterModule.forChild([])] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SearchOutletComponent,
                        SearchComponent,
                        HighlightComponent,
                        SearchResultEmptyComponent,
                        SearchInputComponent
                    ],
                    imports: [
                        A11yModule,
                        FormsModule,
                        BsDropdownModule,
                        CommonModule,
                        SelectModule,
                        ListGroupModule,
                        DeviceStatusModule,
                        RouterModule.forChild([])
                    ],
                    exports: [
                        SearchOutletComponent,
                        SearchComponent,
                        HighlightComponent,
                        SearchResultEmptyComponent,
                        SearchInputComponent
                    ],
                    entryComponents: [SearchInputComponent]
                }]
        }] });
