// @c8y/ngx-components/esm2020/core/search/search.service.mjs
import { Injectable, InjectionToken, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { distinctUntilChanged, shareReplay } from 'rxjs/operators';
import { fromTrigger, hookGeneric, getInjectedHooks, stateToFactory, ExtensionPointForPlugins } from '../common/extension-hooks';
import { PluginsResolveService } from '../plugins';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../plugins";
/**
 * A hook to use for Multi Provider extension.
 * Consider using the `hookSearch` function instead.
 * @deprecated Please don't use the search factory. It will be refactored soon.
 */
export const HOOK_SEARCH = new InjectionToken('HOOK_SEARCH');
/**
 * You can either provide a single `Search` as parameter:
 * ```typescript
 *  hookSearch(...)
 * ```
 *
 * Or an array to directly register multiple:
 * ```typescript
 *  hookSearch([...])
 * ```
 *
 * Or you provide an Service that implements `ExtensionFactory<Search>`
 * ```typescript
 *  export class MySearchFactory implements ExtensionFactory<Search> {...}
 *  ...
 *  hookSearch(MySearchFactory)
 * ```
 * A typed alternative to `HOOK_SEARCH`.
 * @param search The `Search`'s or `ExtensionFactory` to be provided.
 * @returns An `Provider` to be provided in your module.
 * @deprecated Please don't use the search factory. It will be refactored soon.
 */
export function hookSearch(search, options) {
    return hookGeneric(search, HOOK_SEARCH, options);
}
export var SearchFilters;
(function (SearchFilters) {
    SearchFilters["ONLY_DEVICES"] = "onlyDevices";
    SearchFilters["ONLY_GROUPS_AND_ASSETS"] = "onlyGroupsAndAssets";
    SearchFilters["ALL_FILTERS"] = "allFilters";
})(SearchFilters || (SearchFilters = {}));
/**
 * A service which defines the search function.
 */
export class SearchService extends ExtensionPointForPlugins {
    constructor(rootInjector, router, plugins) {
        super(rootInjector, plugins);
        this.router = router;
        this.items$ = this.setupItemsObservable();
    }
    /**
     * Returns the current state.
     * @readonly
     * @returns The current set of actions.
     */
    get state() {
        return this.state$.value;
    }
    /**
     * Adds a new item to the search in the header.
     * @param item The item to add.
     */
    add(item) {
        this.state.add(item);
        this.emitNewState();
    }
    /**
     * Removes an search from the header.
     * @param item The item to remove.
     */
    remove(item) {
        this.state.delete(item);
        this.emitNewState();
    }
    setupItemsObservable() {
        return fromTrigger(this.router, this.refresh$, [
            getInjectedHooks(HOOK_SEARCH, this.injectors),
            () => this.factories,
            stateToFactory(this.state$)
        ]).pipe(shareReplay(1), distinctUntilChanged());
    }
}
SearchService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchService, deps: [{ token: i0.Injector }, { token: i1.Router }, { token: i2.PluginsResolveService }], target: i0.ɵɵFactoryTarget.Injectable });
SearchService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i0.Injector }, { type: i1.Router }, { type: i2.PluginsResolveService }]; } });
