// @c8y/ngx-components/esm2020/core/search/search-result-empty.component.mjs
import { Component, Input } from '@angular/core';
import { gettext } from '../i18n/gettext';
import * as i0 from "@angular/core";
import * as i1 from "../common/icon.directive";
import * as i2 from "../i18n/c8y-translate.directive";
import * as i3 from "../i18n/c8y-translate.pipe";
export class SearchResultEmptyComponent {
    constructor() {
        this.NO_RESULT_FOUND = gettext('No results found.');
        this.msg = this.NO_RESULT_FOUND;
    }
    set _message(msg) {
        if (msg && msg.length > 0) {
            this.msg = msg;
        }
        else {
            this.msg = this.NO_RESULT_FOUND;
        }
    }
    get message() {
        return this.msg;
    }
}
SearchResultEmptyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchResultEmptyComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
SearchResultEmptyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: SearchResultEmptyComponent, selector: "c8y-search-result-empty", inputs: { _message: ["message", "_message"] }, ngImport: i0, template: "<div class=\"no-results\">\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div class=\"p-t-8\">\n      <p><strong>{{ message | translate }}</strong></p>\n      <small translate>Refine your search terms.</small>\n    </div>\n  </div>\n</div>\n", dependencies: [{ kind: "directive", type: i1.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "directive", type: i2.C8yTranslateDirective, selector: "[translate],[ngx-translate]" }, { kind: "pipe", type: i3.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchResultEmptyComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-search-result-empty', template: "<div class=\"no-results\">\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div class=\"p-t-8\">\n      <p><strong>{{ message | translate }}</strong></p>\n      <small translate>Refine your search terms.</small>\n    </div>\n  </div>\n</div>\n" }]
        }], propDecorators: { _message: [{
                type: Input,
                args: ['message']
            }] } });
