// @c8y/ngx-components/esm2020/register-device/register-device-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, Permissions } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class RegisterDeviceNavigationFactory {
    constructor(permissions) {
        this.permissions = permissions;
    }
    get() {
        const userHasPermission = this.permissions.hasAnyRole([
            'ROLE_DEVICE_CONTROL_ADMIN',
            'ROLE_DEVICE_CONTROL_READ'
        ]);
        const nav = [];
        if (userHasPermission) {
            nav.push(new NavigatorNode({
                label: gettext('Registration'),
                icon: 'c8y-device-connect',
                path: '/deviceregistration',
                parent: gettext('Devices'),
                preventDuplicates: true,
                priority: 3000
            }));
        }
        return nav;
    }
}
RegisterDeviceNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceNavigationFactory, deps: [{ token: i1.Permissions }], target: i0.ɵɵFactoryTarget.Injectable });
RegisterDeviceNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Permissions }]; } });
