// @c8y/ngx-components/esm2020/register-device/register-device.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, DeviceBootstrapRealtimeService, DynamicFormsModule, hookNavigator, StepperModule } from '@c8y/ngx-components';
import { DeviceRegistrationViewComponent } from './device-registration-view.component';
import { GeneralDeviceRegistrationComponent } from './general/general-device-registration.component';
import { GeneralDeviceRegistrationButtonComponent } from './general/general-device-registration-button.component';
import { RegisterDeviceNavigationFactory } from './register-device-navigation.factory';
import { RegisterDeviceService } from './register-device.service';
import { ExtensibleDeviceRegistrationButtonComponent } from './extensible/single/extensible-device-registration-button.component';
import { ExtensibleDeviceRegistrationModalComponent } from './extensible/single/extensible-device-registration-modal.component';
import { ExtensibleDeviceRegistrationService } from './extensible/single/extensible-device-registration.service';
import { ExtensibleDeviceRegistrationStepperComponent } from './extensible/single/extensible-device-registration-stepper.component';
import { ExtensibleBulkDeviceRegistrationModalComponent } from './extensible/bulk/extensible-bulk-device-registration-modal.component';
import { BulkExtensibleDeviceRegistrationService } from './extensible/bulk/extensible-bulk-device-registration.service';
import { BulkDeviceRegistrationModalComponent } from './bulk/bulk-device-registration-modal.component';
import { ExtensibleBulkDeviceRegistrationButtonComponent } from './extensible/bulk/extensible-bulk-device-registration-button.component';
import { BulkDeviceRegistrationButtonComponent } from './bulk/bulk-device-registration-button.component';
import { RegisterDeviceDropdownComponent } from './dropdown/register-device-dropdown.component';
import { hookDeviceRegistration } from './dropdown/register-device-extension.service';
import { DeviceRegistrationFactory } from './register-device.factory';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
const routes = [
    {
        path: 'deviceregistration',
        component: DeviceRegistrationViewComponent
    }
];
export class RegisterDeviceModule {
}
RegisterDeviceModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RegisterDeviceModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceModule, declarations: [DeviceRegistrationViewComponent,
        GeneralDeviceRegistrationComponent,
        GeneralDeviceRegistrationButtonComponent,
        BulkDeviceRegistrationModalComponent,
        BulkDeviceRegistrationButtonComponent,
        ExtensibleDeviceRegistrationButtonComponent,
        ExtensibleDeviceRegistrationModalComponent,
        ExtensibleDeviceRegistrationStepperComponent,
        ExtensibleBulkDeviceRegistrationModalComponent,
        ExtensibleBulkDeviceRegistrationButtonComponent,
        RegisterDeviceDropdownComponent], imports: [BsDropdownModule,
        PopoverModule,
        CommonModule,
        CoreModule,
        StepperModule,
        DynamicFormsModule, i1.RouterModule] });
RegisterDeviceModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceModule, providers: [
        hookDeviceRegistration(DeviceRegistrationFactory),
        hookNavigator(RegisterDeviceNavigationFactory),
        DeviceBootstrapRealtimeService,
        RegisterDeviceService,
        ExtensibleDeviceRegistrationService,
        BulkExtensibleDeviceRegistrationService
    ], imports: [BsDropdownModule,
        PopoverModule,
        CommonModule,
        CoreModule,
        StepperModule,
        DynamicFormsModule,
        RouterModule.forChild(routes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        BsDropdownModule,
                        PopoverModule,
                        CommonModule,
                        CoreModule,
                        StepperModule,
                        DynamicFormsModule,
                        RouterModule.forChild(routes)
                    ],
                    declarations: [
                        DeviceRegistrationViewComponent,
                        GeneralDeviceRegistrationComponent,
                        GeneralDeviceRegistrationButtonComponent,
                        BulkDeviceRegistrationModalComponent,
                        BulkDeviceRegistrationButtonComponent,
                        ExtensibleDeviceRegistrationButtonComponent,
                        ExtensibleDeviceRegistrationModalComponent,
                        ExtensibleDeviceRegistrationStepperComponent,
                        ExtensibleBulkDeviceRegistrationModalComponent,
                        ExtensibleBulkDeviceRegistrationButtonComponent,
                        RegisterDeviceDropdownComponent
                    ],
                    providers: [
                        hookDeviceRegistration(DeviceRegistrationFactory),
                        hookNavigator(RegisterDeviceNavigationFactory),
                        DeviceBootstrapRealtimeService,
                        RegisterDeviceService,
                        ExtensibleDeviceRegistrationService,
                        BulkExtensibleDeviceRegistrationService
                    ],
                    entryComponents: [
                        GeneralDeviceRegistrationComponent,
                        ExtensibleBulkDeviceRegistrationButtonComponent,
                        BulkDeviceRegistrationModalComponent,
                        ExtensibleDeviceRegistrationModalComponent,
                        RegisterDeviceDropdownComponent
                    ]
                }]
        }] });
