// @c8y/ngx-components/esm2020/register-device/extensible/bulk/extensible-bulk-device-registration.service.mjs
import { Injectable } from '@angular/core';
import { ApplicationService, FetchClient } from '@c8y/client';
import { AppStateService, gettext } from '@c8y/ngx-components';
import { BaseExtensibleDeviceRegistrationService } from '../base-extensible-device-registration.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "@c8y/ngx-components";
export class BulkExtensibleDeviceRegistrationService extends BaseExtensibleDeviceRegistrationService {
    constructor(applicationService, stateService, client) {
        super(applicationService, stateService);
        this.applicationService = applicationService;
        this.stateService = stateService;
        this.client = client;
    }
    getDeviceRegistrationProviders() {
        return super.getDeviceRegistrationProviders(true);
    }
    async getBulkDescription(contextPath) {
        const res = await this.client.fetch(this.getMetadataUrl(contextPath), {
            params: { lang: this.stateService.state.lang }
        });
        const { c8y_DeviceRegistration } = await res.json();
        if (!c8y_DeviceRegistration || !c8y_DeviceRegistration.bulk) {
            throw new Error(gettext('The response of the microservice is incomplete. The bulk registration information is missing.'));
        }
        return c8y_DeviceRegistration.bulk;
    }
    async upload(csv, contextPath) {
        const method = 'POST';
        const body = this.createBinaryRequestBody(csv);
        const res = await this.client.fetch(this.getBulkUrl(contextPath), { method, body });
        const data = await res.json();
        return { res, data };
    }
    getBulkUrl(contextPath) {
        return `${this.getServiceBaseUrl(contextPath)}/bulk`;
    }
    createBinaryRequestBody(file) {
        const body = new FormData();
        body.append('file', file, file.name);
        return body;
    }
}
BulkExtensibleDeviceRegistrationService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkExtensibleDeviceRegistrationService, deps: [{ token: i1.ApplicationService }, { token: i2.AppStateService }, { token: i1.FetchClient }], target: i0.ɵɵFactoryTarget.Injectable });
BulkExtensibleDeviceRegistrationService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkExtensibleDeviceRegistrationService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkExtensibleDeviceRegistrationService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.ApplicationService }, { type: i2.AppStateService }, { type: i1.FetchClient }]; } });
