// @c8y/ngx-components/esm2020/auth-configuration/sso-configuration/templates/custom-configuration-mapper.mjs
import { TemplateType } from '../sso-configuration.model';
import { GrantType, TenantLoginOptionType } from '@c8y/client';
import { RequestConfiguration } from '../template-parts/request-configuration.model';
import { CertificateType, SignatureConfiguration } from '../template-parts/signature-configuration.model';
import { Injectable } from '@angular/core';
import { ExternalToken } from '../template-parts/external-token-config.model';
import * as i0 from "@angular/core";
export class CustomConfigurationMapper {
    mapFrom(templateModel) {
        const ssoConfiguration = {
            accessTokenToUserDataMappings: templateModel.accessTokenToUserDataMappings,
            audience: templateModel.audience,
            authorizationRequest: templateModel.authorizationRequest.toRequest(),
            buttonName: templateModel.buttonName,
            clientId: templateModel.clientId,
            issuer: templateModel.issuer,
            logoutRequest: templateModel.logoutRequest.toRequest(),
            onNewUser: templateModel.onNewUser,
            providerName: templateModel.providerName,
            redirectToPlatform: templateModel.redirectToPlatform,
            refreshRequest: templateModel.refreshRequest.toRequest(),
            signatureVerificationConfig: templateModel.signatureVerificationConfig.toSignatureVerificationConfig(),
            template: TemplateType.CUSTOM,
            tokenRequest: templateModel.tokenRequest.toRequest(),
            userIdConfig: templateModel.userIdConfig,
            userManagementSource: templateModel.userManagementSource,
            visibleOnLoginPage: templateModel.visibleOnLoginPage,
            type: TenantLoginOptionType.OAUTH2,
            grantType: GrantType.AUTHORIZATION_CODE,
            externalTokenConfig: templateModel.externalTokenConfig.toExternalTokenConfig()
        };
        return ssoConfiguration;
    }
    mapTo(ssoConfiguration) {
        return {
            accessTokenToUserDataMappings: ssoConfiguration.accessTokenToUserDataMappings,
            audience: ssoConfiguration.audience,
            authorizationRequest: new RequestConfiguration(ssoConfiguration.authorizationRequest),
            buttonName: ssoConfiguration.buttonName,
            clientId: ssoConfiguration.clientId,
            issuer: ssoConfiguration.issuer,
            logoutRequest: new RequestConfiguration(ssoConfiguration.logoutRequest),
            onNewUser: ssoConfiguration.onNewUser,
            providerName: ssoConfiguration.providerName,
            redirectToPlatform: ssoConfiguration.redirectToPlatform,
            refreshRequest: new RequestConfiguration(ssoConfiguration.refreshRequest),
            signatureVerificationConfig: new SignatureConfiguration(ssoConfiguration.signatureVerificationConfig),
            template: TemplateType.CUSTOM,
            tokenRequest: new RequestConfiguration(ssoConfiguration.tokenRequest),
            userIdConfig: ssoConfiguration.userIdConfig,
            userManagementSource: ssoConfiguration.userManagementSource,
            visibleOnLoginPage: ssoConfiguration.visibleOnLoginPage,
            certificateType: CertificateType.CUSTOM,
            externalTokenConfig: new ExternalToken(ssoConfiguration.externalTokenConfig)
        };
    }
}
CustomConfigurationMapper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CustomConfigurationMapper, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
CustomConfigurationMapper.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CustomConfigurationMapper, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CustomConfigurationMapper, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
