// @c8y/ngx-components/esm2020/auth-configuration/sso-configuration/templates/template.component.mjs
import { Input, Output, EventEmitter, Component } from '@angular/core';
import { SsoConfigurationMapper } from './sso-configuration.mapper';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./sso-configuration.mapper";
export class TemplateComponent {
    constructor(configurationMapper) {
        this.configurationMapper = configurationMapper;
        this.ssoConfigurationChange = new EventEmitter();
    }
    ngOnInit() {
        this.triggerSubscription = this.ssoConfigurationChangeTrigger.subscribe(() => {
            this.emitSsoConfiguration();
        });
    }
    ngOnChanges(changes) {
        if (changes.ssoConfiguration && changes.ssoConfiguration.currentValue) {
            this.mapSsoConfiguration(changes.ssoConfiguration.currentValue);
        }
    }
    ngOnDestroy() {
        this.triggerSubscription.unsubscribe();
    }
    mapSsoConfiguration(ssoConfiguration) {
        this.templateModel = this.configurationMapper.mapTo(ssoConfiguration);
    }
    emitSsoConfiguration() {
        const ssoConfiguration = this.configurationMapper.mapFrom(this.templateModel);
        if (this.ssoConfiguration.id) {
            ssoConfiguration.id = this.ssoConfiguration.id;
        }
        this.ssoConfigurationChange.emit(ssoConfiguration);
    }
}
TemplateComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TemplateComponent, deps: [{ token: i1.SsoConfigurationMapper }], target: i0.ɵɵFactoryTarget.Component });
TemplateComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: TemplateComponent, selector: "ng-component", inputs: { apps: "apps", groups: "groups", inventoryRoles: "inventoryRoles", ssoConfiguration: "ssoConfiguration", ssoConfigurationChangeTrigger: "ssoConfigurationChangeTrigger" }, outputs: { ssoConfigurationChange: "ssoConfigurationChange" }, usesOnChanges: true, ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TemplateComponent, decorators: [{
            type: Component,
            args: [{
                    template: ''
                }]
        }], ctorParameters: function () { return [{ type: i1.SsoConfigurationMapper }]; }, propDecorators: { apps: [{
                type: Input
            }], groups: [{
                type: Input
            }], inventoryRoles: [{
                type: Input
            }], ssoConfiguration: [{
                type: Input
            }], ssoConfigurationChangeTrigger: [{
                type: Input
            }], ssoConfigurationChange: [{
                type: Output
            }] } });
