// @c8y/ngx-components/esm2020/core/tabs/tab.component.mjs
import { Component, ViewChild, Input, ViewContainerRef, Injector, EnvironmentInjector } from '@angular/core';
import { TabsService } from './tabs.service';
import * as i0 from "@angular/core";
import * as i1 from "./tabs.service";
export class TabComponent {
    constructor(tabs, _injector) {
        this.tabs = tabs;
        this.label = '';
        this.injector = _injector;
    }
    ngAfterContentInit() {
        setTimeout(() => this.tabs.add(this));
    }
    ngOnDestroy() {
        this.tabs.remove(this);
    }
}
TabComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabComponent, deps: [{ token: i1.TabsService }, { token: i0.EnvironmentInjector }], target: i0.ɵɵFactoryTarget.Component });
TabComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: TabComponent, selector: "c8y-tab", inputs: { path: "path", label: "label", icon: "icon", priority: "priority", orientation: "orientation", injector: "injector" }, viewQueries: [{ propertyName: "template", first: true, predicate: ["tpl"], descendants: true }], ngImport: i0, template: '<ng-template #tpl><ng-content></ng-content></ng-template>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-tab',
                    template: '<ng-template #tpl><ng-content></ng-content></ng-template>'
                }]
        }], ctorParameters: function () { return [{ type: i1.TabsService }, { type: i0.EnvironmentInjector }]; }, propDecorators: { template: [{
                type: ViewChild,
                args: ['tpl', { static: false }]
            }], path: [{
                type: Input
            }], label: [{
                type: Input
            }], icon: [{
                type: Input
            }], priority: [{
                type: Input
            }], orientation: [{
                type: Input
            }], injector: [{
                type: Input
            }] } });
