// @c8y/ngx-components/esm2020/core/tabs/tabs.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule } from '../common/common.module';
import { TabComponent } from './tab.component';
import { TabsOutletComponent } from './tabs-outlet.component';
import { ConditionalTabsOutletComponent } from './conditional-tabs-outlet.component';
import { HeaderModule } from '../header/header.module';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class TabsModule {
}
TabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: TabsModule, declarations: [ConditionalTabsOutletComponent, TabsOutletComponent, TabComponent], imports: [CommonModule, i1.RouterModule, HeaderModule], exports: [ConditionalTabsOutletComponent, TabsOutletComponent, TabComponent] });
TabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsModule, imports: [CommonModule, RouterModule.forChild([]), HeaderModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ConditionalTabsOutletComponent, TabsOutletComponent, TabComponent],
                    imports: [CommonModule, RouterModule.forChild([]), HeaderModule],
                    exports: [ConditionalTabsOutletComponent, TabsOutletComponent, TabComponent]
                }]
        }] });
