// @c8y/ngx-components/esm2020/core/login/login.module.mjs
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { ChangePasswordComponent } from './change-password.component';
import { CredentialsComponent } from './credentials.component';
import { LoginComponent } from './login.component';
import { LoginService } from './login.service';
import { RecoverPasswordComponent } from './recover-password.component';
import { AuthenticationModule } from '../authentication/authentication.module';
import { AlertModule } from '../alert/alert.module';
import { TotpAuthComponent } from './totp-auth.component';
import { PasswordStrengthValidatorDirective } from './password-strength-validator.directive';
import { TenantIdSetupComponent } from './tenant-id-setup.component';
import { CredentialsFromQueryParamsService } from './credentials-from-query-params.service';
import * as i0 from "@angular/core";
/**
 * The angular module definition for the login.
 * @exports LoginComponent
 */
export class LoginModule {
    static forRoot() {
        return {
            ngModule: LoginModule,
            providers: LoginModule.providers()
        };
    }
    static providers() {
        return [LoginService, CredentialsFromQueryParamsService];
    }
}
LoginModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoginModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LoginModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: LoginModule, declarations: [LoginComponent,
        RecoverPasswordComponent,
        ChangePasswordComponent,
        CredentialsComponent,
        TotpAuthComponent,
        PasswordStrengthValidatorDirective,
        TenantIdSetupComponent], imports: [FormsModule,
        TooltipModule,
        CommonModule,
        AuthenticationModule,
        AlertModule,
        C8yFormsModule], exports: [LoginComponent] });
LoginModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoginModule, imports: [FormsModule,
        TooltipModule,
        CommonModule,
        AuthenticationModule,
        AlertModule,
        C8yFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoginModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        LoginComponent,
                        RecoverPasswordComponent,
                        ChangePasswordComponent,
                        CredentialsComponent,
                        TotpAuthComponent,
                        PasswordStrengthValidatorDirective,
                        TenantIdSetupComponent
                    ],
                    imports: [
                        FormsModule,
                        TooltipModule,
                        CommonModule,
                        AuthenticationModule,
                        AlertModule,
                        C8yFormsModule
                    ],
                    exports: [LoginComponent]
                }]
        }] });
