// @c8y/ngx-components/esm2020/core/login/password-strength-validator.directive.mjs
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { StrengthValidatorService } from './strength-validator-service';
import * as i0 from "@angular/core";
import * as i1 from "./strength-validator-service";
export class PasswordStrengthValidatorDirective {
    set passwordStrengthEnforced(value) {
        this.forced = value;
    }
    constructor(passwordService) {
        this.passwordService = passwordService;
    }
    validate(control) {
        const strengthFulfilled = this.passwordService.isStrong(control.value || '');
        const enforcementForcedAndNotFulfilled = this.forced && !strengthFulfilled;
        return enforcementForcedAndNotFulfilled ? { passwordStrength: true } : null;
    }
}
PasswordStrengthValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordStrengthValidatorDirective, deps: [{ token: i1.StrengthValidatorService }], target: i0.ɵɵFactoryTarget.Directive });
PasswordStrengthValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: PasswordStrengthValidatorDirective, selector: "[passwordStrengthEnforced]", inputs: { passwordStrengthEnforced: "passwordStrengthEnforced" }, providers: [
        { provide: NG_VALIDATORS, useExisting: PasswordStrengthValidatorDirective, multi: true }
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordStrengthValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[passwordStrengthEnforced]',
                    providers: [
                        { provide: NG_VALIDATORS, useExisting: PasswordStrengthValidatorDirective, multi: true }
                    ]
                }]
        }], ctorParameters: function () { return [{ type: i1.StrengthValidatorService }]; }, propDecorators: { passwordStrengthEnforced: [{
                type: Input
            }] } });
