// @c8y/ngx-components/esm2020/core/login/strength-validator-service.mjs
import { Injectable } from '@angular/core';
import { PasswordService } from '../authentication';
import { PasswordStrength } from '@c8y/client';
import * as i0 from "@angular/core";
import * as i1 from "../authentication";
export class StrengthValidatorService {
    constructor(passwordService) {
        this.passwordService = passwordService;
    }
    isStrong(password) {
        return this.isPasswordGreen(this.passwordService.getStrengthColor(password).passwordStrength);
    }
    isPasswordGreen(strength) {
        return strength === PasswordStrength.GREEN;
    }
}
StrengthValidatorService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StrengthValidatorService, deps: [{ token: i1.PasswordService }], target: i0.ɵɵFactoryTarget.Injectable });
StrengthValidatorService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StrengthValidatorService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StrengthValidatorService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.PasswordService }]; } });
