// @c8y/ngx-components/esm2020/register-device/bulk/bulk-device-registration-button.component.mjs
import { Component } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BulkDeviceRegistrationModalComponent } from './bulk-device-registration-modal.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
import * as i2 from "@c8y/ngx-components";
export class BulkDeviceRegistrationButtonComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    click() {
        this.modalService.show(BulkDeviceRegistrationModalComponent, {
            class: 'modal-sm',
            ariaDescribedby: 'modal-body',
            ariaLabelledBy: 'modal-title',
            ignoreBackdropClick: true
        });
    }
}
BulkDeviceRegistrationButtonComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkDeviceRegistrationButtonComponent, deps: [{ token: i1.BsModalService }], target: i0.ɵɵFactoryTarget.Component });
BulkDeviceRegistrationButtonComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: BulkDeviceRegistrationButtonComponent, selector: "extensible-device-registration-component", ngImport: i0, template: "<button title=\"{{ 'General' | translate }}\" type=\"button\" (click)=\"click()\">\n  <i c8yIcon=\"upload\"></i>\n  {{ 'General' | translate }}\n</button>\n", dependencies: [{ kind: "directive", type: i2.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "pipe", type: i2.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkDeviceRegistrationButtonComponent, decorators: [{
            type: Component,
            args: [{ selector: 'extensible-device-registration-component', template: "<button title=\"{{ 'General' | translate }}\" type=\"button\" (click)=\"click()\">\n  <i c8yIcon=\"upload\"></i>\n  {{ 'General' | translate }}\n</button>\n" }]
        }], ctorParameters: function () { return [{ type: i1.BsModalService }]; } });
