// @c8y/ngx-components/esm2020/core/clipboard/clipboard.service.mjs
import { gettext } from '../i18n';
import { Injectable } from '@angular/core';
import { AlertService } from '../alert/alert.service';
import * as i0 from "@angular/core";
import * as i1 from "../alert/alert.service";
export class ClipboardService {
    constructor(alertService) {
        this.alertService = alertService;
    }
    async writeText(text) {
        try {
            await navigator.clipboard.writeText(text);
            this.alertService.success(gettext('Copied to clipboard.'));
        }
        catch (e) {
            this.alertService.danger(gettext('Could not copy to clipboard.'));
        }
    }
    async readText() {
        try {
            const text = await navigator.clipboard.readText();
            this.alertService.success(gettext('Pasted from clipboard.'));
            return text;
        }
        catch (e) {
            this.alertService.danger(gettext('Could not paste from clipboard.'));
        }
    }
}
ClipboardService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ClipboardService, deps: [{ token: i1.AlertService }], target: i0.ɵɵFactoryTarget.Injectable });
ClipboardService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ClipboardService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ClipboardService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.AlertService }]; } });
