// @c8y/ngx-components/esm2020/reports/reports.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, FormsModule } from '@c8y/ngx-components';
import { ReportsService } from './reports.service';
import { CronService } from './cron.service';
import { ExportSchedulesComponent } from './export-schedules.component';
import { ScheduleModalComponent } from './schedule-modal.component';
import { CronComponent } from './cron.component';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import * as i0 from "@angular/core";
/**
 * The angular module definition for reports.
 * @exports ExportSchedulesComponent
 * @exports ScheduleModalComponent
 * @exports CronComponent
 */
export class ReportsModule {
}
ReportsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ReportsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ReportsModule, declarations: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent], imports: [CoreModule, FormsModule, BsDropdownModule, TooltipModule], exports: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent] });
ReportsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportsModule, providers: [ReportsService, CronService], imports: [CoreModule, FormsModule, BsDropdownModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent],
                    imports: [CoreModule, FormsModule, BsDropdownModule, TooltipModule],
                    entryComponents: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent],
                    providers: [ReportsService, CronService],
                    exports: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent]
                }]
        }] });
