// @c8y/ngx-components/esm2020/widgets/definitions/cockpit-welcome/index.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';
// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
export const cockpitWelcomeWidgetDefinition = {
    id: defaultWidgetIds.COCKPIT_WELCOME,
    label: gettext('Welcome to Cockpit'),
    description: gettext('Displays a welcome message for Cockpit'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/cockpit-welcome').then(m => m.WelcomeToCockpitComponent),
    previewImage: 'c8y-style-assets/welcome-widget-pr.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 12,
                _height: 6
            },
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};
