// @c8y/ngx-components/esm2020/core/select-modal/select-modal.module.mjs
import { NgModule } from '@angular/core';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { FormsModule } from '../forms/forms.module';
import { ListGroupModule } from '../list-group/list-group.module';
import { ProductExperienceModule } from '../product-experience/product-experience.module';
import { SearchModule } from '../search/search.module';
import { SelectModalFilterPipe } from './select-modal-filter.pipe';
import { SelectModalComponent } from './select-modal.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/dropdown";
export class SelectModalModule {
}
SelectModalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SelectModalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SelectModalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SelectModalModule, declarations: [SelectModalComponent, SelectModalFilterPipe], imports: [i1.BsDropdownModule, CommonModule,
        ListGroupModule,
        SearchModule,
        FormsModule,
        ProductExperienceModule], exports: [SelectModalComponent, SelectModalFilterPipe] });
SelectModalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SelectModalModule, imports: [BsDropdownModule.forRoot(),
        CommonModule,
        ListGroupModule,
        SearchModule,
        FormsModule,
        ProductExperienceModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SelectModalModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        BsDropdownModule.forRoot(),
                        CommonModule,
                        ListGroupModule,
                        SearchModule,
                        FormsModule,
                        ProductExperienceModule
                    ],
                    exports: [SelectModalComponent, SelectModalFilterPipe],
                    declarations: [SelectModalComponent, SelectModalFilterPipe]
                }]
        }] });
