// @c8y/ngx-components/esm2020/register-device/dropdown/register-device-extension.service.mjs
import { Injectable, InjectionToken, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { ExtensionPointWithoutStateForPlugins, fromTriggerOnce, hookGeneric, PluginsResolveService } from '@c8y/ngx-components';
import { flatten } from 'lodash';
import { shareReplay, startWith } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@c8y/ngx-components";
/**
 * A hook to use for Multi Provider extension.
 * @deprecated Consider using the `hookDeviceRegistration` function instead.
 */
export const HOOK_DEVICE_REGISTRATION = new InjectionToken('HOOK_DEVICE_REGISTRATION');
/**
 * You can either provide a single `RegisterDeviceExtension` as parameter:
 * ```typescript
 *  hookDeviceRegistration(...)
 * ```
 *
 * Or an array to directly register multiple:
 * ```typescript
 *  hookDeviceRegistration([...])
 * ```
 *
 * Or you provide an Service that implements `ExtensionFactory<RegisterDeviceExtension>`
 * ```typescript
 *  export class MyDeviceRegistrationFactory implements ExtensionFactory<RegisterDeviceExtension> {...}
 *  ...
 *  hookDeviceRegistration(MyDeviceRegistrationFactory)
 * ```
 * A typed alternative to `HOOK_DEVICE_REGISTRATION`.
 * @param registration The `RegisterDeviceExtension`'s or `ExtensionFactory` to be provided.
 * @returns An `Provider` to be provided in your module.
 */
export function hookDeviceRegistration(registration, options) {
    return hookGeneric(registration, HOOK_DEVICE_REGISTRATION, options);
}
/**
 * A service which defines device registration options.
 */
export class RegisterDeviceExtensionService extends ExtensionPointWithoutStateForPlugins {
    constructor(rootInjector, router, plugins) {
        super(rootInjector, plugins);
        this.router = router;
        this.items$ = this.setupItemsObservable();
    }
    setupItemsObservable() {
        return fromTriggerOnce(this.router, this.refresh$, [
            () => flatten(this.injectors.map(injector => injector.get(HOOK_DEVICE_REGISTRATION, [], { self: true }))),
            () => this.factories
        ]).pipe(startWith([]), shareReplay(1));
    }
}
RegisterDeviceExtensionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceExtensionService, deps: [{ token: i0.Injector }, { token: i1.Router }, { token: i2.PluginsResolveService }], target: i0.ɵɵFactoryTarget.Injectable });
RegisterDeviceExtensionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceExtensionService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RegisterDeviceExtensionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i0.Injector }, { type: i1.Router }, { type: i2.PluginsResolveService }]; } });
