// @c8y/ngx-components/esm2020/operations/operation-details/operation-details.module.mjs
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationListItemService } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { OperationsListModule } from '@c8y/ngx-components/operations/operations-list';
import { OperationDetailsService } from './operation-details.service';
import { OperationDetailsComponent } from './operation-details.component';
import { OperationDetailsModalComponent } from './operation-details-modal.component';
import * as i0 from "@angular/core";
/**
 * This module allows to display details of a single operation in a modal.
 */
export class OperationDetailsModule {
}
OperationDetailsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OperationDetailsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsModule, declarations: [OperationDetailsComponent, OperationDetailsModalComponent], imports: [CoreModule, PopoverModule, TooltipModule, CollapseModule, OperationsListModule], exports: [OperationDetailsComponent] });
OperationDetailsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsModule, providers: [OperationDetailsService, BulkOperationListItemService], imports: [CoreModule, PopoverModule, TooltipModule, CollapseModule, OperationsListModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, PopoverModule, TooltipModule, CollapseModule, OperationsListModule],
                    providers: [OperationDetailsService, BulkOperationListItemService],
                    declarations: [OperationDetailsComponent, OperationDetailsModalComponent],
                    entryComponents: [OperationDetailsComponent, OperationDetailsModalComponent],
                    exports: [OperationDetailsComponent]
                }]
        }] });
