// @c8y/ngx-components/esm2020/operations/operation-details/operation-details.service.mjs
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { OperationService } from '@c8y/client';
import { OperationDetailsModalComponent } from './operation-details-modal.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "ngx-bootstrap/modal";
export class OperationDetailsService {
    constructor(operationService, modalService) {
        this.operationService = operationService;
        this.modalService = modalService;
    }
    /**
     * Opens a modal window with the details of given single operation.
     * @param operationOrOperationId Operation's object or id.
     * @param options Additional options:
     * - `collapsed`: boolean - whether the details pane should be initially collapsed
     * - `readOnly`: boolean - whether the details should be displayed in read-only mode
     */
    async openDetails(operationOrOperationId, options) {
        const operation = operationOrOperationId.id
            ? operationOrOperationId
            : (await this.operationService.detail(operationOrOperationId)).data;
        const initialState = { operation, ...options };
        this.modalService.show(OperationDetailsModalComponent, {
            initialState,
            class: 'modal-lg',
            ariaDescribedby: 'modal-body',
            ariaLabelledBy: 'modal-title'
        });
    }
}
OperationDetailsService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsService, deps: [{ token: i1.OperationService }, { token: i2.BsModalService }], target: i0.ɵɵFactoryTarget.Injectable });
OperationDetailsService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationDetailsService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OperationService }, { type: i2.BsModalService }]; } });
