// @c8y/ngx-components/esm2020/loriot-device-registration/loriot-device-registration.factory.mjs
import { Injectable } from '@angular/core';
import { TenantUiService } from '@c8y/ngx-components';
import { LoriotDeviceRegistrationButtonComponent } from './loriot-device-registration-button.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class LoriotDeviceRegistrationFactory {
    constructor(tenantService) {
        this.tenantService = tenantService;
    }
    get() {
        const items = [];
        if (this.tenantService.isMicroserviceSubscribedInCurrentTenant('loriot')) {
            items.push({
                template: LoriotDeviceRegistrationButtonComponent,
                priority: 97,
                category: 'single'
            });
        }
        return items;
    }
}
LoriotDeviceRegistrationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationFactory, deps: [{ token: i1.TenantUiService }], target: i0.ɵɵFactoryTarget.Injectable });
LoriotDeviceRegistrationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.TenantUiService }]; } });
