// @c8y/ngx-components/esm2020/loriot-device-registration/loriot-device-registration.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule } from '@c8y/ngx-components';
import { hookDeviceRegistration } from '@c8y/ngx-components/register-device';
import { LoriotDeviceRegistrationButtonComponent } from './loriot-device-registration-button.component';
import { LoriotDeviceRegistrationComponent } from './loriot-device-registration.component';
import { LoriotDeviceRegistrationFactory } from './loriot-device-registration.factory';
import { LoriotProviderService } from './loriot-provider.service';
import * as i0 from "@angular/core";
export class LoriotDeviceRegistrationModule {
}
LoriotDeviceRegistrationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LoriotDeviceRegistrationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationModule, declarations: [LoriotDeviceRegistrationButtonComponent, LoriotDeviceRegistrationComponent], imports: [CoreModule, CommonModule] });
LoriotDeviceRegistrationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationModule, providers: [hookDeviceRegistration(LoriotDeviceRegistrationFactory), LoriotProviderService], imports: [CoreModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotDeviceRegistrationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, CommonModule],
                    declarations: [LoriotDeviceRegistrationButtonComponent, LoriotDeviceRegistrationComponent],
                    providers: [hookDeviceRegistration(LoriotDeviceRegistrationFactory), LoriotProviderService]
                }]
        }] });
