// @c8y/ngx-components/esm2020/operations/stepper-bulk-type-software/stepper-bulk-type-software.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { baseUrl, BulkOperationType, hookListBulkType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperFramesModule } from '@c8y/ngx-components/operations/stepper-frames';
import { ConfirmSoftwareSelectionStepComponent } from './confirm-software-selection-step.component';
import { SelectSoftwareStepComponent } from './select-software-step.component';
import { StepperBulkTypeSoftwareComponent } from './stepper-bulk-type-software.component';
import * as i0 from "@angular/core";
/** Module for the 'Software update' operation type stepper */
export class StepperBulkTypeSoftwareModule {
}
StepperBulkTypeSoftwareModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeSoftwareModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperBulkTypeSoftwareModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeSoftwareModule, declarations: [StepperBulkTypeSoftwareComponent,
        ConfirmSoftwareSelectionStepComponent,
        SelectSoftwareStepComponent], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        SharedRepositoryModule,
        StepperFramesModule], exports: [StepperBulkTypeSoftwareComponent] });
StepperBulkTypeSoftwareModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeSoftwareModule, providers: [
        hookListBulkType({
            type: BulkOperationType.SOFTWARE,
            c8yIcon: 'c8y-tools',
            name: gettext('Software update'),
            path: `${baseUrl}software`,
            component: StepperBulkTypeSoftwareComponent,
            fragments: ['c8y_SoftwareList', 'c8y_SoftwareUpdate'],
            selected: false
        })
    ], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        SharedRepositoryModule,
        StepperFramesModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeSoftwareModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        StepperBulkTypeSoftwareComponent,
                        ConfirmSoftwareSelectionStepComponent,
                        SelectSoftwareStepComponent
                    ],
                    imports: [
                        CoreModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BulkOperationStepperModule,
                        SharedRepositoryModule,
                        StepperFramesModule
                    ],
                    providers: [
                        hookListBulkType({
                            type: BulkOperationType.SOFTWARE,
                            c8yIcon: 'c8y-tools',
                            name: gettext('Software update'),
                            path: `${baseUrl}software`,
                            component: StepperBulkTypeSoftwareComponent,
                            fragments: ['c8y_SoftwareList', 'c8y_SoftwareUpdate'],
                            selected: false
                        })
                    ],
                    entryComponents: [StepperBulkTypeSoftwareComponent],
                    exports: [StepperBulkTypeSoftwareComponent]
                }]
        }] });
