// @c8y/ngx-components/esm2020/device-shell/shared/device-shell.service.mjs
import { Injectable } from '@angular/core';
import { OperationService } from '@c8y/client';
import { gettext, OptionsService } from '@c8y/ngx-components';
import { CommandDeliveryType } from './device-shell.model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/client";
export class DeviceShellService {
    constructor(optionsService, operationService) {
        this.optionsService = optionsService;
        this.operationService = operationService;
    }
    getDeliveryTypes() {
        return [
            {
                name: gettext(CommandDeliveryType.DEFAULT),
                default: true
            },
            {
                name: gettext(CommandDeliveryType.SMS)
            }
        ];
    }
    async canSendCommandsViaSMS() {
        return !!(await this.optionsService.getSystemOption('messaging', 'provider', false));
    }
    createCommandOperation(deviceId, command, deliveryType) {
        const operation = {
            deviceId,
            description: gettext('Execute shell command') + (command.name ? `: ${command.name}` : ''),
            deliveryType: deliveryType === CommandDeliveryType.SMS ? CommandDeliveryType.SMS : undefined,
            c8y_Command: {
                text: command.text
            }
        };
        return this.operationService.create(operation);
    }
}
DeviceShellService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellService, deps: [{ token: i1.OptionsService }, { token: i2.OperationService }], target: i0.ɵɵFactoryTarget.Injectable });
DeviceShellService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OptionsService }, { type: i2.OperationService }]; } });
