// @c8y/ngx-components/esm2020/sensor-phone/connect-smartphone.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, OptionsService } from '@c8y/ngx-components';
import { BsModalService } from 'ngx-bootstrap/modal';
import { of } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "ngx-bootstrap/modal";
export class ConnectSmartphoneFactory {
    constructor(optionsService, bsModalService) {
        this.optionsService = optionsService;
        this.bsModalService = bsModalService;
    }
    get() {
        if (this.optionsService.get('sensorPhone', false)) {
            return of([
                {
                    icon: 'c8y-icon c8y-icon-mobile-add',
                    type: 'quicklink',
                    label: gettext('Connect smartphone'),
                    url: undefined,
                    click: async () => {
                        const { SensorPhoneModalComponent } = await import('@c8y/ngx-components/sensor-phone/sensor-phone-modal');
                        this.bsModalService.show(SensorPhoneModalComponent, {
                            backdrop: 'static',
                            ariaDescribedby: 'modal-body',
                            ariaLabelledBy: 'modal-title',
                            class: 'lg'
                        });
                    }
                }
            ]);
        }
    }
}
ConnectSmartphoneFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectSmartphoneFactory, deps: [{ token: i1.OptionsService }, { token: i2.BsModalService }], target: i0.ɵɵFactoryTarget.Injectable });
ConnectSmartphoneFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectSmartphoneFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectSmartphoneFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OptionsService }, { type: i2.BsModalService }]; } });
