// @c8y/ngx-components/esm2020/repository/firmware/device-tab/firmware-repository-device-tab.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FirmwareDeviceTabComponent } from './firmware-device-tab.component';
import { FirmwareDeviceTabGuard } from './firmware-device-tab.guard';
import { CoreModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
import * as i0 from "@angular/core";
export class FirmwareRepositoryDeviceTabModule {
    static forRoot() {
        return {
            ngModule: FirmwareRepositoryDeviceTabModule,
            providers: [
                FirmwareDeviceTabGuard,
                hookRoute({
                    context: ViewContext.Device,
                    path: 'firmware',
                    component: FirmwareDeviceTabComponent,
                    label: gettext('Firmware'),
                    icon: 'c8y-firmware',
                    priority: 500,
                    canActivate: [FirmwareDeviceTabGuard]
                })
            ]
        };
    }
}
FirmwareRepositoryDeviceTabModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryDeviceTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FirmwareRepositoryDeviceTabModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryDeviceTabModule, declarations: [FirmwareDeviceTabComponent], imports: [CommonModule, CoreModule, SharedRepositoryModule, OperationDetailsModule] });
FirmwareRepositoryDeviceTabModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryDeviceTabModule, imports: [CommonModule, CoreModule, SharedRepositoryModule, OperationDetailsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryDeviceTabModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CoreModule, SharedRepositoryModule, OperationDetailsModule],
                    declarations: [FirmwareDeviceTabComponent],
                    entryComponents: [FirmwareDeviceTabComponent]
                }]
        }] });
