// @c8y/ngx-components/esm2020/protocol-lwm2m/components/bootstrap-parameters/lwm2m-bootstrap-parameters.service.mjs
import { Injectable } from '@angular/core';
import { FetchClient } from '@c8y/client';
import { Lwm2mClientService } from '@c8y/ngx-components/protocol-lwm2m/services';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
export class Lwm2mBootstrapParametersService extends Lwm2mClientService {
    constructor(client) {
        super(client);
    }
    detail(id) {
        return super.detail(id);
    }
    update(entity) {
        return super.update(entity);
    }
    onBeforeUpdate(entity) {
        delete entity.id;
        return entity;
    }
    getDetailUrl(entity) {
        return `${this.listUrl}/${super.getEntityId(entity)}/bootstrapParams`;
    }
}
Lwm2mBootstrapParametersService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mBootstrapParametersService, deps: [{ token: i1.FetchClient }], target: i0.ɵɵFactoryTarget.Injectable });
Lwm2mBootstrapParametersService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mBootstrapParametersService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mBootstrapParametersService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.FetchClient }]; } });
