// @c8y/ngx-components/esm2020/repository/shared/advanced-software.service.mjs
import { Injectable } from '@angular/core';
import { AppStateService } from '@c8y/ngx-components';
import { FetchClient, Service } from '@c8y/client';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "@c8y/ngx-components";
export class AdvancedSoftwareService extends Service {
    constructor(client, appState) {
        super(client);
        this.appState = appState;
        this.baseUrl = 'service/advanced-software-mgmt';
        this.listUrl = 'software';
        this.propertyName = 'softwareList';
    }
    list(filter = {}) {
        return super.list(filter);
    }
    isASMAvailable() {
        if (this.isAvailable === undefined) {
            this.isAvailable = this.appState.isApplicationAvailable(AdvancedSoftwareService.applicationName);
        }
        return this.isAvailable;
    }
}
AdvancedSoftwareService.applicationName = 'advanced-software-mgmt';
AdvancedSoftwareService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AdvancedSoftwareService, deps: [{ token: i1.FetchClient }, { token: i2.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
AdvancedSoftwareService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AdvancedSoftwareService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AdvancedSoftwareService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.FetchClient }, { type: i2.AppStateService }]; } });
