// @c8y/ngx-components/esm2020/repository/shared/repository.model.mjs
export var RepositoryType;
(function (RepositoryType) {
    RepositoryType["FIRMWARE"] = "c8y_Firmware";
    RepositoryType["SOFTWARE"] = "c8y_Software";
    RepositoryType["CONFIGURATION"] = "c8y_ConfigurationDump";
})(RepositoryType || (RepositoryType = {}));
export const REPOSITORY_BINARY_TYPES = {
    [RepositoryType.SOFTWARE]: 'c8y_SoftwareBinary',
    [RepositoryType.FIRMWARE]: 'c8y_FirmwareBinary',
    [RepositoryType.CONFIGURATION]: 'c8y_ConfigurationDumpBinary'
};
export var DeviceConfigurationOperation;
(function (DeviceConfigurationOperation) {
    DeviceConfigurationOperation["UPLOAD_CONFIG"] = "c8y_UploadConfigFile";
    DeviceConfigurationOperation["DOWNLOAD_CONFIG"] = "c8y_DownloadConfigFile";
    DeviceConfigurationOperation["CONFIG"] = "c8y_Configuration";
    DeviceConfigurationOperation["SEND_CONFIG"] = "c8y_SendConfiguration";
})(DeviceConfigurationOperation || (DeviceConfigurationOperation = {}));
export const PRODUCT_EXPERIENCE = {
    SOFTWARE: {
        EVENTS: {
            REPOSITORY: 'softwareRepository',
            DEVICE_TAB: 'deviceSoftware'
        },
        COMPONENTS: {
            ADD_SOFTWARE_MODAL: 'add-software-modal',
            DEVICE_SOFTWARE_CHANGES: 'device-software-changes',
            DEVICE_SOFTWARE_LIST: 'device-software-list'
        },
        ACTIONS: {
            APPLY_SOFTWARE_CHANGES: 'applySoftwareChanges',
            CLEAR_SOFTWARE_CHANGES: 'clearSoftwareChanges',
            OPEN_INSTALL_SOFTWARE: 'openInstallSoftwareModal',
            OPEN_UPDATE_SOFTWARE: 'openUpdateSoftwareModal',
            DELETE_SOFTWARE: 'deleteSoftware'
        },
        RESULTS: {
            ADD_SOFTWARE: 'addSoftware',
            ADD_SOFTWARE_VERSION: 'addSoftwareVersion',
            EDIT_SOFTWARE: 'editSoftware'
        }
    },
    FIRMWARE: {
        EVENTS: {
            REPOSITORY: 'firmwareRepository',
            DEVICE_TAB: 'deviceFirmware'
        },
        COMPONENTS: {
            ADD_FIRMWARE_MODAL: 'add-firmware-modal',
            ADD_FIRMWAR_PATCH_MODAL: 'add-firmware-patch-modal',
            FIRMWARE_DEVICE_TAB: 'firmware-device-tab',
            DEVICE_FIRMWARE_LIST: 'device-firmware-list'
        },
        ACTIONS: {
            FILTER_FIRMWARE_LIST: 'filterFirmwareList',
            OPEN_INSTALL_FIRMWARE_DIALOG: 'openInstallFirmwareDialog',
            OPEN_REPLACE_FIRMWARE_DIALOG: 'openReplaceFirmwareDialog',
            OPEN_INSTALL_FIRMWARE_PATCH_DIALOG: 'openInstallFirmwarePatchDialog'
        },
        RESULTS: {
            ADD_FIRMWARE: 'addFirmware',
            ADD_FIRMWARE_VERSION: 'addFirmwareVersion',
            ADD_FIRMWARE_PATCH: 'addFirmwarePatch',
            EDIT_FIRMWARE: 'editFirmware',
            CREATE_FIRMWARE_UPDATE_OPERATION: 'createFirmwareUpdateOperation'
        }
    },
    SHARED: {
        COMPONENTS: {
            REPOSITORY_SELECT_MODAL: 'repository-select-modal',
            SELECT_CONFIGURATION_MODAL: 'select-configuration-modal'
        }
    }
};
