// @c8y/ngx-components/esm2020/repository/shared/shared-repository.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CoreModule, FormsModule, OperationRealtimeService } from '@c8y/ngx-components';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { RepositoryService } from './repository.service';
import { RepositorySelectModalComponent } from './select-modal/repository-select-modal.component';
import { FileDownloadComponent } from './file-download/file-download.component';
import { AdvancedSoftwareService } from './advanced-software.service';
import { SoftwareTypeComponent } from './software-type/software-type.component';
import * as i0 from "@angular/core";
export class SharedRepositoryModule {
}
SharedRepositoryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SharedRepositoryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SharedRepositoryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SharedRepositoryModule, declarations: [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent], imports: [CommonModule, CoreModule, FormsModule, PopoverModule], exports: [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent] });
SharedRepositoryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SharedRepositoryModule, providers: [AdvancedSoftwareService, RepositoryService, OperationRealtimeService], imports: [CommonModule, CoreModule, FormsModule, PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SharedRepositoryModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CoreModule, FormsModule, PopoverModule],
                    declarations: [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent],
                    entryComponents: [RepositorySelectModalComponent],
                    providers: [AdvancedSoftwareService, RepositoryService, OperationRealtimeService],
                    exports: [RepositorySelectModalComponent, FileDownloadComponent, SoftwareTypeComponent]
                }]
        }] });
