// @c8y/ngx-components/esm2020/auth-configuration/sso-configuration/sso-configuration.module.mjs
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { SsoConfigurationComponent } from './sso-configuration.component';
import { CoreModule, hookRoute } from '@c8y/ngx-components';
import { SsoConfigurationService } from './sso-configuration.service';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { DatePipe } from '@angular/common';
import { CustomTemplateComponent } from './templates/custom-template.component';
import { KeyCloakTemplateComponent } from './templates/key-cloak-template.component';
import { BasicConfigurationComponent } from './template-parts/basic-configuration.component';
import { RequestConfigurationComponent } from './template-parts/request-configuration.component';
import { UserIdConfigurationComponent } from './template-parts/user-id-configuration.component';
import { UserDataMappingComponent } from './template-parts/user-data-mapping.component';
import { AccessMappingComponent } from './template-parts/access-mapping/access-mapping.component';
import { DynamicAccessMappingComponent } from './template-parts/access-mapping/dynamic-access-mapping.component';
import { SignatureConfigurationComponent } from './template-parts/signature-configuration.component';
import { AadTemplateComponent } from './templates/aad-template.component';
import { LogoutConfigurationComponent } from './template-parts/logout-configuration.component';
import { AuthConfigurationGuard } from '../auth-configuration.guard';
import { CustomConfigurationMapper } from './templates/custom-configuration-mapper';
import { KeyCloakConfigurationMapper } from './templates/key-cloak-configuration-mapper';
import { AadConfigurationMapper } from './templates/aad-configuration-mapper';
import { SsoConfigurationGuard } from './sso-configuration.guard';
import { ExternalTokenConfigComponent } from './template-parts/external-token-config.component';
import { ChildPredicatesComponent } from './template-parts/access-mapping/child-predicates.component';
import { InventoryRolesMappingComponent } from './template-parts/access-mapping/inventory-roles-mapping.component';
import { InventoryRolesModalComponent } from './template-parts/access-mapping/inventory-roles-modal.component';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/collapse";
export class SsoConfigurationModule {
}
SsoConfigurationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SsoConfigurationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SsoConfigurationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SsoConfigurationModule, declarations: [SsoConfigurationComponent,
        CustomTemplateComponent,
        KeyCloakTemplateComponent,
        AadTemplateComponent,
        BasicConfigurationComponent,
        RequestConfigurationComponent,
        UserIdConfigurationComponent,
        UserDataMappingComponent,
        AccessMappingComponent,
        DynamicAccessMappingComponent,
        SignatureConfigurationComponent,
        LogoutConfigurationComponent,
        ChildPredicatesComponent,
        InventoryRolesMappingComponent,
        InventoryRolesModalComponent,
        ExternalTokenConfigComponent], imports: [CoreModule,
        TooltipModule,
        PopoverModule, i1.CollapseModule, BsDatepickerModule,
        AssetSelectorModule], exports: [SsoConfigurationComponent] });
SsoConfigurationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SsoConfigurationModule, providers: [
        SsoConfigurationGuard,
        DatePipe,
        SsoConfigurationService,
        AadConfigurationMapper,
        CustomConfigurationMapper,
        KeyCloakConfigurationMapper,
        hookRoute({
            path: 'auth-configuration/single_sign-on',
            component: SsoConfigurationComponent,
            canActivate: [AuthConfigurationGuard, SsoConfigurationGuard]
        })
    ], imports: [CoreModule,
        TooltipModule,
        PopoverModule,
        CollapseModule.forRoot(),
        BsDatepickerModule,
        AssetSelectorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SsoConfigurationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SsoConfigurationComponent,
                        CustomTemplateComponent,
                        KeyCloakTemplateComponent,
                        AadTemplateComponent,
                        BasicConfigurationComponent,
                        RequestConfigurationComponent,
                        UserIdConfigurationComponent,
                        UserDataMappingComponent,
                        AccessMappingComponent,
                        DynamicAccessMappingComponent,
                        SignatureConfigurationComponent,
                        LogoutConfigurationComponent,
                        ChildPredicatesComponent,
                        InventoryRolesMappingComponent,
                        InventoryRolesModalComponent,
                        ExternalTokenConfigComponent
                    ],
                    imports: [
                        CoreModule,
                        TooltipModule,
                        PopoverModule,
                        CollapseModule.forRoot(),
                        BsDatepickerModule,
                        AssetSelectorModule
                    ],
                    exports: [SsoConfigurationComponent],
                    providers: [
                        SsoConfigurationGuard,
                        DatePipe,
                        SsoConfigurationService,
                        AadConfigurationMapper,
                        CustomConfigurationMapper,
                        KeyCloakConfigurationMapper,
                        hookRoute({
                            path: 'auth-configuration/single_sign-on',
                            component: SsoConfigurationComponent,
                            canActivate: [AuthConfigurationGuard, SsoConfigurationGuard]
                        })
                    ]
                }]
        }] });
