// @c8y/ngx-components/esm2020/sub-assets/shared/sub-assets.model.mjs
export const PRODUCT_EXPERIENCE = {
    EVENT: 'subAssets',
    DELETE_ASSET: {
        COMPONENTS: { DELETE_ASSETS_MODAL: 'delete-assets-modal' },
        ACTIONS: { CASCADE_DELETE: 'cascadeDelete', DELETE_DEVICE_OWNER: 'deleteDeviceOwner' },
        RESULTS: { CANCELED: 'canceled', DELETED: 'deleted' }
    },
    ASSIGN_DEVICES: {
        COMPONENTS: { ASSIGN_DEVICES: 'assign-devices' },
        ACTIONS: { ASSIGN: 'assign', CANCEL: 'cancel', DISPLAY_CHILD_DEVICES: 'displayChildDevices' }
    },
    UNASSIGN_MODAL: {
        COMPONENTS: { UNASSIGN_MODAL: 'unassign-modal' },
        RESULTS: { ASSET_UNASSIGNED: 'asset-unassigned' },
        ACTIONS: { CANCEL: 'cancel' }
    },
    GROUP_INFO: {
        COMPONENTS: { GROUP_INFO: 'group-info' },
        ACTIONS: { EDIT: 'edit' },
        RESULTS: { EDIT_SAVED: 'edit-saved' },
        PROPERTIES: { NAME: 'name', DESCRIPTION: 'description' }
    },
    ADD_GROUP: {
        COMPONENTS: { ADD_GROUP: 'add-group' },
        ACTIONS: { ADD: 'add' },
        RESULTS: { ADD_SUCCESS: 'group-added' }
    }
};
