// @c8y/ngx-components/esm2020/ecosystem/microservices/add-microservice.component.mjs
import { Component } from '@angular/core';
import { gettext, ModalService, Status } from '@c8y/ngx-components';
import { ApplicationType, TenantService } from '@c8y/client';
import { EcosystemService, ERROR_TYPE } from '@c8y/ngx-components/ecosystem/shared';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components/ecosystem/shared";
import * as i2 from "@c8y/ngx-components";
import * as i3 from "@c8y/client";
export class AddMicroserviceComponent {
    constructor(ecosystemService, modal, tenantService) {
        this.ecosystemService = ecosystemService;
        this.modal = modal;
        this.tenantService = tenantService;
        this.headerText = gettext('Add microservice');
        this.successText = gettext('Microservice created');
        this.createMicroserviceApplicationHandler = (f) => this.createMicroserviceApplication(f);
        this.uploadMicroserviceHandler = (f, app) => this.uploadMicroservice(f, app);
    }
    async createMicroserviceApplication(file) {
        await this.ecosystemService.isValidAppType(file, ApplicationType.MICROSERVICE);
        return this.ecosystemService.createAppForArchive(file);
    }
    async uploadMicroservice(file, microservice) {
        const subscribeToCurrentTenant = await this.askIfActivationAfterUploadNeeded();
        await this.ecosystemService.uploadArchiveToApp(file, microservice);
        await this.subscribeMicroservice(microservice, subscribeToCurrentTenant);
    }
    async askIfActivationAfterUploadNeeded() {
        try {
            await this.modal.confirm(gettext('Subscribe to microservice'), gettext('You are about to subscribe to the microservice after upload. Do you want to subscribe to it?'), Status.INFO, { ok: gettext('Subscribe'), cancel: gettext("Don't subscribe") });
            return true;
        }
        catch (ex) {
            return false;
        }
    }
    async subscribeMicroservice(app, subscribeToCurrentTenant) {
        const tenant = (await this.tenantService.current()).data;
        const applications = tenant.applications.references;
        const isSubscribed = applications.some(({ application }) => application.id === app.id);
        if (!isSubscribed && subscribeToCurrentTenant) {
            try {
                return await this.tenantService.subscribeApplication(tenant, app);
            }
            catch (res) {
                if (res.status === 409) {
                    throw Error(ERROR_TYPE.ALREADY_SUBSCRIBED);
                }
            }
        }
        else if (isSubscribed && !subscribeToCurrentTenant) {
            return this.tenantService.unsubscribeApplication(tenant, app);
        }
    }
}
AddMicroserviceComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddMicroserviceComponent, deps: [{ token: i1.EcosystemService }, { token: i2.ModalService }, { token: i3.TenantService }], target: i0.ɵɵFactoryTarget.Component });
AddMicroserviceComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: AddMicroserviceComponent, selector: "c8y-add-microservice", ngImport: i0, template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'microchip'"
    [successText]="successText"
    [createApplicationHandler]="createMicroserviceApplicationHandler"
    [uploadApplicationHandler]="uploadMicroserviceHandler"
  ></c8y-add-application>`, isInline: true, dependencies: [{ kind: "component", type: i1.AddApplicationComponent, selector: "c8y-add-application", inputs: ["headerText", "headerIcon", "successText", "createApplicationHandler", "uploadApplicationHandler", "canGoBack"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddMicroserviceComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-add-microservice',
                    template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'microchip'"
    [successText]="successText"
    [createApplicationHandler]="createMicroserviceApplicationHandler"
    [uploadApplicationHandler]="uploadMicroserviceHandler"
  ></c8y-add-application>`
                }]
        }], ctorParameters: function () { return [{ type: i1.EcosystemService }, { type: i2.ModalService }, { type: i3.TenantService }]; } });
