// @c8y/ngx-components/esm2020/device-shell/shell/shell.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { OperationsTimelineModule } from '@c8y/ngx-components/operations/operations-timeline';
import { CommandTemplatesModule } from '../command-templates';
import { DeviceShellSharedModule } from '../shared';
import { DeviceShellComponent } from './shell.component';
import * as i0 from "@angular/core";
export class ShellModule {
}
ShellModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ShellModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ShellModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ShellModule, declarations: [DeviceShellComponent], imports: [CommonModule,
        CoreModule,
        OperationsTimelineModule,
        CommandTemplatesModule,
        DeviceShellSharedModule], exports: [DeviceShellComponent] });
ShellModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ShellModule, imports: [CommonModule,
        CoreModule,
        OperationsTimelineModule,
        CommandTemplatesModule,
        DeviceShellSharedModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ShellModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        CoreModule,
                        OperationsTimelineModule,
                        CommandTemplatesModule,
                        DeviceShellSharedModule
                    ],
                    declarations: [DeviceShellComponent],
                    exports: [DeviceShellComponent]
                }]
        }] });
