// @c8y/ngx-components/esm2020/operations/operations-timeline/operations-timeline.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { OperationsSharedModule } from '@c8y/ngx-components/operations/shared';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { OperationsTimelineComponent } from './operations-timeline.component';
import * as i0 from "@angular/core";
export class OperationsTimelineModule {
}
OperationsTimelineModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsTimelineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OperationsTimelineModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: OperationsTimelineModule, declarations: [OperationsTimelineComponent], imports: [CommonModule, CoreModule, CollapseModule, TooltipModule, OperationsSharedModule], exports: [OperationsTimelineComponent] });
OperationsTimelineModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsTimelineModule, imports: [CommonModule, CoreModule, CollapseModule, TooltipModule, OperationsSharedModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsTimelineModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CoreModule, CollapseModule, TooltipModule, OperationsSharedModule],
                    declarations: [OperationsTimelineComponent],
                    exports: [OperationsTimelineComponent]
                }]
        }] });
