// @c8y/ngx-components/esm2020/connectivity/impact/impact.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, FormsModule, hookNavigator, hookTab } from '@c8y/ngx-components';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { ImpactConnectivityService } from './impact-connectivity.service';
import { ImpactNodeFactory } from './impact-node.factory';
import { ImpactProviderSettingsComponent } from './impact-provider-settings.component';
import { ImpactTabFactory } from './impact-tab.factory';
import { ImpactGuard } from './impact.guard';
import { StatusDisplayComponent } from './status-display.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
const impactRoutes = [
    {
        path: 'connectivitySettings/impact_provider_settings',
        component: ImpactProviderSettingsComponent,
        canActivate: [ImpactGuard]
    }
];
export class ImpactModule {
}
ImpactModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ImpactModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ImpactModule, declarations: [ImpactProviderSettingsComponent, StatusDisplayComponent], imports: [CoreModule,
        CommonModule,
        FormsModule, i1.RouterModule, PopoverModule,
        ReactiveFormsModule], exports: [ImpactProviderSettingsComponent] });
ImpactModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactModule, providers: [
        ImpactConnectivityService,
        ImpactGuard,
        hookNavigator(ImpactNodeFactory),
        hookTab(ImpactTabFactory)
    ], imports: [CoreModule,
        CommonModule,
        FormsModule,
        RouterModule.forChild(impactRoutes),
        PopoverModule,
        ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ImpactProviderSettingsComponent, StatusDisplayComponent],
                    exports: [ImpactProviderSettingsComponent],
                    imports: [
                        CoreModule,
                        CommonModule,
                        FormsModule,
                        RouterModule.forChild(impactRoutes),
                        PopoverModule,
                        ReactiveFormsModule
                    ],
                    entryComponents: [ImpactProviderSettingsComponent],
                    providers: [
                        ImpactConnectivityService,
                        ImpactGuard,
                        hookNavigator(ImpactNodeFactory),
                        hookTab(ImpactTabFactory)
                    ]
                }]
        }] });
