// @c8y/ngx-components/esm2020/connectivity/impact/impact-node.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, toObservable } from '@c8y/ngx-components';
import { map } from 'rxjs/operators';
import { ImpactTabFactory } from './impact-tab.factory';
import * as i0 from "@angular/core";
import * as i1 from "./impact-tab.factory";
export class ImpactNodeFactory {
    constructor(tabFactory) {
        this.tabFactory = tabFactory;
        this.connectivityNode = new NavigatorNode({
            parent: gettext('Settings'),
            label: gettext('Connectivity'),
            path: 'connectivitySettings',
            icon: 'plug',
            preventDuplicates: true
        });
    }
    get() {
        return toObservable(this.tabFactory.tab()).pipe(map(tab => {
            return !!tab ? [this.connectivityNode] : [];
        }));
    }
}
ImpactNodeFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactNodeFactory, deps: [{ token: i1.ImpactTabFactory }], target: i0.ɵɵFactoryTarget.Injectable });
ImpactNodeFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactNodeFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactNodeFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.ImpactTabFactory }]; } });
