// @c8y/ngx-components/esm2020/connectivity/impact/impact-tab.factory.mjs
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext, toObservable } from '@c8y/ngx-components';
import { map } from 'rxjs/operators';
import { ImpactGuard } from './impact.guard';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./impact.guard";
export class ImpactTabFactory {
    constructor(router, guard) {
        this.router = router;
        this.guard = guard;
        this.config = {
            label: gettext('IMPACT provider settings'),
            path: '/connectivitySettings/impact_provider_settings',
            icon: 'plug'
        };
    }
    get() {
        return this.router.url.match(/connectivitySettings/g) ? this.tab() : [];
    }
    tab() {
        return toObservable(this.guard.canActivate()).pipe(map(canActivate => (canActivate ? this.config : null)));
    }
}
ImpactTabFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactTabFactory, deps: [{ token: i1.Router }, { token: i2.ImpactGuard }], target: i0.ɵɵFactoryTarget.Injectable });
ImpactTabFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactTabFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactTabFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.ImpactGuard }]; } });
