// @c8y/ngx-components/esm2020/register-device/extensible/single/extensible-device-registration.service.mjs
import { Injectable } from '@angular/core';
import { ApplicationService, FetchClient } from '@c8y/client';
import { AppStateService, gettext } from '@c8y/ngx-components';
import { BaseExtensibleDeviceRegistrationService } from '../base-extensible-device-registration.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "@c8y/ngx-components";
export class ExtensibleDeviceRegistrationService extends BaseExtensibleDeviceRegistrationService {
    constructor(applicationService, stateService, client) {
        super(applicationService, stateService);
        this.applicationService = applicationService;
        this.stateService = stateService;
        this.client = client;
    }
    async getSchema(contextPath) {
        const response = await this.client.fetch(this.getMetadataUrl(contextPath), {
            params: { lang: this.stateService.state.lang }
        });
        const data = await response.json();
        if (!data || !data.c8y_DeviceRegistration) {
            throw new Error(gettext('The response of the microservice is incomplete. The device registration information is missing.'));
        }
        return data.c8y_DeviceRegistration;
    }
    getDeviceRegistrationProviders() {
        return super.getDeviceRegistrationProviders(false);
    }
    async registerDevice(contextPath, model) {
        const method = 'POST';
        const body = JSON.stringify(model);
        const headers = { 'content-type': 'application/json', accept: 'application/json' };
        const res = await this.client.fetch(this.getServiceBaseUrl(contextPath), {
            method,
            body,
            headers
        });
        const data = await res.json();
        return { res, data };
    }
}
ExtensibleDeviceRegistrationService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ExtensibleDeviceRegistrationService, deps: [{ token: i1.ApplicationService }, { token: i2.AppStateService }, { token: i1.FetchClient }], target: i0.ɵɵFactoryTarget.Injectable });
ExtensibleDeviceRegistrationService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ExtensibleDeviceRegistrationService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ExtensibleDeviceRegistrationService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.ApplicationService }, { type: i2.AppStateService }, { type: i1.FetchClient }]; } });
