// @c8y/ngx-components/esm2020/register-device/extensible/single/extensible-device-registration-stepper.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import { C8yStepper, GainsightService, gettext } from '@c8y/ngx-components';
import { FormGroup } from '@angular/forms';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { ExtensibleDeviceRegistrationService } from './extensible-device-registration.service';
import { RegisterDeviceService } from '../../register-device.service';
import { STEP_STATE } from '@angular/cdk/stepper';
import { PRODUCT_EXPERIENCE } from '../base-device-registration.model';
import * as i0 from "@angular/core";
import * as i1 from "./extensible-device-registration.service";
import * as i2 from "../../register-device.service";
import * as i3 from "ngx-bootstrap/modal";
import * as i4 from "@c8y/ngx-components";
import * as i5 from "@angular/common";
import * as i6 from "@angular/cdk/stepper";
import * as i7 from "@ngx-formly/core";
export class ExtensibleDeviceRegistrationStepperComponent {
    constructor(customProviderService, registerDeviceService, bsModalRef, gainsightService) {
        this.customProviderService = customProviderService;
        this.registerDeviceService = registerDeviceService;
        this.bsModalRef = bsModalRef;
        this.gainsightService = gainsightService;
        this.pendingStatus = false;
        this.forms = [];
        this.model = {};
        this.SERVER_ERROR = gettext('Cannot register your device.');
        this.SUCCESSFUL_REGISTRATION = gettext('Your device was successfully registered.');
        this.NO_CONNECTION = gettext('Unable to reach the microservice.');
    }
    ngOnInit() {
        this.steps.forEach(() => this.forms.push(new FormGroup({})));
    }
    async save() {
        this.pendingStatus = true;
        this.errorMessage = null;
        const { res, data } = await this.customProviderService
            .registerDevice(this.contextPath, this.model)
            .catch(() => {
            return {
                res: undefined,
                data: undefined
            };
        })
            .finally(() => {
            this.pendingStatus = false;
            this.stepper.next();
        });
        if (res && data) {
            if (res.status >= 400 && res.status < 500) {
                this.handleError(data.message || this.SERVER_ERROR);
            }
            else if (res.status >= 500) {
                this.handleError(this.SERVER_ERROR);
            }
            else {
                this.gainsightService.triggerEvent(PRODUCT_EXPERIENCE.EVENT, {
                    result: PRODUCT_EXPERIENCE.RESULT.SUCCESS,
                    component: PRODUCT_EXPERIENCE.COMPONENT.EXTENSIBLE_SINGLE
                });
                this.message = this.SUCCESSFUL_REGISTRATION;
            }
        }
        else {
            this.handleError(this.NO_CONNECTION);
        }
    }
    close() {
        this.bsModalRef.hide();
    }
    complete() {
        this.registerDeviceService.list();
        this.bsModalRef.hide();
    }
    goToFirstStep() {
        this.stepper.selectedIndex = 0;
    }
    isRegistrationStep() {
        return this.stepper.selectedIndex === this.steps.length - 1;
    }
    finalStepStatus() {
        if (this.stepper.selectedIndex === this.steps.length) {
            if (this.errorMessage) {
                return STEP_STATE.ERROR;
            }
            else {
                return STEP_STATE.DONE;
            }
        }
        else {
            return STEP_STATE.NUMBER;
        }
    }
    handleError(message) {
        this.gainsightService.triggerEvent(PRODUCT_EXPERIENCE.EVENT, {
            result: PRODUCT_EXPERIENCE.RESULT.FAILURE,
            component: PRODUCT_EXPERIENCE.COMPONENT.EXTENSIBLE_SINGLE
        });
        this.errorMessage = message;
    }
}
ExtensibleDeviceRegistrationStepperComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ExtensibleDeviceRegistrationStepperComponent, deps: [{ token: i1.ExtensibleDeviceRegistrationService }, { token: i2.RegisterDeviceService }, { token: i3.BsModalRef }, { token: i4.GainsightService }], target: i0.ɵɵFactoryTarget.Component });
ExtensibleDeviceRegistrationStepperComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ExtensibleDeviceRegistrationStepperComponent, selector: "c8y-extensible-device-registration-stepper", inputs: { contextPath: "contextPath", labels: "labels", steps: "steps" }, viewQueries: [{ propertyName: "stepper", first: true, predicate: C8yStepper, descendants: true, static: true }], ngImport: i0, template: "<c8y-stepper\n  [disableProgressButtons]=\"true\"\n  [disableDefaultIcons]=\"{ edit: true, done: false }\"\n  [customClasses]=\"['m-40', 'm-t-8', 'm-b-16']\"\n>\n  <cdk-step *ngFor=\"let step of steps; let i = index\" [stepControl]=\"forms[i]\" [label]=\"labels[i]\">\n    <c8y-form-group class=\"form-group p-24 p-b-0 p-t-16\">\n      <formly-form [form]=\"forms[i]\" [fields]=\"step\" [model]=\"model\"></formly-form>\n    </c8y-form-group>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block modal-footer separator-top bg-component\"\n      *ngIf=\"isRegistrationStep()\"\n      [showButtons]=\"{ cancel: true, back: true, next: true }\"\n      [labels]=\"{ next: 'Register' }\"\n      (onCancel)=\"close()\"\n      (onNext)=\"save()\"\n      [pending]=\"pendingStatus\"\n      [disabled]=\"forms[i].invalid\"\n    ></c8y-stepper-buttons>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block modal-footer separator-top bg-component\"\n      *ngIf=\"!isRegistrationStep()\"\n      [showButtons]=\"{ cancel: true, back: i !== 0, next: true }\"\n      [disabled]=\"forms[i].invalid\"\n      (onCancel)=\"close()\"\n    ></c8y-stepper-buttons>\n  </cdk-step>\n  <cdk-step [state]=\"finalStepStatus()\" label=\"{{ 'Summary' | translate }}\">\n    <c8y-operation-result\n      *ngIf=\"!errorMessage; else errorInfo\"\n      text=\"{{ message | translate }}\"\n      [size]=\"84\"\n      [vertical]=\"true\"\n      type=\"success\"\n      class=\"lead\"\n      data-cy=\"success\"\n    ></c8y-operation-result>\n    <ng-template #errorInfo>\n      <c8y-operation-result\n        text=\"{{ errorMessage | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"error\"\n        class=\"lead\"\n        data-cy=\"error\"\n      ></c8y-operation-result>\n    </ng-template>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block modal-footer separator-top bg-component\"\n      [showButtons]=\"{ cancel: errorMessage, back: errorMessage, next: !errorMessage }\"\n      [labels]=\"{ back: 'Fix issue', next: 'Close' }\"\n      (onCancel)=\"close()\"\n      (onNext)=\"complete()\"\n      (onBack)=\"goToFirstStep()\"\n    ></c8y-stepper-buttons>\n  </cdk-step>\n</c8y-stepper>\n", dependencies: [{ kind: "directive", type: i5.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i5.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i4.OperationResultComponent, selector: "c8y-operation-result", inputs: ["text", "vertical", "size", "type"] }, { kind: "component", type: i4.FormGroupComponent, selector: "c8y-form-group", inputs: ["hasError", "hasWarning", "hasSuccess", "novalidation", "status"] }, { kind: "component", type: i4.C8yStepper, selector: "c8y-stepper", inputs: ["disableDefaultIcons", "disableProgressButtons", "customClasses", "hideStepProgress", "useStepLabelsAsTitlesOnly"], outputs: ["onStepChange"] }, { kind: "component", type: i6.CdkStep, selector: "cdk-step", inputs: ["stepControl", "label", "errorMessage", "aria-label", "aria-labelledby", "state", "editable", "optional", "completed", "hasError"], outputs: ["interacted"], exportAs: ["cdkStep"] }, { kind: "component", type: i4.C8yStepperButtons, selector: "c8y-stepper-buttons", inputs: ["labels", "pending", "disabled", "showButtons"], outputs: ["onCancel", "onNext", "onBack", "onCustom"] }, { kind: "component", type: i7.FormlyForm, selector: "formly-form", inputs: ["form", "model", "fields", "options"], outputs: ["modelChange"] }, { kind: "pipe", type: i4.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ExtensibleDeviceRegistrationStepperComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-extensible-device-registration-stepper', template: "<c8y-stepper\n  [disableProgressButtons]=\"true\"\n  [disableDefaultIcons]=\"{ edit: true, done: false }\"\n  [customClasses]=\"['m-40', 'm-t-8', 'm-b-16']\"\n>\n  <cdk-step *ngFor=\"let step of steps; let i = index\" [stepControl]=\"forms[i]\" [label]=\"labels[i]\">\n    <c8y-form-group class=\"form-group p-24 p-b-0 p-t-16\">\n      <formly-form [form]=\"forms[i]\" [fields]=\"step\" [model]=\"model\"></formly-form>\n    </c8y-form-group>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block modal-footer separator-top bg-component\"\n      *ngIf=\"isRegistrationStep()\"\n      [showButtons]=\"{ cancel: true, back: true, next: true }\"\n      [labels]=\"{ next: 'Register' }\"\n      (onCancel)=\"close()\"\n      (onNext)=\"save()\"\n      [pending]=\"pendingStatus\"\n      [disabled]=\"forms[i].invalid\"\n    ></c8y-stepper-buttons>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block modal-footer separator-top bg-component\"\n      *ngIf=\"!isRegistrationStep()\"\n      [showButtons]=\"{ cancel: true, back: i !== 0, next: true }\"\n      [disabled]=\"forms[i].invalid\"\n      (onCancel)=\"close()\"\n    ></c8y-stepper-buttons>\n  </cdk-step>\n  <cdk-step [state]=\"finalStepStatus()\" label=\"{{ 'Summary' | translate }}\">\n    <c8y-operation-result\n      *ngIf=\"!errorMessage; else errorInfo\"\n      text=\"{{ message | translate }}\"\n      [size]=\"84\"\n      [vertical]=\"true\"\n      type=\"success\"\n      class=\"lead\"\n      data-cy=\"success\"\n    ></c8y-operation-result>\n    <ng-template #errorInfo>\n      <c8y-operation-result\n        text=\"{{ errorMessage | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"error\"\n        class=\"lead\"\n        data-cy=\"error\"\n      ></c8y-operation-result>\n    </ng-template>\n    <c8y-stepper-buttons\n      class=\"sticky-bottom d-block modal-footer separator-top bg-component\"\n      [showButtons]=\"{ cancel: errorMessage, back: errorMessage, next: !errorMessage }\"\n      [labels]=\"{ back: 'Fix issue', next: 'Close' }\"\n      (onCancel)=\"close()\"\n      (onNext)=\"complete()\"\n      (onBack)=\"goToFirstStep()\"\n    ></c8y-stepper-buttons>\n  </cdk-step>\n</c8y-stepper>\n" }]
        }], ctorParameters: function () { return [{ type: i1.ExtensibleDeviceRegistrationService }, { type: i2.RegisterDeviceService }, { type: i3.BsModalRef }, { type: i4.GainsightService }]; }, propDecorators: { stepper: [{
                type: ViewChild,
                args: [C8yStepper, { static: true }]
            }], contextPath: [{
                type: Input
            }], labels: [{
                type: Input
            }], steps: [{
                type: Input
            }] } });
