// @c8y/ngx-components/esm2020/connectivity/sigfox/sigfox-tab.factory.mjs
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { AbstractConnectivityTabFactory } from '../abstract-connectivity-tab.factory';
import { SigfoxGuard } from './sigfox.guard';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./sigfox.guard";
export class SigfoxTabFactory extends AbstractConnectivityTabFactory {
    constructor(router, guard) {
        super(router, guard);
        this.router = router;
        this.guard = guard;
        this.tabConfig = {
            label: gettext('Sigfox'),
            path: '/connectivitySettings/multiple_lns_connectors_sigfox',
            icon: 'wall-socket'
        };
    }
    config() {
        return this.tabConfig;
    }
}
SigfoxTabFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxTabFactory, deps: [{ token: i1.Router }, { token: i2.SigfoxGuard }], target: i0.ɵɵFactoryTarget.Injectable });
SigfoxTabFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxTabFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxTabFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.SigfoxGuard }]; } });
