// @c8y/ngx-components/esm2020/upgrade/ng1/downgraded.components.mjs
import { downgradeComponent } from '@angular/upgrade/static';
import { BootstrapComponent, DataGridComponent, UserTotpRevokeComponent, LoadingComponent, HighlightComponent, EmptyStateComponent, PasswordInputComponent } from '@c8y/ngx-components';
import { AppLogsAutoRefreshComponent } from '@c8y/ngx-components/app-logs';
import { RangeDisplayComponent } from '@c8y/ngx-components';
import { HelpComponent } from '@c8y/ngx-components';
import { DatapointSelectionListComponent } from '@c8y/ngx-components/datapoint-selector';
import { PaginationComponent } from 'ngx-bootstrap/pagination';
import { RolesAssetTreeComponent } from '@c8y/ngx-components/user-roles';
export const bootstrapComponent = downgradeComponent({ component: BootstrapComponent });
export const userTotpComponent = downgradeComponent({ component: UserTotpRevokeComponent });
export const appLogsAutoRefreshComponent = downgradeComponent({
    component: AppLogsAutoRefreshComponent
});
export const dataGridComponent = downgradeComponent({ component: DataGridComponent });
export const loadingComponent = downgradeComponent({ component: LoadingComponent });
export const rangeDisplayComponent = downgradeComponent({ component: RangeDisplayComponent });
export const helpComponent = downgradeComponent({ component: HelpComponent });
export const highlightComponent = downgradeComponent({
    component: HighlightComponent,
    inputs: ['pattern', 'text']
});
export const emptyStateComponent = downgradeComponent({ component: EmptyStateComponent });
export const datapointSelectionListComponent = downgradeComponent({
    component: DatapointSelectionListComponent
});
export const paginationComponent = downgradeComponent({ component: PaginationComponent });
export const rolesAssetTreeComponent = downgradeComponent({ component: RolesAssetTreeComponent });
export const passwordInputComponent = downgradeComponent({ component: PasswordInputComponent });
