// @c8y/ngx-components/esm2020/upgrade/ng1/serverMessages.service.mjs
import { Injectable, Inject } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { HOOK_PATTERN_MESSAGES } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
export class ServerMessagesService {
    constructor(translateService, patterns) {
        this.translateService = translateService;
        this.MESSAGE_PATTERNS = patterns;
    }
    translate(s) {
        return this.translateService.instant(s);
    }
}
ServerMessagesService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ServerMessagesService, deps: [{ token: i1.TranslateService }, { token: HOOK_PATTERN_MESSAGES }], target: i0.ɵɵFactoryTarget.Injectable });
ServerMessagesService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ServerMessagesService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ServerMessagesService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.TranslateService }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [HOOK_PATTERN_MESSAGES]
                }] }]; } });
