// @c8y/ngx-components/esm2020/upgrade/ng1/title.decorator.mjs
import * as angular from 'angular';
function c8yTitleDecorator($q, $delegate, $injector, $rootScope, $location, $templateCache, $compile, c8yHeaderService) {
    'ngInject';
    $delegate.changeTitle = changeTitle;
    $delegate.setTitleElement = setTitleElement;
    $rootScope.$on('$routeChangeStart', () => {
        changeTitle({});
    });
    $rootScope.$on('$routeChangeSuccess', event => {
        resolveTitle(event);
    });
    $rootScope.$on('$routeUpdate', event => {
        resolveTitle(event);
    });
    function resolveTitle(event) {
        const title = $delegate.get($location.path());
        if (title) {
            const titleData = $q.when(title.data ? $injector.invoke(title.data) : {});
            titleData.then(data => {
                const { templateUrl } = data;
                if (templateUrl) {
                    const template = $templateCache.get(templateUrl);
                    setTitleElement($compile(template)(event.targetScope));
                }
                else {
                    changeTitle(data);
                }
            });
        }
    }
    function changeTitle(newTitleSubtitleObjOrPromise, options = { skipTitleTranslation: false, skipSubtitleTranslation: false }) {
        $q.when(newTitleSubtitleObjOrPromise).then(newTitleSubtitleObj => changeTitleInAngular(translate(newTitleSubtitleObj, options)));
    }
    function translate({ title = '', subtitle = '' }, { skipTitleTranslation = false, skipSubtitleTranslation = false }) {
        const titleTemplate = skipTitleTranslation ? '{{ title }}' : '{{ title | translate }}';
        const subtitleTemplate = skipSubtitleTranslation
            ? '{{ subtitle }}'
            : '{{ subtitle | translate }}';
        const template = `
      <h1 class="text-truncate">${titleTemplate}
      <small>${subtitleTemplate}</small></h1>
    `;
        const isolatedScope = $rootScope.$new(true);
        isolatedScope.title = title;
        isolatedScope.subtitle = subtitle;
        return $compile(angular.element(template))(isolatedScope)[0];
    }
    function changeTitleInAngular(domElement) {
        c8yHeaderService.changeTitle(domElement);
    }
    function setTitleElement($element) {
        $element.show();
        changeTitleInAngular($element[0]);
    }
    return $delegate;
}
export { c8yTitleDecorator };
