// @c8y/ngx-components/esm2020/assets-navigator/assets-navigator.module.mjs
import { NgModule } from '@angular/core';
import { DeviceStatusComponent, DeviceStatusModule, hookNavigator, ModalModule, ModalService } from '@c8y/ngx-components';
import { ASSET_NAVIGATOR_CONFIG } from './asset-node-config.model';
import { AssetNodeFactory } from './asset-node.factory';
import { AssetNodeService } from './asset-node.service';
import { AssetSelectorModule } from './asset-selector/asset-selector.module';
import * as i0 from "@angular/core";
export class AssetsNavigatorModule {
    static config(config = {}) {
        return {
            ngModule: AssetsNavigatorModule,
            providers: [
                {
                    provide: ASSET_NAVIGATOR_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
AssetsNavigatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AssetsNavigatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AssetsNavigatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AssetsNavigatorModule, imports: [ModalModule, AssetSelectorModule, DeviceStatusModule] });
AssetsNavigatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AssetsNavigatorModule, providers: [ModalService, AssetNodeService, hookNavigator(AssetNodeFactory)], imports: [ModalModule, AssetSelectorModule, DeviceStatusModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AssetsNavigatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [ModalModule, AssetSelectorModule, DeviceStatusModule],
                    providers: [ModalService, AssetNodeService, hookNavigator(AssetNodeFactory)],
                    entryComponents: [DeviceStatusComponent]
                }]
        }] });
