// @c8y/ngx-components/esm2020/assets-navigator/group.service.mjs
import { Injectable } from '@angular/core';
import { GroupService as MigratedDeviceGroupService } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
/**
 * @deprecated
 * Original service was moved to core/common.
 * This service is deprecated and serves as a proxy to maintain backward compatibility.
 */
export class DeviceGroupService {
    constructor(migratedDeviceGroupService) {
        this.migratedDeviceGroupService = migratedDeviceGroupService;
    }
    icon(mo, open = false) {
        return this.migratedDeviceGroupService.getIcon(mo, open);
    }
    isGroup(mo) {
        return this.migratedDeviceGroupService.isGroup(mo);
    }
    isDynamicGroup(mo) {
        return this.migratedDeviceGroupService.isDynamicGroup(mo);
    }
    isDataBroker(mo) {
        return this.migratedDeviceGroupService.isDataBroker(mo);
    }
    isDataBrokerActive(mo) {
        return this.migratedDeviceGroupService.isDataBrokerActive(mo);
    }
    isAsset(mo) {
        return this.migratedDeviceGroupService.isAsset(mo);
    }
    isAnyGroup(mo) {
        return this.migratedDeviceGroupService.isAnyGroup(mo);
    }
    isDevice(mo) {
        return this.migratedDeviceGroupService.isDevice(mo);
    }
}
DeviceGroupService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceGroupService, deps: [{ token: i1.GroupService }], target: i0.ɵɵFactoryTarget.Injectable });
DeviceGroupService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceGroupService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceGroupService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.GroupService }]; } });
