// @c8y/ngx-components/esm2020/context-dashboard/device-info-dashboard/device-info-dashboard.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule, hookRoute, ViewContext, gettext } from '@c8y/ngx-components';
import { ContextDashboardModule } from '../context-dashboard.module';
import { DeviceInfoDashboardComponent } from './device-info-dashboard.component';
import * as i0 from "@angular/core";
export class DeviceInfoDashboardModule {
}
DeviceInfoDashboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceInfoDashboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceInfoDashboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceInfoDashboardModule, declarations: [DeviceInfoDashboardComponent], imports: [CommonModule, CoreModule, ContextDashboardModule] });
DeviceInfoDashboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceInfoDashboardModule, providers: [
        hookRoute({
            context: ViewContext.Device,
            path: 'device-info',
            component: DeviceInfoDashboardComponent,
            label: gettext('Info'),
            priority: 10000,
            icon: 'asterisk'
        })
    ], imports: [CommonModule, CoreModule, ContextDashboardModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceInfoDashboardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DeviceInfoDashboardComponent],
                    imports: [CommonModule, CoreModule, ContextDashboardModule],
                    exports: [],
                    entryComponents: [DeviceInfoDashboardComponent],
                    providers: [
                        hookRoute({
                            context: ViewContext.Device,
                            path: 'device-info',
                            component: DeviceInfoDashboardComponent,
                            label: gettext('Info'),
                            priority: 10000,
                            icon: 'asterisk'
                        })
                    ]
                }]
        }] });
