// @c8y/ngx-components/esm2020/protocol-lwm2m/factories/lwm2m-post-operations.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { Lwm2mMicroserviceSubscribedGuard } from '@c8y/ngx-components/protocol-lwm2m/guards';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components/protocol-lwm2m/guards";
export class LWM2MPostOpertaionsNavigationFactory {
    constructor(guard) {
        this.guard = guard;
        this.node = new NavigatorNode({
            label: gettext('LWM2M post-operations'),
            path: 'post-operations',
            icon: 'terminal',
            parent: {
                label: gettext('Device types'),
                icon: 'c8y-device-protocols'
            },
            priority: 0
        });
    }
    get() {
        return this.guard.canActivate() ? this.node : [];
    }
}
LWM2MPostOpertaionsNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LWM2MPostOpertaionsNavigationFactory, deps: [{ token: i1.Lwm2mMicroserviceSubscribedGuard }], target: i0.ɵɵFactoryTarget.Injectable });
LWM2MPostOpertaionsNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LWM2MPostOpertaionsNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LWM2MPostOpertaionsNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Lwm2mMicroserviceSubscribedGuard }]; } });
