// @c8y/ngx-components/esm2020/upgrade/auth-bridge.service.mjs
import { AppStateService, TenantUiService } from '@c8y/ngx-components';
import { BasicAuth, FetchClient } from '@c8y/client';
export class AuthBridgeService {
    constructor(injector, basicAuth, fetchClient, appState, tenantUiService) {
        this.injector = injector;
        this.basicAuth = basicAuth;
        this.fetchClient = fetchClient;
        this.appState = appState;
        this.tenantUiService = tenantUiService;
        this.hookAuth();
    }
    updateBasicAuth(credentials) {
        const { headers } = this.fetchClient.getFetchOptions({});
        if (headers.Authorization) {
            const token = headers.Authorization.match(/basic\s(.*)$/i)[1];
            if (token) {
                this.basicAuth.updateCredentials(credentials);
                this.fetchClient.setAuth(this.basicAuth);
            }
        }
    }
    hookAuth() {
        this.appState.currentUser.subscribe(u => {
            if (u) {
                this.injector.get('c8yAuth').headers = () => this.fetchClient.getFetchOptions({}).headers;
                const { headers } = this.fetchClient.getFetchOptions({});
                if (headers.Authorization) {
                    const token = headers.Authorization.match(/basic\s(.*)$/i)[1];
                    if (token) {
                        this.setToken(token, headers.tfatoken);
                    }
                }
                else {
                    this.setToken(undefined, headers.tfatoken, 'Oauth');
                }
                this.injector.get('$rootScope').$emit('authStateChange', { hasAuth: true });
            }
            else {
                this.injector.get('$rootScope').$emit('authStateChange', { hasAuth: false });
            }
        });
    }
    setToken(token, tfa, type = 'Basic') {
        const c8yAuth = this.injector.get('c8yAuth');
        if (type === 'Basic') {
            c8yAuth.onSetToken({ token, type });
            if (tfa) {
                c8yAuth.setTFAToken(tfa);
            }
        }
        else {
            c8yAuth.authReady();
        }
    }
    getPreferredLoginOption() {
        return this.tenantUiService.getPreferredLoginOption(this.appState.state.loginOptions);
    }
}
export function authBridgeServiceFactory(injector, basicAuth, fetchClient, appState, tenantUiService) {
    return new AuthBridgeService(injector, basicAuth, fetchClient, appState, tenantUiService);
}
export const authBridgeServiceProvider = {
    provide: AuthBridgeService,
    useFactory: authBridgeServiceFactory,
    deps: ['$injector', BasicAuth, FetchClient, AppStateService, TenantUiService]
};
