// @c8y/ngx-components/esm2020/widgets/implementations/image/image-widget.service.mjs
import { Injectable } from '@angular/core';
import { InventoryBinaryService } from '@c8y/client';
import { InventoryService } from '@c8y/client';
import { AlertService, FilesService, gettext } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { defaultObjectPositionValue } from './image-widget.model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "@c8y/ngx-components";
import * as i3 from "@ngx-translate/core";
export class ImageWidgetService {
    constructor(inventory, fileService, alert, translate, binary) {
        this.inventory = inventory;
        this.fileService = fileService;
        this.alert = alert;
        this.translate = translate;
        this.binary = binary;
    }
    async getImageDetails(imageBinaryId) {
        if (!imageBinaryId) {
            return null;
        }
        try {
            const { data: imageBinaryMo } = await this.inventory.detail(imageBinaryId);
            const file = await this.fileService.getFile(imageBinaryMo);
            const base64 = await this.fileService.toBase64(file);
            return {
                file,
                base64: base64
            };
        }
        catch (e) {
            const text = this.translate.instant(gettext('Unable to retrieve image with id: {{ imageBinaryId }}'), { imageBinaryId });
            this.alert.danger(text, e?.data);
        }
        return null;
    }
    async uploadFile(file, options = {}) {
        const hash = await this.fileService.getHashSumOfFile(file);
        const fileDetails = { c8y_SHA256: hash };
        const globalImage = !options?.dashboardMoId || options?.isDeviceTypeDashboard;
        const partialBinaryMo = !globalImage
            ? { ...fileDetails }
            : { ...fileDetails, c8y_Global: {} };
        const { data: mo } = await this.binary.create(file, partialBinaryMo);
        if (options?.dashboardMoId) {
            await this.inventory.childAdditionsAdd(mo, options?.dashboardMoId);
        }
        return mo.id;
    }
    getStyling(config) {
        if (!config.styling?.objectFit || config.styling?.objectFit === 'full-width') {
            return null;
        }
        return {
            'object-fit': config.styling.objectFit,
            'object-position': `${config.styling?.objectPositionX || defaultObjectPositionValue} ${config.styling?.objectPositionY || defaultObjectPositionValue}`
        };
    }
}
ImageWidgetService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImageWidgetService, deps: [{ token: i1.InventoryService }, { token: i2.FilesService }, { token: i2.AlertService }, { token: i3.TranslateService }, { token: i1.InventoryBinaryService }], target: i0.ɵɵFactoryTarget.Injectable });
ImageWidgetService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImageWidgetService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImageWidgetService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.InventoryService }, { type: i2.FilesService }, { type: i2.AlertService }, { type: i3.TranslateService }, { type: i1.InventoryBinaryService }]; } });
