// @c8y/ngx-components/esm2020/core/core.module.mjs
import { NgModule } from '@angular/core';
import { ActionBarModule } from './action-bar/action-bar.module';
import { ActionModule } from './action/action.module';
import { AlertModule } from './alert/alert.module';
import { ApplicationModule } from './application/application.module';
import { AssetTypesModule } from './asset-types/asset-types.module';
import { AuditLogModule } from './audit-log/audit-log.module';
import { AuthenticationModule } from './authentication/authentication.module';
import { BootstrapModule } from './bootstrap/bootstrap.module';
import { BreadcrumbModule } from './breadcrumb/breadcrumb.module';
import { ClipboardModule } from './clipboard/clipboard.module';
import { CommonModule } from './common/common.module';
import { DashboardModule } from './dashboard/dashboard.module';
import { DataGridModule } from './data-grid/data-grid.module';
import { DateTimePickerModule } from './date-time-picker/date-time-picker.module';
import { TimePickerModule } from './time-picker/time-picker.module';
import { DeviceStatusModule } from './device-status/device-status.module';
import { DocsModule } from './docs/docs.module';
import { DrawerModule } from './drawer/drawer.module';
import { DropAreaModule } from './drop-area/drop-area.module';
import { DynamicComponentModule } from './dynamic-component/dynamic-component.module';
import { DynamicFormsModule } from './dynamic-forms/dynamic-forms.module';
import { FilePickerModule } from './file-picker/file-picker.module';
import { FilePickerFormControlModule } from './file-picker-form-control/file-picker-form-control.module';
import { DatePickerModule } from './date-picker/date-picker.module';
import { FormsModule as c8yFormsModule } from './forms/forms.module';
import { HeaderModule } from './header/header.module';
import { HelpModule } from './help/help.module';
import { ListDisplaySwitchModule } from './list-display-switch/list-display-switch.module';
import { ListGroupModule } from './list-group/list-group.module';
import { LoginModule } from './login/login.module';
import { ModalModule } from './modal/modal.module';
import { NavigatorBottomModule } from './navigator/navigator-bottom/navigator-bottom.module';
import { NavigatorTopModule } from './navigator/navigator-top/navigator-top.module';
import { NavigatorModule } from './navigator/navigator.module';
import { PluginsModule } from './plugins/plugins.module';
import { ProductExperienceModule } from './product-experience/product-experience.module';
import { PropertiesListModule } from './properties-list/properties-list.module';
import { RangeDisplayModule } from './range-display/range-display.module';
import { RealtimeModule } from './realtime/realtime.module';
import { SearchModule } from './search/search.module';
import { SelectModalModule } from './select-modal/select-modal.module';
import { SelectModule } from './select/select.module';
import { StepperModule as c8yStepperModule } from './stepper/stepper.module';
import { TabsModule } from './tabs/tabs.module';
import { UiSettingsModule } from './ui-settings/ui-settings.module';
import { UserModule } from './user/user.module';
import { VersionModule } from './version/version.module';
import { WizardModule } from './wizard/wizard.module';
import * as i0 from "@angular/core";
export class CoreModule {
    static forRoot() {
        return {
            ngModule: CoreModule,
            providers: [...CommonModule.providers(), ...LoginModule.providers()]
        };
    }
}
CoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: CoreModule, exports: [ApplicationModule,
        ActionBarModule,
        ActionModule,
        AlertModule,
        AuditLogModule,
        BootstrapModule,
        BreadcrumbModule,
        CommonModule,
        DataGridModule,
        DropAreaModule,
        HeaderModule,
        LoginModule,
        ModalModule,
        NavigatorModule,
        AuthenticationModule,
        SearchModule,
        SelectModule,
        TabsModule,
        UserModule,
        c8yFormsModule,
        DeviceStatusModule,
        DocsModule,
        c8yStepperModule,
        DynamicComponentModule,
        DashboardModule,
        ListGroupModule,
        SelectModalModule,
        FilePickerModule,
        FilePickerFormControlModule,
        DatePickerModule,
        ProductExperienceModule,
        HelpModule,
        RealtimeModule,
        RangeDisplayModule,
        ListDisplaySwitchModule,
        WizardModule,
        PluginsModule,
        DynamicFormsModule,
        PropertiesListModule,
        AssetTypesModule,
        DateTimePickerModule,
        TimePickerModule,
        ClipboardModule,
        VersionModule,
        DrawerModule,
        NavigatorTopModule,
        NavigatorBottomModule,
        UiSettingsModule] });
CoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CoreModule, imports: [ApplicationModule,
        ActionBarModule,
        ActionModule,
        AlertModule,
        AuditLogModule,
        BootstrapModule,
        BreadcrumbModule,
        CommonModule,
        DataGridModule,
        DropAreaModule,
        HeaderModule,
        LoginModule,
        ModalModule,
        NavigatorModule,
        AuthenticationModule,
        SearchModule,
        SelectModule,
        TabsModule,
        UserModule,
        c8yFormsModule,
        DeviceStatusModule,
        DocsModule,
        c8yStepperModule,
        DynamicComponentModule,
        DashboardModule,
        ListGroupModule,
        SelectModalModule,
        FilePickerModule,
        FilePickerFormControlModule,
        DatePickerModule,
        ProductExperienceModule,
        HelpModule,
        RealtimeModule,
        RangeDisplayModule,
        ListDisplaySwitchModule,
        WizardModule,
        PluginsModule,
        DynamicFormsModule,
        PropertiesListModule,
        AssetTypesModule,
        DateTimePickerModule,
        TimePickerModule,
        ClipboardModule,
        VersionModule,
        DrawerModule,
        NavigatorTopModule,
        NavigatorBottomModule,
        UiSettingsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CoreModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [
                        ApplicationModule,
                        ActionBarModule,
                        ActionModule,
                        AlertModule,
                        AuditLogModule,
                        BootstrapModule,
                        BreadcrumbModule,
                        CommonModule,
                        DataGridModule,
                        DropAreaModule,
                        HeaderModule,
                        LoginModule,
                        ModalModule,
                        NavigatorModule,
                        AuthenticationModule,
                        SearchModule,
                        SelectModule,
                        TabsModule,
                        UserModule,
                        c8yFormsModule,
                        DeviceStatusModule,
                        DocsModule,
                        c8yStepperModule,
                        DynamicComponentModule,
                        DashboardModule,
                        ListGroupModule,
                        SelectModalModule,
                        FilePickerModule,
                        FilePickerFormControlModule,
                        DatePickerModule,
                        ProductExperienceModule,
                        HelpModule,
                        RealtimeModule,
                        RangeDisplayModule,
                        ListDisplaySwitchModule,
                        WizardModule,
                        PluginsModule,
                        DynamicFormsModule,
                        PropertiesListModule,
                        AssetTypesModule,
                        DateTimePickerModule,
                        TimePickerModule,
                        ClipboardModule,
                        VersionModule,
                        DrawerModule,
                        NavigatorTopModule,
                        NavigatorBottomModule,
                        UiSettingsModule
                    ]
                }]
        }] });
