// @c8y/ngx-components/esm2020/operations/stepper-frames/preview-block.directive.mjs
import { Directive, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class PreviewBlockDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.show = true;
    }
}
PreviewBlockDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PreviewBlockDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
PreviewBlockDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: PreviewBlockDirective, selector: "[previewBlock]", inputs: { label: ["previewBlock", "label"], show: ["previewBlockIf", "show"] }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PreviewBlockDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[previewBlock]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; }, propDecorators: { label: [{
                type: Input,
                args: ['previewBlock']
            }], show: [{
                type: Input,
                args: ['previewBlockIf']
            }] } });
