// @c8y/ngx-components/esm2020/operations/stepper-frames/stepper-frames.modules.mjs
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { PreviewBlockDirective } from './preview-block.directive';
import { PreviewSelectionFrameComponent } from './preview-selection-frame.component';
import { SelectStepFrameComponent } from './select-step-frame.component';
import * as i0 from "@angular/core";
export class StepperFramesModule {
}
StepperFramesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperFramesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperFramesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StepperFramesModule, declarations: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective], imports: [CoreModule], exports: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective] });
StepperFramesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperFramesModule, imports: [CoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperFramesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule],
                    declarations: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective],
                    exports: [PreviewSelectionFrameComponent, SelectStepFrameComponent, PreviewBlockDirective]
                }]
        }] });
