// @c8y/ngx-components/fesm2015/c8y-ngx-components-widgets-definitions-help-and-service.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';

// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
const helpAndServiceWidgetDefinition = {
    id: defaultWidgetIds.HELP_AND_SERVICE,
    label: gettext('Help and service'),
    description: gettext(`Displays links to help and service resources.`),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/help-and-service-widget').then(m => m.HelpAndServiceViewComponent),
    previewImage: 'c8y-style-assets/help-and-service-widget-pr.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 4,
                _height: 4
            },
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};

/**
 * Generated bundle index. Do not edit.
 */

export { helpAndServiceWidgetDefinition };
