// @c8y/ngx-components/fesm2015/c8y-ngx-components-widgets-definitions-image.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';

// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
const imageWidgetDefinition = {
    id: defaultWidgetIds.IMAGE,
    label: gettext('Image'),
    description: gettext(`Display a single image selected from the user's file system.`),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/image').then(m => m.ImageWidgetViewComponent),
    loadConfigComponent: () => import('@c8y/ngx-components/widgets/implementations/image').then(m => m.ImageWidgetConfigComponent),
    previewImage: 'c8y-style-assets/image-widget.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 4,
                _height: 4
            },
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};

/**
 * Generated bundle index. Do not edit.
 */

export { imageWidgetDefinition };
