// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-product-experience.mjs
const DEVICE_CONTROL_EVENT_PREFIX = 'deviceControl';
const DEVICE_CONTROL_EVENT_SINGLE_PREFIX = `${DEVICE_CONTROL_EVENT_PREFIX}:single`;
/* Event names */
const SINGLE_OPERATION_EVENT_OVERVIEW = `${DEVICE_CONTROL_EVENT_SINGLE_PREFIX}:overview`;
const SINGLE_OPERATION_EVENT_DEVICE_TAB = `${DEVICE_CONTROL_EVENT_SINGLE_PREFIX}:deviceTab`;
const BULK_OPERATION_EVENT = `${DEVICE_CONTROL_EVENT_PREFIX}:bulk`;
/* Actions */
const ACTIONS = {
    FILTER: 'filter',
    OPEN_OPERATION_DETAILS: 'openOperationDetails',
    RETRY: 'retry'
};
const ACTIONS_SINGLE = {
    SCHEDULE_AS_BULK_OPERATION: 'scheduleAsBulkOperation'
};
const ACTIONS_BULK = {
    OPEN_ADD_BULK_OPERATION_DIALOG: 'openAddBulkOperationDialog',
    CANCEL_ADD_BULK_OPERATION_DIALOG: 'cancelAddBulkOperationDialog',
    SELECT_BULK_OPERTION_TYPE: 'selectBulkOperationType'
};
/* Components */
const COMPONENTS = {
    OPERATION_DETAILS: 'operation-details',
    SINGLE_OPERATION_ITEM: 'single-operation-item'
};
/* Add bulk operation dialog steps */
const BULK_OPERATION_DIALOG_STEPS = {
    SELECT_OPERATION_TYPE: 'selectOperationType'
};

/**
 * Generated bundle index. Do not edit.
 */

export { ACTIONS, ACTIONS_BULK, ACTIONS_SINGLE, BULK_OPERATION_DIALOG_STEPS, BULK_OPERATION_EVENT, COMPONENTS, SINGLE_OPERATION_EVENT_DEVICE_TAB, SINGLE_OPERATION_EVENT_OVERVIEW };
