// @c8y/ngx-components/fesm2015/c8y-ngx-components-widgets-definitions-cockpit-legacy-welcome.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';

// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
const legacyCockpitWelcomeWidgetDefinition = {
    id: defaultWidgetIds.LEGACY_COCKPIT_WELCOME,
    label: gettext('Welcome to Cockpit'),
    description: gettext('Displays a welcome message for Cockpit'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/cockpit-legacy-welcome').then(m => m.CockpitLegacyWelcomeWidgetComponent),
    data: {
        settings: {
            noNewWidgets: true,
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};

/**
 * Generated bundle index. Do not edit.
 */

export { legacyCockpitWelcomeWidgetDefinition };
