// @c8y/ngx-components/esm2020/connectivity/abstract-connectivity-node.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, toObservable } from '@c8y/ngx-components';
import { map } from 'rxjs/operators';
import { AbstractConnectivityTabFactory } from './abstract-connectivity-tab.factory';
import * as i0 from "@angular/core";
import * as i1 from "./abstract-connectivity-tab.factory";
export class AbstractConnectivityNodeFactory {
    constructor(tabFactory) {
        this.tabFactory = tabFactory;
        this.connectivityNode = new NavigatorNode({
            parent: gettext('Settings'),
            label: gettext('Connectivity'),
            path: 'connectivitySettings',
            icon: 'plug',
            preventDuplicates: true
        });
    }
    get() {
        return toObservable(this.tabFactory.tab()).pipe(map(tab => {
            return !!tab ? [this.connectivityNode] : [];
        }));
    }
}
AbstractConnectivityNodeFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AbstractConnectivityNodeFactory, deps: [{ token: i1.AbstractConnectivityTabFactory }], target: i0.ɵɵFactoryTarget.Injectable });
AbstractConnectivityNodeFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AbstractConnectivityNodeFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AbstractConnectivityNodeFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.AbstractConnectivityTabFactory }]; } });
