// @c8y/ngx-components/esm2020/connectivity/connectivity.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ConditionalTabsOutletComponent, CoreModule } from '@c8y/ngx-components';
import { ActilityModule } from './actility/actility.module';
import { ImpactModule } from './impact/impact.module';
import { LoriotModule } from './loriot/loriot.module';
import { SigfoxModule } from './sigfox/sigfox.module';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
const connectivityRoutes = [
    {
        path: 'connectivitySettings',
        component: ConditionalTabsOutletComponent
    }
];
export class ConnectivityModule {
}
ConnectivityModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectivityModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ConnectivityModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ConnectivityModule, imports: [CoreModule,
        ActilityModule,
        ImpactModule,
        SigfoxModule,
        LoriotModule, i1.RouterModule] });
ConnectivityModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectivityModule, imports: [CoreModule,
        ActilityModule,
        ImpactModule,
        SigfoxModule,
        LoriotModule,
        RouterModule.forChild(connectivityRoutes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectivityModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        ActilityModule,
                        ImpactModule,
                        SigfoxModule,
                        LoriotModule,
                        RouterModule.forChild(connectivityRoutes)
                    ]
                }]
        }] });
