// @c8y/ngx-components/esm2020/core/common/virtual-scroll/virtual-scroll-window.directive.mjs
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { VIRTUAL_SCROLL_STRATEGY } from '@angular/cdk/scrolling';
import { Directive, forwardRef, Input } from '@angular/core';
import { VirtualScrollWindowStrategy } from './virtual-scroll-window-strategy';
import * as i0 from "@angular/core";
/**
 * Provider factory for `VirtualScrollWindowStrategy` that simply extracts the already created
 * `VirtualScrollWindowStrategy` from the given directive.
 * @param windowDir The instance of `VirtualScrollWindowDirective` to extract the
 *     `VirtualScrollWindowStrategy` from.
 */
export function _virtualScrollWindowStrategyFactory(windowDir) {
    return windowDir._scrollStrategy;
}
// This is an adapted version of the original CdkFixedSizeVirtualScroll
// https://github.com/angular/components/blob/master/src/cdk/scrolling/fixed-size-virtual-scroll.ts
export class VirtualScrollWindowDirective {
    constructor() {
        this._itemSizePx = 20;
        this._minBufferPx = 100;
        this._maxBufferPx = 200;
        this._offsetSizePx = 0;
        /** The scroll strategy used by this directive. */
        this._scrollStrategy = new VirtualScrollWindowStrategy(this.itemSizePx, this.offsetSizePx, this.minBufferPx, this.maxBufferPx);
    }
    /** The size of the items in the list (in pixels). */
    get itemSizePx() {
        return this._itemSizePx;
    }
    set itemSizePx(value) {
        this._itemSizePx = coerceNumberProperty(value);
    }
    /**
     * The minimum amount of buffer rendered beyond the viewport (in pixels).
     * If the amount of buffer dips below this number, more items will be rendered. Defaults to 100px.
     */
    get minBufferPx() {
        return this._minBufferPx;
    }
    set minBufferPx(value) {
        this._minBufferPx = coerceNumberProperty(value);
    }
    /**
     * The number of pixels worth of buffer to render for when rendering new items. Defaults to 200px.
     */
    get maxBufferPx() {
        return this._maxBufferPx;
    }
    set maxBufferPx(value) {
        this._maxBufferPx = coerceNumberProperty(value);
    }
    /** The size of the offset (in pixels). */
    get offsetSizePx() {
        return this._offsetSizePx;
    }
    set offsetSizePx(value) {
        this._offsetSizePx = coerceNumberProperty(value);
    }
    ngOnChanges() {
        this._scrollStrategy.updateItemAndBufferSize(this.itemSizePx, this.offsetSizePx, this.minBufferPx, this.maxBufferPx);
    }
}
VirtualScrollWindowDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VirtualScrollWindowDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
VirtualScrollWindowDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: VirtualScrollWindowDirective, selector: "cdk-virtual-scroll-viewport[windowVirtualScrollStrategy]", inputs: { itemSizePx: "itemSizePx", minBufferPx: "minBufferPx", maxBufferPx: "maxBufferPx", offsetSizePx: "offsetSizePx" }, providers: [
        {
            provide: VIRTUAL_SCROLL_STRATEGY,
            useFactory: _virtualScrollWindowStrategyFactory,
            deps: [forwardRef(() => VirtualScrollWindowDirective)]
        }
    ], usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VirtualScrollWindowDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'cdk-virtual-scroll-viewport[windowVirtualScrollStrategy]',
                    providers: [
                        {
                            provide: VIRTUAL_SCROLL_STRATEGY,
                            useFactory: _virtualScrollWindowStrategyFactory,
                            deps: [forwardRef(() => VirtualScrollWindowDirective)]
                        }
                    ]
                }]
        }], propDecorators: { itemSizePx: [{
                type: Input
            }], minBufferPx: [{
                type: Input
            }], maxBufferPx: [{
                type: Input
            }], offsetSizePx: [{
                type: Input
            }] } });
