// @c8y/ngx-components/esm2020/core/user/user-details-drawer/user-details-drawer.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../../common';
import { hookDrawer } from '../../drawer/drawer.service';
import { UserDetailsDrawerComponent } from './user-details-drawer.component';
import { UserRolesListPipe } from './user-roles-list.pipe';
import * as i0 from "@angular/core";
export class UserDetailsDrawerModule {
}
UserDetailsDrawerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserDetailsDrawerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UserDetailsDrawerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: UserDetailsDrawerModule, declarations: [UserDetailsDrawerComponent, UserRolesListPipe], imports: [CommonModule] });
UserDetailsDrawerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserDetailsDrawerModule, providers: [
        hookDrawer({
            component: UserDetailsDrawerComponent,
            position: 'right',
            priority: 120,
            id: 'userDetailsDrawer'
        })
    ], imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserDetailsDrawerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [UserDetailsDrawerComponent, UserRolesListPipe],
                    providers: [
                        hookDrawer({
                            component: UserDetailsDrawerComponent,
                            position: 'right',
                            priority: 120,
                            id: 'userDetailsDrawer'
                        })
                    ]
                }]
        }] });
