// @c8y/ngx-components/esm2020/core/user/user-details-drawer/user-roles-list.pipe.mjs
import { Pipe } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
export class UserRolesListPipe {
    constructor(translateService) {
        this.translateService = translateService;
    }
    transform(user, translate = true) {
        if (!user) {
            return '';
        }
        const groups = user.groups?.references || [];
        let groupNames = groups.map(ref => ref.group?.name).filter(groupName => !!groupName);
        if (!groupNames.length) {
            return '';
        }
        if (translate) {
            groupNames = groupNames.map(name => this.translateService.instant(name));
        }
        return groupNames.join(', ');
    }
}
UserRolesListPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserRolesListPipe, deps: [{ token: i1.TranslateService }], target: i0.ɵɵFactoryTarget.Pipe });
UserRolesListPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: UserRolesListPipe, name: "userRolesList" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserRolesListPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'userRolesList'
                }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }]; } });
