// @c8y/ngx-components/esm2020/core/date-picker/date-picker.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { DatePickerComponent } from './date-picker.component';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/index';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/dropdown";
import * as i2 from "ngx-bootstrap/datepicker";
/**
 * The angular module definition for a date picker - it allows you to pick either one date or a range of two.
 * @exports DatePickerComponent
 */
export class DatePickerModule {
}
DatePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DatePickerModule, declarations: [DatePickerComponent], imports: [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule, i1.BsDropdownModule, i2.BsDatepickerModule, C8yCommonModule], exports: [DatePickerComponent] });
DatePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatePickerModule, imports: [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        BsDropdownModule.forRoot(),
        BsDatepickerModule.forRoot(),
        C8yCommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        C8yFormsModule,
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BsDropdownModule.forRoot(),
                        BsDatepickerModule.forRoot(),
                        C8yCommonModule
                    ],
                    exports: [DatePickerComponent],
                    declarations: [DatePickerComponent]
                }]
        }] });
