// @c8y/ngx-components/esm2020/core/docs/help-and-support-outlet/help-and-support-outlet.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../../common/common.module';
import { HelpAndSupportOutletComponent } from './help-and-support-outlet.component';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { ProductExperienceModule } from '../../product-experience/product-experience.module';
import { hookDrawer } from '../../drawer/drawer.service';
import { SupportOutletModule } from '../support-outlet/support-outlet.module';
import * as i0 from "@angular/core";
export class HelpAndSupportOutletModule {
}
HelpAndSupportOutletModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndSupportOutletModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HelpAndSupportOutletModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: HelpAndSupportOutletModule, declarations: [HelpAndSupportOutletComponent], imports: [CommonModule, CollapseModule, ProductExperienceModule, SupportOutletModule] });
HelpAndSupportOutletModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndSupportOutletModule, providers: [
        hookDrawer({
            component: HelpAndSupportOutletComponent,
            position: 'right',
            priority: 40,
            id: 'helpAndSupport'
        })
    ], imports: [CommonModule, CollapseModule, ProductExperienceModule, SupportOutletModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndSupportOutletModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CollapseModule, ProductExperienceModule, SupportOutletModule],
                    declarations: [HelpAndSupportOutletComponent],
                    providers: [
                        hookDrawer({
                            component: HelpAndSupportOutletComponent,
                            position: 'right',
                            priority: 40,
                            id: 'helpAndSupport'
                        })
                    ]
                }]
        }] });
