// @c8y/ngx-components/esm2020/register-device/extensible/base-extensible-device-registration.service.mjs
import { ApplicationType } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { flatMap } from 'lodash-es';
export class BaseExtensibleDeviceRegistrationService {
    constructor(applicationService, stateService) {
        this.applicationService = applicationService;
        this.stateService = stateService;
    }
    getMetadataUrl(contextPath) {
        return `${this.getServiceBaseUrl(contextPath)}/metadata`;
    }
    getServiceBaseUrl(contextPath) {
        if (!contextPath || contextPath.length === 0) {
            throw new Error(gettext('The URL of the microservice is misconfigured. A context path is missing.'));
        }
        return `/service/${contextPath}/deviceRegistration`;
    }
    async getDeviceRegistrationProviders(bulk) {
        let providers = [];
        const microservices = await this.listMicroservicesByUser();
        if (microservices && microservices.length > 0) {
            providers = this.getProvidersFromManifestExtensions(microservices, bulk).concat(this.getProvidersFromExtensions(microservices, bulk));
        }
        return providers;
    }
    async listMicroservicesByUser() {
        const { data } = await this.applicationService.listByUser(this.stateService?.currentUser?.value?.userName, {
            dropOverwrittenApps: true,
            noPaging: true,
            pageSize: 100,
            withTotalPages: true
        });
        return data?.filter(({ type }) => type === ApplicationType.MICROSERVICE);
    }
    getProvidersFromManifestExtensions(applications, bulk) {
        return flatMap(applications, application => {
            return application?.manifest?.extensions?.map(extension => {
                return { ...extension, contextPath: application.contextPath };
            });
        }).filter(provider => this.appSupportingExtensibleDeviceRegistrationForm(provider, bulk));
    }
    getProvidersFromExtensions(applications, bulk) {
        return flatMap(applications, application => {
            return application.extensions?.map(extension => {
                return {
                    ...extension,
                    contextPath: application.contextPath
                };
            });
        }).filter(provider => this.appSupportingExtensibleDeviceRegistrationForm(provider, bulk));
    }
    appSupportingExtensibleDeviceRegistrationForm(extension, bulk) {
        if (bulk) {
            return extension?.type === 'extensibleBulkDeviceRegistration';
        }
        else {
            return extension?.type === 'extensibleDeviceRegistration';
        }
    }
}
