// @c8y/ngx-components/esm2020/sigfox-device-registration/sigfox-device-registration.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule } from '@c8y/ngx-components';
import { hookDeviceRegistration } from '@c8y/ngx-components/register-device';
import { SigfoxDeviceRegistrationButtonComponent } from './sigfox-device-registration-button.component';
import { SigfoxDeviceRegistrationComponent } from './sigfox-device-registration.component';
import { SigfoxDeviceRegistrationFactory } from './sigfox-device-registration.factory';
import { SigfoxProviderService } from './sigfox-provider.service';
import * as i0 from "@angular/core";
export class SigfoxDeviceRegistrationModule {
}
SigfoxDeviceRegistrationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SigfoxDeviceRegistrationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationModule, declarations: [SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent], imports: [CoreModule, CommonModule] });
SigfoxDeviceRegistrationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationModule, providers: [hookDeviceRegistration(SigfoxDeviceRegistrationFactory), SigfoxProviderService], imports: [CoreModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, CommonModule],
                    declarations: [SigfoxDeviceRegistrationButtonComponent, SigfoxDeviceRegistrationComponent],
                    providers: [hookDeviceRegistration(SigfoxDeviceRegistrationFactory), SigfoxProviderService]
                }]
        }] });
