// @c8y/ngx-components/esm2020/sigfox-device-registration/sigfox-device-registration.factory.mjs
import { Injectable } from '@angular/core';
import { TenantUiService } from '@c8y/ngx-components';
import { SigfoxDeviceRegistrationButtonComponent } from './sigfox-device-registration-button.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class SigfoxDeviceRegistrationFactory {
    constructor(tenantService) {
        this.tenantService = tenantService;
    }
    get() {
        const items = [];
        if (this.tenantService.isMicroserviceSubscribedInCurrentTenant('sigfox-agent')) {
            items.push({
                template: SigfoxDeviceRegistrationButtonComponent,
                priority: 98,
                category: 'single'
            });
        }
        return items;
    }
}
SigfoxDeviceRegistrationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationFactory, deps: [{ token: i1.TenantUiService }], target: i0.ɵɵFactoryTarget.Injectable });
SigfoxDeviceRegistrationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SigfoxDeviceRegistrationFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.TenantUiService }]; } });
