// @c8y/ngx-components/esm2020/repository/software/device-tab/software-repository-device-tab.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { DeviceSoftwareChangesComponent } from './device-software-changes.component';
import { DeviceSoftwareListComponent } from './device-software-list.component';
import { InstalledSoftwareComponent } from './installed-software.component';
import { SoftwareDeviceTabComponent } from './software-device-tab.component';
import { SoftwareDeviceTabGuard } from './software-device-tab.guard';
import * as i0 from "@angular/core";
export class SoftwareRepositoryDeviceTabModule {
    static forRoot() {
        return {
            ngModule: SoftwareRepositoryDeviceTabModule,
            providers: [
                SoftwareDeviceTabGuard,
                hookRoute({
                    context: ViewContext.Device,
                    path: 'software',
                    component: SoftwareDeviceTabComponent,
                    label: gettext('Software'),
                    icon: 'c8y-tools',
                    priority: 499,
                    canActivate: [SoftwareDeviceTabGuard]
                })
            ]
        };
    }
}
SoftwareRepositoryDeviceTabModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryDeviceTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SoftwareRepositoryDeviceTabModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryDeviceTabModule, declarations: [SoftwareDeviceTabComponent,
        InstalledSoftwareComponent,
        DeviceSoftwareListComponent,
        DeviceSoftwareChangesComponent], imports: [CommonModule,
        CoreModule,
        RouterModule,
        SharedRepositoryModule,
        OperationDetailsModule,
        TooltipModule] });
SoftwareRepositoryDeviceTabModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryDeviceTabModule, imports: [CommonModule,
        CoreModule,
        RouterModule,
        SharedRepositoryModule,
        OperationDetailsModule,
        TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryDeviceTabModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        CoreModule,
                        RouterModule,
                        SharedRepositoryModule,
                        OperationDetailsModule,
                        TooltipModule
                    ],
                    declarations: [
                        SoftwareDeviceTabComponent,
                        InstalledSoftwareComponent,
                        DeviceSoftwareListComponent,
                        DeviceSoftwareChangesComponent
                    ],
                    entryComponents: [
                        SoftwareDeviceTabComponent,
                        DeviceSoftwareListComponent,
                        DeviceSoftwareChangesComponent
                    ]
                }]
        }] });
