// @c8y/ngx-components/esm2020/device-grid/columns/name.cell-renderer.component.mjs
import { Component } from '@angular/core';
import { CellRendererContext } from '@c8y/ngx-components';
import { ColumnUtilService } from './column-util.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "./column-util.service";
export class NameCellRendererComponent {
    constructor(context, columnUtilService) {
        this.context = context;
        this.columnUtilService = columnUtilService;
    }
}
NameCellRendererComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NameCellRendererComponent, deps: [{ token: i1.CellRendererContext }, { token: i2.ColumnUtilService }], target: i0.ɵɵFactoryTarget.Component });
NameCellRendererComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: NameCellRendererComponent, selector: "ng-component", ngImport: i0, template: `
    <a
      [href]="columnUtilService.getHref(context.item)"
      title="{{ columnUtilService.getProperName(context.item) }}"
      class="interact"
    >
      {{ columnUtilService.getProperName(context.item) }}
    </a>
  `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NameCellRendererComponent, decorators: [{
            type: Component,
            args: [{
                    template: `
    <a
      [href]="columnUtilService.getHref(context.item)"
      title="{{ columnUtilService.getProperName(context.item) }}"
      class="interact"
    >
      {{ columnUtilService.getProperName(context.item) }}
    </a>
  `
                }]
        }], ctorParameters: function () { return [{ type: i1.CellRendererContext }, { type: i2.ColumnUtilService }]; } });
