// @c8y/ngx-components/esm2020/widgets/definitions/markdown/index.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';
// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
export const markdownWidgetDefinition = {
    id: defaultWidgetIds.MARKDOWN,
    label: gettext('Markdown widget'),
    description: gettext('Display content provided as Markdown. You can either upload a file, or provide a file path.'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/markdown').then(m => m.MarkdownWidgetViewComponent),
    loadConfigComponent: () => import('@c8y/ngx-components/widgets/implementations/markdown').then(m => m.MarkdownWidgetConfigComponent),
    previewImage: 'c8y-style-assets/markdown-widget.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 4,
                _height: 4
            },
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};
