// @c8y/client/lib/cjs/src/tenant-security-options/TenantSecurityOptionsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenantSecurityOptionsService = void 0;
const core_1 = require("../core");
/**
 * @description
 * This service allows to get tenant security option.
 */
class TenantSecurityOptionsService extends core_1.Service {
    constructor() {
        super(...arguments);
        this.baseUrl = 'tenant';
        this.listUrl = 'security-options';
        this.category = 'password';
        this.propertyName = 'options';
    }
    /**
     * Get a specific tenant's security option.
     *
     * @param {ITenantSecurityOption} Tenant's security option object with key value.
     *
     * @returns Returns promise object that is resolved with the ITenantSecurityOption wrapped by IResult.
     *
     * **Example**
     * ```typescript
     *
     *     const option: ITenantSecurityOption = {
     *        key: 'limit.validity'
     *      };
     *
     *    (async () => {
     *      const {data, res} = await tenantSecurityOptionsService.detail(option);
     *   })();
     * ```
     */
    async detail(tenantSecurityOption) {
        return super.detail(tenantSecurityOption);
    }
    getDetailUrl(tenantSecurityOption) {
        return `${this.listUrl}/${this.category}/${tenantSecurityOption.key}`;
    }
    onBeforeCreate(obj) {
        return obj;
    }
}
exports.TenantSecurityOptionsService = TenantSecurityOptionsService;
