// @c8y/ngx-components/esm2020/connectivity/sim/sim.guard.mjs
import { Injectable } from '@angular/core';
import { AppStateService } from '@c8y/ngx-components';
import { Permissions } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class SimGuard {
    constructor(permissions, appState) {
        this.permissions = permissions;
        this.appState = appState;
    }
    canActivate() {
        if (this.showSimProvider === undefined) {
            const hasRole = Promise.resolve(this.permissions.hasRole('ROLE_CONNECTIVITY_READ'));
            this.showSimProvider = hasRole.then((hasPermissions) => hasPermissions && this.appState.isApplicationAvailable(SimGuard.applicationName));
        }
        return this.showSimProvider;
    }
}
SimGuard.applicationName = 'connectivity';
SimGuard.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SimGuard, deps: [{ token: i1.Permissions }, { token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
SimGuard.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SimGuard });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SimGuard, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Permissions }, { type: i1.AppStateService }]; } });
