// @c8y/ngx-components/esm2020/core/header/header.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ActionModule } from '../action/action.module';
import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import { CommonModule } from '../common/common.module';
import { NavigatorModule } from '../navigator/navigator.module';
import { ProductExperienceModule } from '../product-experience/product-experience.module';
import { SearchModule } from '../search/search.module';
import { UiSettingsModule } from '../ui-settings/ui-settings.module';
import { UserModule } from '../user/user.module';
import { VersionModule } from '../version';
import { HeaderBarComponent } from './header-bar/header-bar.component';
import { TitleOutletComponent } from './title/title-outlet.component';
import { TitleComponent } from './title/title.component';
import { ApplicationModule } from '../application/application.module';
import { DrawerModule } from '../drawer/drawer.module';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/tooltip";
import * as i2 from "ngx-bootstrap/collapse";
/**
 * The angular module definition for the header of the application.
 * @exports HeaderBarComponent
 * @exports TitleComponent
 * @exports TitleOutletComponent
 * @exports AppIconComponent
 * @exports AppSwitcherComponent
 */
export class HeaderModule {
}
HeaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HeaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HeaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: HeaderModule, declarations: [TitleComponent, HeaderBarComponent, TitleOutletComponent], imports: [ApplicationModule,
        A11yModule,
        UserModule,
        SearchModule,
        CommonModule,
        ActionModule,
        BreadcrumbModule,
        NavigatorModule,
        VersionModule, i1.TooltipModule, i2.CollapseModule, ProductExperienceModule,
        DrawerModule,
        UiSettingsModule], exports: [HeaderBarComponent, TitleComponent, TitleOutletComponent] });
HeaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HeaderModule, imports: [ApplicationModule,
        A11yModule,
        UserModule,
        SearchModule,
        CommonModule,
        ActionModule,
        BreadcrumbModule,
        NavigatorModule,
        VersionModule,
        TooltipModule.forRoot(),
        CollapseModule.forRoot(),
        ProductExperienceModule,
        DrawerModule,
        UiSettingsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HeaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TitleComponent, HeaderBarComponent, TitleOutletComponent],
                    imports: [
                        ApplicationModule,
                        A11yModule,
                        UserModule,
                        SearchModule,
                        CommonModule,
                        ActionModule,
                        BreadcrumbModule,
                        NavigatorModule,
                        VersionModule,
                        TooltipModule.forRoot(),
                        CollapseModule.forRoot(),
                        ProductExperienceModule,
                        DrawerModule,
                        UiSettingsModule
                    ],
                    exports: [HeaderBarComponent, TitleComponent, TitleOutletComponent]
                }]
        }] });
