// @c8y/ngx-components/esm2020/core/audit-log/audit-log.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { AuditLogComponent } from './audit-log.component';
import * as i0 from "@angular/core";
/** Audit module allows to show audits list. */
export class AuditLogModule {
}
AuditLogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AuditLogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AuditLogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AuditLogModule, declarations: [AuditLogComponent], imports: [CommonModule], exports: [AuditLogComponent] });
AuditLogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AuditLogModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AuditLogModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AuditLogComponent],
                    imports: [CommonModule],
                    exports: [AuditLogComponent]
                }]
        }] });
