// @c8y/ngx-components/esm2020/auth-configuration/factories/navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, Permissions } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class NavigationFactory {
    constructor(permissions) {
        this.permissions = permissions;
        this.navs = [];
    }
    async get() {
        const userHasPermission = this.permissions.hasAnyRole([
            'ROLE_TENANT_ADMIN',
            'ROLE_TENANT_MANAGEMENT_ADMIN'
        ]);
        if (this.navs.length === 0 && userHasPermission) {
            this.navs.push(new NavigatorNode({
                label: gettext('Authentication'),
                icon: 'c8y-shield',
                path: '/auth-configuration',
                parent: gettext('Settings'),
                routerLinkExact: false,
                priority: 2000
            }));
        }
        return this.navs;
    }
}
NavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigationFactory, deps: [{ token: i1.Permissions }], target: i0.ɵɵFactoryTarget.Injectable });
NavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Permissions }]; } });
