// @c8y/ngx-components/esm2020/auth-configuration/factories/tabs.factory.mjs
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { SsoConfigurationService } from '../sso-configuration/sso-configuration.service';
import { catchError, mapTo, mergeAll, toArray } from 'rxjs/operators';
import { of, EMPTY, merge } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../sso-configuration/sso-configuration.service";
export class TabsFactory {
    constructor(router, ssoConfigurationService) {
        this.router = router;
        this.ssoConfigurationService = ssoConfigurationService;
    }
    get() {
        if (!this.router.url.match(/auth-configuration/g)) {
            return of([]);
        }
        const basicSettingsTab$ = of({
            path: 'auth-configuration/basic_settings',
            label: gettext('Basic settings'),
            icon: 'unlock-alt',
            priority: 1100,
            orientation: 'horizontal'
        });
        const ssoTab$ = this.ssoConfigurationService.getSsoConfiguration$().pipe(mapTo({
            path: 'auth-configuration/single_sign-on',
            label: gettext('Single sign-on'),
            icon: 'sign-in',
            priority: 1050,
            orientation: 'horizontal'
        }), catchError(() => EMPTY));
        return merge([basicSettingsTab$, ssoTab$]).pipe(mergeAll(), toArray());
    }
}
TabsFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsFactory, deps: [{ token: i1.Router }, { token: i2.SsoConfigurationService }], target: i0.ɵɵFactoryTarget.Injectable });
TabsFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.SsoConfigurationService }]; } });
