// @c8y/ngx-components/esm2020/protocol-opcua/opcua-microservice.guard.mjs
import { Injectable } from '@angular/core';
import { AppStateService } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class OpcuaMicroserviceGuard {
    constructor(app) {
        this.app = app;
        this.microserviceName = 'opcua-mgmt-service';
    }
    canActivate() {
        if (!this.checkApp) {
            this.checkApp = this.app.isApplicationAvailable(this.microserviceName);
        }
        return this.checkApp;
    }
}
OpcuaMicroserviceGuard.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OpcuaMicroserviceGuard, deps: [{ token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
OpcuaMicroserviceGuard.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OpcuaMicroserviceGuard });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OpcuaMicroserviceGuard, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.AppStateService }]; } });
