// @c8y/ngx-components/esm2020/protocol-opcua/opcua-protocol.module.mjs
import { CdkTreeModule } from '@angular/cdk/tree';
import { RouterModule } from '@angular/router';
import { NgModule } from '@angular/core';
import { CoreModule, DeviceStatusModule, DropAreaModule, FormsModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { AddressSpaceService } from './address-space.service';
import { OpcuaAddressSpaceDetailComponent } from './opcua-address-space-detail.component';
import { OpcuaDeviceProtocolDetailComponent } from './opcua-device-protocol-detail.component';
import { OpcuaDeviceProtocolDataReportingComponent } from './opcua-device-protocol-data-reporting.component';
import { OpcuaDeviceProtocolMapping } from './opcua-device-protocol-mapping.component';
import { ReactiveFormsModule } from '@angular/forms';
import { OpcuaDeviceProtocolObjectMapping } from './opcua-device-protocol-object-mapping.component';
import { OpcuaDeviceProtocolMappingCustomAction } from './opcua-device-protocol-mapping-customaction.component';
import { OpcuaDeviceProtocolObjectMappingStatus } from './opcua-device-protocol-object-mapping-status-icon.component';
import { OpcuaAddressSpaceComponent } from './opcua-address-space.component';
import { OpcuaAgentGuard } from './opcua-agent.guard';
import { OpcuaMicroserviceGuard } from './opcua-microservice.guard';
import { OpcuaServerConfigComponent } from './opcua-server-config.component';
import { OpcuaServerListComponent } from './opcua-server-list.component';
import { OpcuaServerGuard } from './opcua-server.guard';
import { OpcuaServersComponent } from './opcua-servers.component';
import { OpcuaService } from './opcuaService';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { OpcuaDeviceProtocolDescription } from './opcua-device-protocol-description.component';
import { OpcuaAddressSpaceTreeComponent } from './opcua-address-space-tree.component';
import { OpcuaAutoApplySettingsComponent } from './opcua-auto-apply-settings.component';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { OpcuaDeviceProtocolBrowsePathValidation } from './opcua-device-protocol-browse-path-validation.directive';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "ngx-bootstrap/tooltip";
import * as i3 from "ngx-bootstrap/popover";
import * as i4 from "ngx-bootstrap/collapse";
import * as i5 from "ngx-bootstrap/dropdown";
const routes = [
    {
        context: ViewContext.Device,
        path: 'opcua-servers',
        component: OpcuaServersComponent,
        label: gettext('OPC UA server'),
        icon: 'server',
        canActivate: [OpcuaAgentGuard, OpcuaMicroserviceGuard]
    },
    {
        context: ViewContext.Device,
        path: 'address-space',
        component: OpcuaAddressSpaceComponent,
        label: gettext('Address space'),
        icon: 'cloud-network',
        canActivate: [OpcuaServerGuard, OpcuaMicroserviceGuard]
    }
];
const deviceProtocolRoutes = [
    {
        path: 'deviceprotocols/opcuaV2/:id',
        component: OpcuaDeviceProtocolDetailComponent
    }
];
export class OpcuaProtocolModule {
}
OpcuaProtocolModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OpcuaProtocolModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OpcuaProtocolModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: OpcuaProtocolModule, declarations: [OpcuaServersComponent,
        OpcuaServerListComponent,
        OpcuaServerConfigComponent,
        OpcuaAddressSpaceComponent,
        OpcuaAddressSpaceTreeComponent,
        OpcuaAddressSpaceDetailComponent,
        OpcuaDeviceProtocolDescription,
        OpcuaDeviceProtocolDetailComponent,
        OpcuaDeviceProtocolDataReportingComponent,
        OpcuaDeviceProtocolMapping,
        OpcuaDeviceProtocolObjectMapping,
        OpcuaDeviceProtocolMappingCustomAction,
        OpcuaDeviceProtocolObjectMappingStatus,
        OpcuaAutoApplySettingsComponent,
        OpcuaDeviceProtocolBrowsePathValidation], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        DropAreaModule,
        ButtonsModule,
        CdkTreeModule, i1.RouterModule, i2.TooltipModule, i3.PopoverModule, DeviceStatusModule,
        BrowserAnimationsModule, i4.CollapseModule, i5.BsDropdownModule] });
OpcuaProtocolModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OpcuaProtocolModule, providers: [
        OpcuaAgentGuard,
        OpcuaServerGuard,
        OpcuaMicroserviceGuard,
        OpcuaService,
        AddressSpaceService,
        hookRoute(routes)
    ], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        DropAreaModule,
        ButtonsModule,
        CdkTreeModule,
        RouterModule.forChild(deviceProtocolRoutes),
        TooltipModule.forRoot(),
        PopoverModule.forRoot(),
        DeviceStatusModule,
        BrowserAnimationsModule,
        CollapseModule.forRoot(),
        BsDropdownModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OpcuaProtocolModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        OpcuaServersComponent,
                        OpcuaServerListComponent,
                        OpcuaServerConfigComponent,
                        OpcuaAddressSpaceComponent,
                        OpcuaAddressSpaceTreeComponent,
                        OpcuaAddressSpaceDetailComponent,
                        OpcuaDeviceProtocolDescription,
                        OpcuaDeviceProtocolDetailComponent,
                        OpcuaDeviceProtocolDataReportingComponent,
                        OpcuaDeviceProtocolMapping,
                        OpcuaDeviceProtocolObjectMapping,
                        OpcuaDeviceProtocolMappingCustomAction,
                        OpcuaDeviceProtocolObjectMappingStatus,
                        OpcuaAutoApplySettingsComponent,
                        OpcuaDeviceProtocolBrowsePathValidation
                    ],
                    imports: [
                        CoreModule,
                        FormsModule,
                        ReactiveFormsModule,
                        DropAreaModule,
                        ButtonsModule,
                        CdkTreeModule,
                        RouterModule.forChild(deviceProtocolRoutes),
                        TooltipModule.forRoot(),
                        PopoverModule.forRoot(),
                        DeviceStatusModule,
                        BrowserAnimationsModule,
                        CollapseModule.forRoot(),
                        BsDropdownModule.forRoot()
                    ],
                    entryComponents: [
                        OpcuaAddressSpaceTreeComponent,
                        OpcuaAddressSpaceComponent,
                        OpcuaServersComponent
                    ],
                    providers: [
                        OpcuaAgentGuard,
                        OpcuaServerGuard,
                        OpcuaMicroserviceGuard,
                        OpcuaService,
                        AddressSpaceService,
                        hookRoute(routes)
                    ]
                }]
        }] });
