// @c8y/ngx-components/esm2020/operations/stepper-bulk-type-device-profile/stepper-bulk-type-device-profile.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { DeviceProfileModule } from '@c8y/ngx-components/device-profile';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { baseUrl, BulkOperationType, hookListBulkType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperFramesModule } from '@c8y/ngx-components/operations/stepper-frames';
import { ConfirmDeviceProfileSelectionStepComponent } from './confirm-device-profile-selection-step.component';
import { SelectDeviceProfileStepComponent } from './select-device-profile-step.component';
import { StepperBulkTypeDeviceProfileComponent } from './stepper-bulk-type-device-profile.component';
import * as i0 from "@angular/core";
/** Module for the 'Apply device profile' operation type stepper */
export class StepperBulkTypeDeviceProfileModule {
}
StepperBulkTypeDeviceProfileModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeDeviceProfileModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperBulkTypeDeviceProfileModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeDeviceProfileModule, declarations: [StepperBulkTypeDeviceProfileComponent,
        SelectDeviceProfileStepComponent,
        ConfirmDeviceProfileSelectionStepComponent], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        DeviceProfileModule,
        StepperFramesModule], exports: [StepperBulkTypeDeviceProfileComponent] });
StepperBulkTypeDeviceProfileModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeDeviceProfileModule, providers: [
        hookListBulkType({
            type: BulkOperationType.DEVICE_PROFILE,
            c8yIcon: 'c8y-device-profile',
            name: gettext('Apply device profile'),
            path: `${baseUrl}device-profile`,
            component: StepperBulkTypeDeviceProfileComponent,
            fragments: ['c8y_DeviceProfile'],
            selected: false
        })
    ], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        DeviceProfileModule,
        StepperFramesModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeDeviceProfileModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        StepperBulkTypeDeviceProfileComponent,
                        SelectDeviceProfileStepComponent,
                        ConfirmDeviceProfileSelectionStepComponent
                    ],
                    imports: [
                        CoreModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BulkOperationStepperModule,
                        DeviceProfileModule,
                        StepperFramesModule
                    ],
                    providers: [
                        hookListBulkType({
                            type: BulkOperationType.DEVICE_PROFILE,
                            c8yIcon: 'c8y-device-profile',
                            name: gettext('Apply device profile'),
                            path: `${baseUrl}device-profile`,
                            component: StepperBulkTypeDeviceProfileComponent,
                            fragments: ['c8y_DeviceProfile'],
                            selected: false
                        })
                    ],
                    entryComponents: [StepperBulkTypeDeviceProfileComponent],
                    exports: [StepperBulkTypeDeviceProfileComponent]
                }]
        }] });
