// @c8y/ngx-components/esm2020/core/data-grid/column/cell-renderer.mjs
/* tslint:disable:max-classes-per-file */
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class CellRendererContext {
}
export class CellRendererDefDirective {
    constructor(template) {
        this.template = template;
    }
}
CellRendererDefDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CellRendererDefDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
CellRendererDefDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: CellRendererDefDirective, selector: "[c8yCellRendererDef]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CellRendererDefDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[c8yCellRendererDef]' }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class HeaderCellRendererDefDirective {
    constructor(template) {
        this.template = template;
    }
}
HeaderCellRendererDefDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HeaderCellRendererDefDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
HeaderCellRendererDefDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: HeaderCellRendererDefDirective, selector: "[c8yHeaderCellRendererDef]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HeaderCellRendererDefDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[c8yHeaderCellRendererDef]' }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
