// @c8y/ngx-components/esm2020/core/data-grid/column/custom.column.mjs
import { getBasicInputArrayFormFieldConfig } from '../../dynamic-forms/presets';
import { gettext } from '../../i18n/gettext';
import { BaseColumn } from './base.column';
export class CustomColumn extends BaseColumn {
    constructor(initialColumnConfig) {
        super(initialColumnConfig);
        this.name = this.name || 'custom';
        this.header = this.header || gettext('Custom');
        this.filterable = true;
        this.filteringConfig = {
            fields: [
                {
                    key: 'exists',
                    type: 'switch',
                    defaultValue: false,
                    templateOptions: {
                        label: gettext('Only rows where value is defined')
                    }
                },
                ...getBasicInputArrayFormFieldConfig({
                    key: 'equals',
                    label: gettext('Only rows where value equals to'),
                    addText: gettext('Add next`value`'),
                    tooltip: gettext('Use * as a wildcard character'),
                    placeholder: '10300',
                    optional: true
                })
            ],
            getFilter: (model) => {
                const filter = {};
                if (model.exists) {
                    filter.__has = this.path;
                }
                if (model.equals) {
                    filter[this.path] = { __in: model.equals };
                }
                return filter;
            }
        };
        this.sortable = true;
        this.sortingConfig = {
            pathSortingConfigs: [{ path: this.path }]
        };
    }
}
