// @c8y/ngx-components/esm2020/core/dynamic-component/dynamic-component-alert-aggregator.mjs
import { cloneDeep } from 'lodash-es';
import { BehaviorSubject } from 'rxjs';
import { DismissAlertStrategy } from './dynamic-component-alert.model';
/**
 * Order of widget alerts due to alerts severity.
 */
const WIDGET_ALERTS_ORDER = ['danger', 'warning', 'info'];
export class DynamicComponentAlertAggregator {
    constructor(alerts = []) {
        this.anyAlertExistsSubject = new BehaviorSubject(false);
        this.anyAlertExists$ = this.anyAlertExistsSubject.asObservable();
        this.initAlertsGroups();
        this.setAlerts(alerts);
    }
    /**
     * Adds alert to aggregator.
     *
     * @param {DynamicComponentAlert[]} alerts List of alerts that should be added to aggregator.
     *
     */
    addAlerts(...alerts) {
        this.setAlerts(alerts);
    }
    /**
     * Remove alerts in group by provided alert type or removes all alerts when no param is provided.
     *
     * @param {AlertType} type Type of alerts group that should be removed.
     *
     */
    clear(type) {
        if (type) {
            this.aggregatedAlertsData.get(type).alerts = [];
        }
        else {
            this.aggregatedAlertsData.forEach(value => (value.alerts = []));
        }
        this.onDataChanged();
    }
    /**
     * Sets possibility of alert to be dismissed in specific way.
     *
     * @param {AlertType} alertType Type of alerts for which the dismissal strategy should be set.
     * @param {dismissStrategy} dismissStrategy Type of dismissal strategy that will be set for the alert type.
     *
     */
    setAlertGroupDismissStrategy(alertType, dismissStrategy) {
        this.aggregatedAlertsData.get(alertType).alertDismissal.dismissStrategy = dismissStrategy;
        this.onDataChanged();
    }
    /**
     * Removes alerts in group due to dismissal dismissStrategy and sets userDismissalChoice property.
     * If group has dismissal strategy declared as TEMPORARY_OR_PERMANENT, its userDismissalChoice property can be set to
     * REMIND_ALLOWED or REMIND_DISALLOWED depending on parameter.
     *
     * @param {AlertType} alertType Type of alerts group that should be dismissed.
     * @param {boolean} permanent Determines whether the dismissal of alerts of a given type is permanent.
     *
     */
    dismissAlertGroup(alertType, isPermanent = false) {
        const groupDismissStrategy = this.aggregatedAlertsData.get(alertType).alertDismissal.dismissStrategy;
        switch (groupDismissStrategy) {
            case DismissAlertStrategy.NONE:
                return;
            case DismissAlertStrategy.TEMPORARY:
                this.aggregatedAlertsData.get(alertType).alertDismissal.userDismissalChoice =
                    'REMIND_ALLOWED';
                break;
            case DismissAlertStrategy.TEMPORARY_OR_PERMANENT:
                this.aggregatedAlertsData.get(alertType).alertDismissal.userDismissalChoice = isPermanent
                    ? 'REMIND_DISALLOWED'
                    : 'REMIND_ALLOWED';
        }
        this.clear(alertType);
    }
    /**
     * Alert aggregator as a widget property should not be serialized and stored in database.
     */
    toJSON() {
        return undefined;
    }
    initAlertsGroups() {
        this.aggregatedAlertsData = new Map(WIDGET_ALERTS_ORDER.map(alertType => {
            return [
                alertType,
                {
                    alerts: [],
                    alertDismissal: { dismissStrategy: DismissAlertStrategy.NONE }
                }
            ];
        }));
    }
    setAlerts(alerts) {
        alerts.forEach(alert => {
            if (this.aggregatedAlertsData.get(alert.type).alertDismissal.userDismissalChoice ===
                'REMIND_DISALLOWED') {
                return;
            }
            if (!this.aggregatedAlertsData.get(alert.type)) {
                this.aggregatedAlertsData.get(alert.type).alerts = [];
            }
            this.aggregatedAlertsData.get(alert.type).alerts.push(alert);
        });
        this.onDataChanged();
    }
    onDataChanged() {
        this.alertGroups = [...cloneDeep(this.aggregatedAlertsData)].map(([key, value]) => ({
            type: key,
            value
        }));
        queueMicrotask(() => {
            this.anyAlertExistsSubject.next(this.alertGroups.some(({ value }) => !!value.alerts.length));
        });
    }
}
