// @c8y/ngx-components/esm2020/context-dashboard/add-dashboard.factory.mjs
import { Injectable } from '@angular/core';
import { ContextRouteService, Permissions, ViewContext } from '@c8y/ngx-components';
import { BsModalService } from 'ngx-bootstrap/modal';
import { AddDashboardComponent } from './add-dashboard.component';
import { ContextDashboardService } from './context-dashboard.service';
import { DashboardDetailComponent } from './dashboard-detail.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "./context-dashboard.service";
import * as i3 from "ngx-bootstrap/modal";
export class AddDashboardFactory {
    constructor(permissions, contextDashboardService, bsModal, contextRoute) {
        this.permissions = permissions;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.contextRoute = contextRoute;
    }
    async get(activatedRoute) {
        this.currentContext = this.contextRoute.getContextData(activatedRoute);
        if ([ViewContext.Device, ViewContext.Group].includes(this.currentContext?.context) &&
            (await this.hasPermission(this.currentContext.contextData))) {
            return [
                {
                    component: AddDashboardComponent,
                    priority: -Infinity
                }
            ];
        }
        return [];
    }
    async addDashboard() {
        let initialState;
        if (this.currentContext.context === ViewContext.Device) {
            initialState = {
                deviceType: this.currentContext.contextData.type
            };
        }
        const modal = this.bsModal.show(DashboardDetailComponent, {
            class: 'modal-lg',
            ariaDescribedby: 'modal-body',
            ariaLabelledBy: 'modal-title',
            initialState,
            ignoreBackdropClick: true
        }).content;
        try {
            const dashboardCfg = await modal.result;
            const dashboardMO = await this.contextDashboardService.create(dashboardCfg, this.currentContext);
            await this.contextDashboardService.navigateToDashboard(dashboardMO);
            modal.close();
        }
        catch (ex) {
            // intended empty
        }
    }
    async hasPermission(context) {
        if (context?.id) {
            return await this.permissions.canEdit(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE'], context);
        }
        return this.permissions.hasAnyRole(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE']);
    }
}
AddDashboardFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddDashboardFactory, deps: [{ token: i1.Permissions }, { token: i2.ContextDashboardService }, { token: i3.BsModalService }, { token: i1.ContextRouteService }], target: i0.ɵɵFactoryTarget.Injectable });
AddDashboardFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddDashboardFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddDashboardFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.Permissions }, { type: i2.ContextDashboardService }, { type: i3.BsModalService }, { type: i1.ContextRouteService }]; } });
