// @c8y/ngx-components/esm2020/context-dashboard/paste-dashboard-action.component.mjs
import { Component, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { ViewContext } from '@c8y/ngx-components';
import { ContextDashboardService } from './context-dashboard.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./context-dashboard.service";
import * as i3 from "@c8y/client";
import * as i4 from "@c8y/ngx-components";
export class PasteDashboardActionComponent {
    constructor(vcRef, router, contextDashboardService, inventoryService) {
        this.vcRef = vcRef;
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.inventoryService = inventoryService;
        this.dashboardName = '';
    }
    ngOnInit() {
        this.vcRef.createEmbeddedView(this.templatePaste);
        const clipboard = this.contextDashboardService.copyClipboard;
        if (clipboard) {
            this.dashboardName = clipboard.dashboard.name;
        }
    }
    async pasteDashboard() {
        let context;
        const currentUrl = this.router.url;
        const isDevice = new RegExp(/device\/\d+/).test(currentUrl);
        const isGroup = new RegExp(/group\/\d+/).test(currentUrl);
        const id = currentUrl.match(/\d+/)[0];
        if (isDevice) {
            context = {
                context: ViewContext.Device,
                contextData: await this.getContextMo(id)
            };
        }
        else if (isGroup) {
            context = {
                context: ViewContext.Group,
                contextData: await this.getContextMo(id)
            };
        }
        this.contextDashboardService.pasteDashboard(context);
    }
    async getContextMo(id) {
        return (await this.inventoryService.detail(id)).data;
    }
}
PasteDashboardActionComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasteDashboardActionComponent, deps: [{ token: i0.ViewContainerRef }, { token: i1.Router }, { token: i2.ContextDashboardService }, { token: i3.InventoryService }], target: i0.ɵɵFactoryTarget.Component });
PasteDashboardActionComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: PasteDashboardActionComponent, selector: "c8y-paste-dashboard-action", viewQueries: [{ propertyName: "templatePaste", first: true, predicate: ["templatePaste"], descendants: true, read: TemplateRef, static: true }], ngImport: i0, template: `
    <ng-template #templatePaste>
      <li>
        <button
          class="btn btn-link"
          type="button"
          title="{{ 'Paste dashboard' | translate }}"
          (click)="pasteDashboard()"
        >
          <i class="m-r-4" c8yIcon="clipboard"></i>
          <span ngNonBindable translate [translateParams]="{ dashboardName: dashboardName }">
            Paste dashboard "{{ dashboardName }}"
          </span>
        </button>
      </li>
    </ng-template>
  `, isInline: true, dependencies: [{ kind: "directive", type: i4.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "directive", type: i4.C8yTranslateDirective, selector: "[translate],[ngx-translate]" }, { kind: "pipe", type: i4.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasteDashboardActionComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-paste-dashboard-action',
                    template: `
    <ng-template #templatePaste>
      <li>
        <button
          class="btn btn-link"
          type="button"
          title="{{ 'Paste dashboard' | translate }}"
          (click)="pasteDashboard()"
        >
          <i class="m-r-4" c8yIcon="clipboard"></i>
          <span ngNonBindable translate [translateParams]="{ dashboardName: dashboardName }">
            Paste dashboard "{{ dashboardName }}"
          </span>
        </button>
      </li>
    </ng-template>
  `
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i1.Router }, { type: i2.ContextDashboardService }, { type: i3.InventoryService }]; }, propDecorators: { templatePaste: [{
                type: ViewChild,
                args: ['templatePaste', { read: TemplateRef, static: true }]
            }] } });
