// @c8y/ngx-components/esm2020/context-dashboard/widget.service.mjs
import { Injectable, Inject, Optional } from '@angular/core';
import { sortBy, cloneDeep, get, omit } from 'lodash-es';
import { DynamicComponentService } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { CONTEXT_DASHBOARD_CONFIG } from './context-dashboard.model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@ngx-translate/core";
export class WidgetService {
    constructor(dynamicComponentService, translateService, moduleConfig) {
        this.dynamicComponentService = dynamicComponentService;
        this.translateService = translateService;
        this.moduleConfig = moduleConfig;
        this.dynamicComponentService.items$.subscribe(widgets => {
            this.widgets = widgets;
        });
    }
    async getWidgetDefinition(componentId) {
        return this.dynamicComponentService.getById(componentId);
    }
    getWidgetDefinitions() {
        const translatedComponents = this.widgets.map(cmp => ({
            ...cmp,
            label: this.translateService.instant(cmp.label)
        }));
        // cloneDeep and injector lead to a performance issue. We removing the injectors here,
        // as they are not needed, to avoid use of a lot of CPU for cloning.
        const translatedComponentsWithoutInjector = translatedComponents.map(cmp => omit(cmp, 'injector'));
        return cloneDeep(sortBy(translatedComponentsWithoutInjector, 'label').filter(this.moduleConfig.widgetFilter));
    }
    async mapLegacy(widget) {
        const cmp = await this.getWidgetDefinition(widget.componentId || widget.name);
        widget = this.mapDisplaySettings(widget, cmp);
        if (get(cmp, 'data.settings.upgrade')) {
            widget.widgetComponent = cmp.data.settings.widgetComponent;
            widget.configComponent = cmp.data.settings.configComponent;
            widget.templateUrl = cmp.data.settings.templateUrl;
            widget.configTemplateUrl = cmp.data.settings.configTemplateUrl;
            widget.transformConfigWithContext =
                cmp.data.settings.componentTransformConfigWithContext ||
                    cmp.data.settings.transformConfigWithContext ||
                    widget.transformConfigWithContext;
        }
        else {
            delete widget.templateUrl;
            delete widget.configTemplateUrl;
        }
        return widget;
    }
    mapDisplaySettings(widget, cmp) {
        if (cmp?.data?.displaySettings) {
            widget.config = { displaySettings: cmp.data.displaySettings, ...(widget.config || {}) };
        }
        return widget;
    }
}
WidgetService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetService, deps: [{ token: i1.DynamicComponentService }, { token: i2.TranslateService }, { token: CONTEXT_DASHBOARD_CONFIG, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
WidgetService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.DynamicComponentService }, { type: i2.TranslateService }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [CONTEXT_DASHBOARD_CONFIG]
                }] }]; } });
