// @c8y/ngx-components/esm2020/sub-assets/sub-assets.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { AddGroupModule } from './add-group/add-group.module';
import { AssetTypeCellRendererComponent } from './columns/asset-type.cell-renderer.component';
import { DeleteAssetsModalComponent } from './delete-assets-modal/delete-assets-modal.component';
import { GroupInfoComponent } from './group-info.component';
import { GroupsComponent } from './groups.component';
import { SubAssetsGridComponent } from './sub-assets-grid.component';
import { SubAssetsComponent } from './sub-assets.component';
import { SubAssetsService } from './sub-assets.service';
import { UnassignModalComponent } from './unassign-assets-modal/unassign-modal.component';
import { AssignDevicesComponent } from './assign-devices/assign-devices.component';
import { AssignChildDevicesComponent } from './assign-devices/assign-child-devices.component';
import { AssetPropertiesComponent } from './asset-properties.component';
import { AssetPropertiesItemComponent } from './asset-properties-item.component';
import { SUB_ASSETS_CONFIG, defaultModuleConfig } from './sub-assets.model';
import * as i0 from "@angular/core";
export class SubAssetsModule {
    static config(config = {}) {
        return {
            ngModule: SubAssetsModule,
            providers: [
                {
                    provide: SUB_ASSETS_CONFIG,
                    useValue: { ...defaultModuleConfig, ...config }
                }
            ]
        };
    }
}
SubAssetsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SubAssetsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsModule, declarations: [SubAssetsComponent,
        SubAssetsGridComponent,
        AssetTypeCellRendererComponent,
        GroupsComponent,
        GroupInfoComponent,
        DeleteAssetsModalComponent,
        UnassignModalComponent,
        AssignDevicesComponent,
        AssignChildDevicesComponent,
        AssetPropertiesComponent,
        AssetPropertiesItemComponent], imports: [CoreModule, DeviceGridModule, AddGroupModule, PopoverModule, TooltipModule], exports: [SubAssetsGridComponent] });
SubAssetsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsModule, providers: [
        {
            provide: SUB_ASSETS_CONFIG,
            useValue: defaultModuleConfig
        },
        hookRoute([
            {
                context: ViewContext.Group,
                path: 'subassets',
                priority: 1000,
                icon: 'c8y-group-open',
                label: gettext('Subassets'),
                component: SubAssetsComponent,
                featureId: 'subassets'
            },
            {
                path: 'group',
                component: GroupsComponent
            }
        ]),
        SubAssetsService
    ], imports: [CoreModule, DeviceGridModule, AddGroupModule, PopoverModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SubAssetsComponent,
                        SubAssetsGridComponent,
                        AssetTypeCellRendererComponent,
                        GroupsComponent,
                        GroupInfoComponent,
                        DeleteAssetsModalComponent,
                        UnassignModalComponent,
                        AssignDevicesComponent,
                        AssignChildDevicesComponent,
                        AssetPropertiesComponent,
                        AssetPropertiesItemComponent
                    ],
                    imports: [CoreModule, DeviceGridModule, AddGroupModule, PopoverModule, TooltipModule],
                    entryComponents: [
                        SubAssetsComponent,
                        AssetTypeCellRendererComponent,
                        GroupsComponent,
                        DeleteAssetsModalComponent,
                        UnassignModalComponent
                    ],
                    exports: [SubAssetsGridComponent],
                    providers: [
                        {
                            provide: SUB_ASSETS_CONFIG,
                            useValue: defaultModuleConfig
                        },
                        hookRoute([
                            {
                                context: ViewContext.Group,
                                path: 'subassets',
                                priority: 1000,
                                icon: 'c8y-group-open',
                                label: gettext('Subassets'),
                                component: SubAssetsComponent,
                                featureId: 'subassets'
                            },
                            {
                                path: 'group',
                                component: GroupsComponent
                            }
                        ]),
                        SubAssetsService
                    ]
                }]
        }] });
