// @c8y/ngx-components/esm2020/sub-assets/sub-assets-grid-configuration-strategy.service.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { AbstractConfigurationStrategy, DATA_GRID_CONFIGURATION_CONTEXT, DATA_GRID_CONFIGURATION_CONTEXT_PROVIDER, UserPreferencesConfigurationStrategy } from '@c8y/ngx-components';
import { AssetNodeService } from '@c8y/ngx-components/assets-navigator';
import { SmartGroupGridConfigurationStrategy } from './smart-group-grid-configuration-strategy.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "./smart-group-grid-configuration-strategy.service";
import * as i3 from "@c8y/ngx-components/assets-navigator";
export class SubAssetsGridConfigurationStrategy extends AbstractConfigurationStrategy {
    constructor(userPreferencesConfigurationStrategy, smartGroupGridConfigurationStrategy, assetNodeService, context, contextProvider) {
        super(context, contextProvider);
        this.userPreferencesConfigurationStrategy = userPreferencesConfigurationStrategy;
        this.smartGroupGridConfigurationStrategy = smartGroupGridConfigurationStrategy;
        this.assetNodeService = assetNodeService;
        this.context = context;
        this.contextProvider = contextProvider;
    }
    getConfig$(context) {
        return this.getStrategy(context).getConfig$(context);
    }
    saveConfig$(config, context) {
        return this.getStrategy(context).saveConfig$(config, context);
    }
    getStrategy(ctx) {
        const context = this.retrieveContext(ctx);
        return !!context?.group &&
            this.assetNodeService.isDynamicGroup(context?.group) &&
            context?.group?.c8y_DeviceColumnsConfig
            ? this.smartGroupGridConfigurationStrategy
            : this.userPreferencesConfigurationStrategy;
    }
}
SubAssetsGridConfigurationStrategy.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsGridConfigurationStrategy, deps: [{ token: i1.UserPreferencesConfigurationStrategy }, { token: i2.SmartGroupGridConfigurationStrategy }, { token: i3.AssetNodeService }, { token: DATA_GRID_CONFIGURATION_CONTEXT, optional: true }, { token: DATA_GRID_CONFIGURATION_CONTEXT_PROVIDER, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
SubAssetsGridConfigurationStrategy.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsGridConfigurationStrategy, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SubAssetsGridConfigurationStrategy, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.UserPreferencesConfigurationStrategy }, { type: i2.SmartGroupGridConfigurationStrategy }, { type: i3.AssetNodeService }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DATA_GRID_CONFIGURATION_CONTEXT]
                }, {
                    type: Optional
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DATA_GRID_CONFIGURATION_CONTEXT_PROVIDER]
                }, {
                    type: Optional
                }] }]; } });
