// @c8y/ngx-components/esm2020/widgets/implementations/markdown/markdown-widget-view/markdown-widget-view.component.mjs
import { Component, Input } from '@angular/core';
import { FetchClient } from '@c8y/client';
import { AppStateService } from '@c8y/ngx-components';
import { MarkdownWidgetService } from '../markdown-widget.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/client";
import * as i3 from "../markdown-widget.service";
import * as i4 from "@angular/common";
export class MarkdownWidgetViewComponent {
    constructor(appState, client, markdownWidgetService) {
        this.appState = appState;
        this.client = client;
        this.markdownWidgetService = markdownWidgetService;
        this.headers = { 'Content-Type': 'text/markdown', responseType: 'blob' };
    }
    async ngOnInit() {
        this.contextPath = this.appState.getCurrentContextPath();
        if (this.config.markdownBinaryId) {
            const readmeContent = await (await this.markdownWidgetService.getFile(this.config.markdownBinaryId)).text();
            this.markdown = readmeContent;
        }
        else if (this.config.contentUrl?.toLowerCase() === '/readme.md') {
            this.markdown = await this.getReadmeFileContent();
        }
        else {
            this.setContentFromUrl(this.config.contentUrl);
        }
    }
    setContentFromUrl(url) {
        const req = new XMLHttpRequest();
        req.onreadystatechange = () => this.render(req);
        req.addEventListener('load', () => this.render(req));
        req.open('GET', url);
        req.responseType = 'text';
        req.setRequestHeader('Accept', 'text/html');
        req.send();
    }
    async render(req) {
        if (req.readyState === 4 && req.status === 200) {
            this.markdown = req.response;
        }
    }
    async getReadmeFileContent() {
        const readmeFile = await this.getReadmeFile();
        if (readmeFile.status === 200) {
            return await readmeFile.text();
        }
        return '';
    }
    async getReadmeFile() {
        const options = {
            method: 'GET',
            headers: this.headers
        };
        const result = await this.client.fetch(`/apps/${this.contextPath}${this.config.contentUrl}`, options);
        return result;
    }
}
MarkdownWidgetViewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownWidgetViewComponent, deps: [{ token: i1.AppStateService }, { token: i2.FetchClient }, { token: i3.MarkdownWidgetService }], target: i0.ɵɵFactoryTarget.Component });
MarkdownWidgetViewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: MarkdownWidgetViewComponent, selector: "c8y-markdown-widget-view", inputs: { config: "config" }, ngImport: i0, template: "<div id=\"helpContent\" class=\"p-16 p-t-0 markdown-content\" [innerHTML]=\"markdown | markdownToHtml | async\"></div>\n", dependencies: [{ kind: "pipe", type: i4.AsyncPipe, name: "async" }, { kind: "pipe", type: i1.MarkdownToHtmlPipe, name: "markdownToHtml" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownWidgetViewComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-markdown-widget-view', template: "<div id=\"helpContent\" class=\"p-16 p-t-0 markdown-content\" [innerHTML]=\"markdown | markdownToHtml | async\"></div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.AppStateService }, { type: i2.FetchClient }, { type: i3.MarkdownWidgetService }]; }, propDecorators: { config: [{
                type: Input
            }] } });
