// @c8y/ngx-components/esm2020/protocol-lwm2m/guards/lwm2m-bootstrap-parameters.guard.mjs
import { Injectable } from '@angular/core';
import { TenantUiService, Permissions } from '@c8y/ngx-components';
import { Lwm2mClientService } from '@c8y/ngx-components/protocol-lwm2m/services';
import { Lwm2mBaseGuard } from './lwm2m-base.guard';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/ngx-components/protocol-lwm2m/services";
export class Lwm2mBootstrapParametersGuard extends Lwm2mBaseGuard {
    constructor(tenantService, permissions, lwm2mService) {
        super(tenantService);
        this.tenantService = tenantService;
        this.permissions = permissions;
        this.lwm2mService = lwm2mService;
    }
    additionalChecks(route) {
        const contextData = (route?.data).contextData ||
            (route?.parent.data).contextData;
        return (this.lwm2mService.isLwm2mDevice(contextData) &&
            this.permissions.hasRole('ROLE_INVENTORY_READ'));
    }
}
Lwm2mBootstrapParametersGuard.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mBootstrapParametersGuard, deps: [{ token: i1.TenantUiService }, { token: i1.Permissions }, { token: i2.Lwm2mClientService }], target: i0.ɵɵFactoryTarget.Injectable });
Lwm2mBootstrapParametersGuard.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mBootstrapParametersGuard });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mBootstrapParametersGuard, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.TenantUiService }, { type: i1.Permissions }, { type: i2.Lwm2mClientService }]; } });
