// @c8y/ngx-components/esm2020/data-broker/factories/data-broker.subscription.factory.mjs
import { DeviceRegistrationStatus } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { DataBrokerConnectorStatus } from '../model/data-broker.subscription.model';
export class SubscriptionStatusViewModelFactory {
    build(status) {
        let statusViewModel = null;
        switch (status) {
            case DataBrokerConnectorStatus.ACTIVE: // dataBrokerConnectorsSvc.STATUS.ACTIVE.value
                statusViewModel = {
                    icon: 'check-circle',
                    textClass: 'text-success',
                    text: gettext('Active`data broker subscription`')
                };
                break;
            case DataBrokerConnectorStatus.SUSPENDED: // dataBrokerConnectorsSvc.STATUS.SUSPENDED.value
                statusViewModel = {
                    icon: 'exclamation-triangle',
                    textClass: 'text-warning',
                    text: gettext('Inactive`data broker subscription`')
                };
                break;
            case DataBrokerConnectorStatus.WAITING: // dataBrokerConnectorsSvc.STATUS.WAITING.value
                statusViewModel = {
                    icon: 'spinner',
                    iconClass: 'icon-spin',
                    textClass: 'text-warning',
                    text: gettext('Waiting`data broker subscription`')
                };
                break;
            default:
                break;
        }
        return statusViewModel;
    }
}
export class NewDeviceRequestStatusFactory {
    build(status) {
        let result = null;
        switch (status) {
            case DeviceRegistrationStatus.WAITING_FOR_CONNECTION:
                result = {
                    name: DeviceRegistrationStatus.WAITING_FOR_CONNECTION,
                    label: gettext('Waiting for connection`data broker subscription`'),
                    value: DeviceRegistrationStatus.WAITING_FOR_CONNECTION
                };
                break;
            case DeviceRegistrationStatus.PENDING_ACCEPTANCE:
                result = {
                    name: DeviceRegistrationStatus.PENDING_ACCEPTANCE,
                    label: gettext('Pending acceptance`data broker subscription`'),
                    value: DeviceRegistrationStatus.PENDING_ACCEPTANCE
                };
                break;
            case DeviceRegistrationStatus.ACCEPTED:
                result = {
                    name: DeviceRegistrationStatus.ACCEPTED,
                    label: gettext('Accepted`data broker subscription`'),
                    value: DeviceRegistrationStatus.ACCEPTED
                };
            case DeviceRegistrationStatus.BLOCKED:
                result = {
                    name: DeviceRegistrationStatus.BLOCKED,
                    label: gettext('Blocked`data broker subscription`'),
                    value: DeviceRegistrationStatus.BLOCKED
                };
                break;
        }
        return result;
    }
}
