// @c8y/ngx-components/esm2020/context-dashboard/report-dashboard/report-dashboard.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, hookNavigator, hookRoute, ViewContext } from '@c8y/ngx-components';
import { ContextDashboardModule } from '../context-dashboard.module';
import { ContextDashboardComponent } from '../context-dashboard.component';
import { ReportDashboardGuard } from './report-dashboard.guard';
import { ReportDashboardNavigationFactory } from './report-dashboard-navigation.factory';
import { ReportDashboardListComponent } from './report-dashboard-list.component';
import { ReportDashboardService } from './report-dashboard.service';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/tooltip";
import * as i2 from "@angular/router";
export class ReportDashboardModule {
}
ReportDashboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ReportDashboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardModule, declarations: [ReportDashboardListComponent], imports: [CoreModule,
        ContextDashboardModule, i1.TooltipModule, i2.RouterModule, PopoverModule] });
ReportDashboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardModule, providers: [
        ReportDashboardGuard,
        ReportDashboardService,
        hookRoute({
            path: 'dashboard/:dashboardId',
            component: ContextDashboardComponent,
            canActivate: [ReportDashboardGuard],
            context: ViewContext.Report
        }),
        hookNavigator(ReportDashboardNavigationFactory)
    ], imports: [CoreModule,
        ContextDashboardModule,
        TooltipModule.forRoot(),
        RouterModule.forChild([
            {
                path: 'reports',
                component: ReportDashboardListComponent
            }
        ]),
        PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ReportDashboardListComponent],
                    entryComponents: [],
                    imports: [
                        CoreModule,
                        ContextDashboardModule,
                        TooltipModule.forRoot(),
                        RouterModule.forChild([
                            {
                                path: 'reports',
                                component: ReportDashboardListComponent
                            }
                        ]),
                        PopoverModule
                    ],
                    providers: [
                        ReportDashboardGuard,
                        ReportDashboardService,
                        hookRoute({
                            path: 'dashboard/:dashboardId',
                            component: ContextDashboardComponent,
                            canActivate: [ReportDashboardGuard],
                            context: ViewContext.Report
                        }),
                        hookNavigator(ReportDashboardNavigationFactory)
                    ]
                }]
        }] });
