// @c8y/ngx-components/esm2020/context-dashboard/report-dashboard/report-dashboard-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { InventoryService } from '@c8y/client';
import { NavigatorNode, NavigatorService, gettext } from '@c8y/ngx-components';
import { ReportDashboardService } from './report-dashboard.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/client";
import * as i3 from "./report-dashboard.service";
const REPORTS_NAVIGATOR_NODE_PRIORITY = 40;
export const REPORT_DEFAULT_NAVIGATION_NODE_PRIORITY = 30;
export class ReportDashboardNavigationFactory {
    constructor(navigatorService, inventoryService, reportDashboardService) {
        this.navigatorService = navigatorService;
        this.inventoryService = inventoryService;
        this.reportDashboardService = reportDashboardService;
        this.loaded = false;
    }
    async get() {
        if (!this.loaded) {
            this.loaded = true;
            this.navigatorService.add(new NavigatorNode({
                label: gettext('Reports'),
                path: 'reports',
                icon: 'c8y-reports',
                featureId: 'reports',
                priority: REPORTS_NAVIGATOR_NODE_PRIORITY
            }));
            const reports = (await this.reportDashboardService.listReports({
                filter: { __has: 'c8y_IsNavigatorNode' },
                pageSize: 2000
            })).data;
            reports.forEach(report => {
                const node = new NavigatorNode({
                    label: report.name,
                    path: `reports/${report.id}`,
                    icon: report.icon,
                    priority: report.priority,
                    routerLinkExact: false
                });
                this.navigatorService.add(node);
            });
        }
    }
}
ReportDashboardNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardNavigationFactory, deps: [{ token: i1.NavigatorService }, { token: i2.InventoryService }, { token: i3.ReportDashboardService }], target: i0.ɵɵFactoryTarget.Injectable });
ReportDashboardNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardNavigationFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ReportDashboardNavigationFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.NavigatorService }, { type: i2.InventoryService }, { type: i3.ReportDashboardService }]; } });
