// @c8y/ngx-components/esm2020/files-repository/files-repository.guard.mjs
import { Injectable } from '@angular/core';
import { Permissions } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class FilesRepositoryGuard {
    constructor(permissions) {
        this.permissions = permissions;
    }
    /**
     * Checks if files repository should be active,
     * i.e. whether the current tenant can read or admin files in repository.
     *
     * @returns True, if the feature should be active.
     */
    canActivate() {
        return (this.permissions.hasRole('ROLE_INVENTORY_READ') ||
            this.permissions.hasRole('ROLE_INVENTORY_ADMIN'));
    }
}
FilesRepositoryGuard.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryGuard, deps: [{ token: i1.Permissions }], target: i0.ɵɵFactoryTarget.Injectable });
FilesRepositoryGuard.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryGuard, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryGuard, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.Permissions }]; } });
