// @c8y/ngx-components/esm2020/files-repository/files-repository.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, hookNavigator, hookRoute, DropAreaModule } from '@c8y/ngx-components';
import { FilesRepositoryNavigationFactory } from './files-repository-navigation.factory';
import { FilesRepositoryComponent } from './files-repository.component';
import { FilesRepositoryUploadComponent } from './files-repository-upload.component';
import { FilesRepositoryService } from './files-repository.service';
import { FilesRepositoryGuard } from './files-repository.guard';
import { FilePreviewModule } from '@c8y/ngx-components/file-preview';
import * as i0 from "@angular/core";
export class FilesRepositoryModule {
}
FilesRepositoryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilesRepositoryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryModule, declarations: [FilesRepositoryComponent, FilesRepositoryUploadComponent], imports: [RouterModule, CoreModule, DropAreaModule, FilePreviewModule] });
FilesRepositoryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryModule, providers: [
        FilesRepositoryGuard,
        hookNavigator(FilesRepositoryNavigationFactory),
        hookRoute({
            path: 'files',
            component: FilesRepositoryComponent,
            canActivate: [FilesRepositoryGuard]
        }, { multi: true }),
        FilesRepositoryService
    ], imports: [RouterModule, CoreModule, DropAreaModule, FilePreviewModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [RouterModule, CoreModule, DropAreaModule, FilePreviewModule],
                    exports: [],
                    declarations: [FilesRepositoryComponent, FilesRepositoryUploadComponent],
                    providers: [
                        FilesRepositoryGuard,
                        hookNavigator(FilesRepositoryNavigationFactory),
                        hookRoute({
                            path: 'files',
                            component: FilesRepositoryComponent,
                            canActivate: [FilesRepositoryGuard]
                        }, { multi: true }),
                        FilesRepositoryService
                    ]
                }]
        }] });
