// @c8y/ngx-components/esm2020/files-repository/files-repository-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { FilesRepositoryGuard } from './files-repository.guard';
import * as i0 from "@angular/core";
import * as i1 from "./files-repository.guard";
export class FilesRepositoryNavigationFactory {
    constructor(filesRepositoryGuard) {
        this.filesRepositoryGuard = filesRepositoryGuard;
        this.navs = [];
    }
    async get() {
        if (this.filesRepositoryGuard.canActivate()) {
            if (!this.navs.length) {
                this.navs.push(new NavigatorNode({
                    parent: gettext('Management'),
                    label: gettext('Files repository'),
                    icon: 'c8y-archive',
                    path: 'files',
                    routerLinkExact: false,
                    priority: 1000
                }));
            }
        }
        return this.navs;
    }
}
FilesRepositoryNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryNavigationFactory, deps: [{ token: i1.FilesRepositoryGuard }], target: i0.ɵɵFactoryTarget.Injectable });
FilesRepositoryNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryNavigationFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilesRepositoryNavigationFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.FilesRepositoryGuard }]; } });
