// @c8y/ngx-components/esm2020/child-devices/child-devices.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { ChildDevicesComponent } from './child-devices.component';
import { ChildDevicesTabGuard } from './child-devices.guard';
import * as i0 from "@angular/core";
export class ChildDevicesModule {
}
ChildDevicesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ChildDevicesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ChildDevicesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ChildDevicesModule, declarations: [ChildDevicesComponent], imports: [CoreModule, DeviceGridModule, PopoverModule] });
ChildDevicesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ChildDevicesModule, providers: [
        ChildDevicesTabGuard,
        hookRoute({
            context: ViewContext.Device,
            path: 'child-devices',
            priority: 1000,
            icon: 'code-fork',
            label: gettext('Child devices'),
            component: ChildDevicesComponent,
            canActivate: [ChildDevicesTabGuard]
        })
    ], imports: [CoreModule, DeviceGridModule, PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ChildDevicesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ChildDevicesComponent],
                    imports: [CoreModule, DeviceGridModule, PopoverModule],
                    entryComponents: [ChildDevicesComponent],
                    providers: [
                        ChildDevicesTabGuard,
                        hookRoute({
                            context: ViewContext.Device,
                            path: 'child-devices',
                            priority: 1000,
                            icon: 'code-fork',
                            label: gettext('Child devices'),
                            component: ChildDevicesComponent,
                            canActivate: [ChildDevicesTabGuard]
                        })
                    ]
                }]
        }] });
