// @c8y/ngx-components/esm2020/operations/bulk-operation-scheduler/bulk-operation-scheduler.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule } from '@c8y/ngx-components';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { TimepickerModule } from 'ngx-bootstrap/timepicker';
import { OperationSchedulerComponent } from './operation-scheduler.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/datepicker";
import * as i2 from "ngx-bootstrap/timepicker";
/**
 * This module provides components for scheduling bulk operations.
 */
export class BulkOperationSchedulerModule {
}
BulkOperationSchedulerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationSchedulerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BulkOperationSchedulerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationSchedulerModule, declarations: [OperationSchedulerComponent], imports: [CoreModule,
        ReactiveFormsModule, i1.BsDatepickerModule, i2.TimepickerModule], exports: [OperationSchedulerComponent] });
BulkOperationSchedulerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationSchedulerModule, imports: [CoreModule,
        ReactiveFormsModule,
        BsDatepickerModule.forRoot(),
        TimepickerModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationSchedulerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        ReactiveFormsModule,
                        BsDatepickerModule.forRoot(),
                        TimepickerModule.forRoot()
                    ],
                    providers: [],
                    declarations: [OperationSchedulerComponent],
                    entryComponents: [OperationSchedulerComponent],
                    exports: [OperationSchedulerComponent]
                }]
        }] });
