// @c8y/ngx-components/esm2020/operations/stepper-bulk-type-configuration/stepper-bulk-type-configuration.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { baseUrl, BulkOperationType, hookListBulkType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperFramesModule } from '@c8y/ngx-components/operations/stepper-frames';
import { StepperBulkTypeConfigurationComponent } from './stepper-bulk-type-configuration.component';
import * as i0 from "@angular/core";
/** Module for the 'Configuration update' operation type stepper */
export class StepperBulkTypeConfigurationModule {
}
StepperBulkTypeConfigurationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeConfigurationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperBulkTypeConfigurationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeConfigurationModule, declarations: [StepperBulkTypeConfigurationComponent], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        SharedRepositoryModule,
        StepperFramesModule], exports: [StepperBulkTypeConfigurationComponent] });
StepperBulkTypeConfigurationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeConfigurationModule, providers: [
        hookListBulkType({
            type: BulkOperationType.CONFIGURATION,
            c8yIcon: 'cogs',
            name: gettext('Configuration update'),
            path: `${baseUrl}configuration`,
            component: StepperBulkTypeConfigurationComponent,
            fragments: ['c8y_DownloadConfigFile', 'c8y_Configuration'],
            selected: false
        })
    ], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        SharedRepositoryModule,
        StepperFramesModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeConfigurationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [StepperBulkTypeConfigurationComponent],
                    imports: [
                        CoreModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BulkOperationStepperModule,
                        SharedRepositoryModule,
                        StepperFramesModule
                    ],
                    providers: [
                        hookListBulkType({
                            type: BulkOperationType.CONFIGURATION,
                            c8yIcon: 'cogs',
                            name: gettext('Configuration update'),
                            path: `${baseUrl}configuration`,
                            component: StepperBulkTypeConfigurationComponent,
                            fragments: ['c8y_DownloadConfigFile', 'c8y_Configuration'],
                            selected: false
                        })
                    ],
                    entryComponents: [StepperBulkTypeConfigurationComponent],
                    exports: [StepperBulkTypeConfigurationComponent]
                }]
        }] });
