// @c8y/ngx-components/esm2020/core/dynamic-forms/validation/validation.module.mjs
import { NgModule } from '@angular/core';
import { FormlyModule, FORMLY_CONFIG } from '@ngx-formly/core';
import { TranslateService } from '@ngx-translate/core';
import { C8yTranslateModule } from '../../i18n/c8y-translate.module';
import { c8yValidationConfigFactory } from './validation-config.factory';
import * as i0 from "@angular/core";
export class ValidationModule {
}
ValidationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ValidationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ValidationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ValidationModule, imports: [C8yTranslateModule, FormlyModule] });
ValidationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ValidationModule, providers: [
        {
            provide: FORMLY_CONFIG,
            multi: true,
            useFactory: c8yValidationConfigFactory,
            deps: [TranslateService]
        }
    ], imports: [C8yTranslateModule, FormlyModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ValidationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [C8yTranslateModule, FormlyModule],
                    providers: [
                        {
                            provide: FORMLY_CONFIG,
                            multi: true,
                            useFactory: c8yValidationConfigFactory,
                            deps: [TranslateService]
                        }
                    ]
                }]
        }] });
