// @c8y/ngx-components/esm2020/auth-configuration/sso-configuration/template-parts/external-token-config.model.mjs
import { ValidationMethod } from '../sso-configuration.model';
import { gettext } from '@c8y/ngx-components';
import { RequestConfiguration } from './request-configuration.model';
export const validationMethodConfig = {
    [ValidationMethod.INTROSPECTION]: {
        name: 'INTROSPECTION',
        value: 'INTROSPECTION',
        label: gettext('Introspection`Method of validating access token from external IAM system`'),
        defaults: {
            method: 'POST',
            body: 'token=${accessToken}&client_id=${clientId}&client_secret=',
            url: '',
            headers: {},
            operation: 'EXECUTE',
            requestParams: {}
        }
    },
    [ValidationMethod.USERINFO]: {
        name: 'USERINFO',
        value: 'USERINFO',
        label: gettext('User info`Method of validating access token from external IAM system`'),
        defaults: {
            url: '',
            method: 'GET',
            body: '',
            headers: { Authorization: 'Bearer ${accessToken}' },
            operation: 'EXECUTE',
            requestParams: {}
        }
    }
};
const defaultTokenValidationRequest = {
    body: 'token=${accessToken}&client_id=${clientId}&client_secret=',
    url: '',
    headers: {},
    method: 'POST',
    operation: 'EXECUTE',
    requestParams: {}
};
const defaultUserIdConfig = {
    jwtField: '',
    useConstantValue: true
};
export class ExternalToken {
    constructor(externalTokenConfig) {
        const _externalTokenConfig = Object.assign({ enabled: false }, externalTokenConfig);
        this.userOrAppIdConfig = _externalTokenConfig.userOrAppIdConfig || defaultUserIdConfig;
        this.validationMethod = _externalTokenConfig.validationMethod || ValidationMethod.INTROSPECTION;
        this.validationRequired = _externalTokenConfig.validationRequired;
        this.enabled = _externalTokenConfig.enabled;
        this.tokenValidationRequest = new RequestConfiguration(_externalTokenConfig.tokenValidationRequest || defaultTokenValidationRequest);
        this.accessTokenValidityCheckIntervalInMinutes =
            _externalTokenConfig.accessTokenValidityCheckIntervalInMinutes || 1;
    }
    toExternalTokenConfig() {
        if (!this.enabled) {
            return { enabled: this.enabled };
        }
        const externalTokenConfig = {
            userOrAppIdConfig: this.userOrAppIdConfig,
            validationRequired: this.validationRequired,
            enabled: this.enabled
        };
        if (this.validationRequired) {
            externalTokenConfig.validationMethod = this.validationMethod;
            externalTokenConfig.tokenValidationRequest = this.tokenValidationRequest.toRequest();
            externalTokenConfig.accessTokenValidityCheckIntervalInMinutes =
                this.accessTokenValidityCheckIntervalInMinutes;
        }
        return externalTokenConfig;
    }
}
