// @c8y/ngx-components/esm2020/auth-configuration/sso-configuration/template-parts/request-configuration.model.mjs
import { map, pickBy, identity } from 'lodash-es';
export class RequestConfiguration {
    constructor(requestConfiguration) {
        this.headers = new RequestConfigurationDynamicArray(requestConfiguration.headers);
        this.method = requestConfiguration.method;
        this.operation = requestConfiguration.operation;
        this.requestParams = new RequestConfigurationDynamicArray(requestConfiguration.requestParams);
        this.body = requestConfiguration.body;
        this.url = requestConfiguration.url;
    }
    toRequest() {
        return pickBy({
            body: this.method === 'GET' ? '' : this.body,
            headers: this.headers.toObject(),
            method: this.method,
            operation: this.operation,
            requestParams: this.requestParams.toObject(),
            url: this.url
        }, identity);
    }
}
// tslint:disable-next-line:max-classes-per-file
export class RequestConfigurationDynamicArray extends Array {
    constructor(obj) {
        super();
        this.push(...map(obj, (value, key) => ({
            key,
            value
        })));
    }
    toObject() {
        return this.reduce((obj, element) => {
            obj[element.key] = element.value;
            return obj;
        }, {});
    }
}
