// @c8y/ngx-components/esm2020/device-shell/device-shell.module.mjs
import { NgModule } from '@angular/core';
import { gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { DeviceShellGuard } from './device-shell.guard';
import { DeviceShellComponent, ShellModule } from './shell';
import * as i0 from "@angular/core";
const DEVICE_SHELL_ROUTE = {
    path: 'shell',
    context: ViewContext.Device,
    component: DeviceShellComponent,
    label: gettext('Shell'),
    icon: 'terminal',
    canActivate: [DeviceShellGuard]
};
export class DeviceShellModule {
}
DeviceShellModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceShellModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellModule, imports: [ShellModule] });
DeviceShellModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellModule, providers: [DeviceShellGuard, hookRoute(DEVICE_SHELL_ROUTE)], imports: [ShellModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceShellModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [ShellModule],
                    providers: [DeviceShellGuard, hookRoute(DEVICE_SHELL_ROUTE)]
                }]
        }] });
