// @c8y/ngx-components/esm2020/widgets/definitions/map/index.mjs
import { DynamicManagedObjectResolver, gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';
// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
export const mapWidgetDefinition = {
    id: defaultWidgetIds.MAP,
    label: gettext('Map'),
    description: gettext('Shows the position of a device or all devices in a group.'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/map').then(m => m.MapWidgetComponent),
    loadConfigComponent: () => import('@c8y/ngx-components/widgets/implementations/map').then(m => m.MapWidgetConfigComponent),
    /*providers: [
  
    ],*/
    resolve: {
        device: DynamicManagedObjectResolver
    },
    previewImage: 'c8y-style-assets/map-widget-pr.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 4,
                _height: 4
            },
            ng1: {
                options: {
                    noDeviceTarget: false,
                    groupsSelectable: true,
                    deviceTargetNotRequired: true
                }
            }
        }
    }
};
