// @c8y/ngx-components/esm2020/core/application/application.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { AppHrefPipe } from './app-href.pipe';
import { AppIconComponent } from './app-icon.component';
import { AppSwitcherComponent } from './app-switcher.component';
import { IsActiveAppPipe } from './is-active-app.pipe';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { A11yModule } from '@angular/cdk/a11y';
import { ProductExperienceModule } from '../product-experience/product-experience.module';
import * as i0 from "@angular/core";
export class ApplicationModule {
}
ApplicationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ApplicationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ApplicationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ApplicationModule, declarations: [AppIconComponent, AppSwitcherComponent, AppHrefPipe, IsActiveAppPipe], imports: [CommonModule, BsDropdownModule, A11yModule, ProductExperienceModule], exports: [AppIconComponent, AppSwitcherComponent] });
ApplicationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ApplicationModule, imports: [CommonModule, BsDropdownModule, A11yModule, ProductExperienceModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ApplicationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, BsDropdownModule, A11yModule, ProductExperienceModule],
                    declarations: [AppIconComponent, AppSwitcherComponent, AppHrefPipe, IsActiveAppPipe],
                    exports: [AppIconComponent, AppSwitcherComponent]
                }]
        }] });
