// @c8y/ngx-components/esm2020/core/application/is-active-app.pipe.mjs
import { Pipe } from '@angular/core';
import { AppStateService } from '../common/ui-state.service';
import { map, distinctUntilChanged } from 'rxjs/operators';
import { get } from 'lodash-es';
import * as i0 from "@angular/core";
import * as i1 from "../common/ui-state.service";
export class IsActiveAppPipe {
    constructor(ui) {
        this.ui = ui;
    }
    transform(app) {
        const { contextPath } = app;
        return this.ui.state$.pipe(map(state => get(state, 'app.contextPath')), distinctUntilChanged(), map(currentContextPath => contextPath === currentContextPath));
    }
}
IsActiveAppPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: IsActiveAppPipe, deps: [{ token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Pipe });
IsActiveAppPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: IsActiveAppPipe, name: "isActiveApp" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: IsActiveAppPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'isActiveApp'
                }]
        }], ctorParameters: function () { return [{ type: i1.AppStateService }]; } });
