// @c8y/ngx-components/esm2020/repository/software/list/columns/software-type.grid-column.mjs
import { BaseColumn, gettext } from '@c8y/ngx-components';
import { SoftwareTypeCellRendererComponent } from './software-type.cell-renderer.component';
import { SoftwareTypeFilteringFormRendererComponent } from './software-type.filtering-form-renderer.component';
export class SoftwareTypeGridColumn extends BaseColumn {
    constructor(initialColumnConfig) {
        super(initialColumnConfig);
        this.name = 'type';
        this.path = 'softwareType';
        this.header = gettext('Software type');
        this.cellRendererComponent = SoftwareTypeCellRendererComponent;
        this.filterable = true;
        this.filteringFormRendererComponent = SoftwareTypeFilteringFormRendererComponent;
        this.filteringConfig = {
            getFilter(model) {
                const filter = {};
                if (model.softwareType) {
                    filter.softwareType = { __eq: model.softwareType };
                }
                return filter;
            }
        };
        this.sortable = true;
        this.sortingConfig = {
            pathSortingConfigs: [{ path: this.path }]
        };
    }
}
