// @c8y/ngx-components/esm2020/core/forms/default-validation.directive.mjs
import { NG_VALIDATORS } from '@angular/forms';
import { Directive, Input } from '@angular/core';
import { ValidationPattern } from './validation-pattern';
import * as i0 from "@angular/core";
export class DefaultValidationDirective {
    constructor() {
        this.c8yDefaultValidation = '';
    }
    validate(control) {
        if (!control.value) {
            return null;
        }
        const validation = ValidationPattern.get(this.c8yDefaultValidation);
        if (!validation) {
            throw new Error(`Pattern ${this.c8yDefaultValidation} not found.`);
        }
        let valid = true;
        if (validation.pattern) {
            valid = validation.pattern.test(control.value);
        }
        if (validation.maxLength) {
            valid = control.value.length < validation.maxLength;
        }
        if (validation.minLength) {
            valid = control.value.length >= validation.minLength;
        }
        return valid ? null : { [this.c8yDefaultValidation]: true };
    }
}
DefaultValidationDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultValidationDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
DefaultValidationDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: DefaultValidationDirective, selector: "[c8yDefaultValidation]", inputs: { c8yDefaultValidation: "c8yDefaultValidation" }, providers: [{ provide: NG_VALIDATORS, useExisting: DefaultValidationDirective, multi: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultValidationDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[c8yDefaultValidation]',
                    providers: [{ provide: NG_VALIDATORS, useExisting: DefaultValidationDirective, multi: true }]
                }]
        }], propDecorators: { c8yDefaultValidation: [{
                type: Input
            }] } });
