// @c8y/ngx-components/esm2020/core/forms/max-validation.directive.mjs
import { NG_VALIDATORS } from '@angular/forms';
import { Directive, Input } from '@angular/core';
import { isUndefined, isFinite } from 'lodash-es';
import * as i0 from "@angular/core";
export class MaxValidationDirective {
    constructor() {
        this.max = '';
    }
    validate(control) {
        const parsedMax = typeof this.max === 'number' ? this.max : parseFloat(this.max);
        if (isUndefined(control.value) || !isFinite(parsedMax)) {
            return null;
        }
        const valid = !(control.value > parsedMax);
        return valid ? null : { max: { max: parsedMax, actual: control.value } };
    }
}
MaxValidationDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MaxValidationDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
MaxValidationDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: MaxValidationDirective, selector: "[max]", inputs: { max: "max" }, providers: [{ provide: NG_VALIDATORS, useExisting: MaxValidationDirective, multi: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MaxValidationDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[max]',
                    providers: [{ provide: NG_VALIDATORS, useExisting: MaxValidationDirective, multi: true }]
                }]
        }], propDecorators: { max: [{
                type: Input
            }] } });
