// @c8y/ngx-components/esm2020/core/forms/messages.component.mjs
import { Component, ContentChildren, Input, QueryList } from '@angular/core';
import { MessageDirective } from './message.directive';
import { ValidationPattern } from './validation-pattern';
import { TranslateService } from '@ngx-translate/core';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "@angular/common";
import * as i3 from "../i18n/c8y-translate.pipe";
export class MessagesComponent {
    constructor(translateService) {
        this.translateService = translateService;
        this.show = {};
        this.defaults = ValidationPattern.messages;
        this.messages = [];
        this.messageMap = {};
    }
    ngAfterContentInit() {
        this.messageMap = Object.assign({}, this.defaults, this.customMessages.reduce((messages, directive) => {
            messages[directive.name] = directive.text;
            return messages;
        }, {}));
    }
    changeVisibility(show) {
        this.show = show;
        this.ngOnChanges();
    }
    ngOnChanges() {
        if (this.show) {
            this.messages = Object.keys(this.show)
                .map(key => this.toMessage(key, this.show[key]))
                // if an null message is returned, it will not be displayed
                .filter(message => message !== null);
        }
        else {
            this.messages = [];
        }
    }
    toMessage(key, value) {
        let message = this.messageMap[key];
        if ((!message || message === ValidationPattern.messages.pattern) &&
            key === 'pattern' &&
            typeof value?.requiredPattern === 'string') {
            const pattern = value.requiredPattern.replace(/^\//, '').replace(/\/$/, '');
            const foundPattern = Object.keys(ValidationPattern.rules).find(tmp => ValidationPattern.rules[tmp].pattern?.source === pattern);
            if (foundPattern) {
                key = foundPattern;
                message = this.messageMap[key];
            }
        }
        if (!message) {
            return null;
        }
        return this.translateService.instant(message, value);
    }
}
MessagesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MessagesComponent, deps: [{ token: i1.TranslateService }], target: i0.ɵɵFactoryTarget.Component });
MessagesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: MessagesComponent, selector: "c8y-messages", inputs: { show: "show", defaults: "defaults" }, queries: [{ propertyName: "customMessages", predicate: MessageDirective }], usesOnChanges: true, ngImport: i0, template: "<small class=\"form-control-feedback-message\">\n  <div class=\"d-block\" *ngFor=\"let msg of messages\">\n    {{msg | translate}}\n  </div>\n  <ng-content></ng-content>\n</small>\n", dependencies: [{ kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "pipe", type: i3.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MessagesComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-messages', template: "<small class=\"form-control-feedback-message\">\n  <div class=\"d-block\" *ngFor=\"let msg of messages\">\n    {{msg | translate}}\n  </div>\n  <ng-content></ng-content>\n</small>\n" }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }]; }, propDecorators: { show: [{
                type: Input
            }], defaults: [{
                type: Input
            }], customMessages: [{
                type: ContentChildren,
                args: [MessageDirective]
            }] } });
