// @c8y/ngx-components/esm2020/core/forms/min-validation.directive.mjs
import { NG_VALIDATORS } from '@angular/forms';
import { Directive, Input } from '@angular/core';
import { isUndefined, isFinite } from 'lodash-es';
import * as i0 from "@angular/core";
export class MinValidationDirective {
    constructor() {
        this.min = '';
    }
    validate(control) {
        const parsedMin = typeof this.min === 'number' ? this.min : parseFloat(this.min);
        if (isUndefined(control.value) || !isFinite(parsedMin)) {
            return null;
        }
        const valid = control.value !== null ? !(control.value < parsedMin) : true;
        return valid ? null : { min: { min: parsedMin, actual: control.value } };
    }
}
MinValidationDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MinValidationDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
MinValidationDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: MinValidationDirective, selector: "[min]", inputs: { min: "min" }, providers: [{ provide: NG_VALIDATORS, useExisting: MinValidationDirective, multi: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MinValidationDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[min]',
                    providers: [{ provide: NG_VALIDATORS, useExisting: MinValidationDirective, multi: true }]
                }]
        }], propDecorators: { min: [{
                type: Input
            }] } });
