// @c8y/ngx-components/esm2020/core/forms/phone-validation.directive.mjs
import { NG_VALIDATORS } from '@angular/forms';
import { Directive } from '@angular/core';
import { parsePhoneNumberFromString } from 'libphonenumber-js/max';
import * as i0 from "@angular/core";
export class PhoneValidationDirective {
    constructor() {
        this.allowedCharactersPattern = /^[/\d\-+ ]*$/;
    }
    validate(control) {
        const valid = this.validateInternationalPhoneNumber(control);
        return valid ? null : { internationalPhoneNumber: true };
    }
    validateInternationalPhoneNumber(control) {
        let numberValid;
        try {
            const phoneNumber = parsePhoneNumberFromString(control.value);
            numberValid = this.allowedCharactersPattern.test(control.value) && phoneNumber.isValid();
            if (numberValid && control.value !== phoneNumber.format('E.164')) {
                control.setValue(phoneNumber.format('E.164'), {
                    emitEvent: false,
                    emitModelToViewChange: false
                });
            }
        }
        catch (e) {
            numberValid = !control.value;
        }
        return numberValid;
    }
}
PhoneValidationDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PhoneValidationDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
PhoneValidationDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: PhoneValidationDirective, selector: "[c8yPhoneValidation]", providers: [{ provide: NG_VALIDATORS, useExisting: PhoneValidationDirective, multi: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PhoneValidationDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[c8yPhoneValidation]',
                    providers: [{ provide: NG_VALIDATORS, useExisting: PhoneValidationDirective, multi: true }]
                }]
        }] });
