// @c8y/ngx-components/esm2020/core/forms/range.component.mjs
import { Component, ContentChild, forwardRef, TemplateRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { toNumber } from 'lodash-es';
import { RangeDirective } from './range.directive';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
/**
 * An component which renders a range slider in Cumulocity layout. You need
 * to wrap the default range input in this component.
 *
 * Example:
 * ```
 * <c8y-range>
 *   <input id="range" type="range" min="0" max="100000" value="0" step="1">
 * </c8y-range>
 * ```
 */
export class RangeComponent {
    /**
     * @ignore
     * @param obj The value to update
     */
    writeValue(obj) {
        this.range.elementRef.nativeElement.value = obj;
        this.updateValue();
    }
    /**
     * @ignore
     * @param fn The function to register for changes
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @ignore
     * @param fn The function to register for changes
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @ignore
     */
    ngAfterContentInit() {
        this.updateValue();
        this.range.elementRef.nativeElement.addEventListener('input', () => this.updateValue());
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        this.range.elementRef.nativeElement.removeEventListener('input', () => this.updateValue());
    }
    /**
     * @ignore
     */
    updateValue() {
        const rangeValue = this.range.elementRef.nativeElement.value;
        this.value = toNumber(rangeValue);
        if (this.onChange) {
            this.onChange(this.value);
        }
        this.updatePosition();
    }
    /**
     * Updates the position of the slider
     */
    updatePosition() {
        const range = this.range.elementRef.nativeElement;
        const newValue = Number(((range.value - range.min) * 100) / (range.max - range.min));
        const newPosition = 12 - newValue * 0.24;
        this.left = `calc(${newValue}% + (${newPosition}px))`;
    }
}
RangeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RangeComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
RangeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: RangeComponent, selector: "c8y-range", providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => RangeComponent),
            multi: true
        }
    ], queries: [{ propertyName: "range", first: true, predicate: RangeDirective, descendants: true, static: true }, { propertyName: "rangeValue", first: true, predicate: ["c8yRangeValue"], descendants: true, static: true }], ngImport: i0, template: "<div class=\"range-wrap\" style=\"--track-width: {{left}}\">\n  <div class=\"range-value\" [style.left]=\"left\">\n    <span *ngIf=\"!rangeValue\">{{ value }}</span>\n    <ng-container *ngTemplateOutlet=\"rangeValue\"></ng-container>\n  </div>\n  <ng-content></ng-content>\n</div>\n", dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i1.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RangeComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-range', providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => RangeComponent),
                            multi: true
                        }
                    ], template: "<div class=\"range-wrap\" style=\"--track-width: {{left}}\">\n  <div class=\"range-value\" [style.left]=\"left\">\n    <span *ngIf=\"!rangeValue\">{{ value }}</span>\n    <ng-container *ngTemplateOutlet=\"rangeValue\"></ng-container>\n  </div>\n  <ng-content></ng-content>\n</div>\n" }]
        }], propDecorators: { range: [{
                type: ContentChild,
                args: [RangeDirective, { static: true }]
            }], rangeValue: [{
                type: ContentChild,
                args: ['c8yRangeValue', { static: true }]
            }] } });
