// @c8y/ngx-components/esm2020/core/forms/required-input-placeholder.directive.mjs
import { Directive, ElementRef, Optional, Renderer2 } from '@angular/core';
import { NgControl, Validators } from '@angular/forms';
import { TranslateService } from '@ngx-translate/core';
import { gettext } from '../i18n/gettext';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "@angular/forms";
export class RequiredInputPlaceholderDirective {
    constructor(el, renderer, translationService, ngControl) {
        this.el = el;
        this.renderer = renderer;
        this.translationService = translationService;
        this.ngControl = ngControl;
        this.shouldSkipRequiredHint = false;
        this.shouldSkipRequiredHint = !!this.el.nativeElement.closest('[placeholder-no-required-hint]');
    }
    ngAfterViewChecked() {
        if (this.el?.nativeElement?.required !== undefined || this.ngControl?.control) {
            const isRequired = this.el?.nativeElement?.required ||
                this.ngControl?.control?.hasValidator(Validators.required);
            const willSetRequiredHint = isRequired && !this.shouldSkipRequiredHint;
            const placeholder = this.el.nativeElement.placeholder;
            if (this.placeholder !== placeholder || this.req !== willSetRequiredHint) {
                this.updatePlaceholder(this.getTextForPlaceholder(placeholder, willSetRequiredHint));
            }
        }
    }
    getTextForPlaceholder(placeholder = '', required = false) {
        this.req = required;
        const requiredText = `(${this.translationService.instant(gettext('required'))})`;
        const requiredTextIndex = placeholder.lastIndexOf(requiredText);
        if (requiredTextIndex > 0) {
            placeholder = placeholder.slice(0, requiredTextIndex - 1);
        }
        if (requiredTextIndex === 0) {
            placeholder = '';
        }
        if (!required) {
            return placeholder;
        }
        return placeholder.trim() ? `${placeholder} ${requiredText}` : requiredText;
    }
    updatePlaceholder(placeholder) {
        this.renderer.setAttribute(this.el.nativeElement, 'placeholder', placeholder);
        this.placeholder = placeholder;
    }
}
RequiredInputPlaceholderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RequiredInputPlaceholderDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.TranslateService }, { token: i2.NgControl, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
RequiredInputPlaceholderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: RequiredInputPlaceholderDirective, selector: "input[required], input[formControlName]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RequiredInputPlaceholderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[required], input[formControlName]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.TranslateService }, { type: i2.NgControl, decorators: [{
                    type: Optional
                }] }]; } });
