// @c8y/ngx-components/esm2020/core/forms/simple-json-path-validator.directive.mjs
import { Directive } from '@angular/core';
import { NG_VALIDATORS, Validators } from '@angular/forms';
import * as i0 from "@angular/core";
const SIMPLE_JSON_PATH_REGEX = /^[^.\s]+(\.\S+)*$/;
export function simpleJsonPathValidator(control) {
    const errors = Validators.pattern(SIMPLE_JSON_PATH_REGEX)(control);
    return !!errors
        ? {
            simpleJsonPath: errors.pattern
        }
        : null;
}
export class SimpleJsonPathValidatorDirective {
    validate(control) {
        return simpleJsonPathValidator(control);
    }
}
SimpleJsonPathValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SimpleJsonPathValidatorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
SimpleJsonPathValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: SimpleJsonPathValidatorDirective, selector: "[simpleJsonPath]", providers: [
        {
            provide: NG_VALIDATORS,
            useExisting: SimpleJsonPathValidatorDirective,
            multi: true
        }
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SimpleJsonPathValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[simpleJsonPath]',
                    providers: [
                        {
                            provide: NG_VALIDATORS,
                            useExisting: SimpleJsonPathValidatorDirective,
                            multi: true
                        }
                    ]
                }]
        }] });
