// @c8y/ngx-components/esm2020/core/forms/text-area-row-height.directive.mjs
import { Directive, ElementRef, Input, HostBinding, Renderer2 } from '@angular/core';
import * as i0 from "@angular/core";
export class TextAreaRowHeightDirective {
    constructor(elem, renderer) {
        this.elem = elem;
        this.renderer = renderer;
        this.rows = 1;
        this.minHeightPerRow = 20;
    }
    ngAfterViewInit() {
        const convertedValue = Number(this.heightPerRow);
        if (this.heightPerRow && convertedValue && convertedValue > this.minHeightPerRow) {
            this.setHeight(`${this.rows * convertedValue}px`);
        }
        else {
            this.setHeight(`${this.rows * this.minHeightPerRow}px`);
        }
    }
    setHeight(value) {
        this.renderer.setStyle(this.elem.nativeElement, 'height', value);
    }
}
TextAreaRowHeightDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TextAreaRowHeightDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
TextAreaRowHeightDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: TextAreaRowHeightDirective, selector: "textarea[heightPerRow]", inputs: { rows: "rows", heightPerRow: "heightPerRow" }, host: { properties: { "rows": "this.rows" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TextAreaRowHeightDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'textarea[heightPerRow]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { rows: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['rows']
            }], heightPerRow: [{
                type: Input,
                args: ['heightPerRow']
            }] } });
