// @c8y/ngx-components/esm2020/core/forms/unique-in-collection-by-path-validation.directive.mjs
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { find, get, isEqualWith } from 'lodash-es';
import * as i0 from "@angular/core";
export function uniqueInCollectionByPathValidator(collection, path, comparator) {
    return (control) => {
        const match = find(collection, element => isEqualWith(get(element, path), control.value, comparator));
        return !!match && !!control.value
            ? { uniqueByPath: { key: path, value: get(match, path) } }
            : null;
    };
}
export class UniqueInCollectionByPathValidationDirective {
    validate(control) {
        return uniqueInCollectionByPathValidator(this.collection, this.path, this.comparator)(control);
    }
}
UniqueInCollectionByPathValidationDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UniqueInCollectionByPathValidationDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
UniqueInCollectionByPathValidationDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: UniqueInCollectionByPathValidationDirective, selector: "[uniqueByPath]", inputs: { collection: "collection", path: "path", comparator: "comparator" }, providers: [
        {
            provide: NG_VALIDATORS,
            useExisting: UniqueInCollectionByPathValidationDirective,
            multi: true
        }
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UniqueInCollectionByPathValidationDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[uniqueByPath]',
                    providers: [
                        {
                            provide: NG_VALIDATORS,
                            useExisting: UniqueInCollectionByPathValidationDirective,
                            multi: true
                        }
                    ]
                }]
        }], propDecorators: { collection: [{
                type: Input
            }], path: [{
                type: Input
            }], comparator: [{
                type: Input
            }] } });
