// @c8y/ngx-components/esm2020/widgets/definitions/three-d-rotation/index.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';
// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
export const threeDRotationWidgetDefinition = {
    id: defaultWidgetIds.THREE_D_ROTATION,
    label: gettext('Rotation widget'),
    description: gettext('Displays current device orientation'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/three-d-rotation').then(m => m.ThreeDRotationWidgetViewComponent),
    loadConfigComponent: () => import('@c8y/ngx-components/widgets/implementations/three-d-rotation').then(m => m.ThreeDRotationWidgetConfigComponent),
    previewImage: 'c8y-style-assets/rotation-widget-pr.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 4,
                _height: 4
            },
            ng1: {
                options: {
                    noDeviceTarget: false,
                    groupsSelectable: false
                }
            }
        }
    }
};
