// @c8y/ngx-components/esm2020/core/router/context-route.component.mjs
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { ContextRouteService } from './context-route.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./context-route.service";
export class ContextRouteComponent {
    constructor(route, contextRouteService) {
        this.route = route;
        this.contextRouteService = contextRouteService;
    }
    ngOnInit() {
        this.contextRouteService.init(this.route);
    }
    ngOnDestroy() {
        this.contextRouteService.destroy();
    }
    refreshTabs() {
        this.contextRouteService.refreshContext();
    }
}
ContextRouteComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ContextRouteComponent, deps: [{ token: i1.ActivatedRoute }, { token: i2.ContextRouteService }], target: i0.ɵɵFactoryTarget.Component });
ContextRouteComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ContextRouteComponent, selector: "c8y-context-route", ngImport: i0, template: "<router-outlet></router-outlet>\n", dependencies: [{ kind: "directive", type: i1.RouterOutlet, selector: "router-outlet", inputs: ["name"], outputs: ["activate", "deactivate", "attach", "detach"], exportAs: ["outlet"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ContextRouteComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-context-route', template: "<router-outlet></router-outlet>\n" }]
        }], ctorParameters: function () { return [{ type: i1.ActivatedRoute }, { type: i2.ContextRouteService }]; } });
