// @c8y/ngx-components/esm2020/core/router/router.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule as NgRouterModule } from '@angular/router';
import { RouterService } from './router.service';
import { ContextRouteComponent } from './context-route.component';
import { HeaderModule } from '../header/header.module';
import { CommonModule } from '../common/common.module';
import { ContextRouteGuard } from './context-route.guard';
import { ViewContext } from './router.model';
import { RouterTabsResolver } from './router-tabs.resolver';
import { EmptyComponent } from './empty.component';
import * as i0 from "@angular/core";
import * as i1 from "./router.service";
export const extraRoutes = [
    {
        path: 'users/new',
        component: EmptyComponent
    },
    {
        path: 'tenants/new',
        component: EmptyComponent
    }
];
// tslint:disable-next-line: forin
for (const key in ViewContext) {
    const context = ViewContext[key];
    extraRoutes.push({
        path: context,
        component: ContextRouteComponent,
        canActivate: [ContextRouteGuard],
        data: { context, contextData: {} },
        resolve: {
            tabs: RouterTabsResolver
            // contextData: ContextRouteGuard
        }
    });
}
/**
 * A C8Y router implementation. It implements some default routes needed
 * as well as context routes. Context routes are useful, if you want to
 * use context data on a certain view (e.g. a device), then this routes
 * automatically resolve the associated managed object.
 *
 * You can use this in combination of the Angular default RouterModule
 * or use only this module. Context routes can only be added via the
 * HOOK_ROUTES hook.
 */
export class RouterModule {
    /**
     * Same interface as @angular/router module.
     * Sets `initialNavigation` always to `disabled` and `useHash` always to `true`.
     * @param routes The configured routes.
     * @param config The router default configuration.
     * @returns The module and providers.
     */
    static forRoot(routes = [], config) {
        if (!config) {
            config = {};
        }
        // initialNavigation disabled by default because of MTM-44047 (routes triggered before the actual login)
        config.initialNavigation = config.initialNavigation ?? 'disabled';
        config.useHash = config.useHash ?? true;
        const ngRouterModule = NgRouterModule.forRoot([...routes, ...extraRoutes], config);
        return {
            ngModule: RouterModule,
            providers: [...ngRouterModule.providers]
        };
    }
    constructor(
    // Declared but never used is intended: to always have a instance of RouterService.
    routerService) {
        this.routerService = routerService;
    }
}
RouterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RouterModule, deps: [{ token: i1.RouterService }], target: i0.ɵɵFactoryTarget.NgModule });
RouterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: RouterModule, declarations: [ContextRouteComponent, EmptyComponent], imports: [NgRouterModule, HeaderModule, CommonModule] });
RouterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RouterModule, imports: [NgRouterModule, HeaderModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RouterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ContextRouteComponent, EmptyComponent],
                    imports: [NgRouterModule, HeaderModule, CommonModule]
                }]
        }], ctorParameters: function () { return [{ type: i1.RouterService }]; } });
