// @c8y/ngx-components/esm2020/repository/software/list/software-repository-navigation-factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
export class SoftwareRepositoryNavigationFactory {
    constructor() {
        this.node = new NavigatorNode({
            label: gettext('Software repository'),
            path: 'software',
            icon: 'c8y-tools',
            parent: gettext('Management'),
            priority: 900
        });
    }
    get() {
        return this.node;
    }
}
SoftwareRepositoryNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryNavigationFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
SoftwareRepositoryNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
