// @c8y/ngx-components/esm2020/repository/software/list/software-repository-list.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NameCellRendererComponent } from './columns/name.cell-renderer.component';
import { SoftwareTypeCellRendererComponent } from './columns/software-type.cell-renderer.component';
import { SoftwareTypeFilteringFormRendererComponent } from './columns/software-type.filtering-form-renderer.component';
import { DeviceTypeCellRendererComponent } from './columns/device-type.cell-renderer.component';
import { VersionsCellRendererComponent } from './columns/versions.cell-renderer.component';
import { SoftwareListComponent } from './software-list.component';
import { SoftwareDetailsComponent } from './software-details.component';
import { AddSoftwareModalComponent } from './add-software-modal.component';
import { CoreModule, FormsModule, hookNavigator, hookRoute } from '@c8y/ngx-components';
import { SoftwareRepositoryNavigationFactory } from './software-repository-navigation-factory';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { RouterModule } from '@angular/router';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import * as i0 from "@angular/core";
export class SoftwareRepositoryListModule {
    static forRoot() {
        return {
            ngModule: SoftwareRepositoryListModule,
            providers: [
                hookNavigator(SoftwareRepositoryNavigationFactory),
                hookRoute([
                    {
                        path: 'software',
                        component: SoftwareListComponent
                    },
                    {
                        path: 'software/:id',
                        component: SoftwareDetailsComponent
                    }
                ])
            ]
        };
    }
}
SoftwareRepositoryListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SoftwareRepositoryListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryListModule, declarations: [SoftwareListComponent,
        SoftwareDetailsComponent,
        AddSoftwareModalComponent,
        NameCellRendererComponent,
        SoftwareTypeCellRendererComponent,
        SoftwareTypeFilteringFormRendererComponent,
        DeviceTypeCellRendererComponent,
        VersionsCellRendererComponent], imports: [CommonModule,
        CoreModule,
        FormsModule,
        DeviceGridModule,
        PopoverModule,
        TooltipModule,
        RouterModule,
        SharedRepositoryModule] });
SoftwareRepositoryListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryListModule, imports: [CommonModule,
        CoreModule,
        FormsModule,
        DeviceGridModule,
        PopoverModule,
        TooltipModule,
        RouterModule,
        SharedRepositoryModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SoftwareRepositoryListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        CoreModule,
                        FormsModule,
                        DeviceGridModule,
                        PopoverModule,
                        TooltipModule,
                        RouterModule,
                        SharedRepositoryModule
                    ],
                    declarations: [
                        SoftwareListComponent,
                        SoftwareDetailsComponent,
                        AddSoftwareModalComponent,
                        NameCellRendererComponent,
                        SoftwareTypeCellRendererComponent,
                        SoftwareTypeFilteringFormRendererComponent,
                        DeviceTypeCellRendererComponent,
                        VersionsCellRendererComponent
                    ],
                    entryComponents: [SoftwareListComponent, SoftwareDetailsComponent, AddSoftwareModalComponent]
                }]
        }] });
