// @c8y/ngx-components/esm2020/context-dashboard/device-management-home-dashboard/device-management-home-dashboard.component.mjs
import { Component } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import { WidgetService } from '../widget.service';
import * as i0 from "@angular/core";
import * as i1 from "../widget.service";
import * as i2 from "@c8y/ngx-components";
import * as i3 from "../context-dashboard.component";
export class DeviceManagementHomeDashboardComponent {
    constructor(widgetService) {
        this.widgetService = widgetService;
        this.widgets = [];
        this.dashboardName = 'home-dm1';
        this.defaultWidgets = [
            {
                name: 'devicemanagement.welcome.widget',
                title: gettext('Quick links'),
                _x: 0,
                _y: 0,
                _height: 4,
                _width: 6
            },
            {
                name: 'Recent Alarms',
                title: gettext('Recent alarms'),
                _x: 6,
                _y: 0,
                _height: 4,
                _width: 6
            }
        ];
    }
    ngOnInit() {
        this.defaultWidgets.map(async (widget) => {
            await this.widgetService
                .mapLegacy(widget)
                .then(transformedWidget => this.widgets.push(transformedWidget));
        });
    }
}
DeviceManagementHomeDashboardComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceManagementHomeDashboardComponent, deps: [{ token: i1.WidgetService }], target: i0.ɵɵFactoryTarget.Component });
DeviceManagementHomeDashboardComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: DeviceManagementHomeDashboardComponent, selector: "c8y-device-management-home-dashboard", ngImport: i0, template: "<c8y-title>\n  {{ 'Home' | translate }}\n</c8y-title>\n\n<c8y-context-dashboard\n  [name]=\"dashboardName\"\n  [defaultWidgets]=\"defaultWidgets\"\n  [canDelete]=\"false\"\n></c8y-context-dashboard>\n", dependencies: [{ kind: "component", type: i2.TitleComponent, selector: "c8y-title", inputs: ["pageTitleUpdate"] }, { kind: "component", type: i3.ContextDashboardComponent, selector: "c8y-context-dashboard", inputs: ["name", "childrenClasses", "context", "setTitle", "disabled", "defaultWidgets", "canCopy", "canDelete", "isLoading", "breadcrumbSettings", "showContextHelpButton", "translateWidgetTitle"] }, { kind: "pipe", type: i2.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceManagementHomeDashboardComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-device-management-home-dashboard', template: "<c8y-title>\n  {{ 'Home' | translate }}\n</c8y-title>\n\n<c8y-context-dashboard\n  [name]=\"dashboardName\"\n  [defaultWidgets]=\"defaultWidgets\"\n  [canDelete]=\"false\"\n></c8y-context-dashboard>\n" }]
        }], ctorParameters: function () { return [{ type: i1.WidgetService }]; } });
