// @c8y/ngx-components/esm2020/context-dashboard/device-management-home-dashboard/device-management-home-dashboard.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CoreModule, gettext, hookNavigator, hookRoute } from '@c8y/ngx-components';
import { ContextDashboardModule } from '../context-dashboard.module';
import { DeviceManagementHomeDashboardComponent } from './device-management-home-dashboard.component';
import * as i0 from "@angular/core";
export class DeviceManagementHomeDashboardModule {
}
DeviceManagementHomeDashboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceManagementHomeDashboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceManagementHomeDashboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceManagementHomeDashboardModule, declarations: [DeviceManagementHomeDashboardComponent], imports: [CommonModule, CoreModule, ContextDashboardModule] });
DeviceManagementHomeDashboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceManagementHomeDashboardModule, providers: [
        hookRoute({
            path: '',
            component: DeviceManagementHomeDashboardComponent
        }),
        hookNavigator({ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' })
    ], imports: [CommonModule, CoreModule, ContextDashboardModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceManagementHomeDashboardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DeviceManagementHomeDashboardComponent],
                    imports: [CommonModule, CoreModule, ContextDashboardModule],
                    exports: [],
                    providers: [
                        hookRoute({
                            path: '',
                            component: DeviceManagementHomeDashboardComponent
                        }),
                        hookNavigator({ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' })
                    ]
                }]
        }] });
