// @c8y/ngx-components/esm2020/core/provider-configuration/service/provider-configuration.service.mjs
import { Injectable } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { FetchClient, Service } from '@c8y/client';
import { get } from 'lodash-es';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "@angular/router";
export class ProviderConfigurationService extends Service {
    constructor(client, activatedRoute) {
        super(client);
        this.baseUrl = get(activatedRoute, 'snapshot.data.endpoint.configurationEndpoint.baseUrl');
        this.listUrl = get(activatedRoute, 'snapshot.data.endpoint.configurationEndpoint.listUrl');
    }
    async detail() {
        return super.detail(ProviderConfigurationService.NO_DETAIL_ID);
    }
    async update(entity) {
        return super.update(entity);
    }
    async delete() {
        return super.delete(ProviderConfigurationService.NO_DETAIL_ID);
    }
    /** There is a single provider configuration per tenant and detail URL is not needed */
    getDetailUrl(_) {
        return this.listUrl;
    }
}
ProviderConfigurationService.NO_DETAIL_ID = null;
ProviderConfigurationService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationService, deps: [{ token: i1.FetchClient }, { token: i2.ActivatedRoute }], target: i0.ɵɵFactoryTarget.Injectable });
ProviderConfigurationService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.FetchClient }, { type: i2.ActivatedRoute }]; } });
