// @c8y/ngx-components/esm2020/core/dashboard/wiget-time-context/widget-time-context.model.mjs
import { gettext } from '../../i18n';
export var DateContextQueryParamNames;
(function (DateContextQueryParamNames) {
    DateContextQueryParamNames["DATE_CONTEXT_FROM"] = "dateContextFrom";
    DateContextQueryParamNames["DATE_CONTEXT_TO"] = "dateContextTo";
    DateContextQueryParamNames["DATE_CONTEXT_INTERVAL"] = "dateContextInterval";
    DateContextQueryParamNames["DATE_CONTEXT_REALTIME"] = "dateContextRealtime";
})(DateContextQueryParamNames || (DateContextQueryParamNames = {}));
const todayDate = new Date();
export var TimeSpanInMs;
(function (TimeSpanInMs) {
    TimeSpanInMs[TimeSpanInMs["MINUTE"] = 60000] = "MINUTE";
    TimeSpanInMs[TimeSpanInMs["HOUR"] = 3600000] = "HOUR";
    TimeSpanInMs[TimeSpanInMs["DAY"] = 86400000] = "DAY";
    TimeSpanInMs[TimeSpanInMs["WEEK"] = 604800000] = "WEEK";
    TimeSpanInMs[TimeSpanInMs["MONTH"] = todayDate.valueOf() - new Date(todayDate.setMonth(todayDate.getMonth() - 1)).valueOf()] = "MONTH";
})(TimeSpanInMs || (TimeSpanInMs = {}));
export const INTERVALS = [
    {
        id: 'minutes',
        title: gettext('Last minute'),
        timespanInMs: TimeSpanInMs.MINUTE
    },
    {
        id: 'hours',
        title: gettext('Last hour'),
        timespanInMs: TimeSpanInMs.HOUR
    },
    {
        id: 'days',
        title: gettext('Last day'),
        timespanInMs: TimeSpanInMs.DAY
    },
    {
        id: 'weeks',
        title: gettext('Last week'),
        timespanInMs: TimeSpanInMs.WEEK
    },
    {
        id: 'months',
        title: gettext('Last month'),
        timespanInMs: TimeSpanInMs.MONTH
    },
    { id: 'custom', title: gettext('Custom') }
];
