// @c8y/ngx-components/esm2020/core/dashboard/wiget-time-context/widget-time-context-helper.service.mjs
import { Injectable } from '@angular/core';
import { INTERVALS } from './widget-time-context.model';
import * as i0 from "@angular/core";
export class WidgetTimeContextHelperService {
    /**
     * Validates provided param for being selectable Interval id.
     * @param intervalId Interval id to be validated.
     * @returns True if provided id is valid, selectable Interval id and false if it's not.
     */
    isSelectableInterval(intervalId) {
        if (intervalId === 'custom') {
            return false;
        }
        return INTERVALS.some(({ id }) => id === intervalId);
    }
    /**
     * Validates provided date "from" and date "to":
     * - if both dates are proper date strings
     * - if provided date from is earlier than date to.
     * @param stringifiedDateFrom Date "from" that should be validated.
     * @param stringifiedDateTo Date "to" that should be validated.
     * @returns Result of validation of dates range.
     */
    isValidDateRange(stringifiedDateFrom, stringifiedDateTo) {
        const dateFrom = new Date(stringifiedDateFrom);
        const dateTo = new Date(stringifiedDateTo);
        if (dateFrom instanceof Date &&
            !isNaN(dateFrom?.getTime()) &&
            dateTo instanceof Date &&
            !isNaN(dateTo?.getTime())) {
            return dateFrom < dateTo;
        }
        return false;
    }
    /**
     * Calculates date time context according to provided interval.
     * @param intervalId Interval id indicating time range.
     * @returns Tuple of dates- the first one is date "from" according to selected interval, second one is date "to" (now).
     */
    getDateTimeContextByInterval(intervalId) {
        const interval = INTERVALS.find(({ id }) => id === intervalId);
        const dateTo = new Date();
        const dateFrom = new Date(dateTo.valueOf() - interval.timespanInMs);
        return [dateFrom, dateTo];
    }
}
WidgetTimeContextHelperService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetTimeContextHelperService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
WidgetTimeContextHelperService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetTimeContextHelperService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetTimeContextHelperService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
