// @c8y/ngx-components/esm2020/tenants/tenants-navigation.factory.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { TenantListGuard } from './tenant-list/tenant-list.guard';
import { TENANTS_MODULE_CONFIG } from './tenants.model';
import * as i0 from "@angular/core";
import * as i1 from "./tenant-list/tenant-list.guard";
export class TenantsNavigationFactory {
    constructor(tenantListGuard, config) {
        this.tenantListGuard = tenantListGuard;
        this.config = config;
        this.navs = [];
    }
    async get() {
        const canActivateTenantList = await this.tenantListGuard.canActivate();
        if (!this.navs.length) {
            const subtenantsNavigatorNode = this.config?.subtenantsNavigatorNode ?? true;
            if (subtenantsNavigatorNode !== false) {
                this.navs.push(new NavigatorNode({
                    parent: {
                        label: gettext('Tenants'),
                        icon: 'c8y-layers'
                    },
                    label: gettext('Subtenants'),
                    icon: 'c8y-sub-tenants',
                    path: 'tenants',
                    routerLinkExact: false,
                    priority: 4000,
                    hidden: !canActivateTenantList,
                    ...(subtenantsNavigatorNode === true ? {} : subtenantsNavigatorNode)
                }));
            }
        }
        return this.navs;
    }
}
TenantsNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TenantsNavigationFactory, deps: [{ token: i1.TenantListGuard }, { token: TENANTS_MODULE_CONFIG, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
TenantsNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TenantsNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TenantsNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.TenantListGuard }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TENANTS_MODULE_CONFIG]
                }] }]; } });
