// @c8y/ngx-components/esm2020/tenants/tenants.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { CoreModule, hookNavigator, hookRoute } from '@c8y/ngx-components';
import { TenantsNavigationFactory } from './tenants-navigation.factory';
import { TenantListComponent } from './tenant-list/tenant-list.component';
import { CreationTimeFilteringFormRendererComponent } from './tenant-list/creation-time.filtering-form-renderer.component';
import { StatusFilteringFormRendererComponent } from './tenant-list/status.filtering-form-renderer.component';
import { TENANTS_MODULE_CONFIG } from './tenants.model';
import { TenantListGuard } from './tenant-list/tenant-list.guard';
import * as i0 from "@angular/core";
export class TenantsModule {
    static config(config = {}) {
        return {
            ngModule: TenantsModule,
            providers: [
                {
                    provide: TENANTS_MODULE_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
TenantsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TenantsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TenantsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: TenantsModule, declarations: [TenantListComponent,
        CreationTimeFilteringFormRendererComponent,
        StatusFilteringFormRendererComponent], imports: [RouterModule, BsDatepickerModule, CoreModule] });
TenantsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TenantsModule, providers: [
        TenantListGuard,
        hookNavigator(TenantsNavigationFactory),
        hookRoute({
            path: 'tenants',
            icon: 'c8y-layers',
            component: TenantListComponent,
            canActivate: [TenantListGuard]
        })
    ], imports: [RouterModule, BsDatepickerModule, CoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TenantsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [RouterModule, BsDatepickerModule, CoreModule],
                    exports: [],
                    declarations: [
                        TenantListComponent,
                        CreationTimeFilteringFormRendererComponent,
                        StatusFilteringFormRendererComponent
                    ],
                    entryComponents: [
                        TenantListComponent,
                        CreationTimeFilteringFormRendererComponent,
                        StatusFilteringFormRendererComponent
                    ],
                    providers: [
                        TenantListGuard,
                        hookNavigator(TenantsNavigationFactory),
                        hookRoute({
                            path: 'tenants',
                            icon: 'c8y-layers',
                            component: TenantListComponent,
                            canActivate: [TenantListGuard]
                        })
                    ]
                }]
        }] });
