// @c8y/ngx-components/esm2020/operations/operations-list/operations-list.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { OperationsListItemDetailsModule } from '@c8y/ngx-components/operations/operations-list-item-details';
import { BulkOperationListItemService } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { OperationsSharedModule } from '@c8y/ngx-components/operations/shared';
import { OperationsListComponent } from './operations-list.component';
import { OperationsListItemComponent } from './operations-list-item.component';
import { OperationsListService } from './operations-list.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export const singleOperationListRoutes = [
    {
        path: 'devicecontrol/single',
        component: OperationsListComponent
    }
];
/**
 * This module allows for displaying the list item details of single and bulk operations.
 */
export class OperationsListModule {
}
OperationsListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OperationsListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: OperationsListModule, declarations: [OperationsListComponent, OperationsListItemComponent], imports: [CoreModule,
        StatusFilterModule,
        OperationsListItemDetailsModule,
        OperationsSharedModule,
        TooltipModule, i1.RouterModule], exports: [OperationsListComponent, OperationsListItemComponent] });
OperationsListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsListModule, providers: [
        OperationsListService,
        BulkOperationListItemService,
        hookRoute({
            context: ViewContext.Device,
            path: 'control',
            component: OperationsListComponent,
            label: gettext('Control'),
            icon: 'c8y-icon c8y-icon-device-control',
            priority: 600
        })
    ], imports: [CoreModule,
        StatusFilterModule,
        OperationsListItemDetailsModule,
        OperationsSharedModule,
        TooltipModule,
        RouterModule.forChild(singleOperationListRoutes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        StatusFilterModule,
                        OperationsListItemDetailsModule,
                        OperationsSharedModule,
                        TooltipModule,
                        RouterModule.forChild(singleOperationListRoutes)
                    ],
                    declarations: [OperationsListComponent, OperationsListItemComponent],
                    exports: [OperationsListComponent, OperationsListItemComponent],
                    providers: [
                        OperationsListService,
                        BulkOperationListItemService,
                        hookRoute({
                            context: ViewContext.Device,
                            path: 'control',
                            component: OperationsListComponent,
                            label: gettext('Control'),
                            icon: 'c8y-icon c8y-icon-device-control',
                            priority: 600
                        })
                    ]
                }]
        }] });
