// @c8y/ngx-components/esm2020/core/authentication/password-input.component.mjs
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../i18n/c8y-translate.pipe";
/**
 * The component is used to display a password control that includes the functionality of hiding and showing the input value.
 */
export class PasswordInputComponent {
    constructor() {
        /**
         * Id of input
         */
        this.id = '';
        /**
         * Value for autocomplete attribute of input
         */
        this.autocomplete = 'off';
        this.value = '';
        this.disabled = false;
        this.type = 'password';
        // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
        this.onChange = value => { };
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this.onTouched = () => { };
    }
    writeValue(value) {
        this.value = value ? value : '';
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    onInput($event) {
        this.value = $event.currentTarget.value;
        this.onChange(this.value);
    }
    onFocusOut() {
        this.onTouched();
    }
}
PasswordInputComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordInputComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
PasswordInputComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: PasswordInputComponent, selector: "c8y-password-input", inputs: { id: "id", autocomplete: "autocomplete" }, providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => PasswordInputComponent),
            multi: true
        }
    ], ngImport: i0, template: "<div class=\"input-group input-group-password\">\n  <input\n    class=\"form-control\"\n    [type]=\"type\"\n    [value]=\"value\"\n    [id]=\"id\"\n    (input)=\"onInput($event)\"\n    (focusout)=\"onFocusOut()\"\n    [disabled]=\"disabled\"\n    [autocomplete]=\"autocomplete\"\n  />\n  <span class=\"input-group-btn\">\n    <button\n      *ngIf=\"type === 'password'\"\n      class=\"btn btn-clean\"\n      title=\"{{ 'Show password' | translate }}\"\n      type=\"button\"\n      (click)=\"type = 'text'\"\n    >\n      <i class=\"dlt-c8y-icon-eye\"></i>\n    </button>\n    <button\n      *ngIf=\"type === 'text'\"\n      class=\"btn btn-clean\"\n      title=\"{{ 'Hide password' | translate }}\"\n      type=\"button\"\n      (click)=\"type = 'password'\"\n    >\n      <i class=\"dlt-c8y-icon-eye-slash\"></i>\n    </button>\n  </span>\n</div>\n", dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "pipe", type: i2.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordInputComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-password-input', providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => PasswordInputComponent),
                            multi: true
                        }
                    ], template: "<div class=\"input-group input-group-password\">\n  <input\n    class=\"form-control\"\n    [type]=\"type\"\n    [value]=\"value\"\n    [id]=\"id\"\n    (input)=\"onInput($event)\"\n    (focusout)=\"onFocusOut()\"\n    [disabled]=\"disabled\"\n    [autocomplete]=\"autocomplete\"\n  />\n  <span class=\"input-group-btn\">\n    <button\n      *ngIf=\"type === 'password'\"\n      class=\"btn btn-clean\"\n      title=\"{{ 'Show password' | translate }}\"\n      type=\"button\"\n      (click)=\"type = 'text'\"\n    >\n      <i class=\"dlt-c8y-icon-eye\"></i>\n    </button>\n    <button\n      *ngIf=\"type === 'text'\"\n      class=\"btn btn-clean\"\n      title=\"{{ 'Hide password' | translate }}\"\n      type=\"button\"\n      (click)=\"type = 'password'\"\n    >\n      <i class=\"dlt-c8y-icon-eye-slash\"></i>\n    </button>\n  </span>\n</div>\n" }]
        }], propDecorators: { id: [{
                type: Input
            }], autocomplete: [{
                type: Input
            }] } });
