// @c8y/ngx-components/esm2020/core/authentication/password-strength.component.mjs
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { PasswordService } from './password.service';
import * as i0 from "@angular/core";
import * as i1 from "./password.service";
import * as i2 from "../i18n/c8y-translate.directive";
import * as i3 from "../i18n/c8y-translate.pipe";
export class PasswordStrengthComponent {
    set password(password) {
        if (typeof password === 'string') {
            this.model = this.passwordStrengthCheckerService.getStrengthColor(password);
            this.strength.emit(this.model);
        }
    }
    constructor(passwordStrengthCheckerService) {
        this.passwordStrengthCheckerService = passwordStrengthCheckerService;
        this.strength = new EventEmitter();
        this.password = '';
    }
}
PasswordStrengthComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordStrengthComponent, deps: [{ token: i1.PasswordService }], target: i0.ɵɵFactoryTarget.Component });
PasswordStrengthComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: PasswordStrengthComponent, selector: "c8y-password-strength", inputs: { password: "password" }, outputs: { strength: "strength" }, ngImport: i0, template: "<div class=\"password-strength password-{{ model.colorName }}\">\r\n  <label class=\"password-strength-label small\" translate>\r\n    Password strength\r\n  </label>\r\n  <div>\r\n    <div class=\"password-bar\"></div>\r\n  </div>\r\n  <span class=\"small\">\r\n    {{ model.description | translate }}\r\n  </span>\r\n</div>\r\n", dependencies: [{ kind: "directive", type: i2.C8yTranslateDirective, selector: "[translate],[ngx-translate]" }, { kind: "pipe", type: i3.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordStrengthComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-password-strength', template: "<div class=\"password-strength password-{{ model.colorName }}\">\r\n  <label class=\"password-strength-label small\" translate>\r\n    Password strength\r\n  </label>\r\n  <div>\r\n    <div class=\"password-bar\"></div>\r\n  </div>\r\n  <span class=\"small\">\r\n    {{ model.description | translate }}\r\n  </span>\r\n</div>\r\n" }]
        }], ctorParameters: function () { return [{ type: i1.PasswordService }]; }, propDecorators: { password: [{
                type: Input,
                args: ['password']
            }], strength: [{
                type: Output
            }] } });
