// @c8y/ngx-components/esm2020/core/authentication/password-confirm.directive.mjs
import { Directive, Attribute } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import * as i0 from "@angular/core";
export class PasswordConfirm {
    constructor(passwordConfirm) {
        this.passwordConfirm = passwordConfirm;
    }
    validate(abControl) {
        const value = abControl.value;
        const controlToCompareWith = abControl.root.get(this.passwordConfirm);
        if (controlToCompareWith && value !== controlToCompareWith.value) {
            return { passwordConfirm: true };
        }
        return null;
    }
}
PasswordConfirm.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordConfirm, deps: [{ token: 'passwordConfirm', attribute: true }], target: i0.ɵɵFactoryTarget.Directive });
PasswordConfirm.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: PasswordConfirm, selector: "[passwordConfirm]", providers: [
        {
            provide: NG_VALIDATORS,
            useExisting: PasswordConfirm,
            multi: true
        }
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PasswordConfirm, decorators: [{
            type: Directive,
            args: [{
                    selector: '[passwordConfirm]',
                    providers: [
                        {
                            provide: NG_VALIDATORS,
                            useExisting: PasswordConfirm,
                            multi: true
                        }
                    ]
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Attribute,
                    args: ['passwordConfirm']
                }] }]; } });
