// @c8y/ngx-components/esm2020/core/realtime/measurement-realtime.service.mjs
import { Injectable } from '@angular/core';
import { MeasurementService } from '@c8y/client';
import { from, merge } from 'rxjs';
import { filter, map, switchMap } from 'rxjs/operators';
import { RealtimeSubjectService } from './realtime-subject.service';
import { RealtimeService } from './realtime.service';
import * as i0 from "@angular/core";
import * as i1 from "./realtime-subject.service";
import * as i2 from "@c8y/client";
export class MeasurementRealtimeService extends RealtimeService {
    constructor(realtimeSubject, measurementService) {
        super(realtimeSubject);
        this.realtimeSubject = realtimeSubject;
        this.measurementService = measurementService;
    }
    /**
     * Get an Observable of all measurements filtered by fragment and series.
     *
     * @param {string} fragment Measurement fragment
     *
     * @param {string} series Measurement series
     *
     * @param {string | number | IIdentified} entityOrId Entity object or id
     *
     * @returns An [[Observable]] of newly created entity objects.
     */
    onCreateOfSpecificMeasurement$(fragment, series, entityOrId) {
        return this.onCreate$(entityOrId).pipe(filter(msg => msg[fragment] && msg[fragment][series]));
    }
    /**
     * Get an Observable of all measurements filtered by fragment and series.
     * Combines the latest value from measurement API and future values via realtime API.
     *
     * @param {string} fragment Measurement fragment
     *
     * @param {string} series Measurement series
     *
     * @param {string | number | IIdentified} entityOrId Entity object or id
     *
     * @param {number} pageSize Number of measurements to initially retrieve from backend
     *
     * @param {boolean} emitNullIfInitialValuesWereNotFound Flag that if set to true will initially emit value of null in case no measurement could have been retrieved from backend
     *
     * @returns An [[Observable]] of newly created entity objects.
     */
    latestValueOfSpecificMeasurement$(fragment, series, entityOrId, pageSize = 1, emitNullIfInitialValuesWereNotFound = false) {
        const realtime$ = this.onCreateOfSpecificMeasurement$(fragment, series, entityOrId);
        const lastMeasurement$ = this.lastMeasurement$(fragment, series, entityOrId, pageSize, emitNullIfInitialValuesWereNotFound);
        return merge(lastMeasurement$, realtime$);
    }
    lastMeasurement$(fragment, series, entityOrId, pageSize, emitNullIfInitialValuesWereNotFound = false) {
        return from(this.measurementService.list({
            valueFragmentSeries: series,
            valueFragmentType: fragment,
            source: this.getIdString(entityOrId),
            dateFrom: '1970-01-01',
            revert: true,
            pageSize
        })).pipe(
        // emit null in case no value could have been retrieved so we are able to note that in certain cases.
        map(result => !emitNullIfInitialValuesWereNotFound || result.data.length ? result.data : [null]), 
        // reverse required, so that measurements are received ordered by time (ascending).
        // from ensures values are sent as single measurements and not a measurement array.
        switchMap(result => from(result.reverse())));
    }
    channel() {
        return '/measurements/*';
    }
}
MeasurementRealtimeService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MeasurementRealtimeService, deps: [{ token: i1.RealtimeSubjectService }, { token: i2.MeasurementService }], target: i0.ɵɵFactoryTarget.Injectable });
MeasurementRealtimeService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MeasurementRealtimeService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MeasurementRealtimeService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.RealtimeSubjectService }, { type: i2.MeasurementService }]; } });
