// @c8y/ngx-components/esm2020/core/realtime/realtime-button.component.mjs
import { Component, Input } from '@angular/core';
import { gettext } from '../i18n/gettext';
import { RealtimeService } from './realtime.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../i18n/c8y-translate.pipe";
export class RealtimeButtonComponent {
    constructor() {
        this.label = gettext('Realtime');
        this.title = gettext('Toggle realtime');
    }
    get active() {
        return !!this.service && this.service.active;
    }
    toggle() {
        if (this.service) {
            this.active ? this.service.stop() : this.service.start();
        }
    }
}
RealtimeButtonComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RealtimeButtonComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
RealtimeButtonComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: RealtimeButtonComponent, selector: "c8y-realtime-btn", inputs: { service: "service", label: "label", title: "title" }, ngImport: i0, template: "<button\n  type=\"button\"\n  class=\"btn btn-link c8y-realtime\"\n  title=\"{{ title | translate }}\"\n  (click)=\"toggle()\"\n>\n  <span class=\"c8y-pulse\" [ngClass]=\"{ active: active, inactive: !active }\"></span>\n  <span>{{ label | translate }}</span>\n</button>\n", dependencies: [{ kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "pipe", type: i2.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RealtimeButtonComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-realtime-btn', template: "<button\n  type=\"button\"\n  class=\"btn btn-link c8y-realtime\"\n  title=\"{{ title | translate }}\"\n  (click)=\"toggle()\"\n>\n  <span class=\"c8y-pulse\" [ngClass]=\"{ active: active, inactive: !active }\"></span>\n  <span>{{ label | translate }}</span>\n</button>\n" }]
        }], propDecorators: { service: [{
                type: Input
            }], label: [{
                type: Input
            }], title: [{
                type: Input
            }] } });
