// @c8y/ngx-components/esm2020/core/realtime/realtime.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { RealtimeButtonComponent } from './realtime-button.component';
import * as i0 from "@angular/core";
export class RealtimeModule {
}
RealtimeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RealtimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RealtimeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: RealtimeModule, declarations: [RealtimeButtonComponent], imports: [CommonModule], exports: [RealtimeButtonComponent] });
RealtimeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RealtimeModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RealtimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [RealtimeButtonComponent],
                    exports: [RealtimeButtonComponent]
                }]
        }] });
