// @c8y/ngx-components/esm2020/operations/bulk-operations-stepper-container/bulk-operations-stepper-container.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationsStepperContainerComponent } from './bulk-operations-stepper-container.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export const bulkOperatioStepperRoutes = [
    {
        path: 'devicecontrol/bulk/creation/:type',
        component: BulkOperationsStepperContainerComponent
    }
];
/**
 * Container module for the new bulk operation steppers. Use this module in your application
 * to provide a step by step guided process for scheduling of bulk operations.
 * The module reads the bulk operation type from the URL and displays the corresponding stepper.
 * You can provide stepper for your custom bulk operation type by registering it with the
 * HOOK_LIST_BULK_TYPE injection token.
 */
export class BulkOperationsStepperContainerModule {
}
BulkOperationsStepperContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsStepperContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BulkOperationsStepperContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsStepperContainerModule, declarations: [BulkOperationsStepperContainerComponent], imports: [CoreModule, i1.RouterModule] });
BulkOperationsStepperContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsStepperContainerModule, imports: [CoreModule, RouterModule.forChild(bulkOperatioStepperRoutes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsStepperContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BulkOperationsStepperContainerComponent],
                    imports: [CoreModule, RouterModule.forChild(bulkOperatioStepperRoutes)],
                    entryComponents: [BulkOperationsStepperContainerComponent]
                }]
        }] });
