// @c8y/ngx-components/esm2020/core/properties-list/properties-list.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { PropertiesListComponent } from './properties-list.component';
import * as i0 from "@angular/core";
export class PropertiesListModule {
}
PropertiesListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PropertiesListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PropertiesListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: PropertiesListModule, declarations: [PropertiesListComponent], imports: [CommonModule], exports: [PropertiesListComponent] });
PropertiesListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PropertiesListModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PropertiesListModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PropertiesListComponent],
                    imports: [CommonModule],
                    exports: [PropertiesListComponent],
                    providers: []
                }]
        }] });
