// @c8y/ngx-components/esm2020/protocol-lpwan/lpwan-set-device-protocol.service.mjs
import { Injectable } from '@angular/core';
import { InventoryService, FetchClient, IdentityService } from '@c8y/client';
import { get, orderBy } from 'lodash-es';
import { MultipleLnsConnectorService } from './multiple-lns-connectors';
import { ConnectionType } from './multiple-lns-connectors/multiple-lns-connector.model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "./multiple-lns-connectors";
export class LpwanSetDeviceProtocolService {
    constructor(inventoryService, client, identityService, lnsService) {
        this.inventoryService = inventoryService;
        this.client = client;
        this.identityService = identityService;
        this.lnsService = lnsService;
        this.supportedDevicesCfgs = [
            {
                name: 'actility',
                match: device => get(device, 'c8y_LpwanDevice.lpwanDeviceType') === 'Lora' &&
                    get(device, 'c8y_LpwanDevice.serviceProvider') === 'Actility',
                protocolTypes: ['c8y_ActilityDeviceType', 'c8y_LoraDeviceType', 'c8y_LpwanDeviceType'],
                externalIdTypes: ['c8y_LoriotEUI', 'c8y_Serial'],
                connectionType: ConnectionType.ACTILITY
            },
            {
                name: 'loriot',
                match: device => get(device, 'c8y_LpwanDevice.lpwanDeviceType') === 'Lora' &&
                    get(device, 'c8y_LpwanDevice.serviceProvider') === 'Loriot',
                protocolTypes: ['c8y_LoraDeviceType', 'c8y_LpwanDeviceType'],
                externalIdTypes: ['c8y_LoriotEUI', 'c8y_Serial'],
                connectionType: ConnectionType.LORIOT
            },
            {
                name: 'sigfox',
                match: device => get(device, 'c8y_LpwanDevice.serviceProvider') === 'Sigfox',
                protocolTypes: ['c8y_SigfoxDeviceType', 'c8y_LpwanDeviceType'],
                externalIdTypes: ['com.sigfox.deviceId'],
                connectionType: ConnectionType.SIGFOX
            }
        ];
        this.header = { 'Content-Type': 'application/json' };
    }
    async refreshCache(device) {
        const externalId = await this.getExternalId(device);
        if (externalId) {
            const url = `${this.getMicroserviceUrl(device)}/refreshCache/${externalId}`;
            const options = {
                method: 'POST',
                headers: this.header,
                body: JSON.stringify({})
            };
            return this.client.fetch(url, options);
        }
    }
    getMicroserviceUrl(device) {
        const { serviceProvider } = device.c8y_LpwanDevice;
        let serviceName = serviceProvider.toLowerCase();
        if (serviceProvider === 'Sigfox') {
            serviceName = 'sigfox-agent';
        }
        return `/service/${serviceName}`;
    }
    isSupportedDevice(device) {
        return this.supportedDevicesCfgs.some(({ match }) => match(device));
    }
    async getCurrentProtocol(device) {
        const lpwanDevice = device.c8y_LpwanDevice;
        let protocolId;
        if (lpwanDevice.typeExternalId) {
            const externalId = (await this.identityService.detail(lpwanDevice.typeExternalId)).data;
            protocolId = externalId.managedObject.id;
        }
        if (!protocolId && lpwanDevice.type) {
            protocolId = lpwanDevice.type.split('/')[2];
        }
        if (!protocolId) {
            return null;
        }
        return (await this.inventoryService.detail(protocolId)).data;
    }
    async applyProtocol(device, selectedProtocol) {
        const [protocolExternalId] = (await this.identityService.list(selectedProtocol.id)).data;
        const { externalId, type } = protocolExternalId;
        device.c8y_LpwanDevice.typeExternalId = { externalId, type };
        device.c8y_LpwanDevice.type = 'inventory/managedObjects/' + selectedProtocol.id;
        device.type = selectedProtocol.name;
        return this.inventoryService.update(device);
    }
    async getAvailableProtocols(device) {
        const query = {
            __filter: {
                type: { __in: this.getProtocolTypesMatchingDevice(device) }
            },
            __orderby: [{ name: 1 }]
        };
        return this.inventoryService.listQuery(query, { withTotalPages: true, pageSize: 5 });
    }
    async getAvailableConnections(device) {
        let list;
        const matchingCfg = this.supportedDevicesCfgs.find(({ match }) => match(device));
        const response = await this.lnsService.list(matchingCfg.connectionType);
        if (response && response.status === 200) {
            list = await response.json();
            list = orderBy(list, ['name'], ['asc']);
            return { res: response, data: list };
        }
    }
    getProtocolTypesMatchingDevice(device) {
        const matchingCfg = this.supportedDevicesCfgs.find(({ match }) => match(device));
        return matchingCfg ? matchingCfg.protocolTypes : [];
    }
    async getExternalId(device) {
        const matchingCfg = this.supportedDevicesCfgs.find(({ match }) => match(device));
        const externalIds = (await this.identityService.list(device.id)).data;
        const externalId = externalIds.find(({ type }) => matchingCfg.externalIdTypes.includes(type));
        return externalId ? externalId.externalId : null;
    }
}
LpwanSetDeviceProtocolService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LpwanSetDeviceProtocolService, deps: [{ token: i1.InventoryService }, { token: i1.FetchClient }, { token: i1.IdentityService }, { token: i2.MultipleLnsConnectorService }], target: i0.ɵɵFactoryTarget.Injectable });
LpwanSetDeviceProtocolService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LpwanSetDeviceProtocolService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LpwanSetDeviceProtocolService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.InventoryService }, { type: i1.FetchClient }, { type: i1.IdentityService }, { type: i2.MultipleLnsConnectorService }]; } });
