// @c8y/ngx-components/esm2020/protocol-lpwan/lpwan-protocol.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, FormsModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { LpwanAssignDeviceProtocolComponent } from './lpwan-set-device-protocol.component';
import { LpwanAgentGuard } from './lpwan-agent.guard';
import { LpwanSetDeviceProtocolService } from './lpwan-set-device-protocol.service';
import { LpwanAssignLnsConnectionsComponent } from './lpwan-set-connections.component';
import * as i0 from "@angular/core";
const routes = [
    {
        context: ViewContext.Device,
        path: 'assign-protocol',
        component: LpwanAssignDeviceProtocolComponent,
        label: gettext('LPWAN'),
        icon: 'c8y-device-protocols',
        canActivate: [LpwanAgentGuard]
    }
];
export class LpwanProtocolModule {
}
LpwanProtocolModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LpwanProtocolModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LpwanProtocolModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: LpwanProtocolModule, declarations: [LpwanAssignDeviceProtocolComponent, LpwanAssignLnsConnectionsComponent], imports: [CoreModule, FormsModule] });
LpwanProtocolModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LpwanProtocolModule, providers: [LpwanAgentGuard, LpwanSetDeviceProtocolService, hookRoute(routes)], imports: [CoreModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LpwanProtocolModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LpwanAssignDeviceProtocolComponent, LpwanAssignLnsConnectionsComponent],
                    imports: [CoreModule, FormsModule],
                    entryComponents: [LpwanAssignDeviceProtocolComponent, LpwanAssignLnsConnectionsComponent],
                    providers: [LpwanAgentGuard, LpwanSetDeviceProtocolService, hookRoute(routes)]
                }]
        }] });
