// @c8y/ngx-components/esm2020/ecosystem/plugin-setup-stepper/plugin-setup-stepper.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PluginSetupStepperComponent } from './plugin-setup-stepper.component';
import { gettext, hookStepper, Steppers } from '@c8y/ngx-components';
import { ApplicationPluginsModule } from '@c8y/ngx-components/ecosystem/application-plugins';
import { LicenseConfirmModule } from '@c8y/ngx-components/ecosystem/license-confirm';
import * as i0 from "@angular/core";
export class PluginSetupStepperModule {
}
PluginSetupStepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PluginSetupStepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, declarations: [PluginSetupStepperComponent], imports: [CommonModule, ApplicationPluginsModule, LicenseConfirmModule] });
PluginSetupStepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, providers: [
        hookStepper([
            {
                stepperId: Steppers.SETUP,
                component: PluginSetupStepperComponent,
                label: gettext('Plugins'),
                setupId: 'plugins',
                priority: 30
            }
        ])
    ], imports: [CommonModule, ApplicationPluginsModule, LicenseConfirmModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ApplicationPluginsModule, LicenseConfirmModule],
                    declarations: [PluginSetupStepperComponent],
                    providers: [
                        hookStepper([
                            {
                                stepperId: Steppers.SETUP,
                                component: PluginSetupStepperComponent,
                                label: gettext('Plugins'),
                                setupId: 'plugins',
                                priority: 30
                            }
                        ])
                    ]
                }]
        }] });
