// @c8y/ngx-components/esm2020/widgets/implementations/markdown/markdown-widget.service.mjs
import { Injectable } from '@angular/core';
import { InventoryBinaryService, InventoryService } from '@c8y/client';
import { FilesService, gettext, AlertService } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/client";
import * as i3 from "@ngx-translate/core";
export class MarkdownWidgetService {
    constructor(fileService, inventory, binary, alert, translate) {
        this.fileService = fileService;
        this.inventory = inventory;
        this.binary = binary;
        this.alert = alert;
        this.translate = translate;
    }
    async getFile(markdownBinaryId) {
        if (!markdownBinaryId) {
            return null;
        }
        try {
            const { data: markdownBinaryMo } = await this.inventory.detail(markdownBinaryId);
            const file = await this.fileService.getFile(markdownBinaryMo);
            return file;
        }
        catch (e) {
            const text = this.translate.instant(gettext('Unable to retrieve binary with ID: {{ markdownBinaryId }}'), { markdownBinaryId });
            this.alert.danger(text, e?.data);
        }
        return null;
    }
    async uploadFile(file) {
        const { data: mo } = await this.binary.create(file);
        return mo.id;
    }
}
MarkdownWidgetService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownWidgetService, deps: [{ token: i1.FilesService }, { token: i2.InventoryService }, { token: i2.InventoryBinaryService }, { token: i1.AlertService }, { token: i3.TranslateService }], target: i0.ɵɵFactoryTarget.Injectable });
MarkdownWidgetService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownWidgetService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownWidgetService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.FilesService }, { type: i2.InventoryService }, { type: i2.InventoryBinaryService }, { type: i1.AlertService }, { type: i3.TranslateService }]; } });
