// @c8y/ngx-components/esm2020/ecosystem/application-plugins/application-plugins.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { SharedEcosystemModule } from '@c8y/ngx-components/ecosystem/shared';
import { ApplicationPluginsComponent } from './application-plugins.component';
import { UpdatePluginOfAppComponent } from './update-plugin-of-app/update-plugin-of-app.component';
import { AppStatePipe } from './appState.pipe';
import { AppsToUpdateRemotesSelectComponent } from './apps-to-update-remotes-select.component';
import { InstallPluginComponent } from './install-plugin.component';
import { OrphanedStatusCellRendererComponent } from './orphaned-status-cell-renderer.component';
import { PluginListItemComponent } from './plugin-list-item.component';
import { PluginListComponent } from './plugin-list.component';
import { LabelCellRendererComponent } from './label-cell-renderer.component';
import { gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { ApplicationPluginsGuard } from './application-plugins.guard';
import { OnlyLatestPluginVersionPipe } from './only-latest-plugin-version.pipe';
import * as i0 from "@angular/core";
export class ApplicationPluginsModule {
}
ApplicationPluginsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ApplicationPluginsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ApplicationPluginsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ApplicationPluginsModule, declarations: [ApplicationPluginsComponent,
        PluginListItemComponent,
        InstallPluginComponent,
        PluginListComponent,
        AppsToUpdateRemotesSelectComponent,
        AppStatePipe,
        LabelCellRendererComponent,
        OrphanedStatusCellRendererComponent,
        UpdatePluginOfAppComponent,
        OnlyLatestPluginVersionPipe], imports: [CoreModule, SharedEcosystemModule], exports: [ApplicationPluginsComponent,
        PluginListItemComponent,
        InstallPluginComponent,
        PluginListComponent,
        AppsToUpdateRemotesSelectComponent,
        AppStatePipe,
        LabelCellRendererComponent,
        OrphanedStatusCellRendererComponent,
        UpdatePluginOfAppComponent] });
ApplicationPluginsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ApplicationPluginsModule, providers: [
        ApplicationPluginsGuard,
        hookRoute([
            {
                path: 'plugins',
                component: ApplicationPluginsComponent,
                icon: 'plugin',
                label: gettext('Plugins'),
                context: ViewContext.Application,
                priority: 3,
                canActivate: [ApplicationPluginsGuard]
            }
        ])
    ], imports: [CoreModule, SharedEcosystemModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ApplicationPluginsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, SharedEcosystemModule],
                    declarations: [
                        ApplicationPluginsComponent,
                        PluginListItemComponent,
                        InstallPluginComponent,
                        PluginListComponent,
                        AppsToUpdateRemotesSelectComponent,
                        AppStatePipe,
                        LabelCellRendererComponent,
                        OrphanedStatusCellRendererComponent,
                        UpdatePluginOfAppComponent,
                        OnlyLatestPluginVersionPipe
                    ],
                    exports: [
                        ApplicationPluginsComponent,
                        PluginListItemComponent,
                        InstallPluginComponent,
                        PluginListComponent,
                        AppsToUpdateRemotesSelectComponent,
                        AppStatePipe,
                        LabelCellRendererComponent,
                        OrphanedStatusCellRendererComponent,
                        UpdatePluginOfAppComponent
                    ],
                    providers: [
                        ApplicationPluginsGuard,
                        hookRoute([
                            {
                                path: 'plugins',
                                component: ApplicationPluginsComponent,
                                icon: 'plugin',
                                label: gettext('Plugins'),
                                context: ViewContext.Application,
                                priority: 3,
                                canActivate: [ApplicationPluginsGuard]
                            }
                        ])
                    ]
                }]
        }] });
