// @c8y/ngx-components/esm2020/ecosystem/application-plugins/appState.pipe.mjs
import { Pipe } from '@angular/core';
import { EcosystemService } from '@c8y/ngx-components/ecosystem/shared';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components/ecosystem/shared";
export class AppStatePipe {
    constructor(ecosystemService) {
        this.ecosystemService = ecosystemService;
    }
    transform(app, arg) {
        const appState = this.ecosystemService.getAppState(app);
        return appState[arg];
    }
}
AppStatePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AppStatePipe, deps: [{ token: i1.EcosystemService }], target: i0.ɵɵFactoryTarget.Pipe });
AppStatePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AppStatePipe, name: "appState" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AppStatePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'appState',
                    pure: true
                }]
        }], ctorParameters: function () { return [{ type: i1.EcosystemService }]; } });
