// @c8y/ngx-components/esm2020/ecosystem/application-plugins/orphaned-status-cell-renderer.component.mjs
import { Component } from '@angular/core';
import { ApplicationPluginStatus, CellRendererContext, gettext } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@angular/common";
export class OrphanedStatusCellRendererComponent {
    constructor(context) {
        this.context = context;
        this.label = this.getLabel(context.value);
    }
    getLabel(statusValue) {
        switch (statusValue) {
            case ApplicationPluginStatus.OUTDATED:
                return {
                    value: statusValue,
                    text: gettext('OUTDATED`plugin status`'),
                    class: 'label-warning'
                };
            case ApplicationPluginStatus.ORPHANED:
                return {
                    value: statusValue,
                    text: gettext('ORPHANED`plugin status`'),
                    class: 'label-danger'
                };
            case ApplicationPluginStatus.REVOKED:
                return {
                    value: statusValue,
                    text: gettext('REVOKED`plugin status`'),
                    class: 'label-danger'
                };
            case ApplicationPluginStatus.LATEST:
                return {
                    value: statusValue,
                    text: gettext('LATEST`plugin status`'),
                    class: 'label-success'
                };
            default:
                return null;
        }
    }
}
OrphanedStatusCellRendererComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OrphanedStatusCellRendererComponent, deps: [{ token: i1.CellRendererContext }], target: i0.ɵɵFactoryTarget.Component });
OrphanedStatusCellRendererComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: OrphanedStatusCellRendererComponent, selector: "c8y-orphaned-status-cell-renderer", ngImport: i0, template: "<span\n  *ngIf=\"label\"\n  class=\"label\"\n  [ngClass]=\"label.class\"\n  [title]=\"label.text | translate\"\n>\n  {{ label.text | translate }}\n</span>\n", dependencies: [{ kind: "directive", type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "pipe", type: i1.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OrphanedStatusCellRendererComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-orphaned-status-cell-renderer', template: "<span\n  *ngIf=\"label\"\n  class=\"label\"\n  [ngClass]=\"label.class\"\n  [title]=\"label.text | translate\"\n>\n  {{ label.text | translate }}\n</span>\n" }]
        }], ctorParameters: function () { return [{ type: i1.CellRendererContext }]; } });
