// @c8y/ngx-components/esm2020/core/ui-settings/ui-settings.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from './../common/common.module';
import { UiSettingsComponent } from './ui-settings.component';
import { FormsModule } from '@angular/forms';
import { hookDrawer } from '../drawer/drawer.service';
import * as i0 from "@angular/core";
export class UiSettingsModule {
}
UiSettingsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UiSettingsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UiSettingsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: UiSettingsModule, declarations: [UiSettingsComponent], imports: [CommonModule, FormsModule], exports: [UiSettingsComponent] });
UiSettingsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UiSettingsModule, providers: [
        hookDrawer({
            component: UiSettingsComponent,
            position: 'right',
            priority: 90,
            id: 'uiSettings'
        })
    ], imports: [CommonModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UiSettingsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [UiSettingsComponent],
                    imports: [CommonModule, FormsModule],
                    exports: [UiSettingsComponent],
                    providers: [
                        hookDrawer({
                            component: UiSettingsComponent,
                            position: 'right',
                            priority: 90,
                            id: 'uiSettings'
                        })
                    ]
                }]
        }] });
