// @c8y/ngx-components/esm2020/file-preview/file-preview.module.mjs
import { EventEmitter, NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { FilePreviewComponent } from './file-preview.component';
import { CommonModule } from '@c8y/ngx-components';
import { DOWNLOAD_EMITTER } from './download-emitter.token';
import * as i0 from "@angular/core";
const downloadEmitter = new EventEmitter();
export class FilePreviewModule {
}
FilePreviewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePreviewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilePreviewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FilePreviewModule, declarations: [FilePreviewComponent], imports: [CoreModule, CommonModule], exports: [FilePreviewComponent] });
FilePreviewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePreviewModule, providers: [
        {
            provide: DOWNLOAD_EMITTER,
            useValue: downloadEmitter
        }
    ], imports: [CoreModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePreviewModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FilePreviewComponent],
                    imports: [CoreModule, CommonModule],
                    exports: [FilePreviewComponent],
                    providers: [
                        {
                            provide: DOWNLOAD_EMITTER,
                            useValue: downloadEmitter
                        }
                    ]
                }]
        }] });
