// @c8y/ngx-components/esm2020/core/date-time-picker/date-time-picker.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/index';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { DateTimePickerComponent } from './date-time-picker.component';
import { TimePickerModule } from '../time-picker/time-picker.module';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/datepicker";
/**
 * The angular module definition for a date-time picker - it provides FormControl that handles date and time.
 * @exports DateTimePickerComponent
 */
export class DateTimePickerModule {
}
DateTimePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DateTimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateTimePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DateTimePickerModule, declarations: [DateTimePickerComponent], imports: [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule, i1.BsDatepickerModule, C8yCommonModule,
        TimePickerModule], exports: [DateTimePickerComponent] });
DateTimePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DateTimePickerModule, imports: [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        BsDatepickerModule.forRoot(),
        C8yCommonModule,
        TimePickerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DateTimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        C8yFormsModule,
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BsDatepickerModule.forRoot(),
                        C8yCommonModule,
                        TimePickerModule
                    ],
                    exports: [DateTimePickerComponent],
                    declarations: [DateTimePickerComponent]
                }]
        }] });
