// @c8y/ngx-components/esm2020/core/data-grid/data-grid.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { CdkTableModule } from '@angular/cdk/table';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PaginationModule } from 'ngx-bootstrap/pagination';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule as C8YCommonModule } from '../common/common.module';
import { DynamicFormsModule } from '../dynamic-forms/dynamic-forms.module';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { ModalModule } from '../modal/modal.module';
import { ProductExperienceModule } from '../product-experience/product-experience.module';
import { BaseFilteringFormRendererComponent } from './column/base.filtering-form-renderer.component';
import { CellRendererDefDirective, HeaderCellRendererDefDirective } from './column/cell-renderer';
import { CellRendererComponent } from './column/cell-renderer.component';
import { ColumnDirective } from './column/column.directive';
import { FilteringFormRendererDefDirective } from './column/filtering-form-renderer';
import { FilteringFormRendererComponent } from './column/filtering-form-renderer.component';
import { ConfigureCustomColumnComponent } from './configure-custom-column/configure-custom-column.component';
import { DataGridComponent } from './data-grid.component';
import { FilterMapperModule } from './filter-chip/filter-mapper.module';
import { IsControlVisiblePipe } from './is-control-visible.pipe';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/dropdown";
import * as i2 from "ngx-bootstrap/popover";
import * as i3 from "ngx-bootstrap/tooltip";
import * as i4 from "ngx-bootstrap/pagination";
export class DataGridModule {
}
DataGridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataGridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DataGridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DataGridModule, declarations: [CellRendererDefDirective,
        HeaderCellRendererDefDirective,
        BaseFilteringFormRendererComponent,
        CellRendererComponent,
        FilteringFormRendererDefDirective,
        FilteringFormRendererComponent,
        ColumnDirective,
        DataGridComponent,
        IsControlVisiblePipe,
        ConfigureCustomColumnComponent], imports: [CommonModule,
        CdkTableModule,
        DragDropModule,
        FormsModule,
        ReactiveFormsModule,
        DynamicFormsModule,
        C8YCommonModule,
        C8yFormsModule,
        ModalModule, i1.BsDropdownModule, i2.PopoverModule, i3.TooltipModule, i4.PaginationModule, ProductExperienceModule,
        A11yModule,
        FilterMapperModule], exports: [CellRendererDefDirective,
        HeaderCellRendererDefDirective,
        FilteringFormRendererDefDirective,
        BaseFilteringFormRendererComponent,
        ColumnDirective,
        DataGridComponent,
        IsControlVisiblePipe] });
DataGridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataGridModule, imports: [CommonModule,
        CdkTableModule,
        DragDropModule,
        FormsModule,
        ReactiveFormsModule,
        DynamicFormsModule,
        C8YCommonModule,
        C8yFormsModule,
        ModalModule,
        BsDropdownModule.forRoot(),
        PopoverModule.forRoot(),
        TooltipModule.forRoot(),
        PaginationModule.forRoot(),
        ProductExperienceModule,
        A11yModule,
        FilterMapperModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataGridModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        CdkTableModule,
                        DragDropModule,
                        FormsModule,
                        ReactiveFormsModule,
                        DynamicFormsModule,
                        C8YCommonModule,
                        C8yFormsModule,
                        ModalModule,
                        BsDropdownModule.forRoot(),
                        PopoverModule.forRoot(),
                        TooltipModule.forRoot(),
                        PaginationModule.forRoot(),
                        ProductExperienceModule,
                        A11yModule,
                        FilterMapperModule
                    ],
                    declarations: [
                        CellRendererDefDirective,
                        HeaderCellRendererDefDirective,
                        BaseFilteringFormRendererComponent,
                        CellRendererComponent,
                        FilteringFormRendererDefDirective,
                        FilteringFormRendererComponent,
                        ColumnDirective,
                        DataGridComponent,
                        IsControlVisiblePipe,
                        ConfigureCustomColumnComponent
                    ],
                    exports: [
                        CellRendererDefDirective,
                        HeaderCellRendererDefDirective,
                        FilteringFormRendererDefDirective,
                        BaseFilteringFormRendererComponent,
                        ColumnDirective,
                        DataGridComponent,
                        IsControlVisiblePipe
                    ],
                    entryComponents: [DataGridComponent]
                }]
        }] });
