// @c8y/ngx-components/esm2020/core/data-grid/user-preferences-configuration-strategy.service.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { isEmpty } from 'lodash-es';
import { from, of } from 'rxjs';
import { UserPreferencesService } from '../common/user-preferences/user-preferences.service';
import { AbstractConfigurationStrategy } from './abstract-configuration-strategy';
import { DATA_GRID_CONFIGURATION_CONTEXT, DATA_GRID_CONFIGURATION_CONTEXT_PROVIDER } from './data-grid-configuration.model';
import { LegacyGridConfigMapperService } from './legacy-grid-config-mapper.service';
import * as i0 from "@angular/core";
import * as i1 from "../common/user-preferences/user-preferences.service";
import * as i2 from "./legacy-grid-config-mapper.service";
/**
 * Stores data grid configuration in form of user preferences.
 */
export class UserPreferencesConfigurationStrategy extends AbstractConfigurationStrategy {
    constructor(userPreferencesService, context, contextProvider, legacyGridMapperService) {
        super(context, contextProvider);
        this.userPreferencesService = userPreferencesService;
        this.context = context;
        this.contextProvider = contextProvider;
        this.legacyGridMapperService = legacyGridMapperService;
    }
    getConfig$(context) {
        const configContext = this.retrieveContext(context);
        if (isEmpty(configContext)) {
            return of(undefined);
        }
        else {
            return configContext.legacyConfigKey || configContext.legacyFilterKey
                ? this.legacyGridMapperService.getMappedGridConfig(configContext)
                : this.userPreferencesService.get(configContext?.key);
        }
    }
    saveConfig$(config, context) {
        const configContext = this.retrieveContext(context);
        return isEmpty(configContext)
            ? of(undefined)
            : from(this.userPreferencesService.set(configContext?.key, config));
    }
}
UserPreferencesConfigurationStrategy.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserPreferencesConfigurationStrategy, deps: [{ token: i1.UserPreferencesService }, { token: DATA_GRID_CONFIGURATION_CONTEXT, optional: true }, { token: DATA_GRID_CONFIGURATION_CONTEXT_PROVIDER, optional: true }, { token: i2.LegacyGridConfigMapperService }], target: i0.ɵɵFactoryTarget.Injectable });
UserPreferencesConfigurationStrategy.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserPreferencesConfigurationStrategy, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserPreferencesConfigurationStrategy, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.UserPreferencesService }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DATA_GRID_CONFIGURATION_CONTEXT]
                }, {
                    type: Optional
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [DATA_GRID_CONFIGURATION_CONTEXT_PROVIDER]
                }, {
                    type: Optional
                }] }, { type: i2.LegacyGridConfigMapperService }]; } });
