// @c8y/ngx-components/esm2020/core/modal/modal.module.mjs
import { NgModule } from '@angular/core';
import { ModalModule as BsModalModule } from 'ngx-bootstrap/modal';
import { ConfirmModalComponent } from './modal-confirm.component';
import { ModalComponent } from './modal.component';
import { ModalService } from './modal.service';
import { CommonModule } from '../common/common.module';
import { PopoverConfirmComponent } from './popover-confirm.component';
import { PopoverModule } from 'ngx-bootstrap/popover';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
import * as i2 from "ngx-bootstrap/popover";
/**
 * The angular module definition for modal.
 * @exports ConfirmModalComponent
 * @exports ModalComponent
 * @exports PopoverConfirmComponent
 */
export class ModalModule {
}
ModalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ModalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ModalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ModalModule, declarations: [ConfirmModalComponent, ModalComponent, PopoverConfirmComponent], imports: [i1.ModalModule, CommonModule, i2.PopoverModule], exports: [ConfirmModalComponent, ModalComponent, PopoverConfirmComponent] });
ModalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ModalModule, providers: [ModalService], imports: [BsModalModule.forRoot(), CommonModule, PopoverModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ModalModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [BsModalModule.forRoot(), CommonModule, PopoverModule.forRoot()],
                    exports: [ConfirmModalComponent, ModalComponent, PopoverConfirmComponent],
                    declarations: [ConfirmModalComponent, ModalComponent, PopoverConfirmComponent],
                    providers: [ModalService],
                    entryComponents: [ConfirmModalComponent]
                }]
        }] });
