// @c8y/ngx-components/esm2020/core/common/device.service.mjs
import { Injectable } from '@angular/core';
import { find, get, isUndefined, some } from 'lodash-es';
import * as i0 from "@angular/core";
export class DeviceService {
    static isAnyDevice(mo) {
        const isDevice = DeviceService.isRootDevice(mo) || !DeviceService.hasFragmentOrTypeFromBlacklist(mo);
        return isDevice;
    }
    static hasFragmentOrTypeFromBlacklist(mo) {
        return DeviceService.hasTypeFromBlacklist(mo) || DeviceService.hasFragmentFromBlacklist(mo);
    }
    static hasTypeFromBlacklist(mo) {
        const moType = get(mo, 'type', '');
        return some(DeviceService.TYPES_FOR_NO_DEVICE, (type) => moType.includes(type));
    }
    static hasFragmentFromBlacklist(mo) {
        const result = find(DeviceService.FRAGMENTS_FOR_NO_DEVICE, (fragment) => !isUndefined(mo[fragment]));
        return !!result;
    }
    static isRootDevice(mo) {
        return !!mo[DeviceService.DEVICE_FRAGMENT_TYPE];
    }
}
DeviceService.FRAGMENTS_FOR_NO_DEVICE = [
    'c8y_Dashboard',
    'c8y_Report',
    'c8y_Kpi',
    'c8y_ExportConfiguration',
    'c8y_IsBinary',
    'c8y_NoDevice',
    'c8y_IsDeviceGroup',
    'c8y_IsDynamicGroup',
    'c8y_Group',
    'com_cumulocity_model_smartrest_SmartRestTemplate',
    'com_cumulocity_model_devicesimulator_SensorTemplate',
    '_attachments',
    'c8y_IsDeviceType',
    'c8y_objectmapping_ObjectMapping',
    'c8y_IsAssetType'
];
DeviceService.TYPES_FOR_NO_DEVICE = [
    'c8y_ConfigurationDump',
    'c8y_Firmware',
    'c8y_SmartRule',
    'c8y_Software',
    'impact_object_mapping',
    'c8y_UserPreference',
    'c8y_TenantPolicy',
    'c8y_PrivateSmartRule',
    'c8y_SmartRest2Template',
    'c8y_JsonSchema',
    'c8y_DeviceShellTemplate',
    'c8y_DemoStatus',
    'c8y_DataBroker',
    'c8y_Application_',
    'brandingVariables',
    'c8y_DeviceSimulator',
    'c8y_CertificateMetadata',
    'lwm2m_post_registration',
    'c8y_microservice_manifest_'
];
DeviceService.DEVICE_FRAGMENT_TYPE = 'c8y_IsDevice';
DeviceService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DeviceService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
