// @c8y/ngx-components/esm2020/core/common/dropdown-direction.directive.mjs
import { Directive, ElementRef, HostListener, Renderer2 } from '@angular/core';
import { forEach } from 'lodash-es';
import * as i0 from "@angular/core";
/**
 * A directive to determine if a dropdown should open to the top or as usual to the bottom.
 * This directive has a direct dependency to the bootstrap dropdown (!)
 *
 * Preconditions:
 *  - Bootstrap dropdown is used
 *  - Bootstrap dropdown is not used with container="body"
 *  - Bootstrap dropdown is used like it is intended and documented for e.g.
 * ```html
 * <div dropdown class="dropdown">
 *  <button class="dropdown-toggle" dropdownToggle>Toggle</button>
 *  <ul class="dropdown-menu" *dropdownMenu>
 *    <!-- Content here -->
 *  </ul>
 * </div>
 * ```
 *
 * Example:
 * ```html
 * <div dropdown c8yBsDropdownDirection class="dropdown">
 *  <button class="dropdown-toggle" dropdownToggle>Toggle</button>
 *  <ul class="dropdown-menu" *dropdownMenu>
 *    <!-- Content here -->
 *  </ul>
 * </div>
 * ```
 * You make use of the bootstrap dropdown as usual and additionally add the direction directive to it.
 */
export class DropdownDirectionDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.MARGIN_BOTTOM = 10;
        this.VISIBILITY = 'visibility';
        this.DROPUP = 'dropup';
    }
    onClick() {
        const open = this.elementRef.nativeElement.classList.contains('open');
        if (open) {
            let menu;
            let toggle;
            forEach(this.elementRef.nativeElement.children, (element) => {
                if (element.classList.contains('dropdown-menu')) {
                    menu = element;
                }
                if (element.classList.contains('dropdown-toggle')) {
                    toggle = element;
                }
            });
            if (menu && toggle) {
                this.renderer.setStyle(menu, this.VISIBILITY, 'hidden');
                this.shouldDropup(menu, toggle);
                this.renderer.removeStyle(menu, this.VISIBILITY);
            }
        }
    }
    shouldDropup(container, position) {
        const containerHeight = container.getBoundingClientRect().height;
        const containerDropPosition = position.getBoundingClientRect().bottom;
        const layoutViewport = window.innerHeight;
        const sum = containerHeight + containerDropPosition + this.MARGIN_BOTTOM;
        if (sum >= layoutViewport) {
            this.renderer.addClass(this.elementRef.nativeElement, this.DROPUP);
        }
        else if (this.elementRef.nativeElement.classList.contains(this.DROPUP)) {
            this.renderer.removeClass(this.elementRef.nativeElement, this.DROPUP);
        }
    }
}
DropdownDirectionDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DropdownDirectionDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
DropdownDirectionDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: DropdownDirectionDirective, selector: "[dropdown][c8yBsDropdownDirection],[dropdown][c8yDropdownDirection]", host: { listeners: { "click": "onClick()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DropdownDirectionDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[dropdown][c8yBsDropdownDirection],[dropdown][c8yDropdownDirection]' }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, propDecorators: { onClick: [{
                type: HostListener,
                args: ['click']
            }] } });
