// @c8y/ngx-components/esm2020/core/common/markdown-to-html.pipe.mjs
import { Pipe, SecurityContext } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { from, of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class MarkdownToHtmlPipe {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    transform(markdown, args) {
        return markdown
            ? from(import('marked')).pipe(map(marked => {
                const renderer = new marked.Renderer();
                const baseUrl = args?.baseUrl || '';
                if (baseUrl) {
                    const originalRendererImage = renderer.image;
                    renderer.image = (href, title, text) => {
                        href = baseUrl + href;
                        return originalRendererImage.call(renderer, href, title, text);
                    };
                }
                const originalRendererLink = renderer.link;
                renderer.link = (href, title, text) => {
                    const { hostname, protocol } = window.location;
                    const localLink = href.startsWith(`${protocol}//${hostname}`);
                    const html = originalRendererLink.call(renderer, href, title, text);
                    return localLink
                        ? html
                        : html.replace(/^<a /, `<a target="_blank" rel="noreferrer noopener nofollow" `);
                };
                return marked.parse(markdown, { renderer });
            }), map(html => this.sanitizer.sanitize(SecurityContext.HTML, html)), catchError(() => {
                return of('<p>Failed to process markdown file.</p>');
            }))
            : of('');
    }
}
MarkdownToHtmlPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownToHtmlPipe, deps: [{ token: i1.DomSanitizer }], target: i0.ɵɵFactoryTarget.Pipe });
MarkdownToHtmlPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: MarkdownToHtmlPipe, name: "markdownToHtml" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MarkdownToHtmlPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'markdownToHtml'
                }]
        }], ctorParameters: function () { return [{ type: i1.DomSanitizer }]; } });
