// @c8y/ngx-components/esm2020/core/common/mo-name.pipe.mjs
import { Pipe } from '@angular/core';
import { MoNameCacheService } from './mo-name-cache.service';
import * as i0 from "@angular/core";
import * as i1 from "./mo-name-cache.service";
export class MoNamePipe {
    constructor(moNamePipeService) {
        this.moNamePipeService = moNamePipeService;
    }
    async transform(entityOrId) {
        if (typeof entityOrId === 'object') {
            return entityOrId.name ? entityOrId.name : '';
        }
        this.moNamePipeService.resetCacheTimer(entityOrId);
        return await this.moNamePipeService.getMoName(entityOrId);
    }
}
MoNamePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MoNamePipe, deps: [{ token: i1.MoNameCacheService }], target: i0.ɵɵFactoryTarget.Pipe });
MoNamePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: MoNamePipe, name: "moName" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MoNamePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'moName'
                }]
        }], ctorParameters: function () { return [{ type: i1.MoNameCacheService }]; } });
