// @c8y/ngx-components/esm2020/core/common/mo-name-cache.service.mjs
import { Injectable } from '@angular/core';
import { InventoryService } from '@c8y/client';
import { memoize } from 'lodash-es';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
export class MoNameCacheService {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
        this.clearCacheTimer = [];
        this.cacheLifeSpan = 5 * 60 * 1000; // 5 minutes
        this.getMoName = memoize(async (id) => {
            const { data } = await this.inventoryService.detail(id);
            return data.name ? data.name : '';
        });
    }
    resetCacheTimer(id) {
        clearTimeout(this.clearCacheTimer[id]);
        this.clearCacheTimer[id] = setTimeout(() => {
            this.getMoName.cache.delete(id);
        }, this.cacheLifeSpan);
    }
}
MoNameCacheService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MoNameCacheService, deps: [{ token: i1.InventoryService }], target: i0.ɵɵFactoryTarget.Injectable });
MoNameCacheService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MoNameCacheService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MoNameCacheService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.InventoryService }]; } });
