// @c8y/ngx-components/esm2020/core/common/should-show-mo.pipe.mjs
import { Pipe } from '@angular/core';
import { ManagedObjectType } from './managed-object-type';
import { DeviceService } from './device.service';
import { GroupService } from './group.service';
import * as i0 from "@angular/core";
import * as i1 from "./group.service";
export class ShouldShowMoPipe {
    constructor(groupService) {
        this.groupService = groupService;
    }
    transform(mo, option) {
        if (option === ManagedObjectType.DEVICE) {
            return DeviceService.isAnyDevice(mo);
        }
        if (option === ManagedObjectType.GROUP) {
            return this.groupService.isGroup(mo) || this.groupService.isDynamicGroup(mo);
        }
        return false;
    }
}
ShouldShowMoPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ShouldShowMoPipe, deps: [{ token: i1.GroupService }], target: i0.ɵɵFactoryTarget.Pipe });
ShouldShowMoPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ShouldShowMoPipe, name: "shouldShowMo" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ShouldShowMoPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'shouldShowMo'
                }]
        }], ctorParameters: function () { return [{ type: i1.GroupService }]; } });
