// @c8y/ngx-components/esm2020/core/common/stringify-object.pipe.mjs
import { Pipe } from '@angular/core';
import { DatePipe } from './date.pipe';
import * as i0 from "@angular/core";
import * as i1 from "./date.pipe";
export class StringifyObjectPipe {
    constructor(c8yDatePipe) {
        this.c8yDatePipe = c8yDatePipe;
    }
    transform(value) {
        if (!value) {
            return '';
        }
        if (typeof value === 'object') {
            return JSON.stringify(value);
        }
        if (this.checkDateValidity(value)) {
            return this.c8yDatePipe.transform(value);
        }
        return value;
    }
    checkDateValidity(value) {
        if (!value)
            return false;
        const isNumericOrFloating = /^-?\d+(\.\d+)?$/.test(value);
        const parsedDate = new Date(value);
        const isValidDate = !isNaN(Number(parsedDate)) && Object.prototype.toString.call(parsedDate) === '[object Date]';
        return isValidDate && !isNumericOrFloating;
    }
}
StringifyObjectPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StringifyObjectPipe, deps: [{ token: i1.DatePipe }], target: i0.ɵɵFactoryTarget.Pipe });
StringifyObjectPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StringifyObjectPipe, name: "stringifyObject" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StringifyObjectPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'stringifyObject' }]
        }], ctorParameters: function () { return [{ type: i1.DatePipe }]; } });
