// @c8y/ngx-components/esm2020/core/common/tabset-aria.directive.mjs
import { Directive, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Adds necessary attributes to ngx-bootstrap's tabset markup to comply with a11y requirements.
 */
export class TabsetAriaDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
    ngAfterViewInit() {
        let owned = '';
        this.elementRef.nativeElement.querySelectorAll('[role="tab"]')?.forEach((tab, idx) => {
            tab.setAttribute('id', `tab${idx}`);
            owned = owned.concat(`tab${idx} `);
        });
        this.elementRef.nativeElement
            .querySelectorAll('li')
            ?.forEach(li => li.setAttribute('role', 'presentation'));
        this.elementRef.nativeElement
            .querySelector('[role="tablist"]')
            ?.setAttribute('aria-owns', owned);
    }
}
TabsetAriaDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsetAriaDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TabsetAriaDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: TabsetAriaDirective, selector: "tabset", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TabsetAriaDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tabset'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; } });
