// @c8y/ngx-components/esm2020/core/common/textarea-autoresize.directive.mjs
import { Directive, HostListener, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
export class TextareaAutoresizeDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.focused = false;
        this.resizeObserver = new window.ResizeObserver(() => !this.focused && this.resize());
    }
    onInput() {
        this.resize();
    }
    onFocus() {
        this.focused = true;
    }
    onBlur() {
        this.focused = false;
    }
    ngOnInit() {
        this.resizeObserver.observe(this.elementRef?.nativeElement);
        setTimeout(() => this.resize());
    }
    ngOnDestroy() {
        this.resizeObserver.unobserve(this.elementRef?.nativeElement);
    }
    resize() {
        this.elementRef.nativeElement.style.height = '32px';
        this.elementRef.nativeElement.style.height = this.elementRef?.nativeElement.scrollHeight + 'px';
    }
}
TextareaAutoresizeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TextareaAutoresizeDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TextareaAutoresizeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: TextareaAutoresizeDirective, selector: "[c8y-textarea-autoresize]", host: { listeners: { ":input": "onInput()", ":focus": "onFocus()", ":blur": "onBlur()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TextareaAutoresizeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[c8y-textarea-autoresize]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { onInput: [{
                type: HostListener,
                args: [':input']
            }], onFocus: [{
                type: HostListener,
                args: [':focus']
            }], onBlur: [{
                type: HostListener,
                args: [':blur']
            }] } });
