// @c8y/ngx-components/esm2020/core/common/zip.service.mjs
import { Injectable } from '@angular/core';
import { Observable, throwError } from 'rxjs';
import { concatMap, map, switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class ZipService {
    getEntries(zipFile) {
        return new Observable(subscriber => {
            const reader = new zip.BlobReader(zipFile);
            zip.createReader(reader, zipReader => {
                zipReader.getEntries(entries => {
                    subscriber.next(entries);
                    subscriber.complete();
                });
            }, message => {
                subscriber.error({ message });
            });
        });
    }
    getData(entry) {
        const data = new Observable(subscriber => {
            const writer = new zip.BlobWriter();
            entry.getData(writer, blob => {
                subscriber.next(blob);
                subscriber.complete();
            });
        });
        return data;
    }
    getJsonData(zipFile, fileEntryFilter) {
        return this.getEntries(zipFile).pipe(map(entries => {
            const entryFile = entries.find(entry => entry.filename === fileEntryFilter.filename);
            if (!entryFile) {
                throwError(`File not found ${JSON.stringify(fileEntryFilter)}`);
            }
            return entryFile;
        }), concatMap(entry => this.getData(entry)), switchMap(async (blob) => await blob.text()), map(text => JSON.parse(text)));
    }
}
ZipService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ZipService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ZipService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ZipService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ZipService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
