// @c8y/ngx-components/esm2020/core/docs/support-outlet/support-outlet.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../../common';
import { ProductExperienceModule } from '../../product-experience';
import { hookDrawer } from '../../drawer';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { SupportOutletComponent } from './support-outlet.component';
import { UserModule } from '../../user';
import * as i0 from "@angular/core";
export class SupportOutletModule {
}
SupportOutletModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SupportOutletModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SupportOutletModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SupportOutletModule, declarations: [SupportOutletComponent], imports: [CommonModule, CollapseModule, ProductExperienceModule, UserModule], exports: [SupportOutletComponent] });
SupportOutletModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SupportOutletModule, providers: [
        hookDrawer({
            component: SupportOutletComponent,
            position: 'right',
            priority: 40,
            id: 'support'
        })
    ], imports: [CommonModule, CollapseModule, ProductExperienceModule, UserModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SupportOutletModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CollapseModule, ProductExperienceModule, UserModule],
                    declarations: [SupportOutletComponent],
                    exports: [SupportOutletComponent],
                    providers: [
                        hookDrawer({
                            component: SupportOutletComponent,
                            position: 'right',
                            priority: 40,
                            id: 'support'
                        })
                    ]
                }]
        }] });
