// @c8y/ngx-components/esm2020/core/list-group/list-group.component.mjs
import { Component, HostBinding, ContentChild } from '@angular/core';
import { ListItemTimelineComponent } from './list-item-timeline.component';
import * as i0 from "@angular/core";
/**
 * List groups are flexible and powerful components for displaying
 * both simple lists of elements and complex ones with custom content.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li>
 *     Plain <strong>Text</strong>
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-icon [icon]="'rocket'"></c8y-li-icon>
 *     Icon
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-radio></c8y-li-radio>
 *     Radio
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-checkbox></c8y-li-checkbox>
 *     Checkbox
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-footer>
 *       Footer
 *     </c8y-li-footer>
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-action icon="rocket">
 *       Launch
 *     </c8y-li-action>
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-collapse>
 *       Collapsed content
 *     </c8y-li-collapse>
 *   </c8y-li>
 * </c8y-list-group>
 * ```
 */
export class ListGroupComponent {
    constructor() {
        /**
         * @ignore
         */
        this.timeline = false;
    }
    /**
     * @ignore
     */
    ngAfterContentChecked() {
        this.timeline = !!this.timelineComponent;
    }
}
ListGroupComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListGroupComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
ListGroupComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ListGroupComponent, selector: "c8y-list-group", host: { properties: { "class.c8y-list--timeline": "this.timeline" }, classAttribute: "c8y-list__group" }, queries: [{ propertyName: "timelineComponent", first: true, predicate: ListItemTimelineComponent, descendants: true }], ngImport: i0, template: "<ng-content></ng-content>\n" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListGroupComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-list-group', host: {
                        class: 'c8y-list__group'
                    }, template: "<ng-content></ng-content>\n" }]
        }], propDecorators: { timeline: [{
                type: HostBinding,
                args: ['class.c8y-list--timeline']
            }], timelineComponent: [{
                type: ContentChild,
                args: [ListItemTimelineComponent, { static: false }]
            }] } });
