// @c8y/ngx-components/esm2020/core/list-group/list-item-action.component.mjs
import { Component, EventEmitter, forwardRef, HostListener, Input, Output, ViewChild } from '@angular/core';
import { assignWith } from 'lodash-es';
import { GainsightService } from '../product-experience/gainsight.service';
import { PRODUCT_EXPERIENCE_EVENT_SOURCE } from '../product-experience/product-experience.model';
import * as i0 from "@angular/core";
import * as i1 from "../product-experience/gainsight.service";
import * as i2 from "../common/icon.directive";
import * as i3 from "@angular/common";
import * as i4 from "../i18n/c8y-translate.pipe";
/**
 * Actions are functions that can be executed on the
 * current list item. They are displayed next to the
 * item and can be grouped in a dropdown.
 *
 * @example
 * ```html
 * <c8y-li>
 *   <c8y-li-action (click)="deleteDevice(device.id)" icon="times">
 *     Delete
 *   </c8y-li-action>
 *   <c8y-li-action (click)="(false)" icon="rocket">
 *     Launch to space
 *   </c8y-li-action>
 * </c8y-li>
 * ```
 */
export class ListItemActionComponent {
    constructor(gainsightService) {
        this.gainsightService = gainsightService;
        /**
         * An EventEmitter to determine what to do if the
         * action is triggered.
         */
        this.click = new EventEmitter();
    }
    onClick(event) {
        const element = event.closest('[title]') || event.closest('[uib-tooltip]');
        if (!element) {
            return;
        }
        const eventName = this.productExperienceEvent?.eventName || 'listItemAction';
        const actionName = element.getAttribute('title') || element.getAttribute('uib-tooltip');
        const action = this.gainsightService.translateToEnglish(actionName);
        // Assign url and action to event data only if they are not already defined
        const eventData = assignWith(this.productExperienceEvent?.data, {
            url: element.baseURI,
            action
        }, (targetVal, sourceVal) => sourceVal ?? targetVal);
        this.gainsightService.triggerEvent(eventName, eventData);
    }
}
ListItemActionComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListItemActionComponent, deps: [{ token: i1.GainsightService }], target: i0.ɵɵFactoryTarget.Component });
ListItemActionComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ListItemActionComponent, selector: "c8y-list-item-action, c8y-li-action", inputs: { label: "label", icon: "icon" }, outputs: { click: "click" }, host: { listeners: { "click": "onClick($event.target)" } }, providers: [
        {
            provide: PRODUCT_EXPERIENCE_EVENT_SOURCE,
            useExisting: forwardRef(() => ListItemActionComponent)
        }
    ], viewQueries: [{ propertyName: "template", first: true, predicate: ["template"], descendants: true }], ngImport: i0, template: "<ng-template #template>\n  <li>\n    <button\n      title=\"{{ label | translate }}\"\n      (click)=\"click.emit($event)\"\n      data-cy=\"c8y-li-action--btn\"\n    >\n      <i\n        [c8yIcon]=\"icon\"\n        *ngIf=\"icon\"\n      ></i>\n      &nbsp;\n      <span *ngIf=\"label\">{{ label | translate }}</span>\n      <ng-content></ng-content>\n    </button>\n  </li>\n</ng-template>\n", dependencies: [{ kind: "directive", type: i2.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "pipe", type: i4.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListItemActionComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-list-item-action, c8y-li-action', providers: [
                        {
                            provide: PRODUCT_EXPERIENCE_EVENT_SOURCE,
                            useExisting: forwardRef(() => ListItemActionComponent)
                        }
                    ], template: "<ng-template #template>\n  <li>\n    <button\n      title=\"{{ label | translate }}\"\n      (click)=\"click.emit($event)\"\n      data-cy=\"c8y-li-action--btn\"\n    >\n      <i\n        [c8yIcon]=\"icon\"\n        *ngIf=\"icon\"\n      ></i>\n      &nbsp;\n      <span *ngIf=\"label\">{{ label | translate }}</span>\n      <ng-content></ng-content>\n    </button>\n  </li>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i1.GainsightService }]; }, propDecorators: { template: [{
                type: ViewChild,
                args: ['template', { static: false }]
            }], label: [{
                type: Input
            }], icon: [{
                type: Input
            }], click: [{
                type: Output
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event.target']]
            }] } });
