// @c8y/ngx-components/esm2020/core/list-group/list-group.module.mjs
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { ListGroupComponent } from './list-group.component';
import { ListItemActionComponent } from './list-item-action.component';
import { ListItemBodyComponent } from './list-item-body.component';
import { ListItemCheckboxComponent } from './list-item-checkbox.component';
import { ListItemCollapseComponent } from './list-item-collapse.component';
import { ListItemDragHandleComponent } from './list-item-drag-handle.component';
import { ListItemFooterComponent } from './list-item-footer.component';
import { ListItemIconComponent } from './list-item-icon.component';
import { ListItemRadioComponent } from './list-item-radio.component';
import { ListItemTimelineComponent } from './list-item-timeline.component';
import { ListItemComponent } from './list-item.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/collapse";
/**
 * The list-group module allows you to compose different styles
 * of lists used in Cumulocity IoT. They can consist of plain
 * text, actions, icons and more.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li>Item 1<c8y-li>
 *   <c8y-li>Item 2<c8y-li>
 * </c8y-list-group>
 * ```
 */
export class ListGroupModule {
}
ListGroupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ListGroupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ListGroupModule, declarations: [ListGroupComponent,
        ListItemComponent,
        ListItemIconComponent,
        ListItemBodyComponent,
        ListItemFooterComponent,
        ListItemActionComponent,
        ListItemCollapseComponent,
        ListItemCheckboxComponent,
        ListItemRadioComponent,
        ListItemTimelineComponent,
        ListItemDragHandleComponent], imports: [CommonModule, i1.CollapseModule, BsDropdownModule], exports: [ListGroupComponent,
        ListItemComponent,
        ListItemIconComponent,
        ListItemBodyComponent,
        ListItemFooterComponent,
        ListItemActionComponent,
        ListItemCollapseComponent,
        ListItemCheckboxComponent,
        ListItemRadioComponent,
        ListItemTimelineComponent,
        ListItemDragHandleComponent] });
ListGroupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListGroupModule, imports: [CommonModule, CollapseModule.forRoot(), BsDropdownModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CollapseModule.forRoot(), BsDropdownModule],
                    declarations: [
                        ListGroupComponent,
                        ListItemComponent,
                        ListItemIconComponent,
                        ListItemBodyComponent,
                        ListItemFooterComponent,
                        ListItemActionComponent,
                        ListItemCollapseComponent,
                        ListItemCheckboxComponent,
                        ListItemRadioComponent,
                        ListItemTimelineComponent,
                        ListItemDragHandleComponent
                    ],
                    exports: [
                        ListGroupComponent,
                        ListItemComponent,
                        ListItemIconComponent,
                        ListItemBodyComponent,
                        ListItemFooterComponent,
                        ListItemActionComponent,
                        ListItemCollapseComponent,
                        ListItemCheckboxComponent,
                        ListItemRadioComponent,
                        ListItemTimelineComponent,
                        ListItemDragHandleComponent
                    ]
                }]
        }] });
