// @c8y/ngx-components/esm2020/repository/configuration/device-tab/configuration-repository-device-tab.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CoreModule, ViewContext, gettext, hookRoute } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { DeviceConfigurationGuard } from './device-configuration.guard';
import { DeviceConfigurationComponent } from './device-configuration.component';
import { ConfigurationPreviewComponent } from './configuration-preview.component';
import { ConfigurationFilterPipe } from './configuration-filter.pipe';
import { DeviceConfigurationService } from './device-configuration.service';
import { DeviceConfigurationListComponent } from './device-configuration-list.component';
import { TextBasedConfigurationComponent } from './text-based-configuration.component';
import { SaveToRepositoryComponent } from './save-to-repository.component';
import { SourceCodePreviewComponent } from './source-code-preview.component';
import { TabsModule } from 'ngx-bootstrap/tabs';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/tabs";
export class ConfigurationRepositoryDeviceTabModule {
    static forRoot() {
        return {
            ngModule: ConfigurationRepositoryDeviceTabModule,
            providers: [
                DeviceConfigurationGuard,
                hookRoute({
                    context: ViewContext.Device,
                    path: 'device-configuration',
                    component: DeviceConfigurationComponent,
                    label: gettext('Configuration'),
                    icon: 'gears',
                    priority: 600,
                    canActivate: [DeviceConfigurationGuard]
                })
            ]
        };
    }
}
ConfigurationRepositoryDeviceTabModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryDeviceTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ConfigurationRepositoryDeviceTabModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryDeviceTabModule, declarations: [DeviceConfigurationComponent,
        DeviceConfigurationListComponent,
        ConfigurationPreviewComponent,
        ConfigurationFilterPipe,
        SaveToRepositoryComponent,
        SourceCodePreviewComponent,
        TextBasedConfigurationComponent], imports: [CommonModule,
        CoreModule,
        SharedRepositoryModule,
        OperationDetailsModule, i1.TabsModule] });
ConfigurationRepositoryDeviceTabModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryDeviceTabModule, providers: [DeviceConfigurationService], imports: [CommonModule,
        CoreModule,
        SharedRepositoryModule,
        OperationDetailsModule,
        TabsModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryDeviceTabModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        CoreModule,
                        SharedRepositoryModule,
                        OperationDetailsModule,
                        TabsModule.forRoot()
                    ],
                    declarations: [
                        DeviceConfigurationComponent,
                        DeviceConfigurationListComponent,
                        ConfigurationPreviewComponent,
                        ConfigurationFilterPipe,
                        SaveToRepositoryComponent,
                        SourceCodePreviewComponent,
                        TextBasedConfigurationComponent
                    ],
                    entryComponents: [
                        DeviceConfigurationComponent,
                        SaveToRepositoryComponent,
                        TextBasedConfigurationComponent
                    ],
                    providers: [DeviceConfigurationService]
                }]
        }] });
