// @c8y/ngx-components/esm2020/core/docs/legal-notices-outlet/legal-notices-outlet.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../../common/common.module';
import { LegalNoticesOutletComponent } from './legal-notices-outlet.component';
import { hookDrawer } from '../../drawer/drawer.service';
import { ProductExperienceModule } from '../../product-experience/product-experience.module';
import * as i0 from "@angular/core";
export class LegalNoticesOutletModule {
}
LegalNoticesOutletModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LegalNoticesOutletModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LegalNoticesOutletModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: LegalNoticesOutletModule, declarations: [LegalNoticesOutletComponent], imports: [CommonModule, ProductExperienceModule] });
LegalNoticesOutletModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LegalNoticesOutletModule, providers: [
        hookDrawer({
            component: LegalNoticesOutletComponent,
            position: 'right',
            priority: 40,
            id: 'legalNotices'
        })
    ], imports: [CommonModule, ProductExperienceModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LegalNoticesOutletModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ProductExperienceModule],
                    declarations: [LegalNoticesOutletComponent],
                    providers: [
                        hookDrawer({
                            component: LegalNoticesOutletComponent,
                            position: 'right',
                            priority: 40,
                            id: 'legalNotices'
                        })
                    ]
                }]
        }] });
