// @c8y/ngx-components/esm2020/protocol-lwm2m/components/post-registration-operations/lwm2m-post-operations.service.mjs
import { Injectable } from '@angular/core';
import { FetchClient } from '@c8y/client';
import { Lwm2mClientService } from '@c8y/ngx-components/protocol-lwm2m/services';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
export class LWM2MPostOperationsParametersService extends Lwm2mClientService {
    constructor(client) {
        super(client);
        this.detailUrl = '/postRegistrationOptions';
    }
    get() {
        return super.fetch(this.detailUrl);
    }
    put(data) {
        const headers = { 'content-type': 'application/json', accept: 'application/json' };
        return super.fetch(this.detailUrl, {
            method: 'PUT',
            body: JSON.stringify(data),
            headers: headers
        });
    }
}
LWM2MPostOperationsParametersService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LWM2MPostOperationsParametersService, deps: [{ token: i1.FetchClient }], target: i0.ɵɵFactoryTarget.Injectable });
LWM2MPostOperationsParametersService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LWM2MPostOperationsParametersService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LWM2MPostOperationsParametersService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.FetchClient }]; } });
