// @c8y/ngx-components/esm2020/ecosystem/packages/add-package.component.mjs
import { Component } from '@angular/core';
import { ApplicationAvailability, ApplicationType } from '@c8y/client';
import { EcosystemService } from '@c8y/ngx-components/ecosystem/shared';
import { AppStateService, gettext } from '@c8y/ngx-components';
import { PackageAvailabilityService } from './package-availability.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components/ecosystem/shared";
import * as i2 from "@c8y/ngx-components";
import * as i3 from "./package-availability.service";
export class AddPackageComponent {
    constructor(ecosystemService, appState, packageAvailability) {
        this.ecosystemService = ecosystemService;
        this.appState = appState;
        this.packageAvailability = packageAvailability;
        this.headerText = gettext('Add extension package');
        this.successText = gettext('Extension package created');
        this.createPackageAppHandler = (f) => this.createPackageApp(f);
        this.uploadPackageHandler = (f, app) => this.uploadPackage(f, app);
    }
    async createPackageApp(file) {
        await this.ecosystemService.isValidAppType(file, ApplicationType.HOSTED);
        const applicationPackage = await this.ecosystemService.createAppForArchive(file, true);
        return await this.packageAvailability.askIfAvailabilityShouldBeSetTo(applicationPackage, ApplicationAvailability.SHARED);
    }
    uploadPackage(file, pckg) {
        return this.ecosystemService.uploadArchiveToApp(file, pckg, true);
    }
}
AddPackageComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddPackageComponent, deps: [{ token: i1.EcosystemService }, { token: i2.AppStateService }, { token: i3.PackageAvailabilityService }], target: i0.ɵɵFactoryTarget.Component });
AddPackageComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: AddPackageComponent, selector: "c8y-add-package", ngImport: i0, template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'big-parcel'"
    [successText]="successText"
    [createApplicationHandler]="createPackageAppHandler"
    [uploadApplicationHandler]="uploadPackageHandler"
  ></c8y-add-application>`, isInline: true, dependencies: [{ kind: "component", type: i1.AddApplicationComponent, selector: "c8y-add-application", inputs: ["headerText", "headerIcon", "successText", "createApplicationHandler", "uploadApplicationHandler", "canGoBack"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddPackageComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-add-package',
                    template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'big-parcel'"
    [successText]="successText"
    [createApplicationHandler]="createPackageAppHandler"
    [uploadApplicationHandler]="uploadPackageHandler"
  ></c8y-add-application>`
                }]
        }], ctorParameters: function () { return [{ type: i1.EcosystemService }, { type: i2.AppStateService }, { type: i3.PackageAvailabilityService }]; } });
