// @c8y/ngx-components/esm2020/ecosystem/packages/package-availability.service.mjs
import { Injectable } from '@angular/core';
import { ApplicationAvailability, ApplicationService } from '@c8y/client';
import { AlertService, AppStateService, GainsightService, ModalService, gettext } from '@c8y/ngx-components';
import { PRODUCT_EXPERIENCE } from '@c8y/ngx-components/ecosystem/shared';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/client";
export class PackageAvailabilityService {
    constructor(appState, alert, modal, application, gainsightService) {
        this.appState = appState;
        this.alert = alert;
        this.modal = modal;
        this.application = application;
        this.gainsightService = gainsightService;
        this.CURRENT_LOCATION = location.href;
    }
    async askIfAvailabilityShouldBeSetTo(applicationPackage, availability) {
        // availability does not matter for tenant that do not have or can create subtenants
        if (!this.appState.currentTenant.value?.allowCreateTenants) {
            return applicationPackage;
        }
        try {
            await this.openAvailabilityModal(availability);
        }
        catch {
            // modal closed
            return applicationPackage;
        }
        try {
            const { data: app } = await this.application.updateApplicationAvailability(applicationPackage, availability);
            this.alert.success(gettext('Updated package availability.'));
            this.gainsightService.triggerEvent(PRODUCT_EXPERIENCE.APPLICATIONS.EVENTS.AVAILABILITY, {
                action: PRODUCT_EXPERIENCE.APPLICATIONS.ACTIONS.AVAILABILITY_CHANGE,
                result: availability.toString().toLocaleLowerCase(),
                url: this.CURRENT_LOCATION
            });
            return app;
        }
        catch (e) {
            this.alert.warning(gettext('Failed to set package availability.'));
            this.alert.addServerFailure(e);
            this.gainsightService.triggerEvent(PRODUCT_EXPERIENCE.APPLICATIONS.EVENTS.AVAILABILITY, {
                action: PRODUCT_EXPERIENCE.APPLICATIONS.ACTIONS.AVAILABILITY_CHANGE,
                result: PRODUCT_EXPERIENCE.APPLICATIONS.RESULTS.SERVER_FAILURE,
                url: this.CURRENT_LOCATION
            });
            return applicationPackage;
        }
    }
    async openAvailabilityModal(availability) {
        const body = this.getConfirmationBody(availability);
        await this.modal.confirm(gettext('Package availability'), body, 'info', {
            ok: gettext('Yes'),
            cancel: gettext('No')
        });
    }
    getConfirmationBody(availability) {
        switch (availability) {
            case ApplicationAvailability.SHARED:
                return gettext('Do you want to set the package availability to "Shared"? This will make the package available to all subtenants without explicitly subscribing the package.');
            case ApplicationAvailability.MARKET:
                return gettext('Do you want to set the package availability to "Market"? This will make the package available to your own tenant and needs to be subscribed to subtenants individually.');
            case ApplicationAvailability.PRIVATE:
                return gettext('Do you want to set the package availability to "Private"? This will make the package only available to your own tenant.');
            default:
                return '';
        }
    }
}
PackageAvailabilityService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PackageAvailabilityService, deps: [{ token: i1.AppStateService }, { token: i1.AlertService }, { token: i1.ModalService }, { token: i2.ApplicationService }, { token: i1.GainsightService }], target: i0.ɵɵFactoryTarget.Injectable });
PackageAvailabilityService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PackageAvailabilityService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PackageAvailabilityService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.AppStateService }, { type: i1.AlertService }, { type: i1.ModalService }, { type: i2.ApplicationService }, { type: i1.GainsightService }]; } });
