// @c8y/ngx-components/esm2020/core/help/help.service.mjs
import { Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { BehaviorSubject } from 'rxjs';
import { OptionsService } from '../common/options.service';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
import * as i2 from "../common/options.service";
export class HelpService {
    constructor(translateService, options) {
        this.translateService = translateService;
        this.options = options;
        this.contextHelp = new BehaviorSubject(false);
        this.isWarningAlertViewed = false;
        this.SUPPORTED_LANGUAGES = ['en'];
        this.handleContextHelp();
    }
    isSupportedLanguage() {
        return this.SUPPORTED_LANGUAGES.indexOf(this.translateService.currentLang) > -1;
    }
    handleContextHelp() {
        const option = this.getContextHelpOption();
        if (option) {
            this.contextHelp.next(option);
            return;
        }
        this.contextHelp.next(false);
    }
    getContextHelpOption() {
        return this.options.contextHelp;
    }
}
HelpService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpService, deps: [{ token: i1.TranslateService }, { token: i2.OptionsService }], target: i0.ɵɵFactoryTarget.Injectable });
HelpService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.TranslateService }, { type: i2.OptionsService }]; } });
