// @c8y/ngx-components/esm2020/core/help/help.module.mjs
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ActionBarModule } from '../action-bar/action-bar.module';
import { ProductExperienceModule } from '../product-experience/product-experience.module';
import { CommonModule } from '../common/common.module';
import { HelpComponent } from './help.component';
import { HelpService } from './help.service';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/collapse";
import * as i2 from "ngx-bootstrap/tooltip";
/**
 * Provides components which allow to display help elements
 * inside the app.
 * @exports HelpComponent A component which shows a context help in
 *                        the action bar.
 */
export class HelpModule {
}
HelpModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HelpModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: HelpModule, declarations: [HelpComponent], imports: [CommonModule, i1.CollapseModule, i2.TooltipModule, ActionBarModule,
        ProductExperienceModule], exports: [HelpComponent] });
HelpModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpModule, providers: [HelpService], imports: [CommonModule,
        CollapseModule.forRoot(),
        TooltipModule.forRoot(),
        ActionBarModule,
        ProductExperienceModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [HelpComponent],
                    imports: [
                        CommonModule,
                        CollapseModule.forRoot(),
                        TooltipModule.forRoot(),
                        ActionBarModule,
                        ProductExperienceModule
                    ],
                    exports: [HelpComponent],
                    entryComponents: [HelpComponent],
                    providers: [HelpService]
                }]
        }] });
