// @c8y/ngx-components/esm2020/core/file-picker/file-picker.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FilePickerComponent } from './file-picker.component';
import { DropAreaModule } from '../drop-area/drop-area.module';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { PopoverModule } from 'ngx-bootstrap/popover';
import * as i0 from "@angular/core";
/**
 * The angular module definition for a drop-area.
 * @exports FilePickerComponent
 */
export class FilePickerModule {
}
FilePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FilePickerModule, declarations: [FilePickerComponent], imports: [C8yFormsModule,
        CommonModule,
        DropAreaModule,
        FormsModule,
        C8yCommonModule,
        PopoverModule], exports: [FilePickerComponent] });
FilePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePickerModule, imports: [C8yFormsModule,
        CommonModule,
        DropAreaModule,
        FormsModule,
        C8yCommonModule,
        PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        C8yFormsModule,
                        CommonModule,
                        DropAreaModule,
                        FormsModule,
                        C8yCommonModule,
                        PopoverModule
                    ],
                    exports: [FilePickerComponent],
                    declarations: [FilePickerComponent]
                }]
        }] });
