// @c8y/ngx-components/esm2020/core/bootstrap/bootstrap.module.mjs
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ActionBarModule } from '../action-bar/action-bar.module';
import { AlertModule } from '../alert/alert.module';
import { CommonModule } from '../common/common.module';
import { DrawerModule } from '../drawer/drawer.module';
import { HeaderModule } from '../header/header.module';
import { LoginModule } from '../login/login.module';
import { ModalModule } from '../modal/modal.module';
import { NavigatorModule } from '../navigator/navigator.module';
import { SetupModule } from '../setup/setup.module';
import { TabsModule } from '../tabs/tabs.module';
import { BootstrapComponent } from './bootstrap.component';
import { CookieBannerPreferencesModalComponent } from './cookie-banner/cookie-banner-preferences-modal/cookie-banner-preferences-modal.component';
import { CookieBannerComponent } from './cookie-banner/cookie-banner.component';
import { CookieBannerService } from './cookie-banner/cookie-banner.service';
import * as i0 from "@angular/core";
/**
 * The module which allows to bootstrap a custom application.
 * Use it as `bootstrap: []` component in your main app module
 * or as `<c8y-bootstrap>` in your main app.
 */
export class BootstrapModule {
}
BootstrapModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BootstrapModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BootstrapModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BootstrapModule, declarations: [BootstrapComponent, CookieBannerComponent, CookieBannerPreferencesModalComponent], imports: [FormsModule,
        CommonModule,
        HeaderModule,
        LoginModule,
        RouterModule,
        TabsModule,
        ActionBarModule,
        AlertModule,
        NavigatorModule,
        TooltipModule,
        ModalModule,
        SetupModule,
        DrawerModule], exports: [BootstrapComponent] });
BootstrapModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BootstrapModule, providers: [CookieBannerService], imports: [FormsModule,
        CommonModule,
        HeaderModule,
        LoginModule,
        RouterModule,
        TabsModule,
        ActionBarModule,
        AlertModule,
        NavigatorModule,
        TooltipModule,
        ModalModule,
        SetupModule,
        DrawerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BootstrapModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BootstrapComponent, CookieBannerComponent, CookieBannerPreferencesModalComponent],
                    imports: [
                        FormsModule,
                        CommonModule,
                        HeaderModule,
                        LoginModule,
                        RouterModule,
                        TabsModule,
                        ActionBarModule,
                        AlertModule,
                        NavigatorModule,
                        TooltipModule,
                        ModalModule,
                        SetupModule,
                        DrawerModule
                    ],
                    exports: [BootstrapComponent],
                    entryComponents: [BootstrapComponent, CookieBannerPreferencesModalComponent],
                    providers: [CookieBannerService]
                }]
        }] });
