// @c8y/ngx-components/esm2020/context-dashboard/cockpit-dashboard/cockpit-dashboard.module.mjs
import { NgModule } from '@angular/core';
import { ContextDashboardModule } from '../context-dashboard.module';
import { ContextDashboardComponent } from '../context-dashboard.component';
import { DeviceDashboardGuard } from '../device-dashboard.guard';
import { GroupDashboardGuard } from '../group-dashboard.guard';
import { ViewContext, hookNavigator, gettext, CoreModule } from '@c8y/ngx-components';
import { CockpitDashboardComponent } from './cockpit-dashboard.component';
import { COCKPIT_HOME_DASHBOARD_CONFIG } from './cockpit-dashboard.model';
import * as i0 from "@angular/core";
import * as i1 from "../context-dashboard.module";
export class CockpitDashboardModule {
    static config(config = {}) {
        return {
            ngModule: CockpitDashboardModule,
            providers: [
                {
                    provide: COCKPIT_HOME_DASHBOARD_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
CockpitDashboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CockpitDashboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CockpitDashboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: CockpitDashboardModule, declarations: [CockpitDashboardComponent], imports: [CoreModule, i1.ContextDashboardModule] });
CockpitDashboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CockpitDashboardModule, providers: [hookNavigator({ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' })], imports: [CoreModule,
        ContextDashboardModule.config({
            routes: [
                {
                    path: 'dashboard/:dashboardId',
                    component: ContextDashboardComponent,
                    canActivate: [DeviceDashboardGuard],
                    context: ViewContext.Device
                },
                {
                    path: 'dashboard/:dashboardId',
                    component: ContextDashboardComponent,
                    canActivate: [GroupDashboardGuard],
                    context: ViewContext.Group
                },
                {
                    path: '',
                    component: CockpitDashboardComponent
                }
            ]
        })] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CockpitDashboardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        ContextDashboardModule.config({
                            routes: [
                                {
                                    path: 'dashboard/:dashboardId',
                                    component: ContextDashboardComponent,
                                    canActivate: [DeviceDashboardGuard],
                                    context: ViewContext.Device
                                },
                                {
                                    path: 'dashboard/:dashboardId',
                                    component: ContextDashboardComponent,
                                    canActivate: [GroupDashboardGuard],
                                    context: ViewContext.Group
                                },
                                {
                                    path: '',
                                    component: CockpitDashboardComponent
                                }
                            ]
                        })
                    ],
                    declarations: [CockpitDashboardComponent],
                    providers: [hookNavigator({ label: gettext('Home'), icon: 'home', priority: 10000, path: '/' })]
                }]
        }] });
