// @c8y/ngx-components/esm2020/ecosystem/ecosystem-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { Permissions } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class EcosystemNavigationFactory {
    constructor(permissions) {
        this.permissions = permissions;
        this.navs = [];
        this.showNavigatorNodes = this.permissions.hasAnyRole([
            'ROLE_APPLICATION_MANAGEMENT_READ',
            'ROLE_APPLICATION_MANAGEMENT_ADMIN'
        ]);
    }
    async get() {
        /*
    The NavigatorNodes are added in AngularJS code to keep the Quick links widget working.
    After the Quick links widget is migrated to Angular, the logic of adding the NavigatorNodes can be migrated as well.
        if (this.navs.length === 0 && this.showNavigatorNodes) {
          const applicationsNode: NavigatorNode = new NavigatorNode({
            label: gettext('Applications'),
            icon: 'c8y-modules',
            path: '/ecosystem',
            priority: 300,
            routerLinkExact: false
          });
    
          const microservicesNode = new NavigatorNode({
            label: gettext('Microservices'),
            icon: 'microchip',
            path: '/microservices',
            priority: 200,
            routerLinkExact: false
          });
    
          this.navs.push(
            new NavigatorNode({
              label: gettext('Ecosystem'),
              icon: 'c8y-atom',
              priority: 3200,
              children: [applicationsNode, microservicesNode]
            })
          );
        }
    */
        return this.navs;
    }
}
EcosystemNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: EcosystemNavigationFactory, deps: [{ token: i1.Permissions }], target: i0.ɵɵFactoryTarget.Injectable });
EcosystemNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: EcosystemNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: EcosystemNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Permissions }]; } });
