// @c8y/ngx-components/esm2020/ecosystem/ecosystem-tabs.factory.mjs
import { Injectable } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import { Router } from '@angular/router';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class EcosystemTabs {
    constructor(router) {
        this.router = router;
        this.ecosystemTabs = [
            {
                icon: 'c8y-modules',
                priority: 1000,
                label: gettext('Applications'),
                path: 'ecosystem/application/applications'
            },
            {
                icon: 'tab',
                priority: 998,
                label: gettext('Features'),
                path: 'ecosystem/application/features'
            }
        ];
    }
    get() {
        if (this.router.url === '/ecosystem/application/applications' ||
            this.router.url === '/ecosystem/application/features') {
            return this.ecosystemTabs;
        }
        return [];
    }
}
EcosystemTabs.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: EcosystemTabs, deps: [{ token: i1.Router }], target: i0.ɵɵFactoryTarget.Injectable });
EcosystemTabs.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: EcosystemTabs });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: EcosystemTabs, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Router }]; } });
