// @c8y/ngx-components/esm2020/core/version/backend-version.factory.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { of } from 'rxjs';
import { distinctUntilChanged, filter, map, shareReplay } from 'rxjs/operators';
import { AppStateService } from '../common';
import { gettext } from '../i18n';
import { VERSION_MODULE_CONFIG } from './version.model';
import * as i0 from "@angular/core";
import * as i1 from "../common";
/**
 * Will provide the backend versions of the tenant.
 * The system option 'system' -> 'version' is used for this.
 */
export class BackendVersionFactory {
    constructor(config, appState) {
        this.config = config;
        this.appState = appState;
        this.backendVersion$ = of([]);
        if (this.config?.disableBackendVersionFactory) {
            return;
        }
        this.backendVersion$ = this.appState.state$.pipe(map(state => state?.versions?.backend), filter(backendVersion => !!backendVersion), distinctUntilChanged(), map(backendVersion => this.buildBackendVersion(backendVersion)), shareReplay(1));
    }
    get() {
        return this.backendVersion$;
    }
    buildBackendVersion(backendVersion) {
        return { label: gettext('Backend'), version: backendVersion, priority: 20, type: 'SYSTEM' };
    }
}
BackendVersionFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BackendVersionFactory, deps: [{ token: VERSION_MODULE_CONFIG, optional: true }, { token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
BackendVersionFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BackendVersionFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BackendVersionFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [VERSION_MODULE_CONFIG]
                }] }, { type: i1.AppStateService }]; } });
