// @c8y/ngx-components/esm2020/core/version/version.service.mjs
import { Injectable, Injector } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { distinctUntilChanged, map, shareReplay } from 'rxjs/operators';
import { fromTriggerOnce, getInjectedHooks, sortByPriority, StateService, stateToFactory } from '../common';
import { HOOK_VERSION } from './version.model';
import { pick } from 'lodash-es';
import * as i0 from "@angular/core";
export class VersionService extends StateService {
    constructor(rootInjector) {
        super();
        this.rootInjector = rootInjector;
        this.refreshTrigger = new Subject();
        this.state$ = new BehaviorSubject(new Set());
        this.factories = [];
        this.injectors = [this.rootInjector];
        this.items$ = fromTriggerOnce(undefined, this.refreshTrigger, [
            getInjectedHooks(HOOK_VERSION, this.injectors),
            () => this.factories,
            stateToFactory(this.state$)
        ]).pipe(distinctUntilChanged(), map(versions => sortByPriority(versions)), shareReplay(1));
        this.nonHiddenItems$ = this.items$.pipe(map(versions => versions.filter(version => !version.hidden)));
    }
    get state() {
        return this.state$.value;
    }
    refresh() {
        this.refreshTrigger.next();
    }
    add(version) {
        this.state.add(version);
        this.emitNewState();
    }
    remove(version) {
        this.state.delete(version);
        this.emitNewState();
    }
    cleanUpVersions(versions) {
        return versions.map(version => pick(version, 'label', 'version', 'type', 'custom'));
    }
}
VersionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionService, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable });
VersionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i0.Injector }]; } });
