// @c8y/ngx-components/esm2020/core/version/version.module.mjs
import { NgModule } from '@angular/core';
import { ClipboardModule } from '../clipboard/clipboard.module';
import { HookProviderTypes } from '../common';
import { CommonModule } from '../common/common.module';
import { hookDrawer } from '../drawer/drawer.service';
import { PluginsModule } from '../plugins/plugins.module';
import { BackendVersionFactory } from './backend-version.factory';
import { VersionDetailsModalComponent } from './version-details-modal/version-details-modal.component';
import { VersionDrawerFactory } from './version-drawer.factory';
import { VersionListComponent } from './version-list/version-list.component';
import { hookVersion, VERSION_MODULE_CONFIG } from './version.model';
import { WebSDKPluginVersionFactory } from './websdk-plugin-version.factory';
import { WebSDKVersionFactory } from './websdk-version.factory';
import * as i0 from "@angular/core";
export class VersionModule {
    static config(config = {}) {
        return {
            ngModule: VersionModule,
            providers: [
                {
                    provide: VERSION_MODULE_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
VersionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
VersionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: VersionModule, declarations: [VersionListComponent, VersionDetailsModalComponent], imports: [CommonModule, PluginsModule, ClipboardModule], exports: [VersionListComponent] });
VersionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionModule, providers: [
        hookVersion(BackendVersionFactory, { providerType: HookProviderTypes.ExistingProvider }),
        hookVersion(WebSDKVersionFactory, { providerType: HookProviderTypes.ExistingProvider }),
        hookVersion(WebSDKPluginVersionFactory, { providerType: HookProviderTypes.ExistingProvider }),
        hookDrawer(VersionDrawerFactory)
    ], imports: [CommonModule, PluginsModule, ClipboardModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PluginsModule, ClipboardModule],
                    declarations: [VersionListComponent, VersionDetailsModalComponent],
                    exports: [VersionListComponent],
                    providers: [
                        hookVersion(BackendVersionFactory, { providerType: HookProviderTypes.ExistingProvider }),
                        hookVersion(WebSDKVersionFactory, { providerType: HookProviderTypes.ExistingProvider }),
                        hookVersion(WebSDKPluginVersionFactory, { providerType: HookProviderTypes.ExistingProvider }),
                        hookDrawer(VersionDrawerFactory)
                    ]
                }]
        }] });
