// @c8y/ngx-components/esm2020/core/version/version-drawer.factory.mjs
import { Injectable } from '@angular/core';
import { OptionsService } from '../common';
import { VersionListComponent } from './version-list/version-list.component';
import * as i0 from "@angular/core";
import * as i1 from "../common";
export class VersionDrawerFactory {
    constructor(options) {
        this.options = options;
        this.drawerItem = {
            component: VersionListComponent,
            position: 'right',
            priority: 80,
            id: 'platformInformation'
        };
    }
    get() {
        if (this.shouldShowPlatformInformation()) {
            return this.drawerItem;
        }
        return [];
    }
    shouldShowPlatformInformation() {
        const hidePlatformInformation = this.options.get('hidePlatformInformation');
        if (hidePlatformInformation === undefined) {
            return !this.options.get('hidePowered');
        }
        return !hidePlatformInformation;
    }
}
VersionDrawerFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionDrawerFactory, deps: [{ token: i1.OptionsService }], target: i0.ɵɵFactoryTarget.Injectable });
VersionDrawerFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionDrawerFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: VersionDrawerFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OptionsService }]; } });
