// @c8y/ngx-components/esm2020/core/version/websdk-plugin-version.factory.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { AppStateService } from '../common';
import { VERSION_MODULE_CONFIG } from './version.model';
import { of } from 'rxjs';
import { distinctUntilChanged, filter, map } from 'rxjs/operators';
import { PluginsService } from '../plugins';
import * as i0 from "@angular/core";
import * as i1 from "../common";
import * as i2 from "../plugins";
/**
 * Will provide the version of plugins installed to the current application.
 */
export class WebSDKPluginVersionFactory {
    constructor(config, appState, plugins) {
        this.config = config;
        this.appState = appState;
        this.plugins = plugins;
        this.webSDKPluginVersions$ = of([]);
        if (this.config?.disableWebSDKPluginVersionFactory) {
            return;
        }
        this.webSDKPluginVersions$ = this.appState.currentApplication.pipe(map(app => this.plugins.getMFRemotes(app)), filter(remotes => !!remotes), distinctUntilChanged(), map(remotes => this.getVersionsFromRemotes(remotes)));
    }
    get() {
        return this.webSDKPluginVersions$;
    }
    getVersionsFromRemotes(remotes) {
        const versions = new Array();
        for (const key of Object.keys(remotes)) {
            const modules = remotes[key];
            if (!modules?.length) {
                continue;
            }
            const [contextPath, version] = key.split('@', 2);
            if (!version) {
                continue;
            }
            const versionsBehindKey = modules.map(module => {
                return {
                    label: `${contextPath} => ${module}`,
                    version,
                    type: 'INSTALLED_MF_PLUGINS',
                    hideInRightDrawer: true,
                    custom: {
                        key,
                        contextPath,
                        module
                    }
                };
            });
            versions.push(...versionsBehindKey);
        }
        return versions;
    }
}
WebSDKPluginVersionFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WebSDKPluginVersionFactory, deps: [{ token: VERSION_MODULE_CONFIG, optional: true }, { token: i1.AppStateService }, { token: i2.PluginsService }], target: i0.ɵɵFactoryTarget.Injectable });
WebSDKPluginVersionFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WebSDKPluginVersionFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WebSDKPluginVersionFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [VERSION_MODULE_CONFIG]
                }] }, { type: i1.AppStateService }, { type: i2.PluginsService }]; } });
