// @c8y/ngx-components/esm2020/core/version/websdk-version.factory.mjs
import { Inject, Injectable, Optional } from '@angular/core';
import { AppStateService } from '../common';
import { VERSION_MODULE_CONFIG } from './version.model';
import * as i0 from "@angular/core";
import * as i1 from "../common";
/**
 * Will provide the application version (taken from package.json's `version` field) and
 * the installed versions of the packages`@c8y/ng1-modules` and `@c8y/ngx-components` if available.
 */
export class WebSDKVersionFactory {
    constructor(config, appState) {
        this.config = config;
        this.appState = appState;
        this.webSDKVersions = [];
        this.defaults = {
            type: 'WebSDK',
            priority: 0,
            hidden: true
        };
        if (this.config?.disableWebSDKVersionFactory) {
            return;
        }
        const { ng1, ngx, package: pkg } = this.appState.state.versions.ui;
        if (ng1) {
            this.webSDKVersions.push({
                ...this.defaults,
                label: '@c8y/ng1-modules',
                version: ng1
            });
        }
        if (ngx) {
            this.webSDKVersions.push({
                ...this.defaults,
                label: '@c8y/ngx-components',
                version: ngx
            });
        }
        if (pkg) {
            this.webSDKVersions.push({
                ...this.defaults,
                label: 'Frontend',
                version: pkg,
                hidden: false,
                priority: 50
            });
        }
    }
    get() {
        return this.webSDKVersions;
    }
}
WebSDKVersionFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WebSDKVersionFactory, deps: [{ token: VERSION_MODULE_CONFIG, optional: true }, { token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
WebSDKVersionFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WebSDKVersionFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WebSDKVersionFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [VERSION_MODULE_CONFIG]
                }] }, { type: i1.AppStateService }]; } });
