// @c8y/ngx-components/esm2020/protocol-lwm2m/services/lwm2m-client.service.mjs
import { Injectable } from '@angular/core';
import { FetchClient, Service } from '@c8y/client';
import { agentName, c8y_lwm2m } from '@c8y/ngx-components/protocol-lwm2m/model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
export class Lwm2mClientService extends Service {
    constructor(client) {
        super(client);
        this.listUrl = 'device';
        this.baseUrl = `/service/${agentName}`;
    }
    isLwm2mDevice(mo) {
        return (!!mo?.c8y_IsLwm2mDevice || mo?.c8y_DeviceTypes?.includes(c8y_lwm2m) || mo?.type === c8y_lwm2m);
    }
    detail(entityOrId, filter = {}) {
        return super.detail(entityOrId, filter);
    }
    update(entity) {
        return super.update(entity);
    }
    fetch(url, init) {
        return super.fetch(url, init);
    }
}
Lwm2mClientService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mClientService, deps: [{ token: i1.FetchClient }], target: i0.ɵɵFactoryTarget.Injectable });
Lwm2mClientService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mClientService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mClientService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.FetchClient }]; } });
