// @c8y/ngx-components/esm2020/core/list-display-switch/list-display-switch.module.mjs
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '../common/common.module';
import { ListDisplaySwitchComponent } from './list-display-switch.component';
import * as i0 from "@angular/core";
/**
 * The angular module definition for the list display switcher.
 * It allows to switch between a list and a grid view.
 * @exports ListDisplaySwitchComponent
 */
export class ListDisplaySwitchModule {
}
ListDisplaySwitchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListDisplaySwitchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ListDisplaySwitchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ListDisplaySwitchModule, declarations: [ListDisplaySwitchComponent], imports: [CommonModule, FormsModule], exports: [ListDisplaySwitchComponent] });
ListDisplaySwitchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListDisplaySwitchModule, imports: [CommonModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ListDisplaySwitchModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ListDisplaySwitchComponent],
                    imports: [CommonModule, FormsModule],
                    exports: [ListDisplaySwitchComponent]
                }]
        }] });
