// @c8y/ngx-components/esm2020/core/provider-configuration/provider-configuration.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { CoreModule } from '../core.module';
import { DynamicFormsModule } from '../dynamic-forms/dynamic-forms.module';
import { hookNavigator } from '../navigator';
import { hookRoute } from '../router';
import { hookTab } from '../tabs';
import { hookDynamicProviderConfig } from './provider-configuration-hook';
import { ProviderConfigurationNodeFactory } from './provider-configuration-node.factory';
import { ProviderConfigurationRouteFactory } from './provider-configuration-route.factory';
import { ProviderConfigurationTabFactory } from './provider-configuration-tab.factory';
import { ProviderConfigurationComponent } from './provider-configuration.component';
import * as i0 from "@angular/core";
export class ProviderConfigurationModule {
    static config(config) {
        return {
            ngModule: ProviderConfigurationModule,
            providers: [hookDynamicProviderConfig(config)]
        };
    }
}
ProviderConfigurationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ProviderConfigurationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationModule, declarations: [ProviderConfigurationComponent], imports: [CoreModule, CommonModule, DynamicFormsModule], exports: [ProviderConfigurationComponent] });
ProviderConfigurationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationModule, providers: [
        hookRoute(ProviderConfigurationRouteFactory),
        hookNavigator(ProviderConfigurationNodeFactory),
        hookTab(ProviderConfigurationTabFactory)
    ], imports: [CoreModule, CommonModule, DynamicFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ProviderConfigurationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ProviderConfigurationComponent],
                    exports: [ProviderConfigurationComponent],
                    imports: [CoreModule, CommonModule, DynamicFormsModule],
                    entryComponents: [ProviderConfigurationComponent],
                    providers: [
                        hookRoute(ProviderConfigurationRouteFactory),
                        hookNavigator(ProviderConfigurationNodeFactory),
                        hookTab(ProviderConfigurationTabFactory)
                    ]
                }]
        }] });
