// @c8y/ngx-components/esm2020/core/action-bar/action-bar-item.component.mjs
import { Component, EnvironmentInjector, Input, ViewChild, Injector } from '@angular/core';
import { ActionBarService } from './action-bar.service';
import * as i0 from "@angular/core";
import * as i1 from "./action-bar.service";
/**
 * A <c8y-action-bar-item> can be added with content projection.
 * It allows to show items in three positions (left, right
 * or more).
 *
 * @example
 * ```html
 * <c8y-action-bar-item [placement]="'right'">
 *   <button class="btn btn-link" (click)="addRandom()"><i c8yIcon="plus-circle"></i> Add random</button>
 * </c8y-action-bar-item>
 * ```
 */
export class ActionBarItemComponent {
    /**
     * @ignore
     */
    constructor(actionBarService, _injector) {
        this.actionBarService = actionBarService;
        /**
         * Where the item should be shown. Use "more" to place
         * it in a dropdown on the right.
         */
        this.placement = 'left';
        /**
         * Priority for sorting the list. Higher priority means
         * an earlier position in the list.
         */
        this.priority = 0;
        /**
         * Adds additional classes to the item.
         */
        this.itemClass = '';
        this.injector = _injector;
    }
    /**
     * @ignore
     */
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.actionBarService.add(this));
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.actionBarService.remove(this);
    }
}
ActionBarItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionBarItemComponent, deps: [{ token: i1.ActionBarService }, { token: i0.EnvironmentInjector }], target: i0.ɵɵFactoryTarget.Component });
ActionBarItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ActionBarItemComponent, selector: "c8y-action-bar-item", inputs: { placement: "placement", priority: "priority", itemClass: "itemClass", injector: "injector", groupId: "groupId" }, viewQueries: [{ propertyName: "component", first: true, predicate: ["template"], descendants: true }], ngImport: i0, template: "<ng-template #template>\n  <li class=\"{{ itemClass }}\">\n    <ng-content></ng-content>\n  </li>\n</ng-template>\n" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionBarItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-action-bar-item', template: "<ng-template #template>\n  <li class=\"{{ itemClass }}\">\n    <ng-content></ng-content>\n  </li>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i1.ActionBarService }, { type: i0.EnvironmentInjector }]; }, propDecorators: { placement: [{
                type: Input
            }], priority: [{
                type: Input
            }], itemClass: [{
                type: Input
            }], injector: [{
                type: Input
            }], groupId: [{
                type: Input
            }], component: [{
                type: ViewChild,
                args: ['template', { static: false }]
            }] } });
