// @c8y/ngx-components/esm2020/core/action-bar/action-bar.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { ActionBarItemComponent } from './action-bar-item.component';
import { ActionBarComponent } from './action-bar.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * An action bar can be added with content projection or multi provider
 * extension. It allows to show them in three positions (left, right
 * or more). The actions placed here are always local actions to the
 * current route like "add device" or "lock dashboard".
 *
 * @example
 * ```html
 * <c8y-action-bar-item [placement]="'right'">
 *   <button class="btn btn-link" (click)="addRandom()"><i c8yIcon="plus-circle"></i> Add random</button>
 * </c8y-action-bar-item>
 * ```
 * @exports ActionBarComponent The outlet component
 * @exports ActionBarItemComponent The items to add with cp.
 */
export class ActionBarModule {
}
ActionBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ActionBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ActionBarModule, declarations: [ActionBarComponent, ActionBarItemComponent], imports: [BsDropdownModule, CommonModule, CollapseModule, i1.RouterModule], exports: [ActionBarComponent, ActionBarItemComponent] });
ActionBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionBarModule, imports: [BsDropdownModule, CommonModule, CollapseModule, RouterModule.forChild([])] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionBarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ActionBarComponent, ActionBarItemComponent],
                    imports: [BsDropdownModule, CommonModule, CollapseModule, RouterModule.forChild([])],
                    exports: [ActionBarComponent, ActionBarItemComponent]
                }]
        }] });
