// @c8y/ngx-components/esm2020/core/drawer/drawer.service.mjs
import { Injector, Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject, Subject } from 'rxjs';
import { distinctUntilChanged, map, shareReplay } from 'rxjs/operators';
import { allEntriesAreEqual, fromTriggerOnce, getInjectedHooks, removeDuplicatesIds, StateService, stateToFactory } from '../common';
import { InjectionToken } from '@angular/core';
import { hookGeneric } from '../common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * A hook to use for Multi Provider extension.
 * Consider using the `hookDrawer` function instead.
 */
const HOOK_DRAWER = new InjectionToken('HOOK_DRAWER');
/**
 * You can either provide a single `DrawerItem` as parameter:
 * ```typescript
 *  hookDrawer(...)
 * ```
 *
 * Or an array to directly register multiple:
 * ```typescript
 *  hookDrawer([...])
 * ```
 *
 * Or you provide a Service that implements `ExtensionFactory<DrawerItem>`
 * ```typescript
 *  export class MyDrawerFactory implements ExtensionFactory<DrawerItem> {...}
 *  ...
 *  hookDrawer(MyDrawerFactory)
 * ```
 * A typed alternative to `HOOK_DRAWER`.
 * @param drawer The `DrawerItem`'s or `ExtensionFactory` to be provided.
 * @returns An `Provider` to be provided in your module.
 */
export function hookDrawer(drawer, options) {
    return hookGeneric(drawer, HOOK_DRAWER, options);
}
export class DrawerService extends StateService {
    constructor(rootInjector, router) {
        super();
        this.rootInjector = rootInjector;
        this.router = router;
        /**
         * Refresh the extension factories subject.
         * @readonly
         */
        this.refreshTrigger = new Subject();
        this.state$ = new BehaviorSubject(new Set());
        /**
         * Additional factories that can be added by plugins.
         */
        this.factories = [];
        this.injectors = [this.rootInjector];
        this.items$ = fromTriggerOnce(router, this.refreshTrigger, [
            getInjectedHooks(HOOK_DRAWER, this.injectors),
            () => this.factories,
            stateToFactory(this.state$)
        ]).pipe(distinctUntilChanged(allEntriesAreEqual), map((drawerItems) => removeDuplicatesIds(drawerItems)), shareReplay(1));
    }
    /**
     * Returns the current state.
     * @readonly
     * @returns The current set of actions.
     */
    get state() {
        return this.state$.value;
    }
    getDrawersForPosition$(position) {
        return this.items$.pipe(map(items => items.filter(drawer => drawer.position === position)));
    }
    /**
     * Refresh the extension factories.
     */
    refresh() {
        this.refreshTrigger.next();
    }
    /**
     * Adds a new node to the navigator.
     * @param {DrawerItem} node Navigator node to add.
     */
    add(node) {
        this.state.add(node);
        this.emitNewState();
    }
    /**
     * Removes a node from the navigator.
     * @param {DrawerItem} node Navigator node to remove.
     */
    remove(node) {
        this.state.delete(node);
        this.emitNewState();
    }
}
DrawerService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DrawerService, deps: [{ token: i0.Injector }, { token: i1.Router }], target: i0.ɵɵFactoryTarget.Injectable });
DrawerService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DrawerService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DrawerService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i0.Injector }, { type: i1.Router }]; } });
