// @c8y/ngx-components/esm2020/core/drawer/drawer.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { DrawerOutletComponent } from './drawer-outlet/drawer-outlet.component';
import * as i0 from "@angular/core";
export class DrawerModule {
}
DrawerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DrawerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DrawerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DrawerModule, declarations: [DrawerOutletComponent], imports: [A11yModule, CommonModule], exports: [DrawerOutletComponent] });
DrawerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DrawerModule, imports: [A11yModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DrawerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [A11yModule, CommonModule],
                    declarations: [DrawerOutletComponent],
                    exports: [DrawerOutletComponent]
                }]
        }] });
