// @c8y/ngx-components/esm2020/upgrade/dashboard/component.factory.mjs
import { Injectable } from '@angular/core';
import { DynamicDatapointsResolver, DynamicManagedObjectResolver } from '@c8y/ngx-components';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetComponent } from './widget.component';
import { DeviceSelectorComponent } from './device-selector.component';
import { of } from 'rxjs';
import { shareReplay, switchMap, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./dashboard-bridge.service";
export class Ng1ComponentFactoryService {
    constructor(bridge) {
        this.bridge = bridge;
        this.components$ = of(1).pipe(switchMap(() => this.getComponents()), shareReplay(1), take(1));
    }
    get() {
        return this.components$;
    }
    async getComponents() {
        const ng1Components = await this.bridge.ng1Components.list();
        const components = ng1Components.map(cmp => {
            const resolve = {};
            if (cmp.options?.datapointAttributes && Array.isArray(cmp.options?.datapointAttributes)) {
                cmp.options?.datapointAttributes.forEach(attribute => (resolve[attribute] = DynamicDatapointsResolver));
            }
            if (cmp.options?.managedObjectAttributes &&
                Array.isArray(cmp.options?.managedObjectAttributes)) {
                cmp.options?.managedObjectAttributes.forEach(attribute => (resolve[attribute] = DynamicManagedObjectResolver));
            }
            const data = {
                settings: {
                    ...cmp.options,
                    configTemplateUrl: cmp.configTemplateUrl,
                    widgetComponent: cmp.widgetComponent,
                    configComponent: cmp.configComponent,
                    transformConfigWithContext: cmp.transformConfigWithContext || cmp.componentTransformConfigWithContext,
                    upgrade: true
                }
            };
            data.settings.templateUrl = cmp.templateUrl; // workaround as in object it creates build error
            data.displaySettings = cmp.displaySettings;
            return {
                id: cmp.name,
                label: cmp.nameDisplay || cmp.name,
                description: cmp.description,
                component: WidgetComponent,
                configComponent: WidgetComponent,
                previewImage: cmp.previewImage,
                errorStrategy: cmp.errorStrategy,
                resolve,
                data
            };
        });
        // add the device selector legacy component
        components.push({
            id: 'device.selector.legacy',
            label: 'Device selector',
            description: '',
            data: { settings: { noNewWidgets: true, upgrade: true } },
            component: DeviceSelectorComponent
        });
        return components;
    }
}
Ng1ComponentFactoryService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Ng1ComponentFactoryService, deps: [{ token: i1.DashboardBridgeService }], target: i0.ɵɵFactoryTarget.Injectable });
Ng1ComponentFactoryService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Ng1ComponentFactoryService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Ng1ComponentFactoryService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.DashboardBridgeService }]; } });
