// @c8y/ngx-components/esm2020/upgrade/dashboard/dashboard-upgrade.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule, NgZone, Optional } from '@angular/core';
import { Router } from '@angular/router';
import { ContextDashboardService } from '@c8y/ngx-components/context-dashboard';
import { Ng1ComponentFactoryService } from './component.factory';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetComponent } from './widget.component';
import { DeviceSelectorComponent } from './device-selector.component';
import { C8yTranslateModule, hookComponent } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import * as i0 from "@angular/core";
export class DashboardUpgradeModule {
}
DashboardUpgradeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DashboardUpgradeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DashboardUpgradeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DashboardUpgradeModule, declarations: [DeviceSelectorComponent, WidgetComponent], imports: [CommonModule, C8yTranslateModule, AssetSelectorModule] });
DashboardUpgradeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DashboardUpgradeModule, providers: [
        {
            provide: DashboardBridgeService,
            useFactory: dashboardBridgeFactory,
            deps: ['$injector', NgZone, Router, [new Optional(), ContextDashboardService]]
        },
        hookComponent(Ng1ComponentFactoryService)
    ], imports: [CommonModule, C8yTranslateModule, AssetSelectorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DashboardUpgradeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DeviceSelectorComponent, WidgetComponent],
                    entryComponents: [DeviceSelectorComponent, WidgetComponent],
                    imports: [CommonModule, C8yTranslateModule, AssetSelectorModule],
                    providers: [
                        {
                            provide: DashboardBridgeService,
                            useFactory: dashboardBridgeFactory,
                            deps: ['$injector', NgZone, Router, [new Optional(), ContextDashboardService]]
                        },
                        hookComponent(Ng1ComponentFactoryService)
                    ]
                }]
        }] });
export function dashboardBridgeFactory(injector, ngZone, router, contextDashboardService) {
    return new DashboardBridgeService(injector, ngZone, router, contextDashboardService);
}
