// @c8y/ngx-components/esm2020/upgrade/dashboard/device-selector.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import * as i0 from "@angular/core";
import * as i1 from "./dashboard-bridge.service";
export class DeviceSelectorComponent {
    constructor(dashboardBridgeService, el) {
        this.dashboardBridgeService = dashboardBridgeService;
        this.el = el;
    }
    async ngOnInit() {
        this.ng1Scope = this.dashboardBridgeService.instantiateDeviceSelector(this.el.nativeElement, this.config);
    }
    ngOnDestroy() {
        if (this.ng1Scope) {
            this.ng1Scope.$destroy();
        }
    }
}
DeviceSelectorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceSelectorComponent, deps: [{ token: i1.DashboardBridgeService }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
DeviceSelectorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: DeviceSelectorComponent, selector: "c8y-device-selector-legacy", ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceSelectorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-device-selector-legacy',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return [{ type: i1.DashboardBridgeService }, { type: i0.ElementRef }]; } });
