// @c8y/ngx-components/esm2020/upgrade/dashboard/widget.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef, Optional } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetsDashboardComponent } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "./dashboard-bridge.service";
export class WidgetComponent {
    constructor(dashboard, dashboardBridgeService, el) {
        this.dashboard = dashboard;
        this.dashboardBridgeService = dashboardBridgeService;
        this.el = el;
    }
    onBeforeSave() {
        this.ng1Scope.$broadcast('saveWidget');
        return true;
    }
    async ngOnInit() {
        if (this.config) {
            if (this.config.child) {
                this.config.child.alerts = this.alerts;
            }
            this.ng1Scope = await this.dashboardBridgeService.instantiateComponent(this.config, this.el.nativeElement);
            if (this.dashboard) {
                this.resizeSubscription = this.dashboard.onChangeDashboard.subscribe(() => {
                    this.ng1Scope.$broadcast('dashboardResize');
                });
            }
        }
    }
    ngOnDestroy() {
        if (this.ng1Scope) {
            this.ng1Scope.$destroy();
        }
        if (this.resizeSubscription) {
            this.resizeSubscription.unsubscribe();
        }
    }
}
WidgetComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetComponent, deps: [{ token: i1.WidgetsDashboardComponent, optional: true }, { token: i2.DashboardBridgeService }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
WidgetComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: WidgetComponent, selector: "c8y-widget-legacy", ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-widget-legacy',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return [{ type: i1.WidgetsDashboardComponent, decorators: [{
                    type: Optional
                }] }, { type: i2.DashboardBridgeService }, { type: i0.ElementRef }]; } });
