// @c8y/ngx-components/esm2020/operations/stepper-bulk-type-firmware/stepper-bulk-type-firmware.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule, gettext } from '@c8y/ngx-components';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { baseUrl, BulkOperationType, hookListBulkType } from '@c8y/ngx-components/operations/bulk-operations-service';
import { StepperFramesModule } from '@c8y/ngx-components/operations/stepper-frames';
import { SelectFirmwareComponent } from './select-firmware.component';
import { VersionOrPatchComponent } from './version-or-patch.component';
import { StepperBulkTypeFirmwareComponent } from './stepper-bulk-type-firmware.component';
import * as i0 from "@angular/core";
/** Module for the 'Firmware update' operation type stepper */
export class StepperBulkTypeFirmwareModule {
}
StepperBulkTypeFirmwareModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeFirmwareModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperBulkTypeFirmwareModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeFirmwareModule, declarations: [StepperBulkTypeFirmwareComponent,
        VersionOrPatchComponent,
        SelectFirmwareComponent], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        StepperFramesModule,
        SharedRepositoryModule], exports: [StepperBulkTypeFirmwareComponent] });
StepperBulkTypeFirmwareModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeFirmwareModule, providers: [
        hookListBulkType({
            type: BulkOperationType.FIRMWARE,
            c8yIcon: 'c8y-firmware',
            name: gettext('Firmware update'),
            path: `${baseUrl}firmware`,
            component: StepperBulkTypeFirmwareComponent,
            fragments: ['c8y_Firmware'],
            selected: false
        })
    ], imports: [CoreModule,
        FormsModule,
        ReactiveFormsModule,
        BulkOperationStepperModule,
        StepperFramesModule,
        SharedRepositoryModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StepperBulkTypeFirmwareModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        StepperBulkTypeFirmwareComponent,
                        VersionOrPatchComponent,
                        SelectFirmwareComponent
                    ],
                    imports: [
                        CoreModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BulkOperationStepperModule,
                        StepperFramesModule,
                        SharedRepositoryModule
                    ],
                    providers: [
                        hookListBulkType({
                            type: BulkOperationType.FIRMWARE,
                            c8yIcon: 'c8y-firmware',
                            name: gettext('Firmware update'),
                            path: `${baseUrl}firmware`,
                            component: StepperBulkTypeFirmwareComponent,
                            fragments: ['c8y_Firmware'],
                            selected: false
                        })
                    ],
                    entryComponents: [StepperBulkTypeFirmwareComponent],
                    exports: [StepperBulkTypeFirmwareComponent]
                }]
        }] });
