// @c8y/ngx-components/esm2020/sms-gateway/sms-gateway.module.mjs
import { NgModule } from '@angular/core';
import { gettext, hookPatternMessages, ProviderConfigurationModule } from '@c8y/ngx-components';
import { MESSAGES } from './pattern-messages.data';
import { SmsGatewayGuard } from './sms-gateway.guard';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class SmsGatewayModule {
}
SmsGatewayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SmsGatewayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SmsGatewayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SmsGatewayModule, imports: [i1.ProviderConfigurationModule] });
SmsGatewayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SmsGatewayModule, providers: [SmsGatewayGuard, hookPatternMessages(MESSAGES)], imports: [ProviderConfigurationModule.config([
            {
                navigation: {
                    label: gettext('SMS provider'),
                    path: 'smsgateway',
                    icon: 'envelope-o',
                    parent: gettext('Settings'),
                    priority: 1000,
                    canActivate: [SmsGatewayGuard]
                },
                layout: {
                    pageTitle: gettext('SMS provider'),
                    cardTitle: gettext('Credentials'),
                    description: gettext('Enter your credentials from your SMS provider. This enables platform features that utilize SMS services (for example, two-factor authentication and user notifications).'),
                    providerName: gettext('SMS provider'),
                    providerNamePlaceholder: gettext('Select or type to filter providers'),
                    providerNameNoMatchesHint: gettext('Select one of the existing providers.'),
                    deleteRoles: ['ROLE_SMS_ADMIN'],
                    deleteBtnLabel: gettext('Delete'),
                    saveRoles: ['ROLE_SMS_ADMIN'],
                    saveBtnLabel: gettext('Save'),
                    configurationUpdatedSuccessMsg: gettext('Credentials updated.'),
                    deleteConfigurationModalTitle: gettext('Delete credentials'),
                    deleteConfigurationModalBody: gettext(`You are about to delete SMS provider credentials. Deleting credentials will lock out any users with SMS-based two-factor authentication and deactivate the SMS-based smart rules and device communication. Do you want to proceed?`),
                    deleteConfigurationModalOkBtnLabel: gettext('Delete'),
                    configurationDeletedSuccessMsg: gettext('Credentials deleted.')
                },
                endpoint: {
                    definitionsEndpoint: {
                        baseUrl: 'service/messaging/providers',
                        listUrl: 'definitions'
                    },
                    configurationEndpoint: {
                        baseUrl: 'service/messaging/providers',
                        listUrl: 'configuration'
                    }
                }
            }
        ])] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SmsGatewayModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        ProviderConfigurationModule.config([
                            {
                                navigation: {
                                    label: gettext('SMS provider'),
                                    path: 'smsgateway',
                                    icon: 'envelope-o',
                                    parent: gettext('Settings'),
                                    priority: 1000,
                                    canActivate: [SmsGatewayGuard]
                                },
                                layout: {
                                    pageTitle: gettext('SMS provider'),
                                    cardTitle: gettext('Credentials'),
                                    description: gettext('Enter your credentials from your SMS provider. This enables platform features that utilize SMS services (for example, two-factor authentication and user notifications).'),
                                    providerName: gettext('SMS provider'),
                                    providerNamePlaceholder: gettext('Select or type to filter providers'),
                                    providerNameNoMatchesHint: gettext('Select one of the existing providers.'),
                                    deleteRoles: ['ROLE_SMS_ADMIN'],
                                    deleteBtnLabel: gettext('Delete'),
                                    saveRoles: ['ROLE_SMS_ADMIN'],
                                    saveBtnLabel: gettext('Save'),
                                    configurationUpdatedSuccessMsg: gettext('Credentials updated.'),
                                    deleteConfigurationModalTitle: gettext('Delete credentials'),
                                    deleteConfigurationModalBody: gettext(`You are about to delete SMS provider credentials. Deleting credentials will lock out any users with SMS-based two-factor authentication and deactivate the SMS-based smart rules and device communication. Do you want to proceed?`),
                                    deleteConfigurationModalOkBtnLabel: gettext('Delete'),
                                    configurationDeletedSuccessMsg: gettext('Credentials deleted.')
                                },
                                endpoint: {
                                    definitionsEndpoint: {
                                        baseUrl: 'service/messaging/providers',
                                        listUrl: 'definitions'
                                    },
                                    configurationEndpoint: {
                                        baseUrl: 'service/messaging/providers',
                                        listUrl: 'configuration'
                                    }
                                }
                            }
                        ])
                    ],
                    providers: [SmsGatewayGuard, hookPatternMessages(MESSAGES)]
                }]
        }] });
