// @c8y/ngx-components/esm2020/diagnostics/diagnostics.module.mjs
import { DiagnosticsTabGuard } from './diagnostics-tab.guard';
import { ListGroupModule } from '@c8y/ngx-components';
import { RouterModule } from '@angular/router';
import { DiagnosticsService } from './diagnostics.service';
import { CoreModule, gettext, hookRoute, ViewContext } from '@c8y/ngx-components';
import { DiagnosticsComponent } from './diagnostics.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { BsDropdownModule, BsDropdownConfig } from 'ngx-bootstrap/dropdown';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import * as i0 from "@angular/core";
export class DiagnosticsModule {
}
DiagnosticsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DiagnosticsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DiagnosticsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DiagnosticsModule, declarations: [DiagnosticsComponent], imports: [CoreModule,
        RouterModule,
        CommonModule,
        ListGroupModule,
        SharedRepositoryModule,
        BsDropdownModule,
        TooltipModule], exports: [DiagnosticsComponent] });
DiagnosticsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DiagnosticsModule, providers: [
        DiagnosticsService,
        BsDropdownConfig,
        DiagnosticsTabGuard,
        hookRoute({
            context: ViewContext.Device,
            path: 'diagnostics',
            component: DiagnosticsComponent,
            label: gettext('Diagnostics'),
            icon: 'stethoscope',
            canActivate: [DiagnosticsTabGuard],
            priority: 200
        })
    ], imports: [CoreModule,
        RouterModule,
        CommonModule,
        ListGroupModule,
        SharedRepositoryModule,
        BsDropdownModule,
        TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DiagnosticsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        RouterModule,
                        CommonModule,
                        ListGroupModule,
                        SharedRepositoryModule,
                        BsDropdownModule,
                        TooltipModule
                    ],
                    declarations: [DiagnosticsComponent],
                    exports: [DiagnosticsComponent],
                    providers: [
                        DiagnosticsService,
                        BsDropdownConfig,
                        DiagnosticsTabGuard,
                        hookRoute({
                            context: ViewContext.Device,
                            path: 'diagnostics',
                            component: DiagnosticsComponent,
                            label: gettext('Diagnostics'),
                            icon: 'stethoscope',
                            canActivate: [DiagnosticsTabGuard],
                            priority: 200
                        })
                    ]
                }]
        }] });
