// @c8y/ngx-components/esm2020/services/services-device-tab/services-device-tab.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, gettext, hookBreadcrumb, hookRoute, ViewContext } from '@c8y/ngx-components';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { LastUpdatedDateCellRendererComponent, NameCellRendererComponent, StatusCellRendererComponent } from './columns';
import { ServicesDeviceTabComponent } from './services-device-tab.component';
import { ServicesDeviceTabGuard } from './services-device-tab.guard';
import { ServicesBreadcrumbFactory } from './services.breadcrumb-factory';
import * as i0 from "@angular/core";
export class ServicesDeviceTabModule {
}
ServicesDeviceTabModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ServicesDeviceTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ServicesDeviceTabModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ServicesDeviceTabModule, declarations: [ServicesDeviceTabComponent,
        LastUpdatedDateCellRendererComponent,
        NameCellRendererComponent,
        StatusCellRendererComponent], imports: [CoreModule, RouterModule, TooltipModule] });
ServicesDeviceTabModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ServicesDeviceTabModule, providers: [
        ServicesDeviceTabGuard,
        ServicesBreadcrumbFactory,
        hookRoute({
            context: ViewContext.Device,
            path: 'services',
            component: ServicesDeviceTabComponent,
            label: gettext('Services'),
            icon: 'statistics',
            priority: 500,
            canActivate: [ServicesDeviceTabGuard]
        }),
        hookBreadcrumb(ServicesBreadcrumbFactory)
    ], imports: [CoreModule, RouterModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ServicesDeviceTabModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, RouterModule, TooltipModule],
                    declarations: [
                        ServicesDeviceTabComponent,
                        LastUpdatedDateCellRendererComponent,
                        NameCellRendererComponent,
                        StatusCellRendererComponent
                    ],
                    providers: [
                        ServicesDeviceTabGuard,
                        ServicesBreadcrumbFactory,
                        hookRoute({
                            context: ViewContext.Device,
                            path: 'services',
                            component: ServicesDeviceTabComponent,
                            label: gettext('Services'),
                            icon: 'statistics',
                            priority: 500,
                            canActivate: [ServicesDeviceTabGuard]
                        }),
                        hookBreadcrumb(ServicesBreadcrumbFactory)
                    ]
                }]
        }] });
