// @c8y/ngx-components/esm2020/protocol-impact/impact-protocol.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, hookActionBar } from '@c8y/ngx-components';
import { ImpactActionFactory } from './impact-action.factory';
import { ImpactSubscriptionService } from './impact-subscription.service';
import { RefreshActionComponent } from './refresh-action.component';
import * as i0 from "@angular/core";
export class ImpactProtocolModule {
}
ImpactProtocolModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactProtocolModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ImpactProtocolModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ImpactProtocolModule, declarations: [RefreshActionComponent], imports: [CoreModule] });
ImpactProtocolModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactProtocolModule, providers: [ImpactSubscriptionService, hookActionBar(ImpactActionFactory)], imports: [CoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactProtocolModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RefreshActionComponent],
                    imports: [CoreModule],
                    providers: [ImpactSubscriptionService, hookActionBar(ImpactActionFactory)],
                    entryComponents: [RefreshActionComponent]
                }]
        }] });
