// @c8y/ngx-components/esm2020/protocol-impact/impact-action.factory.mjs
import { Injectable } from '@angular/core';
import { AppStateService } from '@c8y/ngx-components';
import { get } from 'lodash-es';
import { RefreshActionComponent } from './refresh-action.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class ImpactActionFactory {
    constructor(appState) {
        this.appState = appState;
    }
    async get(activeRoute) {
        const actions = [];
        const data = !activeRoute.parent || activeRoute.snapshot.data.context
            ? activeRoute.snapshot.data
            : activeRoute.parent.snapshot.data;
        const { contextData } = data;
        const isDeviceInfoTab = get(activeRoute, 'snapshot.url[0].path') === 'device-info';
        const showRefreshActionButton = isDeviceInfoTab &&
            contextData &&
            contextData.c8y_ImpactResourceInfo &&
            /* call application service only for relevant devices to reduce number of service calls! */
            (await this.appState.isApplicationAvailable(ImpactActionFactory.applicationName));
        if (showRefreshActionButton) {
            actions.push({
                priority: 500,
                placement: 'right',
                component: RefreshActionComponent
            });
        }
        return actions;
    }
}
ImpactActionFactory.applicationName = 'impact';
ImpactActionFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactActionFactory, deps: [{ token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
ImpactActionFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactActionFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ImpactActionFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.AppStateService }]; } });
