// @c8y/ngx-components/esm2020/protocol-impact/refresh-action.component.mjs
import { Component, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { AlertService, gettext } from '@c8y/ngx-components';
import { ImpactSubscriptionService } from './impact-subscription.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./impact-subscription.service";
import * as i3 from "@c8y/ngx-components";
import * as i4 from "@angular/common";
export class RefreshActionComponent {
    constructor(vcRef, router, impactService, alert) {
        this.vcRef = vcRef;
        this.router = router;
        this.impactService = impactService;
        this.alert = alert;
    }
    ngOnInit() {
        this.vcRef.createEmbeddedView(this.templateCopy);
    }
    async refresh() {
        // TODO This is only a dirty hack to retrieve deviceId from URL;
        // In fact contextData should be provided for this component by a resolver?
        const url = this.router &&
            this.router.routerState &&
            this.router.routerState.snapshot &&
            this.router.routerState.snapshot.url;
        const deviceId = url && (/^\/device\/(\d+)\/.*$/gi.exec(url) || [])[1];
        if (deviceId) {
            this.requestInProgress = true;
            try {
                const res = await this.impactService.refreshDeviceResources(deviceId);
                if (res && res.status !== 200) {
                    const data = res.json ? await res.json() : undefined;
                    this.alert.addServerFailure({ data, res });
                }
                else {
                    this.alert.success(gettext('Device resource refresh scheduled.'));
                }
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
            this.requestInProgress = false;
        }
        else {
            this.alert.danger(gettext('Could not find device ID in URL.'));
        }
    }
}
RefreshActionComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RefreshActionComponent, deps: [{ token: i0.ViewContainerRef }, { token: i1.Router }, { token: i2.ImpactSubscriptionService }, { token: i3.AlertService }], target: i0.ɵɵFactoryTarget.Component });
RefreshActionComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: RefreshActionComponent, selector: "c8y-impact-refresh-action", viewQueries: [{ propertyName: "templateCopy", first: true, predicate: ["templateCopy"], descendants: true, read: TemplateRef, static: true }], ngImport: i0, template: "<ng-template #templateCopy>\n  <li>\n    <button\n      class=\"btn btn-link\"\n      title=\"{{ 'Refresh device resources' | translate }}\"\n      (click)=\"refresh()\"\n    >\n      <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': requestInProgress }\"></i>&nbsp;<span\n        translate\n        >Refresh device resources</span\n      >\n    </button>\n  </li>\n</ng-template>\n", dependencies: [{ kind: "directive", type: i3.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "directive", type: i3.C8yTranslateDirective, selector: "[translate],[ngx-translate]" }, { kind: "directive", type: i4.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "pipe", type: i3.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: RefreshActionComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-impact-refresh-action', template: "<ng-template #templateCopy>\n  <li>\n    <button\n      class=\"btn btn-link\"\n      title=\"{{ 'Refresh device resources' | translate }}\"\n      (click)=\"refresh()\"\n    >\n      <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': requestInProgress }\"></i>&nbsp;<span\n        translate\n        >Refresh device resources</span\n      >\n    </button>\n  </li>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i1.Router }, { type: i2.ImpactSubscriptionService }, { type: i3.AlertService }]; }, propDecorators: { templateCopy: [{
                type: ViewChild,
                args: ['templateCopy', { read: TemplateRef, static: true }]
            }] } });
