// @c8y/ngx-components/esm2020/operations/bulk-operation-from-single/bulk-operation-from-single.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationFromSingleComponent } from './bulk-operation-from-single.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export const bulkOperationCreateSingleRoutes = [
    {
        path: 'devicecontrol/single/create-bulk/:operationId',
        component: BulkOperationFromSingleComponent
    }
];
export class BulkOperationFromSingleModule {
}
BulkOperationFromSingleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationFromSingleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BulkOperationFromSingleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationFromSingleModule, declarations: [BulkOperationFromSingleComponent], imports: [CoreModule,
        BulkOperationStepperModule, i1.RouterModule] });
BulkOperationFromSingleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationFromSingleModule, imports: [CoreModule,
        BulkOperationStepperModule,
        RouterModule.forChild(bulkOperationCreateSingleRoutes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationFromSingleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BulkOperationFromSingleComponent],
                    imports: [
                        CoreModule,
                        BulkOperationStepperModule,
                        RouterModule.forChild(bulkOperationCreateSingleRoutes)
                    ]
                }]
        }] });
