// @c8y/ngx-components/esm2020/device-list/device-list-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
export class DeviceListNavigationFactory {
    constructor() {
        this.nav = new NavigatorNode({
            label: gettext('All devices'),
            path: 'device',
            icon: 'c8y-device-management',
            parent: {
                label: gettext('Devices'),
                icon: 'exchange'
            },
            priority: 2000
        });
    }
    async get() {
        return this.nav;
    }
}
DeviceListNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceListNavigationFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DeviceListNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceListNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceListNavigationFactory, decorators: [{
            type: Injectable
        }] });
