// @c8y/ngx-components/esm2020/device-list/device-list.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, hookNavigator } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { ReplaceDeviceDrawerComponent, ReplaceDeviceService } from '@c8y/ngx-components/replace-device';
import { AddSmartGroupComponent } from './add-smart-group.component';
import { DeviceListNavigationFactory } from './device-list-navigation.factory';
import { DeviceListComponent } from './device-list.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
const routes = [
    {
        path: 'device',
        component: DeviceListComponent
    }
];
export class DeviceListModule {
}
DeviceListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceListModule, declarations: [DeviceListComponent, AddSmartGroupComponent], imports: [CoreModule,
        CommonModule,
        DeviceGridModule, i1.RouterModule, ReplaceDeviceDrawerComponent] });
DeviceListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceListModule, providers: [hookNavigator(DeviceListNavigationFactory), ReplaceDeviceService], imports: [CoreModule,
        CommonModule,
        DeviceGridModule,
        RouterModule.forChild(routes),
        ReplaceDeviceDrawerComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        CommonModule,
                        DeviceGridModule,
                        RouterModule.forChild(routes),
                        ReplaceDeviceDrawerComponent
                    ],
                    declarations: [DeviceListComponent, AddSmartGroupComponent],
                    entryComponents: [DeviceListComponent],
                    exports: [],
                    providers: [hookNavigator(DeviceListNavigationFactory), ReplaceDeviceService]
                }]
        }] });
