// @c8y/ngx-components/esm2020/core/user/user.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { ModalModule } from '../modal/modal.module';
import { UserMenuItemComponent } from './user-menu-item.component';
import { UserEditModalComponent } from './user-edit-modal.component';
import { UserEditComponent } from './user-edit.component';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { UserMenuOutletComponent } from './user-menu-outlet.component';
import { AuthenticationModule } from '../authentication/authentication.module';
import { AlertModule } from '../alert/alert.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { RouterModule } from '@angular/router';
import { FormsModule } from '../forms/forms.module';
import { UserTotpSetupComponent } from './user-totp-setup.component';
import { UserTotpRevokeComponent } from './user-totp-revoke.component';
import { ClipboardModule } from '../clipboard/clipboard.module';
import { A11yModule } from '@angular/cdk/a11y';
import { hookDrawer } from '../drawer/drawer.service';
import { UserDetailsDrawerModule } from './user-details-drawer/user-details-drawer.module';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/tooltip";
/**
 * The angular module definition for the user menu.
 * @exports UserMenuComponent
 * @exports UserMenuItemComponent
 * @exports UserEditComponent
 * @exports UserEditModalComponent
 */
export class UserModule {
}
UserModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UserModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: UserModule, declarations: [UserMenuOutletComponent,
        UserEditComponent,
        UserEditModalComponent,
        UserMenuItemComponent,
        UserTotpSetupComponent,
        UserTotpRevokeComponent], imports: [A11yModule,
        FormsModule,
        CommonModule,
        ModalModule,
        BsDropdownModule,
        AuthenticationModule,
        AlertModule,
        RouterModule, i1.TooltipModule, PopoverModule,
        ClipboardModule,
        UserDetailsDrawerModule], exports: [UserMenuOutletComponent,
        UserMenuItemComponent,
        UserEditComponent,
        UserEditModalComponent,
        UserTotpSetupComponent,
        UserTotpRevokeComponent] });
UserModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserModule, providers: [
        hookDrawer({
            component: UserMenuOutletComponent,
            position: 'right',
            priority: 100,
            id: 'userMenuOutlet'
        })
    ], imports: [A11yModule,
        FormsModule,
        CommonModule,
        ModalModule,
        BsDropdownModule,
        AuthenticationModule,
        AlertModule,
        RouterModule,
        TooltipModule.forRoot(),
        PopoverModule,
        ClipboardModule,
        UserDetailsDrawerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        UserMenuOutletComponent,
                        UserEditComponent,
                        UserEditModalComponent,
                        UserMenuItemComponent,
                        UserTotpSetupComponent,
                        UserTotpRevokeComponent
                    ],
                    imports: [
                        A11yModule,
                        FormsModule,
                        CommonModule,
                        ModalModule,
                        BsDropdownModule,
                        AuthenticationModule,
                        AlertModule,
                        RouterModule,
                        TooltipModule.forRoot(),
                        PopoverModule,
                        ClipboardModule,
                        UserDetailsDrawerModule
                    ],
                    exports: [
                        UserMenuOutletComponent,
                        UserMenuItemComponent,
                        UserEditComponent,
                        UserEditModalComponent,
                        UserTotpSetupComponent,
                        UserTotpRevokeComponent
                    ],
                    entryComponents: [UserEditModalComponent, UserTotpSetupComponent, UserTotpRevokeComponent],
                    providers: [
                        hookDrawer({
                            component: UserMenuOutletComponent,
                            position: 'right',
                            priority: 100,
                            id: 'userMenuOutlet'
                        })
                    ]
                }]
        }] });
