// @c8y/ngx-components/esm2020/core/user/user-menu-item.component.mjs
import { Component, Input, ViewChild, EventEmitter, Output } from '@angular/core';
import { UserMenuService } from './user-menu.service';
import { HeaderService } from '../header/header.service';
import * as i0 from "@angular/core";
import * as i1 from "./user-menu.service";
import * as i2 from "../header/header.service";
import * as i3 from "@angular/common";
import * as i4 from "../i18n/c8y-translate.pipe";
export class UserMenuItemComponent {
    constructor(userService, headerService) {
        this.userService = userService;
        this.headerService = headerService;
        this.priority = 0;
        this.click = new EventEmitter();
        this.open$ = this.headerService.rightDrawerOpen$;
    }
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.userService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.userService.remove(this);
    }
    onClick() {
        this.click.emit(this);
    }
}
UserMenuItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserMenuItemComponent, deps: [{ token: i1.UserMenuService }, { token: i2.HeaderService }], target: i0.ɵɵFactoryTarget.Component });
UserMenuItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: UserMenuItemComponent, selector: "c8y-user-menu-item", inputs: { icon: "icon", label: "label", link: "link", target: "target", priority: "priority", dataCy: "dataCy" }, outputs: { click: "click" }, viewQueries: [{ propertyName: "template", first: true, predicate: ["template"], descendants: true }], ngImport: i0, template: "<ng-template #template>\n  <li>\n    <a\n      class=\"c8y-right-drawer__link\"\n      [attr.tabindex]=\"(open$ | async) ? '0' : '-1'\"\n      (click)=\"onClick()\"\n      [attr.data-cy]=\"dataCy\"\n      *ngIf=\"link\"\n      [attr.href]=\"link\"\n      [attr.target]=\"target\"\n    >\n      {{ label | translate }}\n      <ng-content></ng-content>\n    </a>\n    <button\n      class=\"c8y-right-drawer__link\"\n      [attr.tabindex]=\"(open$ | async) ? '0' : '-1'\"\n      type=\"button\"\n      *ngIf=\"!link\"\n      (click)=\"onClick()\"\n      [attr.data-cy]=\"dataCy\"\n    >\n      {{ label | translate }}\n      <ng-content></ng-content>\n    </button>\n  </li>\n</ng-template>\n", dependencies: [{ kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "pipe", type: i4.C8yTranslatePipe, name: "translate" }, { kind: "pipe", type: i3.AsyncPipe, name: "async" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UserMenuItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-user-menu-item', template: "<ng-template #template>\n  <li>\n    <a\n      class=\"c8y-right-drawer__link\"\n      [attr.tabindex]=\"(open$ | async) ? '0' : '-1'\"\n      (click)=\"onClick()\"\n      [attr.data-cy]=\"dataCy\"\n      *ngIf=\"link\"\n      [attr.href]=\"link\"\n      [attr.target]=\"target\"\n    >\n      {{ label | translate }}\n      <ng-content></ng-content>\n    </a>\n    <button\n      class=\"c8y-right-drawer__link\"\n      [attr.tabindex]=\"(open$ | async) ? '0' : '-1'\"\n      type=\"button\"\n      *ngIf=\"!link\"\n      (click)=\"onClick()\"\n      [attr.data-cy]=\"dataCy\"\n    >\n      {{ label | translate }}\n      <ng-content></ng-content>\n    </button>\n  </li>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i1.UserMenuService }, { type: i2.HeaderService }]; }, propDecorators: { icon: [{
                type: Input
            }], label: [{
                type: Input
            }], link: [{
                type: Input
            }], target: [{
                type: Input
            }], priority: [{
                type: Input
            }], dataCy: [{
                type: Input
            }], template: [{
                type: ViewChild,
                args: ['template', { static: false }]
            }], click: [{
                type: Output
            }] } });
