// @c8y/ngx-components/esm2020/cockpit-config/cockpit-config.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, gettext, hookNavigator, hookRoute, hookStepper, Steppers } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import { EcosystemModule } from '@c8y/ngx-components/ecosystem';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CockpitConfigGuard } from './cockpit-config.guard';
import { COCKPIT_CONFIG_PATH } from './cockpit-config.model';
import { CockpitConfigService } from './cockpit-config.service';
import { CockpitConfigurationComponent } from './cockpit-configuration.component';
import { FeatureConfigComponent } from './feature-config.component';
import { HomeDashboardConfigComponent } from './home-dashboard-config.component';
import { MiscConfigComponent } from './misc-config.component';
import { RootNodeConfigComponent } from './root-node-config.component';
import { CockpitSetupStep1Component } from './setup/cockpit-setup-step1.component';
import { CockpitSetupStep2Component } from './setup/cockpit-setup-step2.component';
import { CockpitSetupStep3Component } from './setup/cockpit-setup-step3.component';
import { CockpitSetupStep4Component } from './setup/cockpit-setup-step4.component';
import { CockpitSetupStepperButtonsComponent } from './setup/cockpit-setup-stepper-buttons.component';
import * as i0 from "@angular/core";
import * as i1 from "./cockpit-config.service";
export class CockpitConfigModule {
    constructor(service) {
        this.service = service;
        // only for DI, not used but needed.
    }
}
CockpitConfigModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CockpitConfigModule, deps: [{ token: i1.CockpitConfigService }], target: i0.ɵɵFactoryTarget.NgModule });
CockpitConfigModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: CockpitConfigModule, declarations: [CockpitConfigurationComponent,
        CockpitSetupStep1Component,
        CockpitSetupStep2Component,
        CockpitSetupStep3Component,
        CockpitSetupStep4Component,
        FeatureConfigComponent,
        RootNodeConfigComponent,
        HomeDashboardConfigComponent,
        MiscConfigComponent,
        CockpitSetupStepperButtonsComponent], imports: [CoreModule, AssetSelectorModule, PopoverModule, TooltipModule, EcosystemModule] });
CockpitConfigModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CockpitConfigModule, providers: [
        CockpitConfigGuard,
        hookRoute({
            path: COCKPIT_CONFIG_PATH,
            component: CockpitConfigurationComponent,
            canActivate: [CockpitConfigGuard]
        }),
        hookNavigator(CockpitConfigGuard),
        hookStepper([
            {
                stepperId: Steppers.SETUP,
                component: CockpitSetupStep1Component,
                label: gettext('Features'),
                setupId: 'cockpitFeatures',
                priority: 30
            },
            {
                stepperId: Steppers.SETUP,
                component: CockpitSetupStep2Component,
                label: gettext('Top level nodes'),
                setupId: 'cockpitTopLevelNodes',
                priority: 20
            },
            {
                stepperId: Steppers.SETUP,
                component: CockpitSetupStep3Component,
                label: gettext('Home dashboard'),
                setupId: 'cockpitHomeDashboard',
                priority: 10
            },
            {
                stepperId: Steppers.SETUP,
                component: CockpitSetupStep4Component,
                label: gettext('Misc'),
                setupId: 'cockpitMisc',
                priority: 0
            }
        ])
    ], imports: [CoreModule, AssetSelectorModule, PopoverModule, TooltipModule, EcosystemModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CockpitConfigModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        CockpitConfigurationComponent,
                        CockpitSetupStep1Component,
                        CockpitSetupStep2Component,
                        CockpitSetupStep3Component,
                        CockpitSetupStep4Component,
                        FeatureConfigComponent,
                        RootNodeConfigComponent,
                        HomeDashboardConfigComponent,
                        MiscConfigComponent,
                        CockpitSetupStepperButtonsComponent
                    ],
                    imports: [CoreModule, AssetSelectorModule, PopoverModule, TooltipModule, EcosystemModule],
                    providers: [
                        CockpitConfigGuard,
                        hookRoute({
                            path: COCKPIT_CONFIG_PATH,
                            component: CockpitConfigurationComponent,
                            canActivate: [CockpitConfigGuard]
                        }),
                        hookNavigator(CockpitConfigGuard),
                        hookStepper([
                            {
                                stepperId: Steppers.SETUP,
                                component: CockpitSetupStep1Component,
                                label: gettext('Features'),
                                setupId: 'cockpitFeatures',
                                priority: 30
                            },
                            {
                                stepperId: Steppers.SETUP,
                                component: CockpitSetupStep2Component,
                                label: gettext('Top level nodes'),
                                setupId: 'cockpitTopLevelNodes',
                                priority: 20
                            },
                            {
                                stepperId: Steppers.SETUP,
                                component: CockpitSetupStep3Component,
                                label: gettext('Home dashboard'),
                                setupId: 'cockpitHomeDashboard',
                                priority: 10
                            },
                            {
                                stepperId: Steppers.SETUP,
                                component: CockpitSetupStep4Component,
                                label: gettext('Misc'),
                                setupId: 'cockpitMisc',
                                priority: 0
                            }
                        ])
                    ]
                }]
        }], ctorParameters: function () { return [{ type: i1.CockpitConfigService }]; } });
