// @c8y/ngx-components/esm2020/core/navigator/navigator-bottom/navigator-bottom-drawer.factory.mjs
import { Injectable } from '@angular/core';
import { OptionsService } from '../../common';
import { NavigatorBottomComponent } from './navigator-bottom.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common";
export class NavigatorBottomDrawerFactory {
    constructor(options) {
        this.options = options;
        this.drawerItem = {
            component: NavigatorBottomComponent,
            position: 'left',
            priority: 1001,
            id: 'poweredBy',
            noneRequired: true
        };
    }
    get() {
        if (this.options.get('hidePowered')) {
            return [];
        }
        return this.drawerItem;
    }
}
NavigatorBottomDrawerFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomDrawerFactory, deps: [{ token: i1.OptionsService }], target: i0.ɵɵFactoryTarget.Injectable });
NavigatorBottomDrawerFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomDrawerFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomDrawerFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.OptionsService }]; } });
