// @c8y/ngx-components/esm2020/core/navigator/navigator-bottom/navigator-bottom.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NavigatorBottomComponent } from './navigator-bottom.component';
import { hookDrawer } from '../../drawer/drawer.service';
import { NavigatorBottomDrawerFactory } from './navigator-bottom-drawer.factory';
import * as i0 from "@angular/core";
export class NavigatorBottomModule {
}
NavigatorBottomModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NavigatorBottomModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomModule, declarations: [NavigatorBottomComponent], imports: [CommonModule] });
NavigatorBottomModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomModule, providers: [hookDrawer(NavigatorBottomDrawerFactory)], imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorBottomModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NavigatorBottomComponent],
                    providers: [hookDrawer(NavigatorBottomDrawerFactory)]
                }]
        }] });
