// @c8y/ngx-components/esm2020/datapoint-selector/datapoint-selector.module.mjs
import { DragDropModule } from '@angular/cdk/drag-drop';
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { AssetSelectorModule } from '@c8y/ngx-components/assets-navigator';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { ModalModule } from 'ngx-bootstrap/modal';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { DatapointAttributesFormComponent } from './datapoint-attributes-form/datapoint-attributes-form.component';
import { DatapointSelectionListComponent } from './datapoint-selection-list/datapoint-selection-list.component';
import { DatapointSelectorListItemComponent } from './datapoint-selector-list-item/datapoint-selector-list-item.component';
import { DatapointSelectorModalComponent } from './datapoint-selector-modal/datapoint-selector-modal.component';
import { DatapointSelectorComponent } from './datapoint-selector.component';
import { DatapointTemplatePopoverComponent } from './datapoint-template-popover/datapoint-template-popover.component';
import { DatapointLabelPipe } from './pipes/datapoint-label.pipe';
import { FilterDatapointsPipe } from './pipes/filter-datapoints.pipe';
import { IncludesDatapointPipe } from './pipes/includes-datapoint.pipe';
import * as i0 from "@angular/core";
export class DatapointSelectorModule {
}
DatapointSelectorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatapointSelectorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorModule, declarations: [DatapointSelectorComponent,
        DatapointSelectorModalComponent,
        DatapointSelectorListItemComponent,
        DatapointSelectionListComponent,
        DatapointTemplatePopoverComponent,
        IncludesDatapointPipe,
        DatapointLabelPipe,
        FilterDatapointsPipe,
        DatapointAttributesFormComponent], imports: [AssetSelectorModule,
        CollapseModule,
        CoreModule,
        BsDropdownModule,
        ModalModule,
        TooltipModule,
        PopoverModule,
        DragDropModule], exports: [DatapointSelectorComponent,
        DatapointSelectorModalComponent,
        DatapointSelectorListItemComponent,
        DatapointSelectionListComponent,
        DatapointTemplatePopoverComponent,
        IncludesDatapointPipe,
        DatapointLabelPipe,
        DatapointAttributesFormComponent] });
DatapointSelectorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorModule, imports: [AssetSelectorModule,
        CollapseModule,
        CoreModule,
        BsDropdownModule,
        ModalModule,
        TooltipModule,
        PopoverModule,
        DragDropModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        AssetSelectorModule,
                        CollapseModule,
                        CoreModule,
                        BsDropdownModule,
                        ModalModule,
                        TooltipModule,
                        PopoverModule,
                        DragDropModule
                    ],
                    declarations: [
                        DatapointSelectorComponent,
                        DatapointSelectorModalComponent,
                        DatapointSelectorListItemComponent,
                        DatapointSelectionListComponent,
                        DatapointTemplatePopoverComponent,
                        IncludesDatapointPipe,
                        DatapointLabelPipe,
                        FilterDatapointsPipe,
                        DatapointAttributesFormComponent
                    ],
                    exports: [
                        DatapointSelectorComponent,
                        DatapointSelectorModalComponent,
                        DatapointSelectorListItemComponent,
                        DatapointSelectionListComponent,
                        DatapointTemplatePopoverComponent,
                        IncludesDatapointPipe,
                        DatapointLabelPipe,
                        DatapointAttributesFormComponent
                    ]
                }]
        }] });
