// @c8y/ngx-components/esm2020/datapoint-selector/datapoint-selector.service.mjs
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { DatapointSelectorModalComponent } from './datapoint-selector-modal/datapoint-selector-modal.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
export class DatapointSelectorService {
    constructor(modal) {
        this.modal = modal;
    }
    selectDataPoints(initialState = {}) {
        const modal = this.modal.show(DatapointSelectorModalComponent, {
            ignoreBackdropClick: true,
            keyboard: false,
            initialState,
            ariaDescribedby: 'modal-body',
            ariaLabelledBy: 'modal-title',
            class: 'modal-lg'
        });
        const content = modal.content;
        return content.result;
    }
}
DatapointSelectorService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorService, deps: [{ token: i1.BsModalService }], target: i0.ɵɵFactoryTarget.Injectable });
DatapointSelectorService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointSelectorService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.BsModalService }]; } });
