// @c8y/ngx-components/esm2020/protocol-lpwan/multiple-lns-connectors/connection-info-with-download-csv.component.mjs
import { Component, Input } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { MultipleLnsConnectorService } from './multiple-lns-connector.service';
import { saveAs } from 'file-saver';
import { AlertService, gettext } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
import * as i2 from "./multiple-lns-connector.service";
import * as i3 from "@c8y/ngx-components";
import * as i4 from "@ngx-translate/core";
export class ConnectionInfoWithDownloadCsvComponent {
    constructor(modal, connectorService, alertService, translateService) {
        this.modal = modal;
        this.connectorService = connectorService;
        this.alertService = alertService;
        this.translateService = translateService;
    }
    dismiss() {
        this.modal.hide();
    }
    async download() {
        const url = `/service/${this.appData.contextPath}${this.messageData.attrs.URL}`;
        const res = await this.connectorService.download(url);
        if (res && res.status === 200) {
            const streamData = await res.blob();
            saveAs(streamData, this.translateService.instant(gettext('{{ connectionName }} - devices.csv'), {
                connectionName: this.connectionName
            }));
        }
        else {
            this.alertService.danger(gettext('A server error occurred.'));
        }
    }
}
ConnectionInfoWithDownloadCsvComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectionInfoWithDownloadCsvComponent, deps: [{ token: i1.BsModalRef }, { token: i2.MultipleLnsConnectorService }, { token: i3.AlertService }, { token: i4.TranslateService }], target: i0.ɵɵFactoryTarget.Component });
ConnectionInfoWithDownloadCsvComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ConnectionInfoWithDownloadCsvComponent, selector: "connection-info-with-download-csv", inputs: { messageData: "messageData", appData: "appData", modalTitle: "modalTitle", connectionName: "connectionName" }, ngImport: i0, template: "<div class=\"modal-dialog\">\n  <div class=\"modal-content\">\n    <div class=\"c8y-prompt alert alert-danger\">\n      <h3 class=\"m-b-16\">\n        <i class=\"dlt-c8y-icon-exclamation-circle\"></i>\n        <span>{{ modalTitle | translate }}</span>\n      </h3>\n      <p class=\"text-break-word\">\n        {{ messageData.message | translate }}\n      </p>\n      <br>\n      <span class=\"btn-default\" (click)=\"download()\">{{\n        'Click the link to download the file with the affected devices.' | translate\n      }}</span>\n      <div class=\"alert-footer\">\n        <button\n          title=\"{{ 'Cancel' | translate }}\"\n          type=\"button\"\n          class=\"btn btn-default\"\n          (click)=\"dismiss()\"\n        >\n          {{ 'Cancel' | translate }}\n        </button>\n      </div>\n    </div>\n  </div>\n</div>", dependencies: [{ kind: "pipe", type: i3.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConnectionInfoWithDownloadCsvComponent, decorators: [{
            type: Component,
            args: [{ selector: 'connection-info-with-download-csv', template: "<div class=\"modal-dialog\">\n  <div class=\"modal-content\">\n    <div class=\"c8y-prompt alert alert-danger\">\n      <h3 class=\"m-b-16\">\n        <i class=\"dlt-c8y-icon-exclamation-circle\"></i>\n        <span>{{ modalTitle | translate }}</span>\n      </h3>\n      <p class=\"text-break-word\">\n        {{ messageData.message | translate }}\n      </p>\n      <br>\n      <span class=\"btn-default\" (click)=\"download()\">{{\n        'Click the link to download the file with the affected devices.' | translate\n      }}</span>\n      <div class=\"alert-footer\">\n        <button\n          title=\"{{ 'Cancel' | translate }}\"\n          type=\"button\"\n          class=\"btn btn-default\"\n          (click)=\"dismiss()\"\n        >\n          {{ 'Cancel' | translate }}\n        </button>\n      </div>\n    </div>\n  </div>\n</div>" }]
        }], ctorParameters: function () { return [{ type: i1.BsModalRef }, { type: i2.MultipleLnsConnectorService }, { type: i3.AlertService }, { type: i4.TranslateService }]; }, propDecorators: { messageData: [{
                type: Input
            }], appData: [{
                type: Input
            }], modalTitle: [{
                type: Input
            }], connectionName: [{
                type: Input
            }] } });
