// @c8y/ngx-components/esm2020/protocol-lpwan/multiple-lns-connectors/multiple-lns-connectors.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule, FormsModule } from '@c8y/ngx-components';
import { NoConnectionsFoundComponent } from './no-connections-found.component';
import { RouterModule } from '@angular/router';
import { SigfoxMultipleLnsConnectorComponent } from './sigfox-multiple-lns-connector.component';
import { ActilityMultipleLnsConnectorComponent } from './actility-multiple-lns-connector.component';
import { MultipleLnsConnectorService } from './multiple-lns-connector.service';
import { ConnectionInfoWithDownloadCsvComponent } from './connection-info-with-download-csv.component';
import { LoriotMultipleLnsConnectorComponent } from './loriot/loriot-multiple-lns-connector.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
const settingsRoutes = [
    {
        path: 'connectivitySettings/multiple_lns_connectors_sigfox',
        component: SigfoxMultipleLnsConnectorComponent
    },
    {
        path: 'connectivitySettings/multiple_lns_connectors_actility',
        component: ActilityMultipleLnsConnectorComponent
    },
    {
        path: 'connectivitySettings/multiple_lns_connectors_loriot',
        component: LoriotMultipleLnsConnectorComponent
    }
];
export class MultipleLnsConnectorsModule {
}
MultipleLnsConnectorsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MultipleLnsConnectorsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultipleLnsConnectorsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: MultipleLnsConnectorsModule, declarations: [SigfoxMultipleLnsConnectorComponent,
        NoConnectionsFoundComponent,
        ActilityMultipleLnsConnectorComponent,
        LoriotMultipleLnsConnectorComponent,
        ConnectionInfoWithDownloadCsvComponent], imports: [CoreModule, CommonModule, FormsModule, i1.RouterModule] });
MultipleLnsConnectorsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MultipleLnsConnectorsModule, providers: [MultipleLnsConnectorService], imports: [CoreModule, CommonModule, FormsModule, RouterModule.forChild(settingsRoutes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MultipleLnsConnectorsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SigfoxMultipleLnsConnectorComponent,
                        NoConnectionsFoundComponent,
                        ActilityMultipleLnsConnectorComponent,
                        LoriotMultipleLnsConnectorComponent,
                        ConnectionInfoWithDownloadCsvComponent
                    ],
                    imports: [CoreModule, CommonModule, FormsModule, RouterModule.forChild(settingsRoutes)],
                    providers: [MultipleLnsConnectorService]
                }]
        }] });
