// @c8y/client/lib/cjs/src/operation/OperationBulkGeneralStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationBulkGeneralStatus = void 0;
const gettext_1 = require("../gettext");
/**
 * A bulk operation can have one of the following general statuses.
 */
var OperationBulkGeneralStatus;
(function (OperationBulkGeneralStatus) {
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["SCHEDULED"] = (0, gettext_1.gettext)('SCHEDULED')] = "SCHEDULED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["EXECUTING"] = (0, gettext_1.gettext)('EXECUTING')] = "EXECUTING";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["EXECUTING_WITH_ERROR"] = (0, gettext_1.gettext)('EXECUTING_WITH_ERROR')] = "EXECUTING_WITH_ERROR";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["CANCELED"] = (0, gettext_1.gettext)('CANCELED')] = "CANCELED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["FAILED"] = (0, gettext_1.gettext)('FAILED')] = "FAILED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["SUCCESSFUL"] = (0, gettext_1.gettext)('SUCCESSFUL')] = "SUCCESSFUL";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["DELETED"] = (0, gettext_1.gettext)('DELETED')] = "DELETED";
    OperationBulkGeneralStatus[OperationBulkGeneralStatus["INCORRECT"] = (0, gettext_1.gettext)('INCORRECT')] = "INCORRECT";
})(OperationBulkGeneralStatus = exports.OperationBulkGeneralStatus || (exports.OperationBulkGeneralStatus = {}));
