// @c8y/ngx-components/esm2020/operations/status-filter/status-filter.module.mjs
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CoreModule } from '@c8y/ngx-components';
import { StatusFilterComponent } from './status-filter.component';
import * as i0 from "@angular/core";
/**
 * This module provides a component for selecting status filter.
 */
export class StatusFilterModule {
}
StatusFilterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StatusFilterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StatusFilterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: StatusFilterModule, declarations: [StatusFilterComponent], imports: [CoreModule, TooltipModule, BsDropdownModule], exports: [StatusFilterComponent] });
StatusFilterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StatusFilterModule, imports: [CoreModule, TooltipModule, BsDropdownModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: StatusFilterModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, TooltipModule, BsDropdownModule],
                    declarations: [StatusFilterComponent],
                    entryComponents: [StatusFilterComponent],
                    exports: [StatusFilterComponent]
                }]
        }] });
