// @c8y/ngx-components/esm2020/core/dynamic-forms/dynamic-forms.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyModule } from '@ngx-formly/core';
import { DynamicFormsUIThemeModule } from './dynamic-forms-ui-theme.module';
import { I18nModule } from './i18n/i18n.module';
import { C8yJSONSchema } from './json-schema/c8y-json-schema.service';
import { ValidationModule } from './validation/validation.module';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-formly/core";
export class DynamicFormsModule {
}
DynamicFormsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DynamicFormsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DynamicFormsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DynamicFormsModule, imports: [CommonModule,
        ReactiveFormsModule,
        I18nModule,
        ValidationModule,
        DynamicFormsUIThemeModule, i1.FormlyModule], exports: [FormlyModule, ReactiveFormsModule] });
DynamicFormsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DynamicFormsModule, providers: [C8yJSONSchema], imports: [CommonModule,
        ReactiveFormsModule,
        I18nModule,
        ValidationModule,
        DynamicFormsUIThemeModule,
        FormlyModule.forRoot(), FormlyModule, ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DynamicFormsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                        I18nModule,
                        ValidationModule,
                        DynamicFormsUIThemeModule,
                        FormlyModule.forRoot()
                    ],
                    providers: [C8yJSONSchema],
                    exports: [FormlyModule, ReactiveFormsModule]
                }]
        }] });
