// @c8y/ngx-components/esm2020/core/alert/alert.component.mjs
import { Component, Input, ViewChild, ContentChild } from '@angular/core';
import { AlertService } from './alert.service';
import { AlertDetailsComponent } from './alert-details.component';
import * as i0 from "@angular/core";
import * as i1 from "./alert.service";
/**
 * Alert component allows creating alert with complex view.
 */
export class AlertComponent {
    /**
     * @ignore
     */
    constructor(alertService) {
        this.alertService = alertService;
    }
    /**
     * AlertComponent adds new alert to collection stored in alertService after view init.
     */
    ngAfterViewInit() {
        if (!this.containDetailsComponent) {
            this.detailedData = null;
        }
        this.viewInitTimeout = setTimeout(() => this.alertService.add(this));
    }
    /**
     * AlertComponent removes alert from collection stored in alertService on destroy component.
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.alertService.remove(this);
    }
}
AlertComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AlertComponent, deps: [{ token: i1.AlertService }], target: i0.ɵɵFactoryTarget.Component });
AlertComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: AlertComponent, selector: "c8y-alert", inputs: { type: "type", onDetail: "onDetail", onClose: "onClose" }, queries: [{ propertyName: "containDetailsComponent", first: true, predicate: AlertDetailsComponent, descendants: true }], viewQueries: [{ propertyName: "detailedData", first: true, predicate: ["detailedData"], descendants: true }, { propertyName: "text", first: true, predicate: ["text"], descendants: true }], ngImport: i0, template: "<ng-template #text>\r\n  <ng-content select=\"c8y-alert-text\"></ng-content>\r\n  <ng-content></ng-content>\r\n</ng-template>\r\n<ng-template #detailedData>\r\n  <ng-content select=\"c8y-alert-details\"></ng-content>\r\n</ng-template>" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AlertComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-alert', template: "<ng-template #text>\r\n  <ng-content select=\"c8y-alert-text\"></ng-content>\r\n  <ng-content></ng-content>\r\n</ng-template>\r\n<ng-template #detailedData>\r\n  <ng-content select=\"c8y-alert-details\"></ng-content>\r\n</ng-template>" }]
        }], ctorParameters: function () { return [{ type: i1.AlertService }]; }, propDecorators: { type: [{
                type: Input
            }], onDetail: [{
                type: Input
            }], onClose: [{
                type: Input
            }], detailedData: [{
                type: ViewChild,
                args: ['detailedData', { static: false }]
            }], text: [{
                type: ViewChild,
                args: ['text', { static: false }]
            }], containDetailsComponent: [{
                type: ContentChild,
                args: [AlertDetailsComponent, { static: false }]
            }] } });
