// @c8y/ngx-components/esm2020/core/alert/alert-outlet-base.mjs
import { TemplateRef } from '@angular/core';
export class AlertOutletBase {
    constructor(alertService) {
        this.alertService = alertService;
        /**
         * Currently displayed details of an alert. Only one alert’s details can be displayed at a time.
         */
        this.detailDisplayedAlert = {};
    }
    /**
     * Expands details part of particular alert, and collapses details for another alerts.
     * @param alert The alert which details should be expanded.
     */
    showDetails(alert) {
        if (alert.onDetail) {
            return alert.onDetail();
        }
        this.detailDisplayedAlert = Object.assign({}, alert);
        this.detailDisplayedAlert.mode = 'raw';
        this.setDetailsContent(alert);
    }
    /**
     * Returns true if alert has extra detailedData or defines custom onDetail method.
     */
    hasDetails(alert) {
        return alert.detailedData || alert.onDetail;
    }
    /**
     * Collapses alert's details.
     */
    closeDetails() {
        this.detailDisplayedAlert = {};
    }
    /**
     * Returns true if details part of alert is expanded.
     */
    isDetailsShow(alert) {
        return this.alertService.areSame(alert, this.detailDisplayedAlert);
    }
    /**
     * @ignore
     */
    isTemplateRef(value) {
        return value instanceof TemplateRef;
    }
    /**
     * @ignore
     */
    template(content) {
        return this.isTemplateRef(content) ? content : undefined;
    }
    setDetailsContent(alert) {
        if (this.isTemplateRef(alert.detailedData) || typeof alert.detailedData === 'string') {
            this.detailDisplayedAlert.contents = alert.detailedData;
        }
        else {
            this.detailDisplayedAlert.contents = JSON.stringify(alert.detailedData, null, 2);
        }
    }
}
