// @c8y/ngx-components/esm2020/core/alert/alert.module.mjs
import { NgModule } from '@angular/core';
import { AlertDetailsComponent } from './alert-details.component';
import { AlertOutletComponent } from './alert-outlet.component';
import { AlertTextComponent } from './alert-text.component';
import { AlertComponent } from './alert.component';
import { CommonModule } from '../common/common.module';
import * as i0 from "@angular/core";
/**
 * Alert module is used for displaying status messages to the user e.g. danger, success, warning and info alerts.
 */
export class AlertModule {
}
AlertModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AlertModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AlertModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AlertModule, declarations: [AlertOutletComponent, AlertComponent, AlertDetailsComponent, AlertTextComponent], imports: [CommonModule], exports: [AlertOutletComponent, AlertComponent, AlertDetailsComponent, AlertTextComponent] });
AlertModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AlertModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AlertModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [AlertOutletComponent, AlertComponent, AlertDetailsComponent, AlertTextComponent],
                    declarations: [AlertOutletComponent, AlertComponent, AlertDetailsComponent, AlertTextComponent]
                }]
        }] });
