// @c8y/ngx-components/esm2020/map/map-popup.directive.mjs
import { Directive, ElementRef, TemplateRef, ViewContainerRef } from '@angular/core';
import * as i0 from "@angular/core";
export class MapPopupDirective {
    constructor(template, elementRef, viewContainer) {
        this.template = template;
        this.elementRef = elementRef;
        this.viewContainer = viewContainer;
    }
}
MapPopupDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapPopupDirective, deps: [{ token: i0.TemplateRef }, { token: i0.ElementRef }, { token: i0.ViewContainerRef }], target: i0.ɵɵFactoryTarget.Directive });
MapPopupDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: MapPopupDirective, selector: "[c8yMapPopup]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapPopupDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[c8yMapPopup]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }, { type: i0.ElementRef }, { type: i0.ViewContainerRef }]; } });
