// @c8y/ngx-components/esm2020/map/map.module.mjs
import { CommonModule as NgCommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule as NgFormsModule } from '@angular/forms';
import { CommonModule, FormsModule, RealtimeModule } from '@c8y/ngx-components';
import { MapStatusComponent } from './map-status.component';
import { MapPopupDirective } from './map-popup.directive';
import { MapComponent } from './map.component';
import { ClusterMapComponent } from './cluster-map.component';
import { defaultLayer, defaultMapConfig, MAP_DEFAULT_CONFIG, MAP_TILE_LAYER } from './map.model';
import * as i0 from "@angular/core";
export class MapModule {
}
MapModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MapModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: MapModule, declarations: [MapComponent, MapStatusComponent, ClusterMapComponent, MapPopupDirective], imports: [NgCommonModule, NgFormsModule, CommonModule, FormsModule, RealtimeModule], exports: [MapComponent, MapStatusComponent, ClusterMapComponent, MapPopupDirective] });
MapModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapModule, providers: [
        {
            provide: MAP_TILE_LAYER,
            useValue: [defaultLayer]
        },
        {
            provide: MAP_DEFAULT_CONFIG,
            useValue: defaultMapConfig
        }
    ], imports: [NgCommonModule, NgFormsModule, CommonModule, FormsModule, RealtimeModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MapComponent, MapStatusComponent, ClusterMapComponent, MapPopupDirective],
                    imports: [NgCommonModule, NgFormsModule, CommonModule, FormsModule, RealtimeModule],
                    exports: [MapComponent, MapStatusComponent, ClusterMapComponent, MapPopupDirective],
                    providers: [
                        {
                            provide: MAP_TILE_LAYER,
                            useValue: [defaultLayer]
                        },
                        {
                            provide: MAP_DEFAULT_CONFIG,
                            useValue: defaultMapConfig
                        }
                    ]
                }]
        }] });
