// @c8y/ngx-components/esm2020/map/map.model.mjs
import { InjectionToken } from '@angular/core';
export function getStatus(device) {
    if (!device.c8y_ActiveAlarmsStatus) {
        return 'text-primary';
    }
    if (device.c8y_ActiveAlarmsStatus.critical) {
        return 'status critical';
    }
    if (device.c8y_ActiveAlarmsStatus.major) {
        return 'status major';
    }
    if (device.c8y_ActiveAlarmsStatus.warning) {
        return 'status warning';
    }
    if (device.c8y_ActiveAlarmsStatus.minor) {
        return 'status minor';
    }
    return 'text-primary';
}
export function getC8yMarker(marker, asset) {
    marker.asset = asset;
    return marker;
}
export const MAP_TILE_LAYER = new InjectionToken('MAP_TILE_LAYER');
export var ClusterSize;
(function (ClusterSize) {
    ClusterSize[ClusterSize["NONE"] = 0] = "NONE";
    ClusterSize[ClusterSize["FOUR"] = 1] = "FOUR";
    ClusterSize[ClusterSize["SIXTEEN"] = 2] = "SIXTEEN";
})(ClusterSize || (ClusterSize = {}));
export const MAP_DEFAULT_CONFIG = new InjectionToken('MAP_DEFAULT_CONFIG');
export const defaultLayer = {
    layerUrl: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
    label: 'OpenStreetMap',
    priority: 1000,
    options: {
        maxZoom: 18,
        minZoom: 2,
        attribution: '&copy;<a href="http://www.openstreetmap.org/copyright" rel="noreferrer nofollow">OpenStreetMap</a>',
        noWrap: true
    }
};
export const defaultMapConfig = {
    center: [51.23544, 6.79599],
    zoomLevel: 2
};
