// @c8y/ngx-components/esm2020/core/file-picker-form-control/file-picker-form-control.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FilePickerFormControlComponent } from './file-picker-form-control.component';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import * as i0 from "@angular/core";
/**
 * The angular module definition for a drop-area.
 * @exports FilePickerComponent
 */
export class FilePickerFormControlModule {
}
FilePickerFormControlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePickerFormControlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilePickerFormControlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FilePickerFormControlModule, declarations: [FilePickerFormControlComponent], imports: [C8yFormsModule, CommonModule, FormsModule, C8yCommonModule, TooltipModule], exports: [FilePickerFormControlComponent] });
FilePickerFormControlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePickerFormControlModule, imports: [C8yFormsModule, CommonModule, FormsModule, C8yCommonModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilePickerFormControlModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [C8yFormsModule, CommonModule, FormsModule, C8yCommonModule, TooltipModule],
                    exports: [FilePickerFormControlComponent],
                    declarations: [FilePickerFormControlComponent]
                }]
        }] });
