// @c8y/ngx-components/esm2020/core/file-picker-form-control/file-picker-validators.mjs
import { BytesPipe } from '../common';
import { gettext } from '../i18n';
export class FilePickerValidators {
    static fileMaxSize(maxSizeBytes, filesService) {
        const msg = gettext('The selected file is too large. The size limit is {{ limit }}.');
        return (control) => {
            if (control.value instanceof File) {
                const file = control.value;
                return filesService.size(file) > maxSizeBytes
                    ? { [msg]: { limit: BytesPipe.bytes(maxSizeBytes, 1) } }
                    : null;
            }
            return null;
        };
    }
    static fileNonEmpty(filesService) {
        const msg = gettext('File must not be empty, select another one.');
        return (control) => {
            if (control.value instanceof File) {
                const file = control.value;
                return filesService.size(file) === 0 ? { [msg]: true } : null;
            }
            return null;
        };
    }
    static fileNameMaxLength(filesService) {
        const msg = gettext('The filename exceeds limit of {{ limit }} characters.');
        const filenameLengthLimit = filesService.FILENAME_MAX_LENGTH;
        return (control) => {
            if (control.value instanceof File) {
                const file = control.value;
                return file.name.length > filenameLengthLimit
                    ? { [msg]: { limit: filenameLengthLimit } }
                    : null;
            }
            return null;
        };
    }
    static haveValidExtensions(filesService, accept) {
        const msg = gettext('The selected file is not of a supported type. Supported file types are: {{ accept }}.');
        return (control) => {
            if (control.value instanceof File) {
                const file = control.value;
                return !filesService.haveValidExtensions([file], accept) ? { [msg]: { accept } } : null;
            }
            return null;
        };
    }
    static fileCountMax(fileCountMax) {
        const msg = gettext('Too many files selected. Limit: {{ fileCountMax }}.');
        return (control) => {
            if (Array.isArray(control.value)) {
                return control.value.length > fileCountMax ? { [msg]: { fileCountMax } } : null;
            }
            return null;
        };
    }
    static fileCountMin(fileCountMin) {
        const msg = gettext('Requires min. {{ fileCountMin }} file(s).');
        return (control) => {
            if (Array.isArray(control.value)) {
                return control.value.length < fileCountMin ? { [msg]: { fileCountMin } } : null;
            }
            return null;
        };
    }
}
