// @c8y/ngx-components/esm2020/core/action/action.service.mjs
import { Injector, Injectable, InjectionToken } from '@angular/core';
import { Router } from '@angular/router';
import { fromTrigger, stateToFactory, hookGeneric, getInjectedHooks, ExtensionPointForPlugins } from '../common/extension-hooks';
import { distinctUntilChanged, shareReplay } from 'rxjs/operators';
import { PluginsResolveService } from '../plugins/plugins-resolve.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../plugins/plugins-resolve.service";
/**
 * A hook to add an action using the multi-provider extension concept.
 * @deprecated Consider using the `hookAction` function instead.
 */
export const HOOK_ACTION = new InjectionToken('HOOK_ACTION');
/**
 * A hook to add an action using the multi-provider extension concept.
 *
 * You can either provide a single `Action` as parameter:
 * ```typescript
 *  hookAction(...)
 * ```
 *
 * Or an array to directly register multiple:
 * ```typescript
 *  hookAction([...])
 * ```
 *
 * Or you provide an Service that implements `ExtensionFactory<Action>`
 * ```typescript
 *  export class MyActionFactory implements ExtensionFactory<Action> {...}
 *  ...
 *  hookAction(MyActionFactory)
 * ```
 * A typed alternative to `HOOK_ACTION`.
 * @param actions The `Action`'s or `ExtensionFactory` to be provided.
 * @returns An `Provider` to be provided in your module.
 */
export function hookAction(actions, options) {
    return hookGeneric(actions, HOOK_ACTION, options);
}
/**
 * A service which defines global actions.
 */
export class ActionService extends ExtensionPointForPlugins {
    constructor(rootInjector, router, plugins) {
        super(rootInjector, plugins);
        this.router = router;
        this.items$ = this.setupItemsObservable();
    }
    /**
     * Returns the current state.
     * @readonly
     * @returns The current set of actions.
     */
    get state() {
        return this.state$.value;
    }
    /**
     * Adds a new item to the actions in the header and emits a state change.
     * @param item The item to add.
     */
    add(item) {
        this.state.add(item);
        this.emitNewState();
    }
    /**
     * Removes an action from the header and emits a state change.
     * @param item The item to remove.
     */
    remove(item) {
        this.state.delete(item);
        this.emitNewState();
    }
    setupItemsObservable() {
        return fromTrigger(this.router, this.refresh$, [
            getInjectedHooks(HOOK_ACTION, this.injectors),
            () => this.factories,
            stateToFactory(this.state$)
        ]).pipe(distinctUntilChanged(), shareReplay(1));
    }
}
ActionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionService, deps: [{ token: i0.Injector }, { token: i1.Router }, { token: i2.PluginsResolveService }], target: i0.ɵɵFactoryTarget.Injectable });
ActionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i0.Injector }, { type: i1.Router }, { type: i2.PluginsResolveService }]; } });
