// @c8y/ngx-components/esm2020/core/action/action.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { ActionOutletComponent } from './action-outlet.component';
import { ActionComponent } from './action.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * The angular module definition for actions.
 * @exports ActionComponent
 * @exports ActionOutletComponent
 */
export class ActionModule {
}
ActionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ActionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ActionModule, declarations: [ActionOutletComponent, ActionComponent], imports: [CommonModule, BsDropdownModule, i1.RouterModule], exports: [ActionOutletComponent, ActionComponent] });
ActionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionModule, imports: [CommonModule, BsDropdownModule, RouterModule.forChild([])] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ActionOutletComponent, ActionComponent],
                    imports: [CommonModule, BsDropdownModule, RouterModule.forChild([])],
                    exports: [ActionOutletComponent, ActionComponent]
                }]
        }] });
