// @c8y/ngx-components/esm2020/core/action/action.component.mjs
import { Component, EnvironmentInjector, Injector, Input, ViewChild } from '@angular/core';
import { ActionService } from './action.service';
import * as i0 from "@angular/core";
import * as i1 from "./action.service";
/**
 * A <c8y-action> can be added with content projection.
 *
 * @example
 * ```html
 * <c8y-action>
 *   <button class="btn" (click)="myAction()">My action</button>
 * </c8y-action>
 * ```
 */
export class ActionComponent {
    constructor(actionService, _injector) {
        this.actionService = actionService;
        /**
         * Is this action disabled.
         */
        this.disabled = false;
        /**
         * The priority of this action (the higher the number, the higher the action will be displayed; optional)
         */
        this.priority = 0;
        this.injector = _injector;
    }
    /**
     * After page view is initialized, component adds itself to global action list.
     */
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.actionService.add(this));
    }
    /**
     * Action is removed from global action list, if component is not used anymore.
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.actionService.remove(this);
    }
}
ActionComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionComponent, deps: [{ token: i1.ActionService }, { token: i0.EnvironmentInjector }], target: i0.ɵɵFactoryTarget.Component });
ActionComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: ActionComponent, selector: "c8y-action", inputs: { disabled: "disabled", priority: "priority", icon: "icon", injector: "injector" }, viewQueries: [{ propertyName: "template", first: true, predicate: ["template"], descendants: true }], ngImport: i0, template: "<ng-template #template>\n  <li>\n    <ng-content></ng-content>\n  </li>\n</ng-template>" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActionComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-action', template: "<ng-template #template>\n  <li>\n    <ng-content></ng-content>\n  </li>\n</ng-template>" }]
        }], ctorParameters: function () { return [{ type: i1.ActionService }, { type: i0.EnvironmentInjector }]; }, propDecorators: { disabled: [{
                type: Input
            }], priority: [{
                type: Input
            }], icon: [{
                type: Input
            }], template: [{
                type: ViewChild,
                args: ['template', { static: false }]
            }], injector: [{
                type: Input
            }] } });
