// @c8y/ngx-components/esm2020/core/setup/setup-stepper-factory.mjs
import { Injectable } from '@angular/core';
import { gettext } from '../i18n/gettext';
import { Steppers } from '../stepper/stepper.model';
import { SetupCompletedComponent } from './setup-completed.component';
import { SETUP_FINISHED_STEP_ID } from './setup.model';
import * as i0 from "@angular/core";
/**
 * A factory providing the 'Summary' step to the setup stepper.
 */
export class SetupStepperFactory {
    constructor() {
        this.step = {
            setupId: SETUP_FINISHED_STEP_ID,
            stepperId: Steppers.SETUP,
            component: SetupCompletedComponent,
            label: gettext('Summary'),
            priority: -Infinity
        };
    }
    get() {
        return this.step;
    }
}
SetupStepperFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SetupStepperFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
SetupStepperFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SetupStepperFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SetupStepperFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
