// @c8y/ngx-components/esm2020/core/setup/setup.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { HeaderModule } from '../header/header.module';
import { PluginsModule } from '../plugins/plugins.module';
import { PropertiesListModule } from '../properties-list/properties-list.module';
import { hookStepper } from '../stepper/stepper.service';
import { StepperModule } from '../stepper/stepper.module';
import { SetupCompletedComponent } from './setup-completed.component';
import { SetupComponent } from './setup.component';
import { SetupStepperFactory } from './setup-stepper-factory';
import { HookProviderTypes } from '../common';
import * as i0 from "@angular/core";
export class SetupModule {
}
SetupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SetupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SetupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SetupModule, declarations: [SetupComponent, SetupCompletedComponent], imports: [CommonModule, HeaderModule, StepperModule, PluginsModule, PropertiesListModule], exports: [SetupComponent, SetupCompletedComponent] });
SetupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SetupModule, providers: [
        hookStepper(SetupStepperFactory, { providerType: HookProviderTypes.ExistingProvider })
    ], imports: [CommonModule, HeaderModule, StepperModule, PluginsModule, PropertiesListModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SetupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SetupComponent, SetupCompletedComponent],
                    imports: [CommonModule, HeaderModule, StepperModule, PluginsModule, PropertiesListModule],
                    exports: [SetupComponent, SetupCompletedComponent],
                    entryComponents: [SetupCompletedComponent],
                    providers: [
                        hookStepper(SetupStepperFactory, { providerType: HookProviderTypes.ExistingProvider })
                    ]
                }]
        }] });
