// @c8y/ngx-components/esm2020/ecosystem/shared/ecosystem.model.mjs
/** Wizard types  */
export var EcosystemWizards;
(function (EcosystemWizards) {
    EcosystemWizards["APPLICATION_UPLOAD"] = "ecosystemApplicationUpload";
    EcosystemWizards["MICROSERVICE_UPLOAD"] = "ecosystemMicroserviceUpload";
    EcosystemWizards["PACKAGE_UPLOAD"] = "ecosystemPackageUpload";
    EcosystemWizards["BLUEPRINT_DEPLOYMENT"] = "ecosystemBlueprintDeployment";
    EcosystemWizards["LICENSE_CONFIRM"] = "ecosystemLicenseConfirm";
})(EcosystemWizards || (EcosystemWizards = {}));
export var ERROR_TYPE;
(function (ERROR_TYPE) {
    ERROR_TYPE["TYPE_VALIDATION"] = "TYPE_VALIDATION";
    ERROR_TYPE["ALREADY_SUBSCRIBED"] = "ALREADY_SUBSCRIBED";
    ERROR_TYPE["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    ERROR_TYPE["NO_MANIFEST_FILE"] = "NO_MANIFEST_FILE";
    ERROR_TYPE["INVALID_PACKAGE"] = "INVALID_PACKAGE";
    ERROR_TYPE["INVALID_APPLICATION"] = "INVALID_APPLICATION";
    ERROR_TYPE["MICROSERVICE_NAME_TOO_LONG"] = "MICROSERVICE_NAME_TOO_LONG";
    ERROR_TYPE["APPLICATION_CREATION_FAILED"] = "APPLICATION_CREATION_FAILED";
    ERROR_TYPE["KEY_OR_CONTEXT_PATH_MISMATCH"] = "KEY_OR_CONTEXT_PATH_MISMATCH";
})(ERROR_TYPE || (ERROR_TYPE = {}));
export const PRODUCT_EXPERIENCE = {
    APPLICATIONS: {
        EVENTS: {
            AVAILABILITY: 'availability',
            APPLICATION_CARD: 'applicationCard',
            APPLICATION_PROPERTIES: 'applicationProperties',
            DEPLOY_APPLICATION: 'deployApplication',
            DUPLICATE_APPLICATION: 'duplicateApplication',
            INSTALLED_PLUGINS: 'installedPlugins',
            PACKAGE_PLUGINS: 'packagePlugins',
            PACKAGE_VERSIONS: 'packageVersions',
            FILTER_LIST: 'filterList'
        },
        COMPONENTS: {
            APPLICATION_CARD: 'application-card',
            APPLICATION_PLUGINS: 'application-plugins',
            APPLICATION_PROPERTIES: 'application-properties',
            DEPLOY_APPLICATION: 'deploy-application',
            DUPLICATE_APPLICATION_PROPERTIES: 'duplicate-application-properties',
            PLUGIN_LIST: 'plugin-list',
            PACKAGE_VERSIONS: 'package-versions-list',
            UPDATE_PLUGIN_OF_APP: 'update-plugin-of-app',
            LIST_FILTERS: 'list-filters'
        },
        ACTIONS: {
            AVAILABILITY_CHANGE: 'availabilityChange',
            CANCEL: 'cancel',
            CLONE: 'clone',
            CHANGE_PLUGIN_VERSION: 'changePluginVersion',
            DELETE: 'delete',
            DOWNLOAD: 'download',
            DEPLOY_APPLICATION: 'deployApplication',
            EDIT: 'edit',
            INSTALL_PLUGIN: 'installPlugin',
            INSTALL_PLUGINS: 'installPlugins',
            SELECT_VERSION: 'selectVersion',
            SET_AS_LATEST: 'setAsLatest',
            UPDATE_AVAILABLE: 'updateAvailable',
            UPLOAD: 'upload',
            SET_FILTER_TERM: 'setFilterTerm',
            SET_PREDEFINED_FILTERS: 'setPredefinedFilters',
            RESET_FILTER: 'resetFilter'
        },
        RESULTS: {
            DEPLOYED: 'deployed',
            DUPLICATED: 'duplicated',
            PLUGIN_INSTALLED: 'pluginInstalled',
            PLUGIN_REMOVED: 'pluginRemoved',
            PLUGIN_VERSION_CHANGED: 'pluginVersionChanged',
            SERVER_FAILURE: 'serverFailure',
            SUCCESS: 'success'
        }
    }
};
