// @c8y/ngx-components/esm2020/ecosystem/shared/shared-ecosystem.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, hookWizard } from '@c8y/ngx-components';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { A11yModule } from '@angular/cdk/a11y';
import { AddApplicationComponent } from './add-application.component';
import { ApplicationCardComponent } from './application-card.component';
import { ApplicationPropertiesFormComponent } from './application-properties-form.component';
import { UploadArchiveComponent } from './upload-archive.component';
import { DuplicateApplicationListComponent } from './duplicate-application/duplicate-application-list/duplicate-application-list.component';
import { DuplicateApplicationPropertiesComponent } from './duplicate-application/duplicate-application-properties/duplicate-application-properties.component';
import { DuplicateApplicationComponent } from './duplicate-application/duplicate-application.component';
import { PackageVersionSelectComponent } from './package-version-select/package-version-select.component';
import { EcosystemWizards } from './ecosystem.model';
import { TranslatePackageLabelPipe } from './translate-package-label.pipe';
import { ListFiltersComponent } from './list-filters/list-filters.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class SharedEcosystemModule {
}
SharedEcosystemModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SharedEcosystemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SharedEcosystemModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SharedEcosystemModule, declarations: [AddApplicationComponent,
        ApplicationPropertiesFormComponent,
        UploadArchiveComponent,
        ApplicationCardComponent,
        DuplicateApplicationListComponent,
        DuplicateApplicationPropertiesComponent,
        DuplicateApplicationComponent,
        PackageVersionSelectComponent,
        TranslatePackageLabelPipe,
        ListFiltersComponent], imports: [CoreModule,
        BsDropdownModule,
        TooltipModule,
        A11yModule,
        FormsModule,
        ReactiveFormsModule], exports: [AddApplicationComponent,
        ApplicationPropertiesFormComponent,
        UploadArchiveComponent,
        ApplicationCardComponent,
        DuplicateApplicationListComponent,
        DuplicateApplicationPropertiesComponent,
        DuplicateApplicationComponent,
        PackageVersionSelectComponent,
        TranslatePackageLabelPipe,
        ListFiltersComponent] });
SharedEcosystemModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SharedEcosystemModule, providers: [
        hookWizard({
            wizardId: EcosystemWizards.APPLICATION_UPLOAD,
            component: DuplicateApplicationComponent,
            name: 'Duplicate existing application',
            c8yIcon: 'copy'
        }),
        hookWizard({
            wizardId: 'duplicateApplication',
            component: DuplicateApplicationComponent,
            name: 'Duplicate existing application',
            c8yIcon: 'copy'
        })
    ], imports: [CoreModule,
        BsDropdownModule,
        TooltipModule,
        A11yModule,
        FormsModule,
        ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SharedEcosystemModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        BsDropdownModule,
                        TooltipModule,
                        A11yModule,
                        FormsModule,
                        ReactiveFormsModule
                    ],
                    declarations: [
                        AddApplicationComponent,
                        ApplicationPropertiesFormComponent,
                        UploadArchiveComponent,
                        ApplicationCardComponent,
                        DuplicateApplicationListComponent,
                        DuplicateApplicationPropertiesComponent,
                        DuplicateApplicationComponent,
                        PackageVersionSelectComponent,
                        TranslatePackageLabelPipe,
                        ListFiltersComponent
                    ],
                    exports: [
                        AddApplicationComponent,
                        ApplicationPropertiesFormComponent,
                        UploadArchiveComponent,
                        ApplicationCardComponent,
                        DuplicateApplicationListComponent,
                        DuplicateApplicationPropertiesComponent,
                        DuplicateApplicationComponent,
                        PackageVersionSelectComponent,
                        TranslatePackageLabelPipe,
                        ListFiltersComponent
                    ],
                    providers: [
                        hookWizard({
                            wizardId: EcosystemWizards.APPLICATION_UPLOAD,
                            component: DuplicateApplicationComponent,
                            name: 'Duplicate existing application',
                            c8yIcon: 'copy'
                        }),
                        hookWizard({
                            wizardId: 'duplicateApplication',
                            component: DuplicateApplicationComponent,
                            name: 'Duplicate existing application',
                            c8yIcon: 'copy'
                        })
                    ]
                }]
        }] });
