// @c8y/ngx-components/esm2020/ecosystem/shared/upload-archive.component.mjs
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { EcosystemService } from './ecosystem.service';
import * as i0 from "@angular/core";
import * as i1 from "./ecosystem.service";
import * as i2 from "@c8y/ngx-components";
import * as i3 from "@angular/common";
export class UploadArchiveComponent {
    constructor(ecosystemService, alertService) {
        this.ecosystemService = ecosystemService;
        this.alertService = alertService;
        this.uploadNewVersion = false;
        this.applicationChange = new EventEmitter();
        this.refresh = new EventEmitter();
        this.archives = [];
        this.uploadInProgress = false;
        this.canReactivate = false;
    }
    get uploadProgress() {
        return this.ecosystemService.progress;
    }
    async onFileDroppedEvent(event) {
        if (this.preUploadCallback) {
            try {
                await this.preUploadCallback();
            }
            catch (e) {
                // do nothing, i.e. when confirm modal cancels and throws undefined
                return;
            }
        }
        if (event && event.length > 0) {
            const file = event[0].file;
            this.onFile(file);
        }
    }
    async onFile(file) {
        try {
            await this.ecosystemService.validateArchiveToAppCompatibility(file, this.application);
            const isCustomMicroservice = this.ecosystemService.isCustomMicroservice(this.application);
            this.archives = isCustomMicroservice
                ? []
                : await this.ecosystemService.listArchives(this.application.id);
            this.archives.sort((a, b) => {
                return new Date(b.created) - new Date(a.created);
            });
            this.uploadInProgress = true;
            this.uploadProgress.next(0);
            if (this.archives.length === 6 && !isCustomMicroservice) {
                await this.ecosystemService.removeOldestArchive(this.application, this.archives);
            }
            const app = await this.ecosystemService.uploadArchiveToApp(file, this.application, this.uploadNewVersion);
            this.applicationChange.emit(app);
            this.uploadProgress.next(100);
            this.uploadInProgress = false;
            this.refresh.emit();
            if (this.uploadNewVersion) {
                return;
            }
            this.alertService.success(gettext('Archive uploaded and activated.'));
        }
        catch (ex) {
            this.uploadInProgress = false;
            this.ecosystemService.alertError(ex);
        }
    }
}
UploadArchiveComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UploadArchiveComponent, deps: [{ token: i1.EcosystemService }, { token: i2.AlertService }], target: i0.ɵɵFactoryTarget.Component });
UploadArchiveComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: UploadArchiveComponent, selector: "c8y-upload-archive", inputs: { application: "application", uploadNewVersion: "uploadNewVersion", preUploadCallback: "preUploadCallback" }, outputs: { applicationChange: "applicationChange", refresh: "refresh" }, ngImport: i0, template: "<c8y-drop-area\n  (dropped)=\"onFileDroppedEvent($event)\"\n  [accept]=\"'.zip'\"\n  [loading]=\"uploadInProgress\"\n  [maxAllowedFiles]=\"1\"\n  [message]=\"'Upload a *.zip file' | translate\"\n  [progress]=\"uploadProgress | async\"\n  [forceHideList]=\"true\"\n  class=\"drop-area-sm\"\n>\n</c8y-drop-area>\n", dependencies: [{ kind: "component", type: i2.DropAreaComponent, selector: "c8y-drop-area", inputs: ["formControl", "title", "message", "icon", "loadingMessage", "forceHideList", "alwaysShow", "clickToOpen", "loading", "progress", "maxAllowedFiles", "files", "maxFileSizeInMegaBytes", "accept"], outputs: ["dropped"] }, { kind: "pipe", type: i2.C8yTranslatePipe, name: "translate" }, { kind: "pipe", type: i3.AsyncPipe, name: "async" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: UploadArchiveComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-upload-archive', template: "<c8y-drop-area\n  (dropped)=\"onFileDroppedEvent($event)\"\n  [accept]=\"'.zip'\"\n  [loading]=\"uploadInProgress\"\n  [maxAllowedFiles]=\"1\"\n  [message]=\"'Upload a *.zip file' | translate\"\n  [progress]=\"uploadProgress | async\"\n  [forceHideList]=\"true\"\n  class=\"drop-area-sm\"\n>\n</c8y-drop-area>\n" }]
        }], ctorParameters: function () { return [{ type: i1.EcosystemService }, { type: i2.AlertService }]; }, propDecorators: { application: [{
                type: Input
            }], uploadNewVersion: [{
                type: Input
            }], preUploadCallback: [{
                type: Input
            }], applicationChange: [{
                type: Output
            }], refresh: [{
                type: Output
            }] } });
