// @c8y/ngx-components/esm2020/core/device-status/device-status.module.mjs
import { NgModule } from '@angular/core';
import { DeviceStatusComponent } from './device-status.component';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import * as i0 from "@angular/core";
/**
 * The angular module definition for a device-status.
 * @exports DeviceStatusComponent
 */
export class DeviceStatusModule {
}
DeviceStatusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceStatusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceStatusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceStatusModule, declarations: [DeviceStatusComponent], imports: [CommonModule, TooltipModule], exports: [DeviceStatusComponent] });
DeviceStatusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceStatusModule, imports: [CommonModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceStatusModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TooltipModule],
                    exports: [DeviceStatusComponent],
                    declarations: [DeviceStatusComponent]
                }]
        }] });
