// @c8y/ngx-components/esm2020/core/wizard/wizard.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { BsModalRef, ModalModule as BsModalModule } from 'ngx-bootstrap/modal';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { WizardBodyComponent } from './wizard-body.component';
import { WizardFooterComponent } from './wizard-footer.component';
import { WizardHeaderComponent } from './wizard-header.component';
import { WizardOutletComponent } from './wizard-outlet/wizard-outlet.component';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import { WizardModalService } from './wizard-modal.service';
import * as i0 from "@angular/core";
export class WizardModule {
}
WizardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
WizardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: WizardModule, declarations: [WizardComponent,
        WizardOutletComponent,
        WizardHeaderComponent,
        WizardBodyComponent,
        WizardFooterComponent], imports: [BsModalModule, CommonModule, C8yCommonModule], exports: [WizardComponent,
        WizardOutletComponent,
        WizardHeaderComponent,
        WizardBodyComponent,
        WizardFooterComponent] });
WizardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardModule, providers: [WizardService, BsModalRef, WizardModalService], imports: [BsModalModule, CommonModule, C8yCommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [BsModalModule, CommonModule, C8yCommonModule],
                    exports: [
                        WizardComponent,
                        WizardOutletComponent,
                        WizardHeaderComponent,
                        WizardBodyComponent,
                        WizardFooterComponent
                    ],
                    declarations: [
                        WizardComponent,
                        WizardOutletComponent,
                        WizardHeaderComponent,
                        WizardBodyComponent,
                        WizardFooterComponent
                    ],
                    entryComponents: [
                        WizardComponent,
                        WizardHeaderComponent,
                        WizardBodyComponent,
                        WizardFooterComponent
                    ],
                    providers: [WizardService, BsModalRef, WizardModalService]
                }]
        }] });
