// @c8y/ngx-components/esm2020/core/wizard/wizard-header.component.mjs
import { Component, TemplateRef, ViewChild } from '@angular/core';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import * as i0 from "@angular/core";
import * as i1 from "./wizard.component";
import * as i2 from "./wizard.service";
export class WizardHeaderComponent {
    constructor(wizard, wizardService) {
        this.wizard = wizard;
        this.wizardService = wizardService;
    }
    ngOnInit() {
        setTimeout(() => {
            this.wizardService.headerTemplate$.next({
                template: this.headerContent,
                wizardId: this.wizard.id
            });
        });
    }
}
WizardHeaderComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardHeaderComponent, deps: [{ token: i1.WizardComponent }, { token: i2.WizardService }], target: i0.ɵɵFactoryTarget.Component });
WizardHeaderComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: WizardHeaderComponent, selector: "c8y-wizard-header", viewQueries: [{ propertyName: "headerContent", first: true, predicate: ["headerContent"], descendants: true, static: true }], ngImport: i0, template: `
    <ng-template #headerContent>
      <div class="modal-header dialog-header animated fadeIn">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardHeaderComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-wizard-header',
                    template: `
    <ng-template #headerContent>
      <div class="modal-header dialog-header animated fadeIn">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `
                }]
        }], ctorParameters: function () { return [{ type: i1.WizardComponent }, { type: i2.WizardService }]; }, propDecorators: { headerContent: [{
                type: ViewChild,
                args: ['headerContent', { static: true }]
            }] } });
