// @c8y/ngx-components/esm2020/core/wizard/wizard-footer.component.mjs
import { Component, TemplateRef, ViewChild } from '@angular/core';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import * as i0 from "@angular/core";
import * as i1 from "./wizard.component";
import * as i2 from "./wizard.service";
export class WizardFooterComponent {
    constructor(wizard, wizardService) {
        this.wizard = wizard;
        this.wizardService = wizardService;
    }
    ngOnInit() {
        setTimeout(() => {
            this.wizardService.footerTemplate$.next({
                template: this.footerContent,
                wizardId: this.wizard.id
            });
        });
    }
}
WizardFooterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardFooterComponent, deps: [{ token: i1.WizardComponent }, { token: i2.WizardService }], target: i0.ɵɵFactoryTarget.Component });
WizardFooterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: WizardFooterComponent, selector: "c8y-wizard-footer", viewQueries: [{ propertyName: "footerContent", first: true, predicate: ["footerContent"], descendants: true, static: true }], ngImport: i0, template: `
    <ng-template #footerContent>
      <div class="modal-footer">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardFooterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-wizard-footer',
                    template: `
    <ng-template #footerContent>
      <div class="modal-footer">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `
                }]
        }], ctorParameters: function () { return [{ type: i1.WizardComponent }, { type: i2.WizardService }]; }, propDecorators: { footerContent: [{
                type: ViewChild,
                args: ['footerContent', { static: true }]
            }] } });
