// @c8y/ngx-components/esm2020/core/wizard/wizard-body.component.mjs
import { Component, TemplateRef, ViewChild } from '@angular/core';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import * as i0 from "@angular/core";
import * as i1 from "./wizard.component";
import * as i2 from "./wizard.service";
export class WizardBodyComponent {
    constructor(wizard, wizardService) {
        this.wizard = wizard;
        this.wizardService = wizardService;
    }
    ngOnInit() {
        setTimeout(() => {
            this.wizardService.bodyTemplate$.next({
                template: this.bodyContent,
                wizardId: this.wizard.id
            });
        });
    }
}
WizardBodyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardBodyComponent, deps: [{ token: i1.WizardComponent }, { token: i2.WizardService }], target: i0.ɵɵFactoryTarget.Component });
WizardBodyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: WizardBodyComponent, selector: "c8y-wizard-body", viewQueries: [{ propertyName: "bodyContent", first: true, predicate: ["bodyContent"], descendants: true, static: true }], ngImport: i0, template: `
    <ng-template #bodyContent>
      <div class="modal-inner-scroll animated fadeIn">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardBodyComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-wizard-body',
                    template: `
    <ng-template #bodyContent>
      <div class="modal-inner-scroll animated fadeIn">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `
                }]
        }], ctorParameters: function () { return [{ type: i1.WizardComponent }, { type: i2.WizardService }]; }, propDecorators: { bodyContent: [{
                type: ViewChild,
                args: ['bodyContent', { static: true }]
            }] } });
