// @c8y/ngx-components/esm2020/core/wizard/wizard-modal.service.mjs
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { WizardComponent } from './wizard.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
export class WizardModalService {
    constructor(bsModalService) {
        this.bsModalService = bsModalService;
    }
    /**
     * Shows a wizard modal.
     * @param modalOptions Modal options.
     * @returns a reference to the modal.
     */
    show(modalOptions = {}) {
        const options = {
            class: 'modal-sm',
            ariaDescribedby: 'modal-body',
            ariaLabelledBy: 'modal-title',
            backdrop: 'static',
            ...modalOptions
        };
        return this.bsModalService.show(WizardComponent, options);
    }
}
WizardModalService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardModalService, deps: [{ token: i1.BsModalService }], target: i0.ɵɵFactoryTarget.Injectable });
WizardModalService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardModalService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WizardModalService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.BsModalService }]; } });
