// @c8y/ngx-components/esm2020/data-broker/data-broker.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, ManagedObjectRealtimeService, DeviceBootstrapRealtimeService } from '@c8y/ngx-components';
import { DataBrokerSubscriptionsService } from './data-broker-subscriptions.service';
import * as i0 from "@angular/core";
export class DataBrokerModule {
}
DataBrokerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataBrokerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DataBrokerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DataBrokerModule, imports: [CoreModule] });
DataBrokerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataBrokerModule, providers: [
        DataBrokerSubscriptionsService,
        ManagedObjectRealtimeService,
        DeviceBootstrapRealtimeService
    ], imports: [CoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataBrokerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CoreModule],
                    entryComponents: [],
                    providers: [
                        DataBrokerSubscriptionsService,
                        ManagedObjectRealtimeService,
                        DeviceBootstrapRealtimeService
                    ],
                    exports: []
                }]
        }] });
