// @c8y/ngx-components/esm2020/operations/device-selector/device-selector.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { DeviceSelectorComponent } from './device-selector.component';
import * as i0 from "@angular/core";
/**
 * This module provides a component for selecting devices based on filters.
 */
export class DeviceSelectorModule {
}
DeviceSelectorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceSelectorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceSelectorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceSelectorModule, declarations: [DeviceSelectorComponent], imports: [CoreModule, DeviceGridModule], exports: [DeviceSelectorComponent] });
DeviceSelectorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceSelectorModule, imports: [CoreModule, DeviceGridModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceSelectorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, DeviceGridModule],
                    declarations: [DeviceSelectorComponent],
                    exports: [DeviceSelectorComponent]
                }]
        }] });
