// @c8y/ngx-components/esm2020/core/quicklink/quick-link.module.mjs
import { CoreModule } from '../core.module';
import { NgModule } from '@angular/core';
import { QuickLinkComponent } from './quick-link.component';
import * as i0 from "@angular/core";
export class QuickLinkModule {
}
QuickLinkModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: QuickLinkModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
QuickLinkModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: QuickLinkModule, declarations: [QuickLinkComponent], imports: [CoreModule], exports: [QuickLinkComponent] });
QuickLinkModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: QuickLinkModule, imports: [CoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: QuickLinkModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [QuickLinkComponent],
                    imports: [CoreModule],
                    exports: [QuickLinkComponent],
                    providers: []
                }]
        }] });
