// @c8y/ngx-components/esm2020/actility-device-registration/actility-device-registration.factory.mjs
import { Injectable } from '@angular/core';
import { TenantUiService } from '@c8y/ngx-components';
import { ActilityDeviceRegistrationButtonComponent } from './actility-device-registration-button.component';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class ActilityDeviceRegistrationFactory {
    constructor(tenantService) {
        this.tenantService = tenantService;
    }
    get() {
        const actions = [];
        if (this.tenantService.isMicroserviceSubscribedInCurrentTenant('actility')) {
            actions.push({
                template: ActilityDeviceRegistrationButtonComponent,
                priority: 99,
                category: 'single'
            });
        }
        return actions;
    }
}
ActilityDeviceRegistrationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationFactory, deps: [{ token: i1.TenantUiService }], target: i0.ɵɵFactoryTarget.Injectable });
ActilityDeviceRegistrationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationFactory, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.TenantUiService }]; } });
