// @c8y/ngx-components/esm2020/actility-device-registration/actility-device-registration.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule, CoreModule } from '@c8y/ngx-components';
import { hookDeviceRegistration } from '@c8y/ngx-components/register-device';
import { ActilityDeviceRegistrationButtonComponent } from './actility-device-registration-button.component';
import { ActilityDeviceRegistrationFactory } from './actility-device-registration.factory';
import { ActilityDeviceRegistrationComponent } from './actility-device-registration.component';
import { ActilityDeviceRegistrationService } from './actility-device-registration.service';
import * as i0 from "@angular/core";
export class ActilityDeviceRegistrationModule {
}
ActilityDeviceRegistrationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ActilityDeviceRegistrationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationModule, declarations: [ActilityDeviceRegistrationButtonComponent, ActilityDeviceRegistrationComponent], imports: [CoreModule, CommonModule] });
ActilityDeviceRegistrationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationModule, providers: [
        ActilityDeviceRegistrationService,
        hookDeviceRegistration(ActilityDeviceRegistrationFactory)
    ], imports: [CoreModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityDeviceRegistrationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, CommonModule],
                    declarations: [ActilityDeviceRegistrationButtonComponent, ActilityDeviceRegistrationComponent],
                    providers: [
                        ActilityDeviceRegistrationService,
                        hookDeviceRegistration(ActilityDeviceRegistrationFactory)
                    ]
                }]
        }] });
