// @c8y/ngx-components/esm2020/search/search.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, hookSearch, hookRoute, SearchModule as CoreSearchModule } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { SearchResultsComponent } from './search-results.component';
import { SearchGridComponent } from './search-grid.component';
import { SearchActionComponent } from './search-action.component';
import { AssetSearchService } from './search.service';
import { SearchCustomFiltersComponent } from './search-custom-filters.component';
import { AssetTypeCellRendererComponent } from './columns/asset-type.cell-renderer.component';
import { SEARCH_CONFIG } from './search-config.model';
import * as i0 from "@angular/core";
export class SearchModule {
    static config(config = {}) {
        return {
            ngModule: SearchModule,
            providers: [
                {
                    provide: SEARCH_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
SearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, declarations: [SearchResultsComponent,
        SearchGridComponent,
        SearchActionComponent,
        SearchCustomFiltersComponent,
        AssetTypeCellRendererComponent], imports: [CoreModule, DeviceGridModule, CoreSearchModule], exports: [SearchResultsComponent, SearchGridComponent] });
SearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, providers: [
        AssetSearchService,
        hookRoute({
            path: 'assetsearch',
            component: SearchResultsComponent
        }),
        hookSearch({
            template: SearchActionComponent,
            onSearch: undefined
        })
    ], imports: [CoreModule, DeviceGridModule, CoreSearchModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SearchModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SearchResultsComponent,
                        SearchGridComponent,
                        SearchActionComponent,
                        SearchCustomFiltersComponent,
                        AssetTypeCellRendererComponent
                    ],
                    imports: [CoreModule, DeviceGridModule, CoreSearchModule],
                    entryComponents: [SearchResultsComponent, SearchActionComponent],
                    exports: [SearchResultsComponent, SearchGridComponent],
                    providers: [
                        AssetSearchService,
                        hookRoute({
                            path: 'assetsearch',
                            component: SearchResultsComponent
                        }),
                        hookSearch({
                            template: SearchActionComponent,
                            onSearch: undefined
                        })
                    ]
                }]
        }] });
