// @c8y/ngx-components/esm2020/core/docs/guide-href.directive.mjs
import { Directive, Input, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
export class GuideHrefDirective {
    set baseHref(href) {
        this.htmlAnchor.href = `${href}/${this.c8yGuideHref}`;
    }
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.htmlAnchor = this.elementRef.nativeElement;
        if (!this.htmlAnchor.hasAttribute('target')) {
            this.htmlAnchor.setAttribute('target', '_blank');
        }
    }
}
GuideHrefDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GuideHrefDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
GuideHrefDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: GuideHrefDirective, selector: "[c8y-guide-href]", inputs: { c8yGuideHref: ["c8y-guide-href", "c8yGuideHref"] }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GuideHrefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[c8y-guide-href]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { c8yGuideHref: [{
                type: Input,
                args: ['c8y-guide-href']
            }] } });
