// @c8y/ngx-components/esm2020/api/data.module.mjs
import { NgModule } from '@angular/core';
import { BasicAuth, FetchClient, Realtime, CookieAuth } from '@c8y/client';
import { ApiService } from './api.service';
import * as services from './services';
import * as i0 from "@angular/core";
function toProvider(provide) {
    let deps = [FetchClient, Realtime];
    if (provide === FetchClient) {
        deps = [CookieAuth];
    }
    if (provide === BasicAuth || provide === CookieAuth) {
        deps = [];
    }
    if (provide === Realtime) {
        deps = [FetchClient];
    }
    return { provide, useClass: provide, deps };
}
const providers = Object.keys(services)
    .map(k => toProvider(services[k]))
    .concat([{ provide: ApiService, useClass: ApiService, deps: [FetchClient] }]);
// @dynamic
export class DataModule {
    static providers() {
        return providers;
    }
    static forRoot() {
        return {
            ngModule: DataModule,
            providers
        };
    }
}
DataModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DataModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DataModule });
DataModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataModule, providers: providers });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DataModule, decorators: [{
            type: NgModule,
            args: [{
                    providers
                }]
        }] });
