// @c8y/ngx-components/esm2020/api/http-handler.model.mjs
import { from } from 'rxjs';
import { HttpHandler } from './interceptor.model';
export class HttpInterceptHandler extends HttpHandler {
    constructor(interceptor, nextHandler) {
        super();
        this.interceptor = interceptor;
        this.nextHandler = nextHandler;
    }
    handle(req) {
        return this.interceptor.intercept(req, this.nextHandler);
    }
}
export class HttpRequestHandler extends HttpHandler {
    constructor(fetch, apiService) {
        super();
        this.fetch = fetch;
        this.apiService = apiService;
    }
    handle(req) {
        const { options, url } = req;
        const { method } = options;
        this.apiService?.onStart({ method, options, url });
        let fetchPromise = this.fetch(url, options);
        if (typeof options.responseInterceptor === 'function') {
            fetchPromise = fetchPromise.then(options.responseInterceptor);
        }
        fetchPromise.then((response) => this.apiService?.onFinish({ method, options, url, response }), (response) => this.apiService?.onFinish({ method, options, url, response }));
        return from(fetchPromise);
    }
}
