// @c8y/ngx-components/esm2020/app-logs/app-logs-auto-refresh.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@c8y/ngx-components';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { AppLogsAutoRefreshComponent } from './app-logs-auto-refresh.component';
import { AppLogsService } from './app-logs.service';
import * as i0 from "@angular/core";
export class AppLogsAutoRefreshModule {
}
AppLogsAutoRefreshModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AppLogsAutoRefreshModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AppLogsAutoRefreshModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AppLogsAutoRefreshModule, declarations: [AppLogsAutoRefreshComponent], imports: [CommonModule, ButtonsModule], exports: [AppLogsAutoRefreshComponent] });
AppLogsAutoRefreshModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AppLogsAutoRefreshModule, providers: [AppLogsService], imports: [CommonModule, ButtonsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AppLogsAutoRefreshModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AppLogsAutoRefreshComponent],
                    exports: [AppLogsAutoRefreshComponent],
                    imports: [CommonModule, ButtonsModule],
                    entryComponents: [AppLogsAutoRefreshComponent],
                    providers: [AppLogsService]
                }]
        }] });
