// @c8y/ngx-components/esm2020/auth-configuration/basic-settings/basic-settings.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, hookRoute } from '@c8y/ngx-components';
import { BasicAuthSettingsComponent } from './basic-auth-settings.component';
import { AuthConfigurationGuard } from '../auth-configuration.guard';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { AuthConfigurationService } from './auth-configuration.service';
import { AuthConfigurationComponent } from './auth-configuration.component';
import { TfaSettingsComponent } from './tfa-settings.component';
import { SessionConfigurationComponent } from './session-configuration.component';
import { LoginSettingsComponent } from './login-settings.component';
import { TenantLoginOptionMapper } from './tenant-login-option.mapper';
import * as i0 from "@angular/core";
export class BasicSettingsModule {
}
BasicSettingsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BasicSettingsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BasicSettingsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BasicSettingsModule, declarations: [SessionConfigurationComponent,
        LoginSettingsComponent,
        BasicAuthSettingsComponent,
        AuthConfigurationComponent,
        TfaSettingsComponent], imports: [CoreModule, PopoverModule] });
BasicSettingsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BasicSettingsModule, providers: [
        AuthConfigurationService,
        TenantLoginOptionMapper,
        hookRoute({
            path: 'auth-configuration/basic_settings',
            component: AuthConfigurationComponent,
            canActivate: [AuthConfigurationGuard]
        })
    ], imports: [CoreModule, PopoverModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BasicSettingsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SessionConfigurationComponent,
                        LoginSettingsComponent,
                        BasicAuthSettingsComponent,
                        AuthConfigurationComponent,
                        TfaSettingsComponent
                    ],
                    exports: [],
                    imports: [CoreModule, PopoverModule],
                    providers: [
                        AuthConfigurationService,
                        TenantLoginOptionMapper,
                        hookRoute({
                            path: 'auth-configuration/basic_settings',
                            component: AuthConfigurationComponent,
                            canActivate: [AuthConfigurationGuard]
                        })
                    ]
                }]
        }] });
