// @c8y/ngx-components/esm2020/auth-configuration/basic-settings/typed-option.mjs
import { isString } from 'lodash-es';
export class TypedOption {
    constructor(category, key, type, defaultValue, value) {
        this.category = category;
        this.key = key;
        this.type = type;
        this.defaultValue = defaultValue;
        this.value = value;
    }
    apply(option) {
        this.category = option.category;
        this.key = option.key;
        this.value = option.value;
    }
    getValue() {
        try {
            return this.getValueByType();
        }
        catch (ex) {
            return this.defaultValue;
        }
    }
    getValueByType() {
        switch (this.type) {
            case 'boolean':
                return this.value.toLowerCase() === 'true';
            case 'number':
                return this.parseNumberValue(this.value);
            case 'string':
                return this.parseStringValue(this.value);
            default:
                throw new TypeError();
        }
    }
    parseNumberValue(stringValue) {
        const value = Number.parseInt(stringValue, 10);
        if (typeof value !== 'number' || isNaN(value)) {
            throw new Error();
        }
        return value;
    }
    parseStringValue(value) {
        if (!isString(value)) {
            throw new Error();
        }
        return value;
    }
}
