// @c8y/ngx-components/esm2020/core/time-picker/time-picker.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/index';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { TimePickerComponent } from './time-picker.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/datepicker";
/**
 * The angular module definition for a time picker - it provides FormControl that handles time.
 */
export class TimePickerModule {
}
TimePickerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TimePickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TimePickerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: TimePickerModule, declarations: [TimePickerComponent], imports: [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule, i1.BsDatepickerModule, C8yCommonModule], exports: [TimePickerComponent] });
TimePickerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TimePickerModule, imports: [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        BsDatepickerModule.forRoot(),
        C8yCommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TimePickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        C8yFormsModule,
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        BsDatepickerModule.forRoot(),
                        C8yCommonModule
                    ],
                    exports: [TimePickerComponent],
                    declarations: [TimePickerComponent]
                }]
        }] });
