// @c8y/ngx-components/esm2020/core/breadcrumb/breadcrumb.component.mjs
import { Component, ContentChildren, QueryList } from '@angular/core';
import { BreadcrumbService } from './breadcrumb.service';
import { BreadcrumbItemComponent } from './breadcrumb-item.component';
import * as i0 from "@angular/core";
import * as i1 from "./breadcrumb.service";
export class BreadcrumbComponent {
    constructor(breadcrumbService) {
        this.breadcrumbService = breadcrumbService;
    }
    ngAfterViewInit() {
        this.items = this.contentChildItems.toArray();
        this.viewInitTimeout = setTimeout(() => this.breadcrumbService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.breadcrumbService.remove(this);
    }
}
BreadcrumbComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BreadcrumbComponent, deps: [{ token: i1.BreadcrumbService }], target: i0.ɵɵFactoryTarget.Component });
BreadcrumbComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: BreadcrumbComponent, selector: "c8y-breadcrumb", queries: [{ propertyName: "contentChildItems", predicate: BreadcrumbItemComponent }], ngImport: i0, template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BreadcrumbComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-breadcrumb', template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>" }]
        }], ctorParameters: function () { return [{ type: i1.BreadcrumbService }]; }, propDecorators: { contentChildItems: [{
                type: ContentChildren,
                args: [BreadcrumbItemComponent]
            }] } });
