// @c8y/ngx-components/esm2020/core/breadcrumb/breadcrumb.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { BreadcrumbItemComponent } from './breadcrumb-item.component';
import { BreadcrumbOutletComponent } from './breadcrumb-outlet.component';
import { BreadcrumbComponent } from './breadcrumb.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "ngx-bootstrap/tooltip";
/**
 * The angular module definition for breadcrumbs.
 * @exports BreadcrumbComponent
 * @exports BreadcrumbItemComponent
 * @exports BreadcrumbOutletComponent
 */
export class BreadcrumbModule {
}
BreadcrumbModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BreadcrumbModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BreadcrumbModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BreadcrumbModule, declarations: [BreadcrumbComponent, BreadcrumbItemComponent, BreadcrumbOutletComponent], imports: [i1.RouterModule, CommonModule, i2.TooltipModule], exports: [BreadcrumbComponent, BreadcrumbItemComponent, BreadcrumbOutletComponent] });
BreadcrumbModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BreadcrumbModule, imports: [RouterModule.forChild([]), CommonModule, TooltipModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BreadcrumbModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BreadcrumbComponent, BreadcrumbItemComponent, BreadcrumbOutletComponent],
                    imports: [RouterModule.forChild([]), CommonModule, TooltipModule.forRoot()],
                    exports: [BreadcrumbComponent, BreadcrumbItemComponent, BreadcrumbOutletComponent]
                }]
        }] });
