// @c8y/ngx-components/esm2020/operations/create-bulk-operation-details/create-bulk-operation-details.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule } from '@c8y/ngx-components';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { CreateBulkOperationDetailsComponent } from './create-bulk-operation-details.component';
import * as i0 from "@angular/core";
/**
 * This module provides a component for entering operation details.
 */
export class CreateBulkOperationDetailsModule {
}
CreateBulkOperationDetailsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CreateBulkOperationDetailsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CreateBulkOperationDetailsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: CreateBulkOperationDetailsModule, declarations: [CreateBulkOperationDetailsComponent], imports: [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule], exports: [CreateBulkOperationDetailsComponent] });
CreateBulkOperationDetailsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CreateBulkOperationDetailsModule, imports: [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CreateBulkOperationDetailsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule],
                    declarations: [CreateBulkOperationDetailsComponent],
                    exports: [CreateBulkOperationDetailsComponent]
                }]
        }] });
