// @c8y/ngx-components/esm2020/sub-assets/add-group/add-group.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule, hookRoute } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { AddGroupComponent } from './add-group.component';
import { AddGroupService } from './add-group.service';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class AddGroupModule {
}
AddGroupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AddGroupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AddGroupModule, declarations: [AddGroupComponent], imports: [CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule], exports: [AddGroupComponent] });
AddGroupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddGroupModule, providers: [
        hookRoute({
            path: 'add-group',
            component: AddGroupComponent
        }),
        AddGroupService
    ], imports: [CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AddGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AddGroupComponent],
                    imports: [CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule],
                    entryComponents: [AddGroupComponent],
                    exports: [AddGroupComponent],
                    providers: [
                        hookRoute({
                            path: 'add-group',
                            component: AddGroupComponent
                        }),
                        AddGroupService
                    ]
                }]
        }] });
