// @c8y/ngx-components/esm2020/assets-navigator/asset-selector/group-node.service.mjs
import { Injectable } from '@angular/core';
import { AssetNodeService } from '../asset-node.service';
import { GroupNode } from './group-node';
import * as i0 from "@angular/core";
export class GroupNodeService extends AssetNodeService {
    constructor() {
        super(...arguments);
        this.PAGE_SIZE = 5;
    }
    createGroupNode(config) {
        return new GroupNode(this, config);
    }
    createChildNode(config) {
        return this.createGroupNode(config);
    }
}
GroupNodeService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GroupNodeService, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
GroupNodeService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GroupNodeService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GroupNodeService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
