// @c8y/ngx-components/esm2020/assets-navigator/asset-selector/unassigned-devices-node.mjs
import { gettext } from '@c8y/ngx-components';
import { Action } from '../action.enum';
import { GroupNode } from './group-node';
export class UnassignedDevicesNode extends GroupNode {
    constructor(service, config = {}) {
        super(service, config);
        this.service = service;
        this.label = gettext('Unassigned devices');
        this.icon = 'delete-folder';
        this.groupsSelectable = false;
        this.groupsOnly = false;
        this.showChildDevices = false;
        this.mo = {};
        this.priority = Infinity;
        this.showChildDevices = config.showChildDevices || false;
    }
    click(options = {}) {
        this.hookEvents();
        if (options.open) {
            this.events.next(Action.FETCH);
        }
    }
    addManagedObject(mo) {
        this.add(this.service.createChildNode({ mo, showChildDevices: this.showChildDevices }));
    }
    fetch() {
        return this.service.getUnassignedDevices(this.showChildDevices, this.filterQuery$.value);
    }
    isGroup() {
        return true;
    }
    toString() {
        return UnassignedDevicesNode.NAME;
    }
    setLabel() {
        // do nothing, as this item has a hardcoded label.
        // This is only to overwrite the default inherited method.
        this.label = this.label;
    }
    async updateIcon() {
        // do nothing, as this item has a hardcoded label.
        // This is only to overwrite the default inherited method.
        this.icon = this.icon;
    }
}
UnassignedDevicesNode.NAME = 'UnassignedDevicesNode';
