// @c8y/ngx-components/esm2020/default-subscriptions/default-subscriptions.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, hookNavigator, hookRoute } from '@c8y/ngx-components';
import { DefaultSubscriptionsComponent } from './default-subscriptions.component';
import { DefaultSubscriptionsService } from './default-subscriptions.service';
import { DefaultSubscriptionsGuard } from './default-subscriptions.guard';
import { DefaultSubscriptionsNavigationFactory } from './default-subscriptions-navigation.factory';
import * as i0 from "@angular/core";
export class DefaultSubscriptionsModule {
}
DefaultSubscriptionsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DefaultSubscriptionsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsModule, declarations: [DefaultSubscriptionsComponent], imports: [CoreModule, ReactiveFormsModule], exports: [DefaultSubscriptionsComponent] });
DefaultSubscriptionsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsModule, providers: [
        DefaultSubscriptionsService,
        DefaultSubscriptionsGuard,
        hookNavigator(DefaultSubscriptionsNavigationFactory),
        hookRoute({
            path: 'default-subscriptions',
            component: DefaultSubscriptionsComponent,
            canActivate: [DefaultSubscriptionsGuard]
        })
    ], imports: [CoreModule, ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, ReactiveFormsModule],
                    declarations: [DefaultSubscriptionsComponent],
                    entryComponents: [DefaultSubscriptionsComponent],
                    exports: [DefaultSubscriptionsComponent],
                    providers: [
                        DefaultSubscriptionsService,
                        DefaultSubscriptionsGuard,
                        hookNavigator(DefaultSubscriptionsNavigationFactory),
                        hookRoute({
                            path: 'default-subscriptions',
                            component: DefaultSubscriptionsComponent,
                            canActivate: [DefaultSubscriptionsGuard]
                        })
                    ]
                }]
        }] });
