// @c8y/ngx-components/esm2020/default-subscriptions/default-subscriptions-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { DefaultSubscriptionsGuard } from './default-subscriptions.guard';
import * as i0 from "@angular/core";
import * as i1 from "./default-subscriptions.guard";
export class DefaultSubscriptionsNavigationFactory {
    constructor(defaultSubscriptionsGuard) {
        this.defaultSubscriptionsGuard = defaultSubscriptionsGuard;
        this.navNode = new NavigatorNode({
            label: gettext('Default subscriptions'),
            path: '/default-subscriptions',
            icon: 'c8y-c8y-data',
            parent: gettext('Ecosystem'),
            priority: 100
        });
    }
    /** Returns the navigation node if the feature is active. */
    async get() {
        if (await this.defaultSubscriptionsGuard.canActivate()) {
            return this.navNode;
        }
    }
}
DefaultSubscriptionsNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsNavigationFactory, deps: [{ token: i1.DefaultSubscriptionsGuard }], target: i0.ɵɵFactoryTarget.Injectable });
DefaultSubscriptionsNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DefaultSubscriptionsNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.DefaultSubscriptionsGuard }]; } });
