// @c8y/ngx-components/esm2020/operations/product-experience/operations-product-experience.constants.mjs
const DEVICE_CONTROL_EVENT_PREFIX = 'deviceControl';
const DEVICE_CONTROL_EVENT_SINGLE_PREFIX = `${DEVICE_CONTROL_EVENT_PREFIX}:single`;
/* Event names */
export const SINGLE_OPERATION_EVENT_OVERVIEW = `${DEVICE_CONTROL_EVENT_SINGLE_PREFIX}:overview`;
export const SINGLE_OPERATION_EVENT_DEVICE_TAB = `${DEVICE_CONTROL_EVENT_SINGLE_PREFIX}:deviceTab`;
export const BULK_OPERATION_EVENT = `${DEVICE_CONTROL_EVENT_PREFIX}:bulk`;
/* Actions */
export const ACTIONS = {
    FILTER: 'filter',
    OPEN_OPERATION_DETAILS: 'openOperationDetails',
    RETRY: 'retry'
};
export const ACTIONS_SINGLE = {
    SCHEDULE_AS_BULK_OPERATION: 'scheduleAsBulkOperation'
};
export const ACTIONS_BULK = {
    OPEN_ADD_BULK_OPERATION_DIALOG: 'openAddBulkOperationDialog',
    CANCEL_ADD_BULK_OPERATION_DIALOG: 'cancelAddBulkOperationDialog',
    SELECT_BULK_OPERTION_TYPE: 'selectBulkOperationType'
};
/* Components */
export const COMPONENTS = {
    OPERATION_DETAILS: 'operation-details',
    SINGLE_OPERATION_ITEM: 'single-operation-item'
};
/* Add bulk operation dialog steps */
export const BULK_OPERATION_DIALOG_STEPS = {
    SELECT_OPERATION_TYPE: 'selectOperationType'
};
