// @c8y/ngx-components/esm2020/connectivity/actility/actility-tab.factory.mjs
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { AbstractConnectivityTabFactory } from '../abstract-connectivity-tab.factory';
import { ActilityGuard } from './actility.guard';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./actility.guard";
export class ActilityTabFactory extends AbstractConnectivityTabFactory {
    constructor(router, guard) {
        super(router, guard);
        this.router = router;
        this.guard = guard;
        this.tabConfig = {
            label: gettext('Actility'),
            path: '/connectivitySettings/multiple_lns_connectors_actility',
            icon: 'wall-socket'
        };
    }
    config() {
        return this.tabConfig;
    }
}
ActilityTabFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityTabFactory, deps: [{ token: i1.Router }, { token: i2.ActilityGuard }], target: i0.ɵɵFactoryTarget.Injectable });
ActilityTabFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityTabFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ActilityTabFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.ActilityGuard }]; } });
