// @c8y/ngx-components/esm2020/repository/firmware/list/firmware-repository-list.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FirmwareListComponent } from './firmware-list.component';
import { FirmwareDetailsComponent } from './firmware-details.component';
import { AddFirmwareModalComponent } from './add-firmware-modal.component';
import { AddFirmwarePatchModalComponent } from './add-firmware-patch-modal.component';
import { CoreModule, FormsModule, hookNavigator, hookRoute } from '@c8y/ngx-components';
import { FirmwareRepositoryNavigationFactory } from './firmware-repository-navigation-factory';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import * as i0 from "@angular/core";
export class FirmwareRepositoryListModule {
    static forRoot() {
        return {
            ngModule: FirmwareRepositoryListModule,
            providers: [
                hookNavigator(FirmwareRepositoryNavigationFactory),
                hookRoute([
                    {
                        path: 'firmware',
                        component: FirmwareListComponent
                    },
                    {
                        path: 'firmware/:id',
                        component: FirmwareDetailsComponent
                    }
                ])
            ]
        };
    }
}
FirmwareRepositoryListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FirmwareRepositoryListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryListModule, declarations: [FirmwareListComponent,
        FirmwareDetailsComponent,
        AddFirmwareModalComponent,
        AddFirmwarePatchModalComponent], imports: [CommonModule,
        CoreModule,
        FormsModule,
        PopoverModule,
        TooltipModule,
        SharedRepositoryModule] });
FirmwareRepositoryListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryListModule, imports: [CommonModule,
        CoreModule,
        FormsModule,
        PopoverModule,
        TooltipModule,
        SharedRepositoryModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FirmwareRepositoryListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        CoreModule,
                        FormsModule,
                        PopoverModule,
                        TooltipModule,
                        SharedRepositoryModule
                    ],
                    declarations: [
                        FirmwareListComponent,
                        FirmwareDetailsComponent,
                        AddFirmwareModalComponent,
                        AddFirmwarePatchModalComponent
                    ],
                    entryComponents: [
                        FirmwareListComponent,
                        FirmwareDetailsComponent,
                        AddFirmwareModalComponent,
                        AddFirmwarePatchModalComponent
                    ]
                }]
        }] });
