// @c8y/ngx-components/esm2020/core/select/select.module.mjs
import { NgModule } from '@angular/core';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { SelectComponent } from './select.component';
import { CommonModule } from '../common/common.module';
import { TypeaheadComponent } from './typeahead.component';
import { ListGroupModule } from '../list-group/list-group.module';
import { FormsModule } from '../forms/forms.module';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/dropdown";
/**
 * The angular module definition for a select.
 * @exports SelectComponent
 */
export class SelectModule {
}
SelectModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SelectModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: SelectModule, declarations: [SelectComponent, TypeaheadComponent], imports: [i1.BsDropdownModule, CommonModule, ListGroupModule, FormsModule], exports: [SelectComponent, TypeaheadComponent] });
SelectModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SelectModule, imports: [BsDropdownModule.forRoot(), CommonModule, ListGroupModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [BsDropdownModule.forRoot(), CommonModule, ListGroupModule, FormsModule],
                    exports: [SelectComponent, TypeaheadComponent],
                    declarations: [SelectComponent, TypeaheadComponent]
                }]
        }] });
