// @c8y/ngx-components/esm2020/widgets/implementations/map/map-widget.component.mjs
import { Component, ViewChild } from '@angular/core';
import { ClusterMapComponent } from '@c8y/ngx-components/map';
import { DashboardChildComponent } from '@c8y/ngx-components';
import { filter, takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@angular/common";
import * as i3 from "@c8y/ngx-components/map";
import * as i4 from "@angular/router";
import * as i5 from "./map-event-info.component";
export class MapWidgetComponent {
    constructor(dashboardChild) {
        this.mapConfig = { center: [0, 0] };
        this.destroy$ = new Subject();
        this.listenToWidgetResizeEvent(dashboardChild);
    }
    ngOnInit() {
        if (this.config.device) {
            this.rootNode = this.config.device;
        }
        this.mapConfig = { ...this.config.mapConfig };
        this.savedNode = this.rootNode;
    }
    startFollow(context) {
        if (context.id !== this.rootNode?.id) {
            this.rootNode = context;
        }
        this.mapConfig = {
            ...this.config.mapConfig,
            follow: true,
            realtime: true
        };
    }
    stopFollow() {
        this.mapConfig = {
            ...this.config.mapConfig,
            follow: false
        };
        this.rootNode = this.savedNode;
    }
    ngOnDestroy() {
        this.destroy$.next();
    }
    listenToWidgetResizeEvent(dashboardChild) {
        dashboardChild.changeEnd
            .pipe(filter(child => child.lastChange === 'resize'), takeUntil(this.destroy$))
            .subscribe(() => {
            this.clusterMap.reset();
        });
    }
}
MapWidgetComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapWidgetComponent, deps: [{ token: i1.DashboardChildComponent }], target: i0.ɵɵFactoryTarget.Component });
MapWidgetComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: MapWidgetComponent, selector: "c8y-map-widget", viewQueries: [{ propertyName: "clusterMap", first: true, predicate: ClusterMapComponent, descendants: true }], ngImport: i0, template: "<c8y-map-status\n  [clusterMap]=\"mapWidget\"\n  [(config)]=\"mapConfig\"\n  (onUnfollow)=\"stopFollow()\"\n></c8y-map-status>\n<c8y-cluster-map #mapWidget [rootNode]=\"rootNode\" [config]=\"mapConfig\">\n  <div *c8yMapPopup=\"let context\" class=\"map-marker\">\n    <a class=\"text-truncate deviceLink text-12\" routerLink=\"/device/{{ context.id }}\">\n      <strong>{{ context.name }}</strong>\n    </a>\n    <c8y-map-event-info [asset]=\"context\">\n      <button\n        type=\"button\"\n        class=\"btn btn-default btn-xs btn-block m-t-8\"\n        (click)=\"startFollow(context)\"\n        [title]=\"'Activate realtime on this asset and follow it if it moves' | translate\"\n        *ngIf=\"!mapConfig.follow\"\n        translate\n      >\n        Follow\n      </button>\n      <button\n        type=\"button\"\n        class=\"btn btn-default btn-xs btn-block m-t-8\"\n        (click)=\"stopFollow()\"\n        [title]=\"'Stop following this asset.' | translate\"\n        *ngIf=\"mapConfig.follow\"\n        translate\n      >\n        Unfollow\n      </button>\n    </c8y-map-event-info>\n  </div>\n</c8y-cluster-map>\n", dependencies: [{ kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i1.C8yTranslateDirective, selector: "[translate],[ngx-translate]" }, { kind: "component", type: i3.MapStatusComponent, selector: "c8y-map-status", inputs: ["config", "clusterMap"], outputs: ["configChange", "onUnfollow"] }, { kind: "component", type: i3.ClusterMapComponent, selector: "c8y-cluster-map", inputs: ["config", "rootNode", "asset", "showClusterColor"], outputs: ["mapChange"] }, { kind: "directive", type: i3.MapPopupDirective, selector: "[c8yMapPopup]" }, { kind: "directive", type: i4.RouterLink, selector: "[routerLink]", inputs: ["target", "queryParams", "fragment", "queryParamsHandling", "state", "relativeTo", "preserveFragment", "skipLocationChange", "replaceUrl", "routerLink"] }, { kind: "component", type: i5.MapEventInfoComponent, selector: "c8y-map-event-info", inputs: ["asset"] }, { kind: "pipe", type: i1.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapWidgetComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-map-widget', template: "<c8y-map-status\n  [clusterMap]=\"mapWidget\"\n  [(config)]=\"mapConfig\"\n  (onUnfollow)=\"stopFollow()\"\n></c8y-map-status>\n<c8y-cluster-map #mapWidget [rootNode]=\"rootNode\" [config]=\"mapConfig\">\n  <div *c8yMapPopup=\"let context\" class=\"map-marker\">\n    <a class=\"text-truncate deviceLink text-12\" routerLink=\"/device/{{ context.id }}\">\n      <strong>{{ context.name }}</strong>\n    </a>\n    <c8y-map-event-info [asset]=\"context\">\n      <button\n        type=\"button\"\n        class=\"btn btn-default btn-xs btn-block m-t-8\"\n        (click)=\"startFollow(context)\"\n        [title]=\"'Activate realtime on this asset and follow it if it moves' | translate\"\n        *ngIf=\"!mapConfig.follow\"\n        translate\n      >\n        Follow\n      </button>\n      <button\n        type=\"button\"\n        class=\"btn btn-default btn-xs btn-block m-t-8\"\n        (click)=\"stopFollow()\"\n        [title]=\"'Stop following this asset.' | translate\"\n        *ngIf=\"mapConfig.follow\"\n        translate\n      >\n        Unfollow\n      </button>\n    </c8y-map-event-info>\n  </div>\n</c8y-cluster-map>\n" }]
        }], ctorParameters: function () { return [{ type: i1.DashboardChildComponent }]; }, propDecorators: { clusterMap: [{
                type: ViewChild,
                args: [ClusterMapComponent]
            }] } });
