// @c8y/ngx-components/esm2020/widgets/implementations/map/map-widget.module.mjs
import { CommonModule as NgCommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule as NgFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { CommonModule, FormsModule } from '@c8y/ngx-components';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { MapModule } from '@c8y/ngx-components/map';
import { MapEventInfoComponent } from './map-event-info.component';
import { MapWidgetConfigComponent } from './map-widget-config.component';
import { MapWidgetComponent } from './map-widget.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/tooltip";
export class MapWidgetModule {
}
MapWidgetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapWidgetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MapWidgetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: MapWidgetModule, declarations: [MapWidgetComponent, MapWidgetConfigComponent, MapEventInfoComponent], imports: [NgCommonModule,
        NgFormsModule,
        CommonModule,
        FormsModule,
        MapModule,
        RouterModule, i1.TooltipModule], exports: [MapWidgetComponent, MapWidgetConfigComponent, MapEventInfoComponent] });
MapWidgetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapWidgetModule, imports: [NgCommonModule,
        NgFormsModule,
        CommonModule,
        FormsModule,
        MapModule,
        RouterModule,
        TooltipModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: MapWidgetModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MapWidgetComponent, MapWidgetConfigComponent, MapEventInfoComponent],
                    imports: [
                        NgCommonModule,
                        NgFormsModule,
                        CommonModule,
                        FormsModule,
                        MapModule,
                        RouterModule,
                        TooltipModule.forRoot()
                    ],
                    exports: [MapWidgetComponent, MapWidgetConfigComponent, MapEventInfoComponent]
                }]
        }] });
