// @c8y/ngx-components/esm2020/auth-configuration/auth-configuration.module.mjs
import { NgModule } from '@angular/core';
import { BasicSettingsModule } from './basic-settings/basic-settings.module';
import { SsoConfigurationModule } from './sso-configuration/sso-configuration.module';
import { hookNavigator, hookRoute, hookTab } from '@c8y/ngx-components';
import { TabsFactory } from './factories/tabs.factory';
import { NavigationFactory } from './factories/navigation.factory';
import { AuthConfigurationGuard } from './auth-configuration.guard';
import * as i0 from "@angular/core";
export class AuthConfigurationModule {
}
AuthConfigurationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AuthConfigurationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AuthConfigurationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: AuthConfigurationModule, imports: [BasicSettingsModule, SsoConfigurationModule] });
AuthConfigurationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AuthConfigurationModule, providers: [
        AuthConfigurationGuard,
        hookTab(TabsFactory),
        hookNavigator(NavigationFactory),
        hookRoute({
            path: 'auth-configuration',
            redirectTo: 'auth-configuration/basic_settings',
            pathMatch: 'full'
        })
    ], imports: [BasicSettingsModule, SsoConfigurationModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: AuthConfigurationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [BasicSettingsModule, SsoConfigurationModule],
                    providers: [
                        AuthConfigurationGuard,
                        hookTab(TabsFactory),
                        hookNavigator(NavigationFactory),
                        hookRoute({
                            path: 'auth-configuration',
                            redirectTo: 'auth-configuration/basic_settings',
                            pathMatch: 'full'
                        })
                    ]
                }]
        }] });
