// @c8y/ngx-components/esm2020/operations/operation-summary/operation-summary.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { OperationSummaryComponent } from './operation-summary.component';
import * as i0 from "@angular/core";
/**
 * This module provides a component for displaying the summary of given bulk operation.
 */
export class OperationSummaryModule {
}
OperationSummaryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationSummaryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OperationSummaryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: OperationSummaryModule, declarations: [OperationSummaryComponent], imports: [CoreModule], exports: [OperationSummaryComponent] });
OperationSummaryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationSummaryModule, imports: [CoreModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationSummaryModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule],
                    declarations: [OperationSummaryComponent],
                    entryComponents: [OperationSummaryComponent],
                    exports: [OperationSummaryComponent]
                }]
        }] });
