// @c8y/ngx-components/esm2020/core/data-grid/filter-chip/filter-mapper.factory.mjs
import { Inject, Injectable, InjectionToken, Optional } from '@angular/core';
import { castArray } from 'lodash-es';
import { hookGeneric } from '../../common/extension-hooks';
import * as i0 from "@angular/core";
const HOOK_FILTER_MAPPER = new InjectionToken('HOOK_FILTER_MAPPER');
/**
 * You can either provide a single `FilterMapper` as parameter:
 * ```typescript
 *  hookFilterMapper(...)
 * ```
 *
 * Or an array to directly register multiple:
 * ```typescript
 *  hookFilterMapper([...])
 * ```
 *
 * Or you provide an Service that implements `FilterMapper`
 * ```typescript
 *  export class MyFilterMapper implements FilterMapper {...}
 *  ...
 *  hookFilterMapper(MyFilterMapper)
 * ```
 * A typed alternative to using `HOOK_FILTER_MAPPER`.
 * @param mapper A filter mapper, e.g. `BooleanFilterMapper`, or `ExtensionFactory` to be provided.
 * @param options If this is a multi provider or not (defaults to true) and provider type definition (defaults to ClassProvider) - `HookOptions`.
 * @returns An `Provider` to be provided in your module.
 */
export function hookFilterMapper(mapper, options) {
    return hookGeneric(mapper, HOOK_FILTER_MAPPER, options);
}
export class FilterMapperFactory {
    constructor(fieldMappers) {
        this.map = {};
        fieldMappers.forEach(fieldMapper => {
            castArray(fieldMapper.fieldType).forEach(type => {
                this.map[type] = fieldMapper;
            });
        });
    }
    get(type) {
        if (type === undefined || type === null) {
            throw new Error('Invalid input: type is undefined or null');
        }
        if (typeof type !== 'string' && typeof type !== 'function') {
            throw new Error(`Invalid input: type must be a string or a Type<FieldType>, but received ${typeof type}`);
        }
        const typeName = typeof type === 'string' ? type : type.name;
        const mapper = this.map[typeName];
        if (!mapper) {
            throw new Error(`Could not find FilterMapper for type ${typeName}`);
        }
        return mapper;
    }
}
FilterMapperFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperFactory, deps: [{ token: HOOK_FILTER_MAPPER, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
FilterMapperFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: Array, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [HOOK_FILTER_MAPPER]
                }] }]; } });
