// @c8y/ngx-components/esm2020/core/data-grid/filter-chip/filter-mapper.module.mjs
import { NgModule } from '@angular/core';
import { FilterMapperPipe, GroupedFilterChips } from '..';
import { BooleanFilterMapper, DateFilterMapper, RadioFilterMapper, SelectFilterMapper, StringFilterMapper, TypeaheadFilterMapper } from './filter-mapper-impl.service';
import { hookFilterMapper } from './filter-mapper.factory';
import * as i0 from "@angular/core";
export class FilterMapperModule {
}
FilterMapperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilterMapperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperModule, declarations: [FilterMapperPipe, GroupedFilterChips], exports: [FilterMapperPipe, GroupedFilterChips] });
FilterMapperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperModule, providers: [
        hookFilterMapper(BooleanFilterMapper),
        hookFilterMapper(RadioFilterMapper),
        hookFilterMapper(TypeaheadFilterMapper),
        hookFilterMapper(DateFilterMapper),
        hookFilterMapper(SelectFilterMapper),
        hookFilterMapper(StringFilterMapper)
    ] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [
                        hookFilterMapper(BooleanFilterMapper),
                        hookFilterMapper(RadioFilterMapper),
                        hookFilterMapper(TypeaheadFilterMapper),
                        hookFilterMapper(DateFilterMapper),
                        hookFilterMapper(SelectFilterMapper),
                        hookFilterMapper(StringFilterMapper)
                    ],
                    declarations: [FilterMapperPipe, GroupedFilterChips],
                    exports: [FilterMapperPipe, GroupedFilterChips]
                }]
        }] });
