// @c8y/ngx-components/esm2020/core/data-grid/filter-chip/filter-mapper.pipe.mjs
import { Pipe } from '@angular/core';
import { of } from 'rxjs';
import { FilterMapperService } from './filter-mapper.service';
import * as i0 from "@angular/core";
import * as i1 from "./filter-mapper.service";
export class FilterMapperPipe {
    constructor(filterMapperService) {
        this.filterMapperService = filterMapperService;
    }
    transform(column) {
        const filterMapperService = this.filterMapperService;
        const filterObject = {
            value: '',
            displayValue: '',
            path: [],
            externalFilterQuery: column.externalFilterQuery,
            filteringConfig: column.filteringConfig,
            columnName: column.name,
            remove() {
                return filterMapperService.removeChip(this);
            }
        };
        if (column.filterPredicate) {
            const chip = {
                value: column.filterPredicate,
                displayValue: column.filterPredicate,
                filterPredicate: column.filterPredicate,
                columnName: column.name,
                remove() {
                    return {
                        columnName: this.columnName,
                        filterPredicate: this.filterPredicate
                    };
                }
            };
            return of([chip]);
        }
        return filterMapperService.getMappedFilterValues(filterObject);
    }
}
FilterMapperPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperPipe, deps: [{ token: i1.FilterMapperService }], target: i0.ɵɵFactoryTarget.Pipe });
FilterMapperPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperPipe, name: "mapToFilterChips" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: FilterMapperPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'mapToFilterChips'
                }]
        }], ctorParameters: function () { return [{ type: i1.FilterMapperService }]; } });
