// @c8y/ngx-components/esm2020/core/data-grid/filter-chip/grouped-filter-chips.pipe.mjs
import { Pipe } from '@angular/core';
import { groupBy } from 'lodash-es';
import * as i0 from "@angular/core";
export class GroupedFilterChips {
    transform(chips) {
        const groupedChips = groupBy(chips, 'label');
        const result = Object.entries(groupedChips).map(([label, chips]) => ({
            label: label === 'undefined' ? undefined : label,
            chips: chips
        }));
        return result.reverse();
    }
}
GroupedFilterChips.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GroupedFilterChips, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
GroupedFilterChips.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: GroupedFilterChips, name: "groupedFilterChips" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: GroupedFilterChips, decorators: [{
            type: Pipe,
            args: [{
                    name: 'groupedFilterChips'
                }]
        }] });
