// @c8y/ngx-components/esm2020/repository/configuration/list/configuration-repository-navigation-factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
export class ConfigurationRepositoryNavigationFactory {
    constructor() {
        this.node = new NavigatorNode({
            label: gettext('Configuration repository'),
            path: 'configuration',
            icon: 'gears',
            parent: gettext('Management'),
            priority: 800
        });
    }
    get() {
        return this.node;
    }
}
ConfigurationRepositoryNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryNavigationFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ConfigurationRepositoryNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryNavigationFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
