// @c8y/ngx-components/esm2020/repository/configuration/list/configuration-repository-list.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ConfigurationListComponent } from './configuration-list.component';
import { ConfigurationDetailComponent } from './configuration-detail.component';
import { CoreModule, hookNavigator, hookRoute } from '@c8y/ngx-components';
import { ConfigurationRepositoryNavigationFactory } from './configuration-repository-navigation-factory';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import * as i0 from "@angular/core";
export class ConfigurationRepositoryListModule {
    static forRoot() {
        return {
            ngModule: ConfigurationRepositoryListModule,
            providers: [
                hookNavigator(ConfigurationRepositoryNavigationFactory),
                hookRoute({
                    path: 'configuration',
                    component: ConfigurationListComponent
                })
            ]
        };
    }
}
ConfigurationRepositoryListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ConfigurationRepositoryListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryListModule, declarations: [ConfigurationListComponent, ConfigurationDetailComponent], imports: [CommonModule, CoreModule, SharedRepositoryModule, TooltipModule] });
ConfigurationRepositoryListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryListModule, imports: [CommonModule, CoreModule, SharedRepositoryModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: ConfigurationRepositoryListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CoreModule, SharedRepositoryModule, TooltipModule],
                    declarations: [ConfigurationListComponent, ConfigurationDetailComponent],
                    entryComponents: [ConfigurationListComponent, ConfigurationDetailComponent]
                }]
        }] });
