// @c8y/ngx-components/esm2020/core/header/title/title.component.mjs
import { Component, ViewChild, Input } from '@angular/core';
import { HeaderService } from '../header.service';
import * as i0 from "@angular/core";
import * as i1 from "../header.service";
export class TitleComponent {
    constructor(headerService) {
        this.headerService = headerService;
        this.pageTitleUpdate = true;
    }
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.headerService.changeTitle(this.template, this.pageTitleUpdate));
    }
    ngOnDestroy() {
        if (this.headerService.state.title === this.template) {
            clearTimeout(this.viewInitTimeout);
            this.headerService.changeTitle();
        }
    }
}
TitleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TitleComponent, deps: [{ token: i1.HeaderService }], target: i0.ɵɵFactoryTarget.Component });
TitleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: TitleComponent, selector: "c8y-title", inputs: { pageTitleUpdate: "pageTitleUpdate" }, viewQueries: [{ propertyName: "template", first: true, predicate: ["template"], descendants: true, static: true }], ngImport: i0, template: "<ng-template #template>\n  <h1 class=\"text-truncate\">\n    <ng-content></ng-content>\n  </h1>\n</ng-template>\n" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TitleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-title', template: "<ng-template #template>\n  <h1 class=\"text-truncate\">\n    <ng-content></ng-content>\n  </h1>\n</ng-template>\n" }]
        }], ctorParameters: function () { return [{ type: i1.HeaderService }]; }, propDecorators: { pageTitleUpdate: [{
                type: Input
            }], template: [{
                type: ViewChild,
                args: ['template', { static: true }]
            }] } });
