// @c8y/ngx-components/esm2020/device-shell/command-templates/command-templates.component.mjs
import { Component, EventEmitter } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/modal";
import * as i2 from "./command-templates.directive";
/* https://stackoverflow.com/a/65290658 */
export function rootScopeFactory($injector) {
    return $injector.get('$rootScope').$new();
}
export class CommandTemplatesComponent {
    constructor(modalRef) {
        this.modalRef = modalRef;
        this.onTemplateSelected = new EventEmitter();
    }
}
CommandTemplatesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CommandTemplatesComponent, deps: [{ token: i1.BsModalRef }], target: i0.ɵɵFactoryTarget.Component });
CommandTemplatesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: CommandTemplatesComponent, selector: "c8y-command-templates", providers: [
        {
            deps: ['$injector'],
            provide: '$scope',
            useFactory: rootScopeFactory
        }
    ], ngImport: i0, template: `<c8y-command-templates-directive
    (close)="modalRef.hide(); onTemplateSelected.next($event)"
    (dismiss)="modalRef.hide()"
  ></c8y-command-templates-directive>`, isInline: true, dependencies: [{ kind: "directive", type: i2.CommandTemplates, selector: "c8y-command-templates-directive", outputs: ["close", "dismiss"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CommandTemplatesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c8y-command-templates',
                    template: `<c8y-command-templates-directive
    (close)="modalRef.hide(); onTemplateSelected.next($event)"
    (dismiss)="modalRef.hide()"
  ></c8y-command-templates-directive>`,
                    providers: [
                        {
                            deps: ['$injector'],
                            provide: '$scope',
                            useFactory: rootScopeFactory
                        }
                    ]
                }]
        }], ctorParameters: function () { return [{ type: i1.BsModalRef }]; } });
