// @c8y/ngx-components/esm2020/core/dashboard/widgets-dashboard-event.service.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * A service which can be used, to emit data on a dashboard level.
 * Each Widget Dashboard gets an own instance of this service provided
 * and you can then update all or only one.
 */
export class WidgetsDashboardEventService {
    constructor() {
        /**
         * All events sent on the current dashboard.
         */
        this.events$ = new Subject();
        this.lastValueByEvent = {};
    }
    /**
     * Returns a subscribable event observable of specific event type if eventType is passed, or observable of
     * all the events if no param is passed to method.
     * @param eventType The event you want an observable for.
     * @returns An observable of event.
     */
    getObservable(eventType) {
        return this.events$.pipe(filter(({ type }) => (eventType ? type === eventType : true)), map(({ data }) => data));
    }
    /**
     * Returns the last value of the event.
     * @param eventType The change event type you want the last value from.
     * @returns The last value used by this event, undefined if non was sent so far.
     */
    getLastValue(eventType) {
        return this.lastValueByEvent[eventType];
    }
    /**
     * Emits a new event.
     * @param event The WidgetChangeEvent to emit.
     */
    emit(event) {
        this.lastValueByEvent[event.type] = event.data;
        this.events$.next(event);
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        this.events$.complete();
    }
}
WidgetsDashboardEventService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetsDashboardEventService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
WidgetsDashboardEventService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetsDashboardEventService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: WidgetsDashboardEventService, decorators: [{
            type: Injectable
        }] });
