// @c8y/ngx-components/esm2020/operations/bulk-operation-list-item/bulk-operation-wrapper.component.mjs
import { Component, Inject, Input, Optional, ViewChild } from '@angular/core';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import { assign } from 'lodash-es';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { BULK_OPERATION_WRAPPER_MODAL_BODY_COMPONENT } from './bulk-operation-wrapper.model';
import { ModalBodyHostDirective } from './modal-body-host.directive';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components/operations/bulk-operations-service";
import * as i2 from "ngx-bootstrap/modal";
import * as i3 from "./modal-body-host.directive";
import * as i4 from "@c8y/ngx-components";
export class BulkOperationWrapperComponent {
    constructor(bulkOperationsService, modalRef, modalService, modalBodyConfig) {
        this.bulkOperationsService = bulkOperationsService;
        this.modalRef = modalRef;
        this.modalService = modalService;
        this.modalBodyConfig = modalBodyConfig;
        this.destroyed$ = new Subject();
    }
    ngOnInit() {
        if (this.modalBodyConfig) {
            const componentRef = this.modalBodyHost.viewContainerRef.createComponent(this.modalBodyConfig.component);
            assign(componentRef.instance, this.modalBodyConfig.config);
            componentRef.instance.bulkOperation = this.bulkOperation;
            componentRef.instance.showFailedOperation
                .pipe(takeUntil(this.destroyed$))
                .subscribe(event => this.openDetails(event));
        }
    }
    ngOnDestroy() {
        this.destroyed$.next();
        this.destroyed$.complete();
    }
    // openDetails function needs to be defined in the both places (component and the service) to avoid circular dependency.
    async openDetails(bulkOperationId) {
        const bulkOperation = (await this.bulkOperationsService.getBulkOperationById(bulkOperationId)).data;
        const initialState = { bulkOperation };
        this.modalService.show(BulkOperationWrapperComponent, {
            initialState,
            class: 'modal-lg',
            ariaDescribedby: 'modal-body',
            ariaLabelledBy: 'modal-title'
        });
    }
}
BulkOperationWrapperComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationWrapperComponent, deps: [{ token: i1.BulkOperationsService }, { token: i2.BsModalRef }, { token: i2.BsModalService }, { token: BULK_OPERATION_WRAPPER_MODAL_BODY_COMPONENT, optional: true }], target: i0.ɵɵFactoryTarget.Component });
BulkOperationWrapperComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: BulkOperationWrapperComponent, selector: "c8y-bulk-operations-wrapper", inputs: { bulkOperation: "bulkOperation" }, viewQueries: [{ propertyName: "modalBodyHost", first: true, predicate: ModalBodyHostDirective, descendants: true, static: true }], ngImport: i0, template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header\">\n    <h3 id=\"modal-title\">{{ 'Bulk operation details' | translate }}</h3>\n  </div>\n\n  <div class=\"modal-inner-scroll\" id=\"modal-body\">\n    <ng-container modalBodyHost></ng-container>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      title=\"{{ 'Close' | translate }}\"\n      type=\"button\"\n      class=\"btn btn-default\"\n      (click)=\"modalRef.hide()\"\n    >\n      {{ 'Close' | translate }}\n    </button>\n  </div>\n</div>\n", dependencies: [{ kind: "directive", type: i3.ModalBodyHostDirective, selector: "[modalBodyHost]" }, { kind: "pipe", type: i4.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationWrapperComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-bulk-operations-wrapper', template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header\">\n    <h3 id=\"modal-title\">{{ 'Bulk operation details' | translate }}</h3>\n  </div>\n\n  <div class=\"modal-inner-scroll\" id=\"modal-body\">\n    <ng-container modalBodyHost></ng-container>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      title=\"{{ 'Close' | translate }}\"\n      type=\"button\"\n      class=\"btn btn-default\"\n      (click)=\"modalRef.hide()\"\n    >\n      {{ 'Close' | translate }}\n    </button>\n  </div>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.BulkOperationsService }, { type: i2.BsModalRef }, { type: i2.BsModalService }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [BULK_OPERATION_WRAPPER_MODAL_BODY_COMPONENT]
                }] }]; }, propDecorators: { bulkOperation: [{
                type: Input
            }], modalBodyHost: [{
                type: ViewChild,
                args: [ModalBodyHostDirective, { static: true }]
            }] } });
