// @c8y/ngx-components/esm2020/operations/bulk-operation-list-item/bulk-operation-list-item.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { OperationSummaryModule } from '@c8y/ngx-components/operations/operation-summary';
import { OperationsListItemDetailsModule } from '@c8y/ngx-components/operations/operations-list-item-details';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { BulkOperationListItemComponent } from './bulk-operation-list-item.component';
import { BulkOperationListItemService } from './bulk-operation-list-item.service';
import { BulkOperationWrapperComponent } from './bulk-operation-wrapper.component';
import { BULK_OPERATION_WRAPPER_MODAL_BODY_COMPONENT } from './bulk-operation-wrapper.model';
import { ModalBodyHostDirective } from './modal-body-host.directive';
import { BulkOperationsRescheduleModalComponent } from './modals/bulk-operations-reschedule-modal.component';
import * as i0 from "@angular/core";
/**
 * This module allows for displaying the list item of a bulk operation with its details.
 */
export class BulkOperationListItemModule {
}
BulkOperationListItemModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationListItemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BulkOperationListItemModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationListItemModule, declarations: [BulkOperationListItemComponent,
        BulkOperationsRescheduleModalComponent,
        BulkOperationWrapperComponent,
        ModalBodyHostDirective], imports: [CoreModule,
        PopoverModule,
        TooltipModule,
        OperationSummaryModule,
        BulkOperationSchedulerModule,
        OperationsListItemDetailsModule], exports: [BulkOperationListItemComponent,
        BulkOperationsRescheduleModalComponent,
        BulkOperationWrapperComponent] });
BulkOperationListItemModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationListItemModule, providers: [
        BulkOperationListItemService,
        {
            provide: BULK_OPERATION_WRAPPER_MODAL_BODY_COMPONENT,
            useValue: {
                component: BulkOperationListItemComponent,
                config: { detailsCollapsed: false, readOnly: true }
            }
        }
    ], imports: [CoreModule,
        PopoverModule,
        TooltipModule,
        OperationSummaryModule,
        BulkOperationSchedulerModule,
        OperationsListItemDetailsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationListItemModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        PopoverModule,
                        TooltipModule,
                        OperationSummaryModule,
                        BulkOperationSchedulerModule,
                        OperationsListItemDetailsModule
                    ],
                    providers: [
                        BulkOperationListItemService,
                        {
                            provide: BULK_OPERATION_WRAPPER_MODAL_BODY_COMPONENT,
                            useValue: {
                                component: BulkOperationListItemComponent,
                                config: { detailsCollapsed: false, readOnly: true }
                            }
                        }
                    ],
                    declarations: [
                        BulkOperationListItemComponent,
                        BulkOperationsRescheduleModalComponent,
                        BulkOperationWrapperComponent,
                        ModalBodyHostDirective
                    ],
                    entryComponents: [
                        BulkOperationListItemComponent,
                        BulkOperationsRescheduleModalComponent,
                        BulkOperationWrapperComponent
                    ],
                    exports: [
                        BulkOperationListItemComponent,
                        BulkOperationsRescheduleModalComponent,
                        BulkOperationWrapperComponent
                    ]
                }]
        }] });
