// @c8y/ngx-components/esm2020/core/drop-area/drop-area.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { DropAreaComponent } from './drop-area.component';
import * as i0 from "@angular/core";
/**
 * The angular module definition for a drop-area.
 * @exports DropAreaComponent
 */
export class DropAreaModule {
}
DropAreaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DropAreaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropAreaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DropAreaModule, declarations: [DropAreaComponent], imports: [CommonModule], exports: [DropAreaComponent] });
DropAreaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DropAreaModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DropAreaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [DropAreaComponent],
                    declarations: [DropAreaComponent]
                }]
        }] });
