// @c8y/ngx-components/esm2020/device-profile/device-profile-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
export class DeviceProfileNavigationFactory {
    async get() {
        if (!this.nodeItem) {
            this.nodeItem = new NavigatorNode({
                label: gettext('Device profiles'),
                path: '/device-profiles',
                icon: 'c8y-device-profile',
                parent: gettext('Management')
            });
        }
        return this.nodeItem;
    }
}
DeviceProfileNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileNavigationFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
DeviceProfileNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileNavigationFactory, decorators: [{
            type: Injectable
        }] });
