// @c8y/ngx-components/esm2020/device-profile/device-profile.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, hookNavigator, hookRoute, gettext, ViewContext } from '@c8y/ngx-components';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { ReactiveFormsModule } from '@angular/forms';
import { DeviceProfileComponent } from './device-profile.component';
import { DeviceProfileListComponent } from './device-profile-list.component';
import { DeviceProfileNavigationFactory } from './device-profile-navigation.factory';
import { DeviceProfileService } from './device-profile.service';
import { AddDeviceProfileComponent } from './add-device-profile.component';
import { DeviceProfileGuard } from './device-profile.guard';
import { SharedRepositoryModule } from '@c8y/ngx-components/repository/shared';
import { SelectConfigurationModalComponent } from './select-configuration-modal.component';
import { DeviceTabProfileComponent } from './device-tab-profile/device-tab-profile.component';
import { DeviceTabProfileDetailComponent } from './device-tab-profile/device-tab-profile-detail.component';
import { DeviceProfileItemListComponent } from './device-tab-profile/device-profile-item-list.component';
import { OperationDetailsModule } from '@c8y/ngx-components/operations/operation-details';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "ngx-bootstrap/dropdown";
const deviceProfilesRoutes = [
    {
        path: 'device-profiles/:id',
        component: DeviceProfileComponent
    },
    {
        path: 'device-profiles',
        component: DeviceProfileListComponent
    }
];
const deviceTabProfileRoutes = [
    {
        context: ViewContext.Device,
        path: 'device-profile',
        component: DeviceTabProfileComponent,
        label: gettext('Device profile'),
        icon: 'c8y-device-profile',
        canActivate: [DeviceProfileGuard]
    }
];
export class DeviceProfileModule {
}
DeviceProfileModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DeviceProfileModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileModule, declarations: [DeviceProfileComponent,
        DeviceProfileListComponent,
        AddDeviceProfileComponent,
        SelectConfigurationModalComponent,
        DeviceTabProfileComponent,
        DeviceTabProfileDetailComponent,
        DeviceProfileItemListComponent], imports: [CoreModule,
        CommonModule,
        SharedRepositoryModule, i1.RouterModule, i2.BsDropdownModule, TooltipModule,
        ReactiveFormsModule,
        ButtonsModule,
        PopoverModule,
        OperationDetailsModule] });
DeviceProfileModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileModule, providers: [
        hookNavigator(DeviceProfileNavigationFactory),
        hookRoute(deviceTabProfileRoutes),
        DeviceProfileService,
        DeviceProfileGuard
    ], imports: [CoreModule,
        CommonModule,
        SharedRepositoryModule,
        RouterModule.forChild(deviceProfilesRoutes),
        BsDropdownModule.forRoot(),
        TooltipModule,
        ReactiveFormsModule,
        ButtonsModule,
        PopoverModule,
        OperationDetailsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DeviceProfileModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DeviceProfileComponent,
                        DeviceProfileListComponent,
                        AddDeviceProfileComponent,
                        SelectConfigurationModalComponent,
                        DeviceTabProfileComponent,
                        DeviceTabProfileDetailComponent,
                        DeviceProfileItemListComponent
                    ],
                    exports: [],
                    imports: [
                        CoreModule,
                        CommonModule,
                        SharedRepositoryModule,
                        RouterModule.forChild(deviceProfilesRoutes),
                        BsDropdownModule.forRoot(),
                        TooltipModule,
                        ReactiveFormsModule,
                        ButtonsModule,
                        PopoverModule,
                        OperationDetailsModule
                    ],
                    entryComponents: [
                        AddDeviceProfileComponent,
                        SelectConfigurationModalComponent,
                        DeviceTabProfileComponent
                    ],
                    providers: [
                        hookNavigator(DeviceProfileNavigationFactory),
                        hookRoute(deviceTabProfileRoutes),
                        DeviceProfileService,
                        DeviceProfileGuard
                    ]
                }]
        }] });
