// @c8y/ngx-components/esm2020/ecosystem/license-confirm/license-confirm.module.mjs
import { NgModule } from '@angular/core';
import { LicenseConfirmModalComponent } from './license-confirm-modal.component';
import { LicenseViewComponent } from './license-view.component';
import { CoreModule, FormsModule, hookWizard } from '@c8y/ngx-components';
import { EcosystemWizards, SharedEcosystemModule } from '@c8y/ngx-components/ecosystem/shared';
import * as i0 from "@angular/core";
export class LicenseConfirmModule {
}
LicenseConfirmModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LicenseConfirmModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LicenseConfirmModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: LicenseConfirmModule, declarations: [LicenseConfirmModalComponent, LicenseViewComponent], imports: [CoreModule, FormsModule, SharedEcosystemModule], exports: [LicenseConfirmModalComponent, LicenseViewComponent] });
LicenseConfirmModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LicenseConfirmModule, providers: [
        hookWizard({
            wizardId: EcosystemWizards.LICENSE_CONFIRM,
            component: LicenseConfirmModalComponent,
            name: undefined
        })
    ], imports: [CoreModule, FormsModule, SharedEcosystemModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LicenseConfirmModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LicenseConfirmModalComponent, LicenseViewComponent],
                    imports: [CoreModule, FormsModule, SharedEcosystemModule],
                    exports: [LicenseConfirmModalComponent, LicenseViewComponent],
                    providers: [
                        hookWizard({
                            wizardId: EcosystemWizards.LICENSE_CONFIRM,
                            component: LicenseConfirmModalComponent,
                            name: undefined
                        })
                    ]
                }]
        }] });
