// @c8y/ngx-components/esm2020/operations/operations-list-item-details/operations-list-item-details.module.mjs
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { CoreModule } from '@c8y/ngx-components';
import { BulkSingleOperationsListModule } from '@c8y/ngx-components/operations/bulk-single-operations-list';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { OperationDetailsTabsComponent } from './operation-details-tabs.component';
import { OperationDetailsComponent } from './operation-details.component';
import * as i0 from "@angular/core";
/**
 * This module allows for displaying the list of single operations and their details.
 */
export class OperationsListItemDetailsModule {
}
OperationsListItemDetailsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsListItemDetailsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OperationsListItemDetailsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: OperationsListItemDetailsModule, declarations: [OperationDetailsComponent, OperationDetailsTabsComponent], imports: [CoreModule,
        TooltipModule,
        PopoverModule,
        StatusFilterModule,
        BulkSingleOperationsListModule], exports: [OperationDetailsComponent, OperationDetailsTabsComponent] });
OperationsListItemDetailsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsListItemDetailsModule, imports: [CoreModule,
        TooltipModule,
        PopoverModule,
        StatusFilterModule,
        BulkSingleOperationsListModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: OperationsListItemDetailsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        TooltipModule,
                        PopoverModule,
                        StatusFilterModule,
                        BulkSingleOperationsListModule
                    ],
                    declarations: [OperationDetailsComponent, OperationDetailsTabsComponent],
                    exports: [OperationDetailsComponent, OperationDetailsTabsComponent]
                }]
        }] });
