// @c8y/ngx-components/esm2020/core/dashboard/wiget-time-context/interval-picker/interval-picker.component.mjs
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { INTERVALS } from '../widget-time-context.model';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/forms";
import * as i3 from "../../../i18n/c8y-translate.pipe";
export class IntervalPickerComponent {
    constructor() {
        this.INTERVALS = INTERVALS;
        this.touched = false;
        this.disabled = false;
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this.onChange = _ => { };
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this.onTouched = () => { };
    }
    writeValue(value) {
        this.value = value;
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(onTouched) {
        this.onTouched = onTouched;
    }
    markAsTouched() {
        if (!this.touched) {
            this.onTouched();
            this.touched = true;
        }
    }
    setDisabledState(disabled) {
        this.disabled = disabled;
    }
}
IntervalPickerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: IntervalPickerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
IntervalPickerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: IntervalPickerComponent, selector: "c8y-interval-picker", providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => IntervalPickerComponent),
            multi: true
        }
    ], ngImport: i0, template: "<select\n  class=\"form-control\"\n  [attr.aria-label]=\"'Time range' | translate\"\n  [ngModel]=\"value\"\n  (ngModelChange)=\"onChange($event)\"\n>\n  <option\n    *ngFor=\"let interval of INTERVALS\"\n    [value]=\"interval.id\"\n    [disabled]=\"interval.id === 'custom'\"\n  >\n    {{ interval.title | translate }}\n  </option>\n</select>\n<span></span>\n", dependencies: [{ kind: "directive", type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i2.NgSelectOption, selector: "option", inputs: ["ngValue", "value"] }, { kind: "directive", type: i2.ɵNgSelectMultipleOption, selector: "option", inputs: ["ngValue", "value"] }, { kind: "directive", type: i2.SelectControlValueAccessor, selector: "select:not([multiple])[formControlName],select:not([multiple])[formControl],select:not([multiple])[ngModel]", inputs: ["compareWith"] }, { kind: "directive", type: i2.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i2.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { kind: "pipe", type: i3.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: IntervalPickerComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-interval-picker', providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => IntervalPickerComponent),
                            multi: true
                        }
                    ], template: "<select\n  class=\"form-control\"\n  [attr.aria-label]=\"'Time range' | translate\"\n  [ngModel]=\"value\"\n  (ngModelChange)=\"onChange($event)\"\n>\n  <option\n    *ngFor=\"let interval of INTERVALS\"\n    [value]=\"interval.id\"\n    [disabled]=\"interval.id === 'custom'\"\n  >\n    {{ interval.title | translate }}\n  </option>\n</select>\n<span></span>\n" }]
        }] });
