// @c8y/ngx-components/esm2020/trusted-certificates/trusted-certificates-navigation.factory.mjs
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as i0 from "@angular/core";
export class TrustedCertificatesNavigationFactory {
    constructor() {
        this.navs = [];
    }
    async get() {
        if (this.navs.length === 0) {
            this.navs.push(new NavigatorNode({
                label: gettext('Trusted certificates'),
                icon: 'certificate',
                path: '/trusted-certificates',
                parent: gettext('Management'),
                priority: 100
            }));
        }
        return this.navs;
    }
}
TrustedCertificatesNavigationFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TrustedCertificatesNavigationFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
TrustedCertificatesNavigationFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TrustedCertificatesNavigationFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: TrustedCertificatesNavigationFactory, decorators: [{
            type: Injectable
        }] });
