// @c8y/ngx-components/esm2020/trusted-certificates/trusted-certificate.model.mjs
export const PRODUCT_EXPERIENCE = {
    EVENT: 'trustedCertificate',
    CERTIFICATES: {
        COMPONENTS: {
            TRUSTED_CERTIFICATE: 'trusted-certificate'
        },
        ACTIONS: {
            ADD: 'add',
            UPDATE: 'update',
            DELETE: 'delete'
        },
        RESULTS: {
            ADD_SUCCESS: 'certificateAdded',
            UPDATE_SUCCESS: 'certificateUpdated',
            DELETE_SUCCESS: 'certificateDeleted'
        }
    },
    VERIFICATION_CODE: {
        COMPONENTS: {
            REGENERATE_CODE: 'regenerate-unsigned-verification-code',
            VERIFY_CODE: 'verify-signed-verification-code',
            DOWNLOAD_CODE: 'download-unsigned-verification-code'
        },
        ACTIONS: {
            REGENERATE: 'regenerate',
            VERIFY: 'verify',
            DOWNLOAD: 'download'
        },
        RESULTS: {
            REGENERATE_SUCCESS: 'unsignedVerificationCodeRegenerated',
            VERIFY_SUCCESS: 'signedCodeVerified',
            VERIFY_FAILED: 'signedCodeVerificationFailed'
        }
    }
};
