// @c8y/ngx-components/esm2020/sensor-phone/sensor-phone-modal/sensor-phone.service.mjs
import { Injectable } from '@angular/core';
import { DeviceRegistrationService, InventoryService } from '@c8y/client';
import { AppStateService, gettext } from '@c8y/ngx-components';
import { head } from 'lodash-es';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@c8y/client";
export class SensorPhoneService {
    constructor(ui, inventoryService, deviceRegistrationService) {
        this.ui = ui;
        this.inventoryService = inventoryService;
        this.deviceRegistrationService = deviceRegistrationService;
        this.HTTP_STATUS_CODE = {
            OK: 200,
            CREATED: 201
        };
        this.PHONE_GROUP_FRAGMENT = 'c8y_IsSensorPhoneGroup';
        this.PHONE_GROUP_NAME = gettext('Phones');
        this.ui.currentTenant.subscribe(data => {
            if (data) {
                this.domainName = data.domainName;
                this.tenantId = data.name;
            }
        });
    }
    getHost() {
        return this.domainName.substring(this.domainName.indexOf('.') + 1);
    }
    getTenantName() {
        return this.domainName.split('.')[0];
    }
    getTenantId() {
        return this.tenantId;
    }
    async checkPhoneGroupExists() {
        const { data, res } = await this.inventoryService.list({
            fragmentType: this.PHONE_GROUP_FRAGMENT
        });
        if (data && res && res.status === this.HTTP_STATUS_CODE.OK) {
            return head(data);
        }
        return null;
    }
    async addPhoneGroup() {
        const group = await this.checkPhoneGroupExists();
        if (group) {
            return group;
        }
        return await this.createPhoneGroup();
    }
    async createPhoneGroup() {
        const newPhoneGroup = {
            c8y_Global: {},
            c8y_IsDeviceGroup: {},
            type: 'c8y_DeviceGroup',
            name: this.PHONE_GROUP_NAME,
            [this.PHONE_GROUP_FRAGMENT]: {}
        };
        const { data, res } = await this.inventoryService.create(newPhoneGroup);
        if (data && res && res.status === this.HTTP_STATUS_CODE.CREATED) {
            return data;
        }
        return null;
    }
    generateRegistrationData(deviceId) {
        const res = 'c8y://' + this.getTenantName() + '.' + this.getHost() + '/?deviceId=' + deviceId;
        return res;
    }
    async createPhoneRegistrationRequest(deviceId) {
        const group = await this.addPhoneGroup();
        const device = { id: deviceId, groupId: group.id };
        const { res } = await this.deviceRegistrationService.create(device);
        if (res.status !== this.HTTP_STATUS_CODE.CREATED) {
            throw res;
        }
    }
    async acceptPhoneRegistrationRequest(deviceId) {
        const { res } = await this.deviceRegistrationService.accept(deviceId);
        if (res.status !== this.HTTP_STATUS_CODE.OK) {
            throw res;
        }
    }
}
SensorPhoneService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SensorPhoneService, deps: [{ token: i1.AppStateService }, { token: i2.InventoryService }, { token: i2.DeviceRegistrationService }], target: i0.ɵɵFactoryTarget.Injectable });
SensorPhoneService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SensorPhoneService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: SensorPhoneService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.AppStateService }, { type: i2.InventoryService }, { type: i2.DeviceRegistrationService }]; } });
