// @c8y/ngx-components/esm2020/operations/bulk-operations-list/bulk-operations-list.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { BulkOperationListItemModule } from '@c8y/ngx-components/operations/bulk-operation-list-item';
import { BulkOperationsListComponent } from './bulk-operations-list.component';
import { BulkOperationModalsService } from './modals/bulk-operation-modals.service';
import { BulkOperationsModalComponent } from './modals/bulk-operations-modal.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export const bulkOperationListRoutes = [
    {
        path: 'devicecontrol/bulk',
        component: BulkOperationsListComponent
    }
];
/**
 * This module allows for displaying the list of bulk operations and their details.
 */
export class BulkOperationsListModule {
}
BulkOperationsListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BulkOperationsListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsListModule, declarations: [BulkOperationsListComponent, BulkOperationsModalComponent], imports: [CoreModule,
        PopoverModule,
        TooltipModule,
        StatusFilterModule,
        BulkOperationSchedulerModule,
        BulkOperationListItemModule, i1.RouterModule], exports: [BulkOperationsListComponent, BulkOperationsModalComponent] });
BulkOperationsListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsListModule, providers: [BulkOperationModalsService], imports: [CoreModule,
        PopoverModule,
        TooltipModule,
        StatusFilterModule,
        BulkOperationSchedulerModule,
        BulkOperationListItemModule,
        RouterModule.forChild(bulkOperationListRoutes)] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationsListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        PopoverModule,
                        TooltipModule,
                        StatusFilterModule,
                        BulkOperationSchedulerModule,
                        BulkOperationListItemModule,
                        RouterModule.forChild(bulkOperationListRoutes)
                    ],
                    providers: [BulkOperationModalsService],
                    declarations: [BulkOperationsListComponent, BulkOperationsModalComponent],
                    entryComponents: [BulkOperationsListComponent, BulkOperationsModalComponent],
                    exports: [BulkOperationsListComponent, BulkOperationsModalComponent]
                }]
        }] });
