// @c8y/ngx-components/esm2020/operations/bulk-operation-stepper/bulk-operation-stepper.module.mjs
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { CreateBulkOperationDetailsModule } from '@c8y/ngx-components/operations/create-bulk-operation-details';
import { DeviceSelectorModule } from '@c8y/ngx-components/operations/device-selector';
import { OperationSummaryModule } from '@c8y/ngx-components/operations/operation-summary';
import { BulkOperationStepper } from './bulk-operation-stepper.component';
import { CustomStep } from './custom-step.directive';
import * as i0 from "@angular/core";
/**
 * This module provides base stepper class and stepper wrapper component.
 */
export class BulkOperationStepperModule {
}
BulkOperationStepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BulkOperationStepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationStepperModule, declarations: [BulkOperationStepper, CustomStep], imports: [CoreModule,
        DeviceSelectorModule,
        CreateBulkOperationDetailsModule,
        OperationSummaryModule], exports: [BulkOperationStepper, CustomStep] });
BulkOperationStepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationStepperModule, imports: [CoreModule,
        DeviceSelectorModule,
        CreateBulkOperationDetailsModule,
        OperationSummaryModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BulkOperationStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CoreModule,
                        DeviceSelectorModule,
                        CreateBulkOperationDetailsModule,
                        OperationSummaryModule
                    ],
                    declarations: [BulkOperationStepper, CustomStep],
                    exports: [BulkOperationStepper, CustomStep]
                }]
        }] });
