// @c8y/ngx-components/esm2020/operations/bulk-operation-stepper/custom-step.directive.mjs
import { Directive, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class CustomStep {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.buttonsDisabled = false;
        this.onNext = ({ stepper, step }) => {
            // steps without own `onNext` handler, e.g. preview steps need to mark themselves as `completed`,
            // otherwise stepper will not allow to move forth from them as soon as the user navigates back
            // and the `c8y-stepper-buttons` component marks the step as incomplete.
            step.completed = true;
            stepper.next();
        };
    }
}
CustomStep.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CustomStep, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
CustomStep.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.7", type: CustomStep, selector: "[customStep]", inputs: { label: ["customStep", "label"], completed: ["customStepCompleted", "completed"], buttonsDisabled: ["customStepButtonsDisabled", "buttonsDisabled"], onNext: ["customStepOnNext", "onNext"] }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: CustomStep, decorators: [{
            type: Directive,
            args: [{ selector: '[customStep]' }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; }, propDecorators: { label: [{
                type: Input,
                args: ['customStep']
            }], completed: [{
                type: Input,
                args: ['customStepCompleted']
            }], buttonsDisabled: [{
                type: Input,
                args: ['customStepButtonsDisabled']
            }], onNext: [{
                type: Input,
                args: ['customStepOnNext']
            }] } });
