// @c8y/ngx-components/fesm2020/c8y-ngx-components-datapoint-library.mjs
import * as i0 from '@angular/core';
import { NgModule } from '@angular/core';
import { hookRoute, ViewContext, gettext, hookNavigator, NavigatorNode } from '@c8y/ngx-components';
import { pathToDatapointLibrary, DATAPOINT_LIBRARY_CONFIG } from '@c8y/ngx-components/datapoint-library/model';

class DatapointLibraryModule {
    static forRoot(config = {}) {
        return {
            ngModule: DatapointLibraryModule,
            providers: [
                hookRoute([
                    {
                        path: pathToDatapointLibrary,
                        loadComponent: () => import('@c8y/ngx-components/datapoint-library/list').then(m => m.DatapointLibraryListComponent)
                    },
                    {
                        context: ViewContext.DatapointLibrary,
                        path: 'details',
                        loadComponent: () => import('@c8y/ngx-components/datapoint-library/details').then(m => m.DatapointLibraryDetailsComponent),
                        label: gettext('Details'),
                        icon: 'palette',
                        priority: 500
                    }
                ]),
                hookNavigator([
                    new NavigatorNode({
                        parent: gettext('Configuration'),
                        label: gettext('Data point library'),
                        featureId: 'dataPointLibrary',
                        path: pathToDatapointLibrary,
                        icon: 'c8y-data-points',
                        routerLinkExact: false
                    })
                ]),
                {
                    provide: DATAPOINT_LIBRARY_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
DatapointLibraryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointLibraryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DatapointLibraryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: DatapointLibraryModule });
DatapointLibraryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointLibraryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: DatapointLibraryModule, decorators: [{
            type: NgModule
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { DatapointLibraryModule };
