// @c8y/ngx-components/fesm2020/c8y-ngx-components-widgets-implementations-help-and-service-widget.mjs
import * as i0 from '@angular/core';
import { Component, NgModule } from '@angular/core';
import * as i1 from '@c8y/ngx-components';
import { CoreModule } from '@c8y/ngx-components';
import * as i2 from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';

const PRODUCT_EXPERIENCE = {
    HELP_AND_SERVICE_LINKS: {
        EVENTS: { HELP_AND_SERVICE_LINKS: 'helpAndServiceLinks' },
        COMPONENTS: { HELP_AND_SERVICE_COMPONENT: 'right-drawer' },
        ACTIONS: {},
        RESULTS: {}
    }
};

class HelpAndServiceViewComponent {
    constructor(docs) {
        this.docs = docs;
        this.PRODUCT_EXPERIENCE = PRODUCT_EXPERIENCE;
    }
}
HelpAndServiceViewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndServiceViewComponent, deps: [{ token: i1.DocsService }], target: i0.ɵɵFactoryTarget.Component });
HelpAndServiceViewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: HelpAndServiceViewComponent, selector: "c8y-help-and-service-view", ngImport: i0, template: "<div class=\"card-group-block\">\n  <a\n    *ngFor=\"let link of docs.items$ | async\"\n    title=\"{{ link.label | translate }}\"\n    role=\"button\"\n    class=\"card card--btn\"\n    target=\"_blank\"\n    rel=\"noopener noreferrer\"\n    [href]=\"link.url\"\n    c8yProductExperience\n    [actionName]=\"PRODUCT_EXPERIENCE.HELP_AND_SERVICE_LINKS.EVENTS.HELP_AND_SERVICE_LINKS\"\n    [actionData]=\"{\n      component: PRODUCT_EXPERIENCE.HELP_AND_SERVICE_LINKS.COMPONENTS.HELP_AND_SERVICE_COMPONENT,\n      action: link.label\n    }\"\n  >\n    <div class=\"card-block text-center\">\n      <div class=\"icon-32\">\n        <i [c8yIcon]=\"link.icon\" class=\"c8y-icon-duocolor\"></i>\n      </div>\n      <small class=\"text-muted\">{{ link.label | translate }}</small>\n    </div>\n  </a>\n</div>\n", dependencies: [{ kind: "directive", type: i1.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i1.ProductExperienceDirective, selector: "[c8yProductExperience]", inputs: ["actionName", "actionData", "inherit", "suppressDataOverriding"] }, { kind: "pipe", type: i1.C8yTranslatePipe, name: "translate" }, { kind: "pipe", type: i2.AsyncPipe, name: "async" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndServiceViewComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-help-and-service-view', template: "<div class=\"card-group-block\">\n  <a\n    *ngFor=\"let link of docs.items$ | async\"\n    title=\"{{ link.label | translate }}\"\n    role=\"button\"\n    class=\"card card--btn\"\n    target=\"_blank\"\n    rel=\"noopener noreferrer\"\n    [href]=\"link.url\"\n    c8yProductExperience\n    [actionName]=\"PRODUCT_EXPERIENCE.HELP_AND_SERVICE_LINKS.EVENTS.HELP_AND_SERVICE_LINKS\"\n    [actionData]=\"{\n      component: PRODUCT_EXPERIENCE.HELP_AND_SERVICE_LINKS.COMPONENTS.HELP_AND_SERVICE_COMPONENT,\n      action: link.label\n    }\"\n  >\n    <div class=\"card-block text-center\">\n      <div class=\"icon-32\">\n        <i [c8yIcon]=\"link.icon\" class=\"c8y-icon-duocolor\"></i>\n      </div>\n      <small class=\"text-muted\">{{ link.label | translate }}</small>\n    </div>\n  </a>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.DocsService }]; } });

class HelpAndServiceModule {
}
HelpAndServiceModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndServiceModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HelpAndServiceModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: HelpAndServiceModule, declarations: [HelpAndServiceViewComponent], imports: [CoreModule, ReactiveFormsModule] });
HelpAndServiceModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndServiceModule, imports: [CoreModule, ReactiveFormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: HelpAndServiceModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CoreModule, ReactiveFormsModule],
                    declarations: [HelpAndServiceViewComponent]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { HelpAndServiceModule, HelpAndServiceViewComponent };
