// @c8y/ngx-components/fesm2020/c8y-ngx-components-protocol-lwm2m.mjs
import * as i0 from '@angular/core';
import { NgModule } from '@angular/core';
import { hookDeviceGridAction } from '@c8y/ngx-components/device-grid';
import { hookNavigator, hookRoute, ViewContext, gettext } from '@c8y/ngx-components';
import { Lwm2mMicroserviceSubscribedGuard, Lwm2mBootstrapParametersGuard } from '@c8y/ngx-components/protocol-lwm2m/guards';
import { LWM2MPostOpertaionsNavigationFactory, Lwm2mDeviceGridFactory } from '@c8y/ngx-components/protocol-lwm2m/factories';

class Lwm2mModule {
    static forRoot() {
        return {
            ngModule: Lwm2mModule,
            providers: [
                Lwm2mMicroserviceSubscribedGuard,
                Lwm2mBootstrapParametersGuard,
                hookNavigator(LWM2MPostOpertaionsNavigationFactory),
                hookDeviceGridAction(Lwm2mDeviceGridFactory),
                hookRoute({
                    path: 'post-operations',
                    loadComponent: () => import('@c8y/ngx-components/protocol-lwm2m/components/post-registration-operations').then(m => m.LWM2MPostOperationsComponent),
                    icon: 'terminal',
                    priority: 2000,
                    canActivate: [Lwm2mMicroserviceSubscribedGuard]
                }),
                hookRoute({
                    path: 'lwm2m-bootstrap-parameters',
                    context: ViewContext.Device,
                    loadComponent: () => import('@c8y/ngx-components/protocol-lwm2m/components/bootstrap-parameters').then(m => m.Lwm2mBootstrapParametersComponent),
                    label: gettext('LWM2M bootstrap parameters'),
                    icon: 'c8y-device-protocols',
                    canActivate: [Lwm2mBootstrapParametersGuard]
                })
            ]
        };
    }
}
Lwm2mModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
Lwm2mModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mModule });
Lwm2mModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: Lwm2mModule, decorators: [{
            type: NgModule
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { Lwm2mModule };
