// @c8y/ngx-components/fesm2020/c8y-ngx-components-widgets-definitions-kpi.mjs
import { gettext, DynamicDatapointsResolver } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';

// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
const kpiWidgetDefinition = {
    id: defaultWidgetIds.KPI,
    label: gettext('KPI'),
    description: gettext('Displays the last measurement value of a device.'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/kpi').then(m => m.KpiWidgetViewComponent),
    loadConfigComponent: () => import('@c8y/ngx-components/widgets/implementations/kpi').then(m => m.KpiWidgetConfigComponent),
    previewImage: 'c8y-style-assets/kpi-widget.png',
    resolve: {
        datapoints: DynamicDatapointsResolver
    },
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 2,
                _height: 2
            },
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};

/**
 * Generated bundle index. Do not edit.
 */

export { kpiWidgetDefinition };
