// @c8y/ngx-components/fesm2020/c8y-ngx-components-ecosystem-plugin-setup-stepper.mjs
import * as i0 from '@angular/core';
import { Component, NgModule } from '@angular/core';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i1 from '@c8y/ngx-components';
import { hookStepper, Steppers, gettext } from '@c8y/ngx-components';
import { map, filter } from 'rxjs/operators';
import * as i3 from '@c8y/ngx-components/ecosystem/application-plugins';
import { ApplicationPluginsModule } from '@c8y/ngx-components/ecosystem/application-plugins';
import { LicenseConfirmModule } from '@c8y/ngx-components/ecosystem/license-confirm';

class PluginSetupStepperComponent {
    constructor(appState, stepper, setup) {
        this.appState = appState;
        this.stepper = stepper;
        this.setup = setup;
        this.appId$ = this.appState.currentApplication.pipe(map(app => app?.id), filter(appId => !!appId));
    }
    next() {
        this.setup.stepCompleted(this.stepper.selectedIndex);
        this.stepper.next();
    }
    back() {
        this.stepper.previous();
    }
}
PluginSetupStepperComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperComponent, deps: [{ token: i1.AppStateService }, { token: i1.C8yStepper }, { token: i1.SetupComponent }], target: i0.ɵɵFactoryTarget.Component });
PluginSetupStepperComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.7", type: PluginSetupStepperComponent, selector: "c8y-plugin-setup-stepper", ngImport: i0, template: "<c8y-app-plugins *ngIf=\"appId$ | async as appId\" [appId]=\"appId\"></c8y-app-plugins>\n<div class=\"card-footer separator d-flex j-c-center\">\n  <button\n    class=\"btn btn-default\"\n    type=\"button\"\n    *ngIf=\"stepper.selectedIndex !== 0\"\n    (click)=\"back()\"\n    translate\n  >\n    Previous\n  </button>\n  <button class=\"btn btn-primary\" type=\"submit\" (click)=\"next()\" translate>\n    Continue\n  </button>\n</div>\n", dependencies: [{ kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i3.ApplicationPluginsComponent, selector: "c8y-app-plugins", inputs: ["appId"] }, { kind: "pipe", type: i2.AsyncPipe, name: "async" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-plugin-setup-stepper', template: "<c8y-app-plugins *ngIf=\"appId$ | async as appId\" [appId]=\"appId\"></c8y-app-plugins>\n<div class=\"card-footer separator d-flex j-c-center\">\n  <button\n    class=\"btn btn-default\"\n    type=\"button\"\n    *ngIf=\"stepper.selectedIndex !== 0\"\n    (click)=\"back()\"\n    translate\n  >\n    Previous\n  </button>\n  <button class=\"btn btn-primary\" type=\"submit\" (click)=\"next()\" translate>\n    Continue\n  </button>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.AppStateService }, { type: i1.C8yStepper }, { type: i1.SetupComponent }]; } });

class PluginSetupStepperModule {
}
PluginSetupStepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PluginSetupStepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, declarations: [PluginSetupStepperComponent], imports: [CommonModule, ApplicationPluginsModule, LicenseConfirmModule] });
PluginSetupStepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, providers: [
        hookStepper([
            {
                stepperId: Steppers.SETUP,
                component: PluginSetupStepperComponent,
                label: gettext('Plugins'),
                setupId: 'plugins',
                priority: 30
            }
        ])
    ], imports: [CommonModule, ApplicationPluginsModule, LicenseConfirmModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: PluginSetupStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ApplicationPluginsModule, LicenseConfirmModule],
                    declarations: [PluginSetupStepperComponent],
                    providers: [
                        hookStepper([
                            {
                                stepperId: Steppers.SETUP,
                                component: PluginSetupStepperComponent,
                                label: gettext('Plugins'),
                                setupId: 'plugins',
                                priority: 30
                            }
                        ])
                    ]
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { PluginSetupStepperModule };
