// @c8y/ngx-components/fesm2020/c8y-ngx-components-widgets-definitions-device-management-welcome.mjs
import { gettext } from '@c8y/ngx-components';
import { defaultWidgetIds } from '@c8y/ngx-components/widgets/definitions';

// TODO: with typescript 4.9 add satisfies DynamicComponentDefinition
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator
const deviceManagementWelcomeWidgetDefinition = {
    id: defaultWidgetIds.DEVICE_MANAGEMENT_WELCOME,
    label: gettext('Quick Links'),
    description: gettext('Displays quick links for Device management'),
    loadComponent: () => import('@c8y/ngx-components/widgets/implementations/device-management-welcome').then(m => m.WelcomeToDeviceManagementComponent),
    previewImage: 'c8y-style-assets/dm-welcome-widget-pr.png',
    data: {
        settings: {
            noNewWidgets: false,
            widgetDefaults: {
                _width: 6,
                _height: 4
            },
            ng1: {
                options: {
                    noDeviceTarget: true,
                    groupsSelectable: false
                }
            }
        }
    }
};

/**
 * Generated bundle index. Do not edit.
 */

export { deviceManagementWelcomeWidgetDefinition };
