// @c8y/ngx-components/esm2020/bookmarks/bookmarks.service.mjs
import { Injectable } from '@angular/core';
import { UserPreferencesService } from '@c8y/ngx-components';
import { uniqBy } from 'lodash-es';
import { map } from 'rxjs/internal/operators/map';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
export class BookmarkService {
    constructor(userPreferencesService) {
        this.userPreferencesService = userPreferencesService;
        this.USER_PREFERENCES_BOOKMARKS_KEY = 'bookmarks';
    }
    async updateBookmarksInStorage(newBookmarks) {
        const existingBookmarks = await this.getBookmarks();
        const mergedBookmarks = [...newBookmarks, ...existingBookmarks];
        const cleanedBookmarks = uniqBy(mergedBookmarks, 'id').filter((bookmark) => !bookmark.markToRemove);
        this.setUserPreferencesBookmarks(cleanedBookmarks);
    }
    getCurrentActiveNodeIcon(document) {
        const BOOKMARK = 'bookmark';
        const iconElement = this.getIconElement(document);
        if (!iconElement) {
            return BOOKMARK;
        }
        const iconClassName = iconElement.className;
        const iconName = this.extractIconName(iconClassName);
        return iconName || BOOKMARK;
    }
    extractIconName(input) {
        const regex = /\b(?:dlt-)?c8y-icon-((?:c8y-)?\w+(?:-[\w-]+)*)\b/g;
        let iconName = null;
        input.replace(regex, (matchedIconString, name) => {
            iconName = matchedIconString.startsWith('dlt-') ? name : 'c8y-' + name;
            return matchedIconString;
        });
        return iconName;
    }
    async getBookmarks() {
        return await this.userPreferencesService
            .get(this.USER_PREFERENCES_BOOKMARKS_KEY)
            .pipe(map((bookmarks) => bookmarks ?? []))
            .toPromise();
    }
    generateRandomID() {
        const array = new Uint8Array(16);
        crypto.getRandomValues(array);
        return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
    }
    getIconElement(document) {
        const currentActiveNode = document.querySelectorAll('.link.active');
        if (!currentActiveNode.length) {
            return null;
        }
        const activeNode = currentActiveNode.item(0);
        return activeNode.firstElementChild.firstChild;
    }
    setUserPreferencesBookmarks(bookmarks) {
        this.userPreferencesService.set(this.USER_PREFERENCES_BOOKMARKS_KEY, bookmarks);
        this.updatedBookmarks = bookmarks;
    }
}
BookmarkService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BookmarkService, deps: [{ token: i1.UserPreferencesService }], target: i0.ɵɵFactoryTarget.Injectable });
BookmarkService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BookmarkService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BookmarkService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.UserPreferencesService }]; } });
