// @c8y/ngx-components/esm2020/bookmarks/bookmarks.module.mjs
import { DragDropModule } from '@angular/cdk/drag-drop';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule, CoreModule, FormsModule as c8yFormsModule, hookDrawer, ListGroupModule, ModalModule } from '@c8y/ngx-components';
import { BookmarksComponent } from './bookmarks.component';
import { EditBookmarksComponent } from './edit-bookmarks/edit-bookmarks.component';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/tooltip";
export class BookmarksModule {
}
BookmarksModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BookmarksModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BookmarksModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: BookmarksModule, declarations: [BookmarksComponent, EditBookmarksComponent], imports: [CoreModule,
        CommonModule,
        ListGroupModule,
        DragDropModule,
        FormsModule,
        c8yFormsModule,
        ModalModule, i1.TooltipModule], exports: [BookmarksComponent] });
BookmarksModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BookmarksModule, providers: [
        hookDrawer({
            component: BookmarksComponent,
            position: 'right',
            priority: 50,
            id: 'bookmarks'
        })
    ], imports: [CoreModule,
        CommonModule,
        ListGroupModule,
        DragDropModule,
        FormsModule,
        c8yFormsModule,
        ModalModule,
        TooltipModule.forRoot()] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: BookmarksModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BookmarksComponent, EditBookmarksComponent],
                    imports: [
                        CoreModule,
                        CommonModule,
                        ListGroupModule,
                        DragDropModule,
                        FormsModule,
                        c8yFormsModule,
                        ModalModule,
                        TooltipModule.forRoot()
                    ],
                    exports: [BookmarksComponent],
                    providers: [
                        hookDrawer({
                            component: BookmarksComponent,
                            position: 'right',
                            priority: 50,
                            id: 'bookmarks'
                        })
                    ]
                }]
        }] });
