// @c8y/ngx-components/esm2020/core/navigator/navigator-drawer.factory.mjs
import { Injectable } from '@angular/core';
import { NavigatorOutletComponent } from './navigator-outlet.component';
import { map } from 'rxjs/operators';
import { NavigatorService } from './navigator.service';
import * as i0 from "@angular/core";
import * as i1 from "./navigator.service";
export class NavigatorDrawerFactory {
    constructor(navigatorService) {
        this.navigatorService = navigatorService;
        this.drawerItem = {
            component: NavigatorOutletComponent,
            position: 'left',
            priority: 50,
            id: 'navigatorOutlet'
        };
    }
    get() {
        return this.navigatorService.hasItemsInNavigator$.pipe(map(tmp => (tmp ? this.drawerItem : [])));
    }
}
NavigatorDrawerFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorDrawerFactory, deps: [{ token: i1.NavigatorService }], target: i0.ɵɵFactoryTarget.Injectable });
NavigatorDrawerFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorDrawerFactory });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorDrawerFactory, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.NavigatorService }]; } });
