// @c8y/ngx-components/esm2020/core/navigator/navigator.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { CommonModule } from '../common/common.module';
import { NavigatorIconComponent } from './navigator-icon.component';
import { NavigatorNodeComponent } from './navigator-node.component';
import { NavigatorOutletComponent } from './navigator-outlet.component';
import { ModalModule } from '../modal/modal.module';
import { hookDrawer } from '../drawer/drawer.service';
import { NavigatorDrawerFactory } from './navigator-drawer.factory';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/collapse";
/**
 * The angular module definition for the navigator.
 * @exports NavigatorOutletComponent
 * @exports NavigatorNodeComponent
 */
export class NavigatorModule {
}
NavigatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NavigatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: NavigatorModule, declarations: [NavigatorOutletComponent, NavigatorNodeComponent, NavigatorIconComponent], imports: [RouterModule, i1.CollapseModule, CommonModule, ModalModule], exports: [NavigatorOutletComponent, NavigatorNodeComponent] });
NavigatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorModule, providers: [hookDrawer(NavigatorDrawerFactory)], imports: [RouterModule, CollapseModule.forRoot(), CommonModule, ModalModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NavigatorOutletComponent, NavigatorNodeComponent, NavigatorIconComponent],
                    imports: [RouterModule, CollapseModule.forRoot(), CommonModule, ModalModule],
                    exports: [NavigatorOutletComponent, NavigatorNodeComponent],
                    entryComponents: [NavigatorIconComponent],
                    providers: [hookDrawer(NavigatorDrawerFactory)]
                }]
        }] });
