// @c8y/ngx-components/esm2020/core/navigator/navigator-top/navigator-top.module.mjs
import { NgModule } from '@angular/core';
import { ApplicationModule } from '../../application/application.module';
import { CommonModule } from '../../common';
import { hookDrawer } from '../../drawer/drawer.service';
import { NavigatorTopComponent } from './navigator-top.component';
import * as i0 from "@angular/core";
export class NavigatorTopModule {
}
NavigatorTopModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorTopModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NavigatorTopModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.7", ngImport: i0, type: NavigatorTopModule, declarations: [NavigatorTopComponent], imports: [CommonModule, ApplicationModule] });
NavigatorTopModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorTopModule, providers: [
        hookDrawer({
            component: NavigatorTopComponent,
            position: 'left',
            priority: 100,
            id: 'tenantBrandAndAppIcon',
            noneRequired: true
        })
    ], imports: [CommonModule, ApplicationModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: NavigatorTopModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ApplicationModule],
                    declarations: [NavigatorTopComponent],
                    providers: [
                        hookDrawer({
                            component: NavigatorTopComponent,
                            position: 'left',
                            priority: 100,
                            id: 'tenantBrandAndAppIcon',
                            noneRequired: true
                        })
                    ]
                }]
        }] });
