// @c8y/ngx-components/esm2020/connectivity/loriot/loriot-tab.factory.mjs
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AbstractConnectivityTabFactory } from '../abstract-connectivity-tab.factory';
import { LoriotGuard } from './loriot.guard';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./loriot.guard";
export class LoriotTabFactory extends AbstractConnectivityTabFactory {
    constructor(router, guard) {
        super(router, guard);
        this.router = router;
        this.guard = guard;
        this.tabConfig = {
            label: 'LORIOT',
            path: '/connectivitySettings/multiple_lns_connectors_loriot',
            icon: 'wall-socket'
        };
    }
    config() {
        return this.tabConfig;
    }
}
LoriotTabFactory.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotTabFactory, deps: [{ token: i1.Router }, { token: i2.LoriotGuard }], target: i0.ɵɵFactoryTarget.Injectable });
LoriotTabFactory.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotTabFactory, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.7", ngImport: i0, type: LoriotTabFactory, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.LoriotGuard }]; } });
